// @lumiarq/lumis@1.2.0/dist/commands/commands.command.js
/**
 * Framework/project command bridge for Lumis.
 *
 * Core Lumis commands stay framework-agnostic. Framework-specific commands are
 * provided by packs, while project-specific commands are configured locally in
 * `lumis.config.ts` under `projectCommands`.
 */
import { spawn } from 'node:child_process';
import { once } from 'node:events';
/**
 * Collect normalized command metadata from both pack and project sources.
 *
 * @param context - Shared command runtime context.
 * @returns Combined normalized command entries sorted by source then name.
 */
export async function collectNormalizedCommands(context) {
    const packCommands = await context.resolvePackCommands();
    const projectCommands = context.resolveProjectCommands();
    const normalizedPack = packCommands.map((command) => ({
        source: 'pack',
        group: 'framework',
        name: command.name,
        description: command.description,
        args: command.args.map((arg) => ({
            name: arg.name,
            required: arg.required,
            description: arg.description,
        })),
    }));
    const normalizedProject = projectCommands.map((command) => ({
        source: 'project',
        group: 'project',
        name: command.name,
        description: command.description,
        args: [
            {
                name: 'shell',
                required: true,
                description: command.command,
            },
        ],
    }));
    return [...normalizedPack, ...normalizedProject].sort((left, right) => {
        if (left.source !== right.source) {
            return left.source.localeCompare(right.source);
        }
        return left.name.localeCompare(right.name);
    });
}
/**
 * Format normalized command entries in a stable text layout.
 *
 * @param entries - Normalized command records.
 * @returns Multi-line human-readable command listing.
 */
function formatNormalizedCommands(entries) {
    if (entries.length === 0) {
        return 'No pack or project commands available.\n';
    }
    const lines = ['Commands (normalized):'];
    for (const entry of entries) {
        const args = entry.args.length
            ? entry.args
                .map((arg) => `${arg.name}${arg.required ? '*' : ''}`)
                .join(', ')
            : '(none)';
        lines.push(`- ${entry.name} [source=${entry.source}, group=${entry.group}]`, `  description: ${entry.description}`, `  args: ${args}`);
    }
    return `${lines.join('\n')}\n`;
}
/**
 * List available pack and project commands.
 *
 * @param context - Shared command runtime context.
 * @returns Exit code 0 when listing succeeds.
 */
export async function executeCommandsListCommand(context) {
    const normalized = await collectNormalizedCommands(context);
    const asJson = context.commandArgs.includes('--json');
    if (asJson) {
        context.write(`${JSON.stringify(normalized, null, 2)}\n`);
        return 0;
    }
    context.write(formatNormalizedCommands(normalized));
    return 0;
}
/**
 * Run a pack or project command by name.
 *
 * Usage:
 * - `lumis commands run pack <name> [key=value ...]`
 * - `lumis commands run project <name>`
 *
 * @param context - Shared command runtime context.
 * @returns Exit code 0 on success, otherwise 1.
 */
export async function executeCommandsRunCommand(context) {
    const [scope, name, ...rest] = context.commandArgs;
    if (!scope || !name) {
        context.writeError('Usage: lumis commands run <pack|project> <name> [args]\n');
        return 1;
    }
    if (scope === 'pack') {
        const packCommands = await context.resolvePackCommands();
        const command = packCommands.find((entry) => entry.name === name);
        if (!command) {
            context.writeError(`Pack command not found: ${name}\n`);
            return 1;
        }
        const args = parsePackArgs(rest);
        if (!args) {
            context.writeError('Pack args must be key=value pairs.\n');
            return 1;
        }
        await command.run(args, context.projectRoot);
        context.write(`Executed pack command: ${name}\n`);
        return 0;
    }
    if (scope === 'project') {
        const projectCommands = context.resolveProjectCommands();
        const command = projectCommands.find((entry) => entry.name === name);
        if (!command) {
            context.writeError(`Project command not found: ${name}\n`);
            return 1;
        }
        const ok = await runProjectCommand(command, context);
        return ok ? 0 : 1;
    }
    context.writeError(`Unknown command scope: ${scope}. Use pack or project.\n`);
    return 1;
}
/**
 * Parse `key=value` pairs for pack command argument invocation.
 *
 * @param rawArgs - Raw argument tokens after command name.
 * @returns Parsed key/value map, or null when a token is malformed.
 */
function parsePackArgs(rawArgs) {
    const entries = [];
    for (const token of rawArgs) {
        const equalIndex = token.indexOf('=');
        if (equalIndex <= 0) {
            return null;
        }
        const key = token.slice(0, equalIndex).trim();
        const value = token.slice(equalIndex + 1).trim();
        if (!key) {
            return null;
        }
        entries.push([key, value]);
    }
    return Object.fromEntries(entries);
}
/**
 * Run a project-defined shell command in the project root.
 *
 * @param command - Project command definition from config.
 * @param context - Shared command runtime context.
 * @returns True when command exits with code 0.
 */
async function runProjectCommand(command, context) {
    context.write(`Running project command: ${command.name}\n`);
    const child = spawn(command.command, {
        cwd: context.projectRoot,
        shell: true,
        stdio: 'inherit',
    });
    const [exitCode] = (await once(child, 'close'));
    if (exitCode !== 0) {
        context.writeError(`Project command failed with code ${exitCode ?? 'unknown'}: ${command.name}\n`);
        return false;
    }
    context.write(`Executed project command: ${command.name}\n`);
    return true;
}
