// @lumiarq/lumis@1.2.0/dist/commands/init.command.js
/**
 * Initialize a Lumis project with a default config, scripts, IR build, and doctor checks.
 */
import fs from 'node:fs';
import path from 'node:path';
import child_process from 'node:child_process';
import { saveLumisConfig } from '../config-loader.js';
import { executeDoctorCommand } from './doctor.command.js';
import { executeIrRebuildCommand } from './ir-rebuild.command.js';
const DEFAULT_SCRIPT_ENTRY = 'lumis';
/**
 * Run Lumis project initialization.
 *
 * @param context - Shared command runtime context.
 * @returns Exit code (0 success, 1 error).
 */
export async function executeInitCommand(context) {
    const configPath = path.join(context.projectRoot, 'lumis.config.json');
    if (fs.existsSync(configPath)) {
        context.writeError(`lumis.config.json already exists at ${configPath}. init aborted.\n`);
        return 1;
    }
    const pack = await context.resolvePack();
    const adapterName = pack?.adapter.name;
    const framework = pack?.manifest.frameworks?.[0];
    const config = {
        ...(adapterName ? { adapter: adapterName } : {}),
        presetPrompts: [],
        context: { include: ['src/**'], exclude: ['node_modules/**'] },
        projectCommands: [],
    };
    try {
        saveLumisConfig(context.projectRoot, config);
        context.write(`Created ${configPath}\n`);
    }
    catch (error) {
        context.writeError(`Failed to write Lumis config: ${String(error)}\n`);
        return 1;
    }
    const packageJsonPath = path.join(context.projectRoot, 'package.json');
    let packageJson = null;
    let packageJsonDirty = false;
    const packDependency = pack?.manifest.name ?? '@lumiarq/pack-lumiarq';
    if (fs.existsSync(packageJsonPath)) {
        try {
            packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
            const scripts = packageJson?.scripts ?? {};
            if (!scripts[DEFAULT_SCRIPT_ENTRY]) {
                scripts[DEFAULT_SCRIPT_ENTRY] = 'lumis';
                packageJson.scripts = scripts;
                packageJsonDirty = true;
                context.write(`Added scripts.${DEFAULT_SCRIPT_ENTRY} to package.json\n`);
            }
            else {
                context.write('package.json script lumis already exists, skipping update.\n');
            }
            if (!packageJson?.dependencies) {
                packageJson.dependencies = {};
            }
            if (!packageJson?.dependencies[packDependency]) {
                packageJson.dependencies[packDependency] = '^1.0.0';
                packageJsonDirty = true;
                context.write(`Added ${packDependency} to package.json dependencies.\n`);
            }
            if (packageJsonDirty) {
                fs.writeFileSync(packageJsonPath, JSON.stringify(packageJson, null, 2), 'utf8');
                context.write('Updated package.json with Lumis settings.\n');
            }
        }
        catch (e) {
            context.writeError(`Could not update package.json: ${String(e)}\n`);
        }
    }
    else {
        context.write('No package.json found; skipping script setup.\n');
    }
    if (packageJson && packageJson.dependencies && packageJson.dependencies[packDependency]) {
        const skipAutoInstall = process.env.LUMIS_SKIP_AUTO_INSTALL === '1';
        if (skipAutoInstall) {
            context.write('Auto-install skipped via LUMIS_SKIP_AUTO_INSTALL.\n');
        }
        else {
            try {
                const lockFile = fs.existsSync(path.join(context.projectRoot, 'pnpm-lock.yaml'))
                    ? 'pnpm'
                    : fs.existsSync(path.join(context.projectRoot, 'yarn.lock'))
                        ? 'yarn'
                        : fs.existsSync(path.join(context.projectRoot, 'package-lock.json'))
                            ? 'npm'
                            : 'npm';
                const installCmd = lockFile === 'pnpm' ? 'pnpm install' : lockFile === 'yarn' ? 'yarn install' : 'npm install';
                context.write(`Running ${installCmd} to ensure pack is installed...\n`);
                child_process.execSync(installCmd, {
                    cwd: context.projectRoot,
                    stdio: 'inherit',
                });
                context.write('Dependency installation step completed.\n');
            }
            catch (e) {
                context.writeError(`Failed to install dependencies automatically: ${String(e)}\n`);
            }
        }
    }
    context.write(`Selected pack: ${pack?.manifest.name ?? 'none'} (${framework ?? 'unknown'})\n`);
    context.write(`Selected adapter: ${adapterName ?? 'none'}\n`);
    context.write('Running initial IR rebuild...\n');
    const rebuildCode = await executeIrRebuildCommand(context);
    if (rebuildCode !== 0) {
        context.writeError('ir rebuild failed during init.\n');
        return rebuildCode;
    }
    context.write('Running doctor checks...\n');
    const doctorCode = await executeDoctorCommand(context);
    if (doctorCode !== 0) {
        context.writeError('doctor checks failed during init.\n');
        return doctorCode;
    }
    context.write('Lumis initialization completed successfully.\n');
    return 0;
}
