// @lumiarq/lumis@1.2.0/dist/commands/intent.command.js
/**
 * Intent command.
 *
 * This command parses a textual intent into the generic ParsedIntent shape,
 * resolves adapter/IR context, and previews an execution plan.
 */
import { createEmptyIR, loadIR } from '@lumiarq/context';
import { parseIntentWithAI } from '@lumiarq/agent';
import { createPlanWithAdapter } from '@lumiarq/planner';
import { applyRouterPreferenceOverrides } from '../router-preferences.js';
/**
 * Parse a token into the supported generic intent action enum.
 *
 * @param token - First token from raw intent input.
 * @returns Parsed action when valid, otherwise null.
 */
function parseAction(token) {
    if (!token) {
        return null;
    }
    const lowered = token.toLowerCase();
    if (lowered === 'create' || lowered === 'update' || lowered === 'delete' || lowered === 'query') {
        return lowered;
    }
    return null;
}
/**
 * Lightweight parser for command-line intent text.
 *
 * The adapter remains the source of truth for how intent maps to concrete steps;
 * this parser only normalizes a generic shape so adapters can resolve it.
 *
 * @param raw - Raw intent text passed from the command line.
 * @returns Parsed intent object, or null when input cannot be parsed.
 */
export function parseRawIntent(raw) {
    const trimmed = raw.trim();
    if (!trimmed) {
        return null;
    }
    const tokens = trimmed.split(/\s+/);
    const action = parseAction(tokens[0]);
    if (!action) {
        return null;
    }
    const kind = tokens[1] ?? 'artifact';
    const name = tokens[2];
    const inIndex = tokens.findIndex((token) => token.toLowerCase() === 'in');
    const module = inIndex >= 0 && tokens[inIndex + 1] ? tokens[inIndex + 1] : undefined;
    const withIndex = tokens.findIndex((token) => token.toLowerCase() === 'with');
    const constraints = withIndex >= 0 ? tokens.slice(withIndex + 1).join(' ').split(/\s+and\s+/i).filter(Boolean) : [];
    const subject = {
        kind,
        ...(name ? { name } : {}),
        ...(module ? { attributes: { module } } : {}),
    };
    return {
        raw: trimmed,
        action,
        subjects: [subject],
        constraints,
    };
}
function parseIntentOptions(commandArgs) {
    const filtered = [];
    let useAI = false;
    let provider;
    let model;
    for (let index = 0; index < commandArgs.length; index += 1) {
        const token = commandArgs[index];
        if (token === '--ai') {
            useAI = true;
            continue;
        }
        if (token === '--provider') {
            const providerArg = commandArgs[index + 1];
            if (providerArg) {
                provider = providerArg;
                index += 1;
            }
            continue;
        }
        if (token === '--model') {
            const modelArg = commandArgs[index + 1];
            if (modelArg) {
                model = modelArg;
                index += 1;
            }
            continue;
        }
        if (token) {
            filtered.push(token);
        }
    }
    return {
        useAI,
        provider,
        model,
        rawIntent: filtered.join(' ').trim(),
    };
}
function resolveAIModel(context, options) {
    if (options.model && options.model.trim().length > 0) {
        return options.model;
    }
    const taskModel = context.config.ai?.taskModels?.intentParse;
    if (taskModel && taskModel.trim().length > 0) {
        return taskModel;
    }
    if (context.config.ai?.defaultModel && context.config.ai.defaultModel.trim().length > 0) {
        return context.config.ai.defaultModel;
    }
    if (context.config.ai?.model && context.config.ai.model.trim().length > 0) {
        return context.config.ai.model;
    }
    return undefined;
}
async function parseIntent(context, options) {
    const aiEnabled = options.useAI || context.config.ai?.enabled === true;
    if (!aiEnabled) {
        return { intent: parseRawIntent(options.rawIntent) };
    }
    const provider = options.provider ?? context.config.ai?.provider;
    const model = resolveAIModel(context, options);
    const modelWithProvider = provider && model && !model.includes(':') ? `${provider}:${model}` : model;
    const taskModels = {
        ...(context.config.ai?.taskModels?.intentParse ? { 'intent-parse': context.config.ai.taskModels.intentParse } : {}),
        ...(context.config.ai?.taskModels?.planPropose ? { 'plan-propose': context.config.ai.taskModels.planPropose } : {}),
        ...(context.config.ai?.taskModels?.tokenSuggest ? { 'token-suggest': context.config.ai.taskModels.tokenSuggest } : {}),
        ...(context.config.ai?.taskModels?.templateVariant ? { 'template-variant': context.config.ai.taskModels.templateVariant } : {}),
    };
    const parsed = await parseIntentWithAI(options.rawIntent, {
        ...(modelWithProvider ? { model: modelWithProvider } : {}),
        ...(provider ? { defaultProviderId: provider } : {}),
        ...(context.config.ai?.systemPromptExtra ? { systemPromptExtra: context.config.ai.systemPromptExtra } : {}),
        modelPolicy: {
            ...(context.config.ai?.mode ? { mode: context.config.ai.mode } : {}),
            ...(context.config.ai?.defaultModel || context.config.ai?.model
                ? { defaultModel: context.config.ai.defaultModel ?? context.config.ai.model }
                : {}),
            taskModels,
        },
    });
    const diagnostics = [];
    diagnostics.push(`AI parser: provider=${parsed.diagnostics.provider} model=${parsed.diagnostics.model} confidence=${parsed.diagnostics.confidence.toFixed(2)}`);
    if (parsed.diagnostics.ambiguityReasons.length > 0) {
        diagnostics.push(`AI ambiguity: ${parsed.diagnostics.ambiguityReasons.join('; ')}`);
    }
    return {
        intent: parsed.intent,
        diagnostics,
    };
}
/**
 * Render a concise plan preview to stdout.
 *
 * @param context - Shared command runtime context.
 * @param intent - Parsed intent to display.
 * @param estimatedFiles - Estimated number of files affected by the plan.
 * @param stepCount - Number of execution steps in the generated plan.
 * @returns Nothing. Writes formatted preview text to the command output stream.
 */
function renderPlanPreview(context, intent, estimatedFiles, stepCount) {
    context.write('Intent preview\n');
    context.write(`Raw: ${intent.raw}\n`);
    context.write(`Action: ${intent.action}\n`);
    const subject = intent.subjects[0];
    const module = typeof subject?.attributes?.module === 'string' ? subject.attributes.module : 'n/a';
    context.write(`Subject: kind=${subject?.kind ?? 'n/a'} name=${subject?.name ?? 'n/a'} module=${module}\n`);
    if (intent.constraints.length > 0) {
        context.write(`Constraints: ${intent.constraints.join(', ')}\n`);
    }
    context.write(`Plan steps: ${stepCount}\n`);
    context.write(`Estimated files: ${estimatedFiles}\n`);
}
/**
 * Parse, resolve, and preview an intent using the detected adapter.
 *
 * @param context - Shared command runtime context.
 * @returns Exit code (0 success, 1 when input/adapter state is invalid).
 */
export async function executeIntentCommand(context) {
    const options = parseIntentOptions(context.commandArgs);
    if (!options.rawIntent) {
        context.writeError('Missing intent text.\n');
        context.write('Usage: lumis intent [--ai] [--provider <provider>] [--model <model>] <create|update|delete|query ...>\n');
        return 1;
    }
    const parsed = await parseIntent(context, options);
    const parsedIntent = parsed.intent;
    if (!parsedIntent) {
        context.writeError('Invalid intent. Expected first token to be one of: create, update, delete, query.\n');
        return 1;
    }
    for (const line of parsed.diagnostics ?? []) {
        context.write(`${line}\n`);
    }
    const adapter = await context.resolveAdapter();
    if (!adapter) {
        context.writeError('No compatible adapter/pack detected for this project.\n');
        return 1;
    }
    const ir = applyRouterPreferenceOverrides(loadIR(context.projectRoot) ?? createEmptyIR(context.projectRoot), adapter.name, context.config);
    const plan = await createPlanWithAdapter(parsedIntent, ir, adapter);
    renderPlanPreview(context, parsedIntent, plan.estimatedFiles, plan.steps.length);
    if (plan.steps.length === 0) {
        context.write('Adapter returned no executable steps for this intent.\n');
    }
    return 0;
}
