// @lumiarq/lumis@1.2.0/dist/commands/make.command.js
/**
 * Artifact generation command.
 *
 * This command takes free-form intent text, resolves the active adapter/pack,
 * plans execution steps, and optionally executes the plan.
 */
import { executePlan } from '@lumiarq/agent';
import { createEmptyIR, loadIR, saveIR, updateIRWithAdapter } from '@lumiarq/context';
import { createPlanWithAdapter } from '@lumiarq/planner';
import { parseRawIntent } from './intent.command.js';
import { renderPlanDetail } from '../output/plan-display.js';
import { renderExecutionResult } from '../output/result-display.js';
import { applyRouterPreferenceOverrides } from '../router-preferences.js';
import { runPostMakeRuntimeTasks } from '../runtime/orchestration.js';
/**
 * Parse CLI arguments for the make command.
 *
 * @param commandArgs - Raw command arguments passed after `make`.
 * @returns Parsed make options including dry-run flag and raw intent text.
 */
function parseMakeOptions(commandArgs) {
    const dryRun = commandArgs.includes('--dry-run') || commandArgs.includes('-n');
    const filtered = commandArgs.filter((token) => token !== '--dry-run' && token !== '-n');
    return {
        dryRun,
        rawIntent: filtered.join(' ').trim(),
    };
}
/**
 * Execute the make command pipeline.
 *
 * @param context - Shared command runtime context.
 * @returns Exit code: 0 on success, 1 when validation/resolution/execution fails.
 */
export async function executeMakeCommand(context) {
    const options = parseMakeOptions(context.commandArgs);
    if (!options.rawIntent) {
        context.writeError('Missing intent text.\n');
        context.write('Usage: lumis make [--dry-run|-n] <create|update|delete|query ...>\n');
        return 1;
    }
    const parsedIntent = parseRawIntent(options.rawIntent);
    if (!parsedIntent) {
        context.writeError('Invalid intent. Expected first token to be one of: create, update, delete, query.\n');
        return 1;
    }
    const pack = await context.resolvePack();
    const adapter = await context.resolveAdapter();
    if (!adapter) {
        context.writeError('No compatible adapter/pack detected for this project.\n');
        return 1;
    }
    const ir = applyRouterPreferenceOverrides(loadIR(context.projectRoot) ?? createEmptyIR(context.projectRoot), adapter.name, context.config);
    const plan = await createPlanWithAdapter(parsedIntent, ir, adapter);
    const packFramework = pack ? (pack.manifest.frameworks?.[0] ?? 'unknown') : 'unknown';
    const packLabel = pack ? `${pack.manifest.name} (${packFramework})` : adapter.name;
    renderPlanDetail(context, plan, packLabel);
    if (plan.steps.length === 0) {
        context.write('No executable steps were produced for this intent.\n');
        return 0;
    }
    if (options.dryRun) {
        return 0;
    }
    const execution = await executePlan(plan, ir, adapter, {
        ...(context.onTrace ? { onTrace: context.onTrace } : {}),
        updateIR: async (currentIR, changedFile) => updateIRWithAdapter(adapter, currentIR, changedFile),
    });
    renderExecutionResult(context, execution.summary, plan.steps);
    if (execution.summary.failed || execution.summary.rolledBack) {
        context.writeError('Execution failed and rollback was attempted.\n');
        if (execution.summary.failedStepId) {
            context.writeError(`Failed step: ${execution.summary.failedStepId}\n`);
        }
        return 1;
    }
    saveIR(context.projectRoot, execution.ir);
    context.write('Saved IR: .lumis/ir.json\n');
    const runtimeResult = await runPostMakeRuntimeTasks({
        projectRoot: context.projectRoot,
        adapter,
        config: context.config,
        write: context.write,
        writeError: context.writeError,
        ...(context.onTrace ? { onTrace: context.onTrace } : {}),
    });
    if (!runtimeResult.success) {
        context.writeError('Runtime task execution failed. See transcript files under .lumis/transcripts.\n');
        return 1;
    }
    return 0;
}
