// @lumiarq/lumis@1.2.0/dist/commands/runtime.command.js
import fs from 'node:fs';
import path from 'node:path';
import { executeRuntimeTasks } from '../runtime/orchestration.js';
import { loadRuntimeTranscriptIndex, pruneRuntimeTranscripts } from '../runtime/transcripts.js';
const VALID_TASKS = ['typecheck', 'lint', 'boundaryCheck', 'build', 'test'];
function isRuntimeTask(value) {
    return VALID_TASKS.includes(value);
}
function parseRuntimeRunOptions(args) {
    const tasks = [];
    let profileId;
    let continueOnError = false;
    let approveRemote = false;
    let asJson = false;
    for (let index = 0; index < args.length; index += 1) {
        const token = args[index];
        if (!token) {
            continue;
        }
        if (token === '--profile') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--profile requires a value' };
            }
            profileId = value;
            index += 1;
            continue;
        }
        if (token === '--task') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--task requires a value' };
            }
            if (!isRuntimeTask(value)) {
                return { ok: false, error: `Unsupported task: ${value}. Supported tasks: ${VALID_TASKS.join(', ')}` };
            }
            tasks.push(value);
            index += 1;
            continue;
        }
        if (token === '--continue-on-error') {
            continueOnError = true;
            continue;
        }
        if (token === '--approve-remote') {
            approveRemote = true;
            continue;
        }
        if (token === '--json') {
            asJson = true;
            continue;
        }
        return { ok: false, error: `Unknown runtime option: ${token}` };
    }
    return {
        ok: true,
        value: {
            profileId,
            tasks,
            continueOnError,
            approveRemote,
            asJson,
        },
    };
}
function parseRuntimePruneOptions(args) {
    let keep = 20;
    let maxAgeDays;
    let profileId;
    let label;
    let dryRun = false;
    let asJson = false;
    for (let index = 0; index < args.length; index += 1) {
        const token = args[index];
        if (!token) {
            continue;
        }
        if (token === '--keep') {
            const value = args[index + 1];
            const parsed = Number(value);
            if (!value || !Number.isInteger(parsed) || parsed < 0) {
                return { ok: false, error: '--keep requires a non-negative integer' };
            }
            keep = parsed;
            index += 1;
            continue;
        }
        if (token === '--max-age-days') {
            const value = args[index + 1];
            const parsed = Number(value);
            if (!value || Number.isNaN(parsed) || parsed < 0) {
                return { ok: false, error: '--max-age-days requires a non-negative number' };
            }
            maxAgeDays = parsed;
            index += 1;
            continue;
        }
        if (token === '--profile') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--profile requires a value' };
            }
            profileId = value;
            index += 1;
            continue;
        }
        if (token === '--label') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--label requires a value' };
            }
            label = value;
            index += 1;
            continue;
        }
        if (token === '--dry-run') {
            dryRun = true;
            continue;
        }
        if (token === '--json') {
            asJson = true;
            continue;
        }
        return { ok: false, error: `Unknown prune option: ${token}` };
    }
    return {
        ok: true,
        value: {
            keep,
            maxAgeDays,
            profileId,
            label,
            dryRun,
            asJson,
        },
    };
}
function parseRuntimeListOptions(args) {
    let profileId;
    let label;
    let limit = 20;
    let asJson = false;
    for (let index = 0; index < args.length; index += 1) {
        const token = args[index];
        if (!token) {
            continue;
        }
        if (token === '--profile') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--profile requires a value' };
            }
            profileId = value;
            index += 1;
            continue;
        }
        if (token === '--label') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--label requires a value' };
            }
            label = value;
            index += 1;
            continue;
        }
        if (token === '--limit') {
            const value = args[index + 1];
            const parsed = Number(value);
            if (!value || !Number.isInteger(parsed) || parsed <= 0) {
                return { ok: false, error: '--limit requires a positive integer' };
            }
            limit = parsed;
            index += 1;
            continue;
        }
        if (token === '--json') {
            asJson = true;
            continue;
        }
        return { ok: false, error: `Unknown list option: ${token}` };
    }
    return {
        ok: true,
        value: {
            profileId,
            label,
            limit,
            asJson,
        },
    };
}
function parseRuntimeShowOptions(args) {
    let id;
    let asJson = false;
    for (let index = 0; index < args.length; index += 1) {
        const token = args[index];
        if (!token) {
            continue;
        }
        if (token === '--id') {
            const value = args[index + 1];
            if (!value) {
                return { ok: false, error: '--id requires a value' };
            }
            id = value;
            index += 1;
            continue;
        }
        if (token === '--json') {
            asJson = true;
            continue;
        }
        if (!id && !token.startsWith('--')) {
            id = token;
            continue;
        }
        return { ok: false, error: `Unknown show option: ${token}` };
    }
    if (!id) {
        return { ok: false, error: 'Missing transcript id. Provide <id> or --id <id>.' };
    }
    return {
        ok: true,
        value: { id, asJson },
    };
}
function getRuntimeAdapter(context, adapter) {
    if (adapter) {
        return adapter;
    }
    if (context.config.runtime?.taskCommandOverrides && Object.keys(context.config.runtime.taskCommandOverrides).length > 0) {
        return {
            name: 'runtime-override',
            version: '1.0.0',
            language: 'javascript',
            detect: async () => ({ detected: true, confidence: 1, reason: 'runtime task overrides' }),
            buildIR: async () => {
                throw new Error('runtime buildIR is unavailable for override-only execution');
            },
            updateIR: async (ir) => ir,
            placementRules: [],
            dependencyRules: [],
            namingRules: [],
            resolveIntent: async () => [],
            resolveTemplate: () => '',
            verificationHooks: [],
            buildCommands: {},
        };
    }
    return null;
}
export async function executeRuntimeRunCommand(context) {
    const parsedOptions = parseRuntimeRunOptions(context.commandArgs);
    if (!parsedOptions.ok) {
        context.writeError(`${parsedOptions.error}\n`);
        context.write('Usage: lumis runtime run [--profile <id>] [--task <typecheck|lint|boundaryCheck|build|test>] [--continue-on-error] [--approve-remote] [--json]\n');
        return 1;
    }
    const options = parsedOptions.value;
    const adapter = getRuntimeAdapter(context, await context.resolveAdapter());
    if (!adapter) {
        context.writeError('No compatible adapter/pack detected and no runtime task overrides configured.\n');
        return 1;
    }
    const result = await executeRuntimeTasks({
        projectRoot: context.projectRoot,
        adapter,
        config: context.config,
        write: context.write,
        writeError: context.writeError,
        ...(context.onTrace ? { onTrace: context.onTrace } : {}),
        enabled: true,
        ...(options.profileId ? { profileId: options.profileId } : {}),
        requestedTasks: options.tasks,
        failOnTaskError: !options.continueOnError,
        approveRemote: options.approveRemote,
        labelPrefix: 'runtime',
    });
    if (options.asJson) {
        context.write(`${JSON.stringify(result, null, 2)}\n`);
    }
    return result.success ? 0 : 1;
}
export async function executeRuntimeTranscriptsPruneCommand(context) {
    const parsedOptions = parseRuntimePruneOptions(context.commandArgs);
    if (!parsedOptions.ok) {
        context.writeError(`${parsedOptions.error}\n`);
        context.write('Usage: lumis runtime transcripts prune [--keep <count>] [--max-age-days <days>] [--profile <id>] [--label <prefix>] [--dry-run] [--json]\n');
        return 1;
    }
    const options = parsedOptions.value;
    const result = pruneRuntimeTranscripts(context.projectRoot, {
        keep: options.keep,
        ...(options.maxAgeDays !== undefined ? { maxAgeDays: options.maxAgeDays } : {}),
        ...(options.profileId ? { profileId: options.profileId } : {}),
        ...(options.label ? { label: options.label } : {}),
        dryRun: options.dryRun,
    });
    if (options.asJson) {
        context.write(`${JSON.stringify(result, null, 2)}\n`);
        return 0;
    }
    context.write(`${options.dryRun ? 'Would remove' : 'Removed'} ${result.removedEntries.length} transcript entr${result.removedEntries.length === 1 ? 'y' : 'ies'}; ` +
        `${result.keptEntries.length} retained.\n`);
    if (result.missingFilesRemoved > 0) {
        context.write(`Missing transcript references pruned: ${result.missingFilesRemoved}\n`);
    }
    context.write(`Transcript index: ${result.indexPath}\n`);
    return 0;
}
export async function executeRuntimeTranscriptsListCommand(context) {
    const parsedOptions = parseRuntimeListOptions(context.commandArgs);
    if (!parsedOptions.ok) {
        context.writeError(`${parsedOptions.error}\n`);
        context.write('Usage: lumis runtime transcripts list [--profile <id>] [--label <prefix>] [--limit <n>] [--json]\n');
        return 1;
    }
    const options = parsedOptions.value;
    const index = loadRuntimeTranscriptIndex(context.projectRoot);
    const entries = index.entries
        .filter((entry) => (options.profileId ? entry.profileId === options.profileId : true))
        .filter((entry) => (options.label ? entry.label.startsWith(options.label) : true))
        .slice(0, options.limit);
    if (options.asJson) {
        context.write(`${JSON.stringify({ total: entries.length, entries }, null, 2)}\n`);
        return 0;
    }
    if (entries.length === 0) {
        context.write('No transcript entries matched the current filters.\n');
        return 0;
    }
    context.write(`Runtime transcripts (${entries.length}):\n`);
    for (const entry of entries) {
        context.write(`- ${entry.id} label=${entry.label} profile=${entry.profileId} result=${entry.success ? 'ok' : 'fail'} ` +
            `duration=${entry.durationMs}ms file=${entry.transcriptFile}\n`);
    }
    return 0;
}
export async function executeRuntimeTranscriptsShowCommand(context) {
    const parsedOptions = parseRuntimeShowOptions(context.commandArgs);
    if (!parsedOptions.ok) {
        context.writeError(`${parsedOptions.error}\n`);
        context.write('Usage: lumis runtime transcripts show <id> [--json]\n');
        return 1;
    }
    const options = parsedOptions.value;
    const index = loadRuntimeTranscriptIndex(context.projectRoot);
    const entry = index.entries.find((candidate) => candidate.id === options.id);
    if (!entry) {
        context.writeError(`Transcript id not found: ${options.id}\n`);
        return 1;
    }
    const transcriptPath = path.join(context.projectRoot, entry.transcriptFile);
    if (!fs.existsSync(transcriptPath)) {
        context.writeError(`Transcript file missing: ${entry.transcriptFile}\n`);
        return 1;
    }
    const raw = fs.readFileSync(transcriptPath, 'utf8');
    if (options.asJson) {
        context.write(`${raw}\n`);
        return 0;
    }
    const parsed = JSON.parse(raw);
    context.write(`Transcript: ${entry.id}\n`);
    context.write(`Command: ${parsed.command} ${parsed.args.join(' ')}\n`);
    context.write(`Result: ${parsed.success ? 'ok' : 'fail'} (exit=${parsed.exitCode ?? 'null'})\n`);
    context.write(`Duration: ${parsed.durationMs}ms\n`);
    if (parsed.diagnostics.length > 0) {
        context.write(`Diagnostics: ${parsed.diagnostics.join(' | ')}\n`);
    }
    if (parsed.stdout.length > 0) {
        context.write(`\n--- stdout ---\n${parsed.stdout}\n`);
    }
    if (parsed.stderr.length > 0) {
        context.write(`\n--- stderr ---\n${parsed.stderr}\n`);
    }
    return 0;
}
