// @lumiarq/lumis@1.2.0/dist/commands/runtime.js
import { resolveAdapterForProject, resolvePackForProject } from '../adapter-loader.js';
import { loadLumisConfig } from '../config-loader.js';
/**
 * Build a reusable command execution context.
 *
 * @param options - Runtime inputs such as argv, command path/args, project root, and output writers.
 * @returns Hydrated command context with lazy pack/adapter resolvers and loaded config.
 */
export async function createCommandContext(options) {
    const config = await loadLumisConfig(options.projectRoot, options.onTrace ? { onTrace: options.onTrace } : undefined);
    let cachedPack = null;
    let cachedAdapter = null;
    let cachedPackCommands = null;
    const write = options.write ?? ((message) => process.stdout.write(message));
    const writeError = options.writeError ?? ((message) => process.stderr.write(message));
    return {
        projectRoot: options.projectRoot,
        commandPath: options.commandPath ?? [],
        commandArgs: options.commandArgs ?? [],
        argv: options.argv,
        config,
        write,
        writeError,
        resolvePack: async () => {
            if (!cachedPack) {
                const resolverOptions = config.adapter
                    ? { configuredAdapter: config.adapter }
                    : undefined;
                cachedPack = resolvePackForProject(options.projectRoot, resolverOptions);
            }
            return cachedPack;
        },
        resolveAdapter: async () => {
            if (!cachedAdapter) {
                const resolverOptions = config.adapter
                    ? { configuredAdapter: config.adapter }
                    : undefined;
                cachedAdapter = resolveAdapterForProject(options.projectRoot, resolverOptions);
            }
            return cachedAdapter;
        },
        resolvePackCommands: async () => {
            if (!cachedPackCommands) {
                cachedPackCommands = (async () => {
                    const resolvedPack = await (async () => {
                        if (!cachedPack) {
                            const resolverOptions = config.adapter
                                ? { configuredAdapter: config.adapter }
                                : undefined;
                            cachedPack = resolvePackForProject(options.projectRoot, resolverOptions);
                        }
                        return cachedPack;
                    })();
                    if (!resolvedPack?.adapter.commands) {
                        return [];
                    }
                    return resolvedPack.adapter.commands();
                })();
            }
            return cachedPackCommands;
        },
        resolveProjectCommands: () => config.projectCommands ?? [],
        ...(options.onTrace ? { onTrace: options.onTrace } : {}),
    };
}
