// @lumiarq/lumis@1.2.0/dist/commands/tinker-chat-runtime.js
import { createGatewayFromEnv, createProvidersFromEnv, } from '@lumiarq/agent';
import { AIGatewayEmbedder, BM25Index, createGroundingGenerator, createRAGPipeline, HybridRetriever, MemoryVectorStore, SlidingWindowChunker, } from '@lumiarq/rag';
import { scanProjectFiles } from '../lib/file-scanner.js';
import { appendMemory, projectFingerprint, saveMemoryStore, retrieveMemoryContext, } from '../lib/tinker-memory.js';
import { formatToolResult } from '../tools/tool-dispatch.js';
import { extractToolCalls, stripToolCalls, TOOL_CALL_SYSTEM_SECTION } from '../tools/tool-dispatch.js';
import { executeManagedToolCall, handleAssistantMessage, handleUserMessage, } from './tinker-session-runtime.js';
const MAX_TOOL_CALL_ROUNDS = 12;
export function buildChatSystemMessage(session) {
    const runtime = session.context.config.preferredRuntime ?? 'auto';
    const model = session.chatModel ?? 'provider default';
    const baseParts = [
        'You are Lumis Tinker Chat — an expert coding assistant with direct access to this project.',
        `Project root: ${session.context.projectRoot}`,
        `Framework: ${session.framework}`,
        `Pack: ${session.packName}`,
        `Runtime preference: ${runtime}`,
        `Model preference: ${model}`,
        '',
        'Behavior rules:',
        '- Be concise and practical for software engineering tasks.',
        '- Never fabricate model internals, training data, or platform architecture.',
        '- If asked about model/provider and unknown, say so directly.',
        '- You can receive retrieved repository context from indexed project files. Prefer it over assumptions.',
        '- If source context is missing, use the read-file tool to fetch it.',
    ];
    if (session.toolCallsEnabled) {
        baseParts.push('', TOOL_CALL_SYSTEM_SECTION);
    }
    return { role: 'system', content: baseParts.join('\n') };
}
export function resetChatHistory(session) {
    session.chatHistory = [buildChatSystemMessage(session)];
}
export function looksLikeFilesystemPath(text) {
    const value = text.trim();
    if (!value || value.includes(' '))
        return false;
    if (value.startsWith('/'))
        return true;
    if (value.startsWith('./') || value.startsWith('../'))
        return true;
    if (value.startsWith('~/'))
        return true;
    return value.includes('/');
}
function buildRetrievalContext(chunks) {
    if (chunks.length === 0)
        return '';
    const selected = chunks.slice(0, 6);
    const context = selected
        .map((item, index) => {
        const sourcePath = String(item.chunk.metadata?.path ?? item.chunk.documentId);
        return `[${index + 1}] ${sourcePath}\n${item.chunk.text.trim()}`;
    })
        .join('\n\n');
    return [
        'Repository context for this turn (retrieved from indexed project files):',
        context,
        '',
        'Answer using this context when relevant. If context is insufficient, say what is missing.',
    ].join('\n');
}
export async function ensureChatGateway(session) {
    if (session.chatGateway) {
        return session.chatGateway;
    }
    const preferredRuntime = session.context.config.preferredRuntime;
    const providers = createProvidersFromEnv();
    const availableProviderIds = new Set(providers.map((provider) => provider.id));
    const canUsePreferredRuntime = preferredRuntime ? availableProviderIds.has(preferredRuntime) : false;
    if (canUsePreferredRuntime && preferredRuntime) {
        session.chatGateway = await createGatewayFromEnv({ defaultProviderId: preferredRuntime });
    }
    else {
        session.chatGateway = await createGatewayFromEnv();
    }
    return session.chatGateway;
}
export async function initRepoRag(session) {
    if (session.ragReady || session.ragInitializing) {
        return;
    }
    session.ragInitializing = true;
    session.context.write('Indexing repository context for grounded chat...\n');
    try {
        const gateway = await ensureChatGateway(session);
        const docs = await scanProjectFiles(session.context.projectRoot);
        if (docs.length === 0) {
            session.context.write('Repo grounding skipped: no indexable files found.\n\n');
            return;
        }
        const embedder = new AIGatewayEmbedder(gateway, {
            dimensions: 1536,
            ...(session.chatModel ? { model: session.chatModel } : {}),
        });
        const store = new MemoryVectorStore();
        const retriever = new HybridRetriever(embedder, store, new BM25Index());
        const generator = createGroundingGenerator(async (systemPrompt, userPrompt) => {
            const result = await gateway.complete({
                task: 'answer-ground',
                prompt: userPrompt,
                systemPrompt,
                ...(session.chatModel ? { model: session.chatModel } : {}),
            });
            return result.text;
        });
        const pipeline = createRAGPipeline({
            chunker: new SlidingWindowChunker(),
            embedder,
            store,
            retriever,
            generator,
            topK: 8,
            topN: 5,
        });
        await pipeline.ingest(docs);
        session.ragPipeline = pipeline;
        session.ragIndexedFiles = docs.length;
        session.ragReady = true;
        session.context.write(`Repo grounding ready (${docs.length} files indexed).\n\n`);
    }
    catch (error) {
        session.context.writeError(`Repo grounding unavailable: ${error instanceof Error ? error.message : String(error)}\n\n`);
    }
    finally {
        session.ragInitializing = false;
    }
}
export async function retrieveRepoContext(session, text) {
    if (!session.ragReady) {
        await initRepoRag(session);
    }
    if (!session.ragPipeline) {
        return undefined;
    }
    const chunks = await session.ragPipeline.retrieve(text, { topK: 6, strategy: 'hybrid' });
    if (chunks.length === 0) {
        return undefined;
    }
    return buildRetrievalContext(chunks);
}
async function recordSyntheticAssistantContext(session, turnIndex, message, kind) {
    session.sessionReplayer.recordAssistantMessage(message);
    await session.observer.recordTurn({
        timestamp: new Date().toISOString(),
        sessionId: session.sessionId,
        projectFingerprint: projectFingerprint(session.context.projectRoot),
        turnIndex,
        eventType: 'assistant-message',
        data: { message, kind },
    });
}
export async function sendChatMessage(session, text) {
    const gateway = await ensureChatGateway(session);
    const turnIndex = session.chatHistory.length;
    const nextText = await handleUserMessage(session, text, turnIndex);
    const retrievalContext = await retrieveRepoContext(session, nextText);
    const memoryContext = retrieveMemoryContext(session.memoryStore, session.context.projectRoot, nextText);
    const outbound = [...session.chatHistory];
    if (retrievalContext) {
        outbound.push({ role: 'system', content: retrievalContext });
        await recordSyntheticAssistantContext(session, turnIndex, retrievalContext, 'retrieval-context');
    }
    if (memoryContext) {
        outbound.push({ role: 'system', content: memoryContext });
        await recordSyntheticAssistantContext(session, turnIndex, memoryContext, 'memory-context');
    }
    outbound.push({ role: 'user', content: nextText });
    session.chatHistory.push({ role: 'user', content: nextText });
    let round = 0;
    while (round < MAX_TOOL_CALL_ROUNDS) {
        round++;
        const result = await gateway.complete({
            task: 'answer-ground',
            prompt: nextText,
            messages: outbound,
            ...(session.chatModel ? { model: session.chatModel } : {}),
        });
        const rawReply = result.text.trim();
        if (!rawReply) {
            session.context.write('(no response)\n\n');
            session.chatHistory.push({ role: 'assistant', content: '' });
            await handleAssistantMessage(session, '', turnIndex);
            session.memoryStore = appendMemory(session.memoryStore, session.context.projectRoot, nextText, '(no response)');
            saveMemoryStore(session.memoryStore);
            return;
        }
        if (!session.toolCallsEnabled) {
            session.chatHistory.push({ role: 'assistant', content: rawReply });
            await handleAssistantMessage(session, rawReply, turnIndex);
            session.context.write(`${rawReply}\n\n`);
            session.memoryStore = appendMemory(session.memoryStore, session.context.projectRoot, nextText, rawReply);
            saveMemoryStore(session.memoryStore);
            return;
        }
        const toolCalls = extractToolCalls(rawReply);
        if (toolCalls.length === 0) {
            const prose = stripToolCalls(rawReply);
            session.chatHistory.push({ role: 'assistant', content: rawReply });
            await handleAssistantMessage(session, rawReply, turnIndex);
            if (prose)
                session.context.write(`${prose}\n\n`);
            session.memoryStore = appendMemory(session.memoryStore, session.context.projectRoot, nextText, prose || rawReply);
            saveMemoryStore(session.memoryStore);
            return;
        }
        const prose = stripToolCalls(rawReply);
        if (prose)
            session.context.write(`${prose}\n`);
        await handleAssistantMessage(session, rawReply, turnIndex, toolCalls);
        outbound.push({ role: 'assistant', content: rawReply });
        const resultParts = [];
        for (const call of toolCalls) {
            session.context.write(`  ⚙ ${call.tool}${call.tool === 'read-file' || call.tool === 'write-file' ? ` ${call.path}` : ''}\n`);
            const toolResult = await executeManagedToolCall(session, call, turnIndex);
            resultParts.push(formatToolResult(toolResult));
            session.context.write(toolResult.outcome === 'failure' && toolResult.error
                ? `  → error: ${toolResult.error}\n`
                : `  → ${toolResult.outcome} (${toolResult.durationMs}ms)\n`);
        }
        session.context.write('\n');
        outbound.push({ role: 'user', content: resultParts.join('\n\n---\n\n') });
    }
    session.context.writeError('(tool-call loop exceeded max rounds)\n\n');
}
