// @lumiarq/lumis@1.2.0/dist/commands/tinker-session-runtime.js
import { executeToolCall } from '../tools/index.js';
import { projectFingerprint } from '../lib/tinker-memory.js';
import { enforcePolicy } from '../lib/tinker-policy.js';
function createPluginContext(session) {
    return {
        projectRoot: session.context.projectRoot,
        session: {
            context: { write: session.context.write, writeError: session.context.writeError },
            chatModel: session.chatModel,
            chatEnabled: session.chatEnabled,
        },
    };
}
async function recordObserverEvent(session, turnIndex, eventType, data) {
    try {
        await session.observer.recordTurn({
            timestamp: new Date().toISOString(),
            sessionId: session.sessionId,
            projectFingerprint: projectFingerprint(session.context.projectRoot),
            turnIndex,
            eventType,
            data,
        });
    }
    catch (error) {
        session.context.writeError(`Observer error: ${error instanceof Error ? error.message : String(error)}\n`);
    }
}
export function classifyToolCallPolicyOperation(call) {
    switch (call.tool) {
        case 'write-file':
            return 'write';
        case 'run-command':
        case 'typecheck':
        case 'run-tests':
        case 'lint':
            return 'execute';
        case 'read-file':
        case 'search-repo':
        case 'git-diff':
            return 'read';
    }
}
export async function runSessionLifecycleHooks(session, hookName) {
    const pluginContext = createPluginContext(session);
    for (const plugin of session.pluginRegistry.getAll()) {
        const hook = plugin[hookName];
        if (typeof hook !== 'function')
            continue;
        try {
            await hook.call(plugin, pluginContext);
        }
        catch (error) {
            session.context.writeError(`Plugin ${plugin.name} hook error: ${error instanceof Error ? error.message : String(error)}\n`);
        }
    }
}
export async function applyUserMessageHooks(session, message) {
    let nextMessage = message;
    const pluginContext = createPluginContext(session);
    for (const plugin of session.pluginRegistry.getAll()) {
        if (!plugin.onUserMessage)
            continue;
        try {
            const result = await plugin.onUserMessage(nextMessage, pluginContext);
            if (typeof result === 'string' && result.trim()) {
                nextMessage = result;
            }
        }
        catch (error) {
            session.context.writeError(`Plugin ${plugin.name} hook error: ${error instanceof Error ? error.message : String(error)}\n`);
        }
    }
    return nextMessage;
}
export async function handleAssistantMessage(session, message, turnIndex, toolCalls) {
    const pluginContext = createPluginContext(session);
    for (const plugin of session.pluginRegistry.getAll()) {
        if (!plugin.onAssistantMessage)
            continue;
        try {
            await plugin.onAssistantMessage(message, pluginContext);
        }
        catch (error) {
            session.context.writeError(`Plugin ${plugin.name} hook error: ${error instanceof Error ? error.message : String(error)}\n`);
        }
    }
    session.sessionReplayer.recordAssistantMessage(message, toolCalls?.map((call) => ({ tool: call.tool, args: call })));
    await recordObserverEvent(session, turnIndex, 'assistant-message', {
        message,
        ...(toolCalls ? { toolCalls } : {}),
    });
}
export async function handleUserMessage(session, message, turnIndex) {
    const nextMessage = await applyUserMessageHooks(session, message);
    session.sessionReplayer.recordUserMessage(nextMessage);
    await recordObserverEvent(session, turnIndex, 'user-message', { message: nextMessage });
    return nextMessage;
}
export async function executeManagedToolCall(session, call, turnIndex) {
    const startedAt = Date.now();
    const operation = classifyToolCallPolicyOperation(call);
    const pluginContext = createPluginContext(session);
    try {
        enforcePolicy(session.policyGate, 'tool-call');
        enforcePolicy(session.policyGate, operation);
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        session.auditTrail.recordToolCall(call.tool, 'failure', { operation, error: message });
        await recordObserverEvent(session, turnIndex, 'error', {
            scope: 'tool-policy',
            tool: call.tool,
            operation,
            error: message,
        });
        return {
            tool: call.tool,
            outcome: 'failure',
            output: '',
            error: message,
            durationMs: Date.now() - startedAt,
        };
    }
    await recordObserverEvent(session, turnIndex, 'tool-call', {
        tool: call.tool,
        call,
        operation,
    });
    for (const plugin of session.pluginRegistry.getAll()) {
        if (!plugin.onToolCall)
            continue;
        try {
            const allowed = await plugin.onToolCall(call, pluginContext);
            if (allowed === false) {
                const message = `Tool call '${call.tool}' rejected by plugin '${plugin.name}'`;
                session.auditTrail.recordToolCall(call.tool, 'skipped', {
                    operation,
                    rejectedBy: plugin.name,
                });
                await recordObserverEvent(session, turnIndex, 'tool-result', {
                    tool: call.tool,
                    outcome: 'skipped',
                    rejectedBy: plugin.name,
                });
                return {
                    tool: call.tool,
                    outcome: 'skipped',
                    output: '',
                    error: message,
                    durationMs: Date.now() - startedAt,
                };
            }
        }
        catch (error) {
            session.context.writeError(`Plugin ${plugin.name} hook error: ${error instanceof Error ? error.message : String(error)}\n`);
        }
    }
    let result;
    try {
        result = await executeToolCall(call, session.context.projectRoot, session.writeLog);
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        result = {
            tool: call.tool,
            outcome: 'failure',
            output: '',
            error: message,
            durationMs: Date.now() - startedAt,
        };
    }
    for (const plugin of session.pluginRegistry.getAll()) {
        if (!plugin.onToolResult)
            continue;
        try {
            result = await plugin.onToolResult(call, result, pluginContext);
        }
        catch (error) {
            session.context.writeError(`Plugin ${plugin.name} hook error: ${error instanceof Error ? error.message : String(error)}\n`);
        }
    }
    session.auditTrail.recordToolCall(call.tool, result.outcome, {
        operation,
        durationMs: result.durationMs,
        ...(result.error ? { error: result.error } : {}),
    });
    await recordObserverEvent(session, turnIndex, 'tool-result', {
        tool: call.tool,
        outcome: result.outcome,
        durationMs: result.durationMs,
        ...(result.error ? { error: result.error } : {}),
    });
    return result;
}
