// @lumiarq/lumis@1.2.0/dist/commands/tinker.command.js
/**
 * Interactive tinker command entrypoint.
 *
 * Provides an interactive REPL where developers can:
 * - Explore pack-specific context and commands
 * - Test intent parsing
 * - Preview plans without executing
 * - Inspect IR state
 */
import readline from 'node:readline';
import { loadMemoryStore, } from '../lib/tinker-memory.js';
import { TinkerPluginRegistry } from '../plugins/tinker-plugin.js';
import { createObserver } from '../lib/tinker-observer.js';
import { getPolicyForSession } from '../lib/tinker-policy.js';
import { AuditTrail, SessionReplayer, generateSessionId } from '../lib/tinker-replay.js';
import { looksLikeFilesystemPath, resetChatHistory, sendChatMessage, } from './tinker-chat-runtime.js';
import { handleChatCommand } from './tinker-chat-commands.js';
import { handleClassicInput } from './tinker-classic-mode.js';
import { runSessionLifecycleHooks } from './tinker-session-runtime.js';
import { renderWelcome, runAgentPlan } from './tinker-repl-display.js';
const AGENT_RUNTIMES = new Set(['claude-code', 'copilot']);
function isModelRuntime(runtime) {
    if (!runtime)
        return true;
    if (runtime === 'local')
        return true;
    return !AGENT_RUNTIMES.has(runtime);
}
/**
 * Handle a single tinker input line.
 */
async function handleInput(session, input) {
    const trimmed = input.trim();
    if (!trimmed)
        return;
    if (session.chatEnabled) {
        if (trimmed.startsWith('/')) {
            if (handleChatCommand(session, trimmed, {
                renderWelcome,
                runAgentPlan,
            })) {
                return;
            }
            if (looksLikeFilesystemPath(trimmed)) {
                await sendChatMessage(session, trimmed);
                return;
            }
            session.context.writeError(`Unknown chat command: ${trimmed}\n`);
            session.context.write('Type /help for chat commands.\n\n');
            return;
        }
        if (!trimmed.startsWith('/')) {
            await sendChatMessage(session, trimmed);
            return;
        }
    }
    await handleClassicInput(session, trimmed);
}
/**
 * Run interactive tinker REPL.
 *
 * @param context - Shared command runtime context.
 * @returns Exit code 0 on success, 1 on resolution/setup errors.
 */
export async function executeTinkerCommand(context) {
    const pack = await context.resolvePack();
    if (!pack) {
        context.writeError('No compatible pack detected for this project.\n');
        return 1;
    }
    const packFramework = pack.manifest.frameworks?.[0] ?? 'unknown';
    const sessionId = generateSessionId();
    process.env.TINKER_SESSION_ID = sessionId;
    const session = {
        context,
        packName: pack.manifest.name,
        framework: packFramework,
        rl: readline.createInterface({
            input: process.stdin,
            output: process.stdout,
        }),
        running: true,
        chatEnabled: isModelRuntime(context.config.preferredRuntime),
        ...(context.config.ai?.defaultModel ? { chatModel: context.config.ai.defaultModel } : {}),
        chatHistory: [],
        ragReady: false,
        ragInitializing: false,
        ragIndexedFiles: 0,
        writeLog: [],
        agentRunning: false,
        toolCallsEnabled: true,
        memoryStore: loadMemoryStore(),
        pluginRegistry: new TinkerPluginRegistry(),
        observer: createObserver(sessionId),
        policyGate: getPolicyForSession(),
        auditTrail: new AuditTrail(context.projectRoot),
        sessionReplayer: new SessionReplayer(context.projectRoot, sessionId),
        sessionId,
    };
    if (session.chatEnabled) {
        resetChatHistory(session);
    }
    renderWelcome(session);
    void runSessionLifecycleHooks(session, 'onSessionStart');
    return new Promise((resolve) => {
        let closing = false;
        const closeAndResolve = async () => {
            if (closing)
                return;
            closing = true;
            await runSessionLifecycleHooks(session, 'onSessionEnd');
            // Phase 8: Flush observability
            await session.observer.flush();
            session.rl.close();
            resolve(0);
        };
        session.rl.setPrompt(session.chatEnabled ? '>>> ' : '> ');
        session.rl.prompt();
        session.rl.on('line', async (line) => {
            try {
                await handleInput(session, line);
            }
            catch (error) {
                session.context.writeError(`Error: ${error instanceof Error ? error.message : String(error)}\n`);
            }
            if (!session.running) {
                await closeAndResolve();
                return;
            }
            session.rl.prompt();
        });
        session.rl.on('SIGINT', async () => {
            session.context.write('\nGoodbye.\n');
            session.running = false;
            await closeAndResolve();
        });
        session.rl.on('close', () => {
            closeAndResolve();
        });
    });
}
