// @lumiarq/lumis@1.2.0/dist/commands/tinker-classic-mode.js
import { createEmptyIR, loadIR } from '@lumiarq/context';
import { createPlanWithAdapter } from '@lumiarq/planner';
import { ansi } from '../output/ansi.js';
import { parseRawIntent } from './intent.command.js';
export async function showStatus(session) {
    const ir = loadIR(session.context.projectRoot) ?? createEmptyIR(session.context.projectRoot);
    session.context.write(ansi.label('Project Status'));
    session.context.write('\n');
    session.context.write(`  Root: ${session.context.projectRoot}\n`);
    session.context.write(`  Pack: ${session.packName}\n`);
    session.context.write(`  Framework: ${session.framework}\n`);
    session.context.write(`  IR modules: ${ir.modules.length}\n`);
    session.context.write(`  IR version: ${ir.version}\n`);
    session.context.write('\n');
}
export async function showCommands(session) {
    const commands = await session.context.resolvePackCommands();
    session.context.write(ansi.label('Available Commands'));
    session.context.write('\n');
    if (commands.length === 0) {
        session.context.write('  (no commands)\n');
    }
    else {
        for (const cmd of commands) {
            session.context.write(`  ${cmd.name}  # ${cmd.description ?? '(no description)'}\n`);
        }
    }
    session.context.write('\n');
}
export async function previewIntent(session, rawIntent) {
    const parsed = parseRawIntent(rawIntent);
    if (!parsed) {
        session.context.writeError('Invalid intent. Expected: create|update|delete|query <kind> [<name>] [in <module>]\n');
        return;
    }
    session.context.write(ansi.label('Intent Preview'));
    session.context.write('\n');
    session.context.write(`  Raw: ${parsed.raw}\n`);
    session.context.write(`  Action: ${parsed.action}\n`);
    const subject = parsed.subjects[0];
    if (subject) {
        session.context.write(`  Kind: ${subject.kind}\n`);
        if (subject.name)
            session.context.write(`  Name: ${subject.name}\n`);
        if (subject.attributes?.module)
            session.context.write(`  Module: ${subject.attributes.module}\n`);
    }
    if (parsed.constraints.length > 0) {
        session.context.write(`  Constraints: ${parsed.constraints.join(', ')}\n`);
    }
    session.context.write('\n');
}
export async function previewPlan(session, rawIntent) {
    const parsed = parseRawIntent(rawIntent);
    if (!parsed) {
        session.context.writeError('Invalid intent. Expected: create|update|delete|query <kind> [<name>] [in <module>]\n');
        return;
    }
    const adapter = await session.context.resolveAdapter();
    if (!adapter) {
        session.context.writeError('No adapter available.\n');
        return;
    }
    const ir = loadIR(session.context.projectRoot) ?? createEmptyIR(session.context.projectRoot);
    const plan = await createPlanWithAdapter(parsed, ir, adapter);
    session.context.write(ansi.label('Plan Preview'));
    session.context.write('\n');
    session.context.write(`  Steps: ${plan.steps.length}\n`);
    session.context.write(`  Estimated files: ${plan.estimatedFiles}\n`);
    if (plan.steps.length > 0) {
        session.context.write('\n  Execution steps:\n');
        for (const step of plan.steps) {
            session.context.write(`    ${step.id}: ${step.operation} ${step.artifactKind} "${step.name}" → ${step.targetFile}\n`);
        }
    }
    session.context.write('\n');
}
export async function showIR(session) {
    const ir = loadIR(session.context.projectRoot) ?? createEmptyIR(session.context.projectRoot);
    session.context.write(ansi.label('IR Snapshot'));
    session.context.write('\n');
    session.context.write(`  Framework: ${ir.framework}\n`);
    session.context.write(`  Version: ${ir.version}\n`);
    session.context.write(`  Modules: ${ir.modules.length}\n`);
    if (ir.modules.length > 0) {
        session.context.write('\n  Module list:\n');
        for (const mod of ir.modules.slice(0, 20)) {
            session.context.write(`    - ${mod.name}\n`);
        }
        if (ir.modules.length > 20) {
            session.context.write(`    ... and ${ir.modules.length - 20} more\n`);
        }
    }
    session.context.write('\n');
}
export function showConfig(session) {
    session.context.write(ansi.label('Lumis Config'));
    session.context.write('\n');
    session.context.write(`${JSON.stringify(session.context.config, null, 2)}\n`);
    session.context.write('\n');
}
export async function handleClassicInput(session, input) {
    const trimmed = input.trim();
    if (!trimmed)
        return;
    if (trimmed === '/quit' || trimmed === '/exit') {
        session.running = false;
        session.context.write('Goodbye.\n');
        return;
    }
    if (trimmed === '/help') {
        session.context.write('Commands:\n');
        session.context.write('  /help              Show this help\n');
        session.context.write('  /status            Show project and IR status\n');
        session.context.write('  /commands          List available pack commands\n');
        session.context.write('  /intent <text>     Parse and preview intent\n');
        session.context.write('  /plan <text>       Show plan for intent without executing\n');
        session.context.write('  /ir                Show current IR snapshot\n');
        session.context.write('  /config            Show Lumis config\n');
        session.context.write('  /quit              Exit tinker mode\n\n');
        return;
    }
    if (trimmed === '/status') {
        await showStatus(session);
        return;
    }
    if (trimmed === '/commands') {
        await showCommands(session);
        return;
    }
    if (trimmed === '/ir') {
        await showIR(session);
        return;
    }
    if (trimmed === '/config') {
        showConfig(session);
        return;
    }
    if (trimmed.startsWith('/intent ')) {
        const intentText = trimmed.slice('/intent '.length).trim();
        if (intentText) {
            await previewIntent(session, intentText);
        }
        return;
    }
    if (trimmed.startsWith('/plan ')) {
        const intentText = trimmed.slice('/plan '.length).trim();
        if (intentText) {
            await previewPlan(session, intentText);
        }
        return;
    }
    if (/^(create|update|delete|query)\s+/i.test(trimmed)) {
        await previewIntent(session, trimmed);
        return;
    }
    session.context.writeError(`Unknown command or intent: ${trimmed}\n`);
    session.context.write('Type /help for commands.\n');
}
