// @lumiarq/lumis@1.2.0/dist/commands/tinker-agent-plan.js
import { ansi } from '../output/ansi.js';
export const STATUS_ICON = {
    pending: '○',
    running: '●',
    done: '✓',
    failed: '✗',
    skipped: '–',
};
function tryParseJsonObject(raw) {
    try {
        return JSON.parse(raw);
    }
    catch {
        const fenced = raw.match(/```(?:json)?\s*([\s\S]*?)```/i);
        if (!fenced) {
            return undefined;
        }
        try {
            return JSON.parse(fenced[1] ?? '');
        }
        catch {
            return undefined;
        }
    }
}
export function normalizeAgentPlan(goal, value) {
    const fallback = {
        goal,
        riskLevel: 'medium',
        estimatedFiles: [],
        steps: [
            { id: '1', description: 'Analyze current code and identify change points.', status: 'pending', risk: 'low' },
            { id: '2', description: 'Implement the requested changes in targeted files.', status: 'pending', risk: 'medium' },
            { id: '3', description: 'Run verification checks and summarize outcomes.', status: 'pending', risk: 'low' },
        ],
    };
    if (!value || typeof value !== 'object') {
        return fallback;
    }
    const record = value;
    const riskLevel = record['riskLevel'];
    const estimatedFiles = Array.isArray(record['estimatedFiles'])
        ? record['estimatedFiles'].filter((item) => typeof item === 'string')
        : [];
    const rawSteps = Array.isArray(record['steps']) ? record['steps'] : [];
    const steps = [];
    for (const [index, step] of rawSteps.entries()) {
        if (!step || typeof step !== 'object')
            continue;
        const source = step;
        const description = typeof source['description'] === 'string' ? source['description'].trim() : '';
        if (!description)
            continue;
        const idValue = source['id'];
        const id = typeof idValue === 'string' && idValue.trim() ? idValue : String(index + 1);
        const stepRisk = source['risk'];
        const verifyHint = typeof source['verifyHint'] === 'string' ? source['verifyHint'] : undefined;
        steps.push({
            id,
            description,
            status: 'pending',
            ...(stepRisk === 'low' || stepRisk === 'medium' || stepRisk === 'high' ? { risk: stepRisk } : {}),
            ...(verifyHint ? { verifyHint } : {}),
        });
    }
    if (steps.length === 0) {
        return fallback;
    }
    return {
        goal,
        riskLevel: riskLevel === 'low' || riskLevel === 'medium' || riskLevel === 'high' ? riskLevel : 'medium',
        estimatedFiles,
        steps,
    };
}
export function renderAgentPlan(write, plan) {
    write(ansi.label('Agent Plan'));
    write('\n');
    write(`  Goal: ${plan.goal}\n`);
    write(`  Risk: ${plan.riskLevel}\n`);
    if (plan.estimatedFiles.length > 0) {
        write(`  Estimated files: ${plan.estimatedFiles.join(', ')}\n`);
    }
    write('\n');
    write('  Steps:\n');
    for (const step of plan.steps) {
        write(`    ${step.id}. [${step.status}] ${step.description}\n`);
        if (step.verifyHint) {
            write(`       verify: ${step.verifyHint}\n`);
        }
    }
    write('\n');
}
export async function proposeAgentPlan(options) {
    const result = await options.gateway.complete({
        task: 'plan-propose',
        prompt: [
            'Create a concise coding execution plan as strict JSON only.',
            `Goal: ${options.goal}`,
            `Project root: ${options.projectRoot}`,
            `Framework: ${options.framework}`,
            '',
            'Return JSON shape:',
            '{',
            '  "riskLevel": "low" | "medium" | "high",',
            '  "estimatedFiles": string[],',
            '  "steps": Array<{ "id": string; "description": string; "risk"?: "low" | "medium" | "high"; "verifyHint"?: string }>',
            '}',
        ].join('\n'),
        ...(options.chatModel ? { model: options.chatModel } : {}),
    });
    const parsed = tryParseJsonObject(result.text);
    return normalizeAgentPlan(options.goal, parsed);
}
