// @lumiarq/lumis@1.2.0/dist/commands/tinker-eval.command.js
import fs from 'node:fs';
import path from 'node:path';
import { createGatewayFromEnv } from '@lumiarq/agent';
const DEFAULT_SCENARIOS = [
    {
        id: 'status-helpfulness',
        prompt: 'Give me a concise checklist to debug a failing TypeScript build in this repo.',
        mustInclude: ['typecheck', 'tsconfig'],
        maxLatencyMs: 12_000,
    },
    {
        id: 'safe-file-edits',
        prompt: 'What safe sequence would you follow before editing a production config file?',
        mustInclude: ['backup', 'verify'],
        maxLatencyMs: 12_000,
    },
    {
        id: 'test-gating',
        prompt: 'How do you ensure code changes are validated before commit?',
        mustInclude: ['typecheck', 'test'],
        maxLatencyMs: 12_000,
    },
];
function parseArgs(args) {
    const result = {};
    for (let i = 0; i < args.length; i++) {
        const token = args[i];
        if (token === '--scenarios' && i + 1 < args.length) {
            const val = args[i + 1];
            if (val)
                result.scenarioPath = val;
            i++;
            continue;
        }
        if (token === '--model' && i + 1 < args.length) {
            const val = args[i + 1];
            if (val)
                result.model = val;
            i++;
            continue;
        }
    }
    return result;
}
function loadScenarios(projectRoot, scenarioPath) {
    if (!scenarioPath)
        return DEFAULT_SCENARIOS;
    const absolute = path.isAbsolute(scenarioPath) ? scenarioPath : path.join(projectRoot, scenarioPath);
    const raw = fs.readFileSync(absolute, 'utf8');
    const parsed = JSON.parse(raw);
    if (!Array.isArray(parsed)) {
        throw new Error('Scenario file must be a JSON array');
    }
    const scenarios = [];
    for (const item of parsed) {
        if (typeof item !== 'object' || item === null)
            continue;
        const obj = item;
        if (typeof obj.id !== 'string' || typeof obj.prompt !== 'string')
            continue;
        scenarios.push({
            id: obj.id,
            prompt: obj.prompt,
            ...(Array.isArray(obj.mustInclude)
                ? { mustInclude: obj.mustInclude.filter((x) => typeof x === 'string') }
                : {}),
            ...(typeof obj.maxLatencyMs === 'number' ? { maxLatencyMs: obj.maxLatencyMs } : {}),
        });
    }
    if (scenarios.length === 0) {
        throw new Error('No valid scenarios found in scenario file');
    }
    return scenarios;
}
function evaluateOne(scenario, response, latencyMs) {
    const must = scenario.mustInclude ?? [];
    const responseLower = response.toLowerCase();
    const missing = must.filter((token) => !responseLower.includes(token.toLowerCase()));
    const keywordScore = must.length === 0 ? 1 : must.length - missing.length;
    const keywordMax = must.length === 0 ? 1 : must.length;
    const latencyBonus = scenario.maxLatencyMs && latencyMs <= scenario.maxLatencyMs ? 1 : 0;
    const latencyMax = scenario.maxLatencyMs ? 1 : 0;
    const score = keywordScore + latencyBonus;
    const maxScore = keywordMax + latencyMax;
    return {
        id: scenario.id,
        latencyMs,
        score,
        maxScore,
        passed: score === maxScore,
        missing,
        response,
    };
}
export async function executeTinkerEvalCommand(context) {
    const { scenarioPath, model } = parseArgs(context.commandArgs);
    const scenarios = loadScenarios(context.projectRoot, scenarioPath);
    context.write(`Running tinker eval with ${scenarios.length} scenario(s)...\n`);
    if (model)
        context.write(`Model override: ${model}\n`);
    context.write('\n');
    const gateway = await createGatewayFromEnv({
        ...(context.config.preferredRuntime ? { defaultProviderId: context.config.preferredRuntime } : {}),
    });
    const results = [];
    for (const scenario of scenarios) {
        const start = Date.now();
        const response = await gateway.complete({
            task: 'answer-ground',
            prompt: scenario.prompt,
            messages: [
                {
                    role: 'system',
                    content: [
                        'You are evaluating CLI coding-assistant quality.',
                        'Respond with concise practical guidance.',
                    ].join('\n'),
                },
                { role: 'user', content: scenario.prompt },
            ],
            ...(model ? { model } : {}),
        });
        const latencyMs = Date.now() - start;
        const evalResult = evaluateOne(scenario, response.text.trim(), latencyMs);
        results.push(evalResult);
        context.write(`${evalResult.passed ? '✓' : '✗'} ${scenario.id}  (${latencyMs}ms)  ${evalResult.score}/${evalResult.maxScore}\n`);
        if (evalResult.missing.length > 0) {
            context.write(`  missing: ${evalResult.missing.join(', ')}\n`);
        }
    }
    const totalScore = results.reduce((acc, r) => acc + r.score, 0);
    const maxScore = results.reduce((acc, r) => acc + r.maxScore, 0);
    const passedCount = results.filter((r) => r.passed).length;
    const avgLatency = Math.round(results.reduce((acc, r) => acc + r.latencyMs, 0) / Math.max(results.length, 1));
    context.write('\n');
    context.write(`Summary: ${passedCount}/${results.length} passed | Score ${totalScore}/${maxScore} | Avg latency ${avgLatency}ms\n`);
    const reportDir = path.join(context.projectRoot, '.lumis', 'evals');
    fs.mkdirSync(reportDir, { recursive: true });
    const reportPath = path.join(reportDir, `tinker-eval-${Date.now()}.json`);
    fs.writeFileSync(reportPath, JSON.stringify({
        generatedAt: new Date().toISOString(),
        projectRoot: context.projectRoot,
        model: model ?? null,
        scenarios,
        results,
        summary: { totalScore, maxScore, passedCount, total: results.length, avgLatencyMs: avgLatency },
    }, null, 2), 'utf8');
    context.write(`Report: ${reportPath}\n`);
    return passedCount === results.length ? 0 : 1;
}
