// @lumiarq/lumis@1.2.0/dist/commands/tinker-chat-commands.js
import fs from 'node:fs';
import path from 'node:path';
import { buildChatSystemMessage, ensureChatGateway, resetChatHistory, } from './tinker-chat-runtime.js';
import { clearMemoryStore, getMemoryFilePath, saveMemoryStore, } from '../lib/tinker-memory.js';
import { rollbackLastWrite, runVerificationGates, } from '../tools/index.js';
import { proposeAgentPlan, renderAgentPlan } from './tinker-agent-plan.js';
export const PROVIDER_TOKEN_ENV = {
    anthropic: 'ANTHROPIC_API_KEY',
    openai: 'OPENAI_API_KEY',
    openrouter: 'OPENROUTER_API_KEY',
    groq: 'GROQ_API_KEY',
    deepseek: 'DEEPSEEK_API_KEY',
    mistral: 'MISTRAL_API_KEY',
    gemini: 'GEMINI_API_KEY',
    ollama: 'OLLAMA_BASE_URL',
};
function showChatHelp(session) {
    session.context.write('Available Commands:\n');
    session.context.write('  /back              Return to classic tinker screen\n');
    session.context.write('  /set token <value> Set provider token for this session only\n');
    session.context.write('  /set model <name> [--save]  Set preferred model for this session\n');
    session.context.write('  /show              Show runtime/model info\n');
    session.context.write('  /load <model>      Alias for /set model <name>\n');
    session.context.write('  /agent plan <goal>             Create a structured execution plan\n');
    session.context.write('  /agent status                  Show current agent plan\n');
    session.context.write('  /agent run [strict|off]        Execute plan with verification gates\n');
    session.context.write('  /agent verify                  Run typecheck + tests\n');
    session.context.write('  /agent rollback                Undo last write-file step\n');
    session.context.write('  /agent clear                   Clear current agent plan\n');
    session.context.write('  /tools on|off      Enable/disable LLM tool-call dispatch\n');
    session.context.write('  /memory show       Show persistent memory stats/recent entries\n');
    session.context.write('  /memory clear      Clear persistent memory store\n');
    session.context.write('  /clear             Clear session context\n');
    session.context.write('  /bye               Exit\n');
    session.context.write('  /?, /help          Show this help\n');
    session.context.write('\n');
}
function persistDefaultModel(session, model) {
    const configPath = path.join(session.context.projectRoot, 'lumis.config.json');
    const nextConfig = {
        ...session.context.config,
        ai: {
            ...(session.context.config.ai ?? {}),
            defaultModel: model,
        },
    };
    fs.writeFileSync(configPath, JSON.stringify(nextConfig, null, 2), 'utf8');
    session.context.config = nextConfig;
}
export function handleChatCommand(session, trimmed, handlers) {
    if (trimmed === '/?' || trimmed === '/help') {
        showChatHelp(session);
        return true;
    }
    if (trimmed === '/clear') {
        resetChatSessionHistory(session);
        session.context.write('Cleared chat context.\n\n');
        return true;
    }
    if (trimmed === '/back') {
        session.chatEnabled = false;
        session.rl.setPrompt('> ');
        session.context.write('Returned to classic tinker screen.\n');
        handlers.renderWelcome(session);
        return true;
    }
    if (trimmed === '/show') {
        writeRuntimeStatus(session);
        return true;
    }
    if (trimmed === '/tools on' || trimmed === '/tools off') {
        session.toolCallsEnabled = trimmed === '/tools on';
        if (session.chatHistory.length > 0) {
            session.chatHistory[0] = buildChatSystemMessage(session);
        }
        session.context.write(`Tool-call dispatch ${session.toolCallsEnabled ? 'enabled' : 'disabled'}.\n\n`);
        return true;
    }
    if (trimmed === '/memory show') {
        writeMemoryStatus(session);
        return true;
    }
    if (trimmed === '/memory clear') {
        session.memoryStore = clearMemoryStore();
        saveMemoryStore(session.memoryStore);
        session.context.write('Cleared persistent memory store.\n\n');
        return true;
    }
    if (trimmed.startsWith('/agent plan ')) {
        const goal = trimmed.slice('/agent plan '.length).trim();
        if (!goal) {
            session.context.writeError('Usage: /agent plan <goal>\n');
            return true;
        }
        void (async () => {
            session.context.write('Creating agent plan...\n');
            try {
                const gateway = await ensureChatGateway(session);
                const plan = await proposeAgentPlan({
                    goal,
                    projectRoot: session.context.projectRoot,
                    framework: session.framework,
                    gateway,
                    ...(session.chatModel ? { chatModel: session.chatModel } : {}),
                });
                session.agentPlan = plan;
                renderAgentPlan(session.context.write, plan);
            }
            catch (error) {
                session.context.writeError(`Failed to create plan: ${error instanceof Error ? error.message : String(error)}\n\n`);
            }
        })();
        return true;
    }
    if (trimmed === '/agent status') {
        if (!session.agentPlan) {
            session.context.write('No active agent plan. Use /agent plan <goal>.\n\n');
            return true;
        }
        renderAgentPlan(session.context.write, session.agentPlan);
        return true;
    }
    if (trimmed === '/agent clear') {
        delete session.agentPlan;
        session.context.write('Cleared active agent plan.\n\n');
        return true;
    }
    if (trimmed.startsWith('/agent run')) {
        if (!session.agentPlan) {
            session.context.writeError('No active plan. Use /agent plan <goal> first.\n');
            return true;
        }
        if (session.agentRunning) {
            session.context.writeError('Agent plan is already running.\n');
            return true;
        }
        const rawGate = trimmed.slice('/agent run'.length).trim();
        const gateLevel = rawGate === 'strict' || rawGate === 'standard' || rawGate === 'off' ? rawGate : 'standard';
        void handlers.runAgentPlan(session, gateLevel);
        return true;
    }
    if (trimmed === '/agent verify') {
        void (async () => {
            session.context.write('Running verification gates...\n');
            const report = await runVerificationGates(session.context.projectRoot);
            session.context.write(`${report.passed ? '✓' : '✗'} ${report.summary}\n\n`);
            if (!report.passed) {
                const errors = report.results
                    .flatMap((result) => result.diagnostics ?? [])
                    .filter((diagnostic) => diagnostic.severity === 'error')
                    .slice(0, 10);
                for (const diagnostic of errors) {
                    const location = diagnostic.file
                        ? `${diagnostic.file}${diagnostic.line ? `:${diagnostic.line}` : ''} `
                        : '';
                    session.context.write(`  ${location}${diagnostic.message}\n`);
                }
                session.context.write('\n');
            }
        })();
        return true;
    }
    if (trimmed === '/agent rollback') {
        if (session.writeLog.length === 0) {
            session.context.write('Nothing to rollback.\n\n');
            return true;
        }
        void (async () => {
            const record = await rollbackLastWrite(session.writeLog);
            if (!record) {
                session.context.write('Nothing to rollback.\n\n');
            }
            else {
                session.context.write(`Rolled back: ${record}\n\n`);
            }
        })();
        return true;
    }
    if (trimmed === '/bye') {
        session.running = false;
        session.context.write('Goodbye.\n');
        return true;
    }
    if (trimmed.startsWith('/load ')) {
        const model = trimmed.slice('/load '.length).trim();
        if (!model) {
            session.context.writeError('Usage: /load <model>\n');
            return true;
        }
        session.chatModel = model;
        session.context.write(`Loaded model: ${model}\n\n`);
        return true;
    }
    if (trimmed.startsWith('/set model ')) {
        const raw = trimmed.slice('/set model '.length).trim();
        const save = raw.endsWith(' --save');
        const model = save ? raw.slice(0, -' --save'.length).trim() : raw;
        if (!model) {
            session.context.writeError('Usage: /set model <name> [--save]\n');
            return true;
        }
        session.chatModel = model;
        if (save) {
            persistDefaultModel(session, model);
            session.context.write(`Set model: ${model} (saved to lumis.config.json)\n\n`);
        }
        else {
            session.context.write(`Set model: ${model}\n\n`);
        }
        return true;
    }
    if (trimmed.startsWith('/set token ')) {
        const token = trimmed.slice('/set token '.length).trim();
        if (!token) {
            session.context.writeError('Usage: /set token <value>\n');
            return true;
        }
        const runtime = session.context.config.preferredRuntime;
        const tokenEnv = PROVIDER_TOKEN_ENV[runtime ?? ''];
        if (!tokenEnv || tokenEnv === 'OLLAMA_BASE_URL') {
            session.context.writeError(`No API token variable is mapped for runtime "${runtime ?? 'auto'}". Use lumis use <provider> first.\n`);
            return true;
        }
        process.env[tokenEnv] = token;
        session.chatGateway = undefined;
        session.context.write(`Set ${tokenEnv} for this tinker session.\n\n`);
        return true;
    }
    return false;
}
function resetChatSessionHistory(session) {
    resetChatHistory(session);
}
function writeRuntimeStatus(session) {
    session.context.write(`Runtime: ${session.context.config.preferredRuntime ?? 'auto'}\n`);
    const tokenEnv = PROVIDER_TOKEN_ENV[session.context.config.preferredRuntime ?? ''];
    if (tokenEnv) {
        if (tokenEnv === 'OLLAMA_BASE_URL') {
            session.context.write(`Ollama endpoint: ${process.env[tokenEnv] ?? 'http://localhost:11434/v1'}\n`);
        }
        else {
            session.context.write(`${tokenEnv}: ${process.env[tokenEnv] ? 'set' : 'missing'}\n`);
        }
    }
    session.context.write(`Model: ${session.chatModel ?? 'provider default'}\n`);
    session.context.write(`Tool calls: ${session.toolCallsEnabled ? 'on' : 'off'}\n`);
    session.context.write(`Memory file: ${getMemoryFilePath()}\n`);
    session.context.write(`Memory entries: ${session.memoryStore.entries.length}\n`);
    session.context.write(`Repo grounding: ${session.ragReady ? `ready (${session.ragIndexedFiles} files)` : session.ragInitializing ? 'indexing' : 'not initialized'}\n\n`);
}
function writeMemoryStatus(session) {
    session.context.write(`Memory file: ${getMemoryFilePath()}\n`);
    session.context.write(`Entries: ${session.memoryStore.entries.length}\n`);
    const recent = session.memoryStore.entries.slice(-5);
    if (recent.length > 0) {
        session.context.write('Recent:\n');
        for (const entry of recent) {
            session.context.write(`  - ${entry.timestamp}: ${entry.summary}\n`);
        }
    }
    session.context.write('\n');
}
