// @lumiarq/lumis@1.2.0/dist/adapter-loader.js
/**
 * Adapter/pack loading for the host `lumis` package.
 *
 * This module is intentionally host-owned (not pack-owned) so the CLI/runtime has
 * one place to resolve built-in packs, detect project compatibility, and map to an adapter.
 */
import fs from 'node:fs';
import path from 'node:path';
import { LUMIARQ_SCOPE } from '@lumiarq/context';
import { LUMIARQ_PACK_MANIFEST, LumiarqAdapter } from '@lumiarq/pack-lumiarq';
import { NEXTJS_PACK_MANIFEST, NextjsAdapter } from '@lumiarq/pack-nextjs';
import { DJANGO_PACK_MANIFEST, DjangoAdapter } from '@lumiarq/pack-django';
import { LARAVEL_PACK_MANIFEST, LaravelAdapter } from '@lumiarq/pack-laravel';
const BUILTIN_PACKS = [
    {
        manifest: LUMIARQ_PACK_MANIFEST,
        adapter: LumiarqAdapter,
    },
    {
        manifest: NEXTJS_PACK_MANIFEST,
        adapter: NextjsAdapter,
    },
    {
        manifest: DJANGO_PACK_MANIFEST,
        adapter: DjangoAdapter,
    },
    {
        manifest: LARAVEL_PACK_MANIFEST,
        adapter: LaravelAdapter,
    },
];
async function discoverPacksForProject(projectRoot) {
    const packageJsonPath = path.join(projectRoot, 'package.json');
    if (!fs.existsSync(packageJsonPath)) {
        return [];
    }
    let pkgContent;
    try {
        pkgContent = fs.readFileSync(packageJsonPath, 'utf8');
    }
    catch {
        return [];
    }
    let pkg;
    try {
        pkg = JSON.parse(pkgContent);
    }
    catch {
        return [];
    }
    const dependencies = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {}),
    };
    const packs = [];
    for (const depName of Object.keys(dependencies)) {
        if (!depName.startsWith(`${LUMIARQ_SCOPE}/`)) {
            continue;
        }
        if (depName === `${LUMIARQ_SCOPE}/lumis`) {
            continue;
        }
        try {
            // Attempt to resolve a known pack module by name.
            const module = await import(depName);
            const manifest = module.LUMIARQ_PACK_MANIFEST ??
                module.LUMIS_PACK_MANIFEST ??
                module.packManifest ??
                module.default?.manifest;
            const adapter = module.LumiarqAdapter ??
                module.LumisAdapter ??
                module.adapter ??
                module.default?.adapter;
            if (manifest && adapter) {
                packs.push({ manifest, adapter });
            }
        }
        catch {
            // Fallback: if the dependency is a known built-in pack, use that module.
            const fallbackPack = BUILTIN_PACKS.find((p) => p.manifest.name === depName);
            if (fallbackPack) {
                packs.push(fallbackPack);
            }
        }
    }
    return packs;
}
/**
 * Return built-in packs sorted by detection priority (highest first).
 *
 * @returns A new array of built-in pack modules sorted by priority descending.
 */
function getPackFrameworks(pack) {
    return Array.isArray(pack.manifest.frameworks) ? pack.manifest.frameworks : [];
}
function getPackPrimaryFramework(pack) {
    return getPackFrameworks(pack)[0] ?? '';
}
export function getBuiltinPacks() {
    return [...BUILTIN_PACKS].sort((left, right) => right.manifest.detectionPriority - left.manifest.detectionPriority);
}
/**
 * Convenience helper for callers that still operate on adapters directly.
 *
 * @returns Built-in adapters derived from the built-in pack registry.
 */
export function getBuiltinAdapters() {
    return getBuiltinPacks().map((pack) => pack.adapter);
}
/**
 * Discover and return all pack modules available to the current project.
 * Combines built-in packs with packs declared in package.json dependencies.
 *
 * @param projectRoot - Path to the project for which to discover packs.
 * @returns Array of discovered pack modules.
 */
export async function getAvailablePacks(projectRoot) {
    const discoveredPacks = await discoverPacksForProject(projectRoot);
    const merged = new Map();
    for (const pack of getBuiltinPacks()) {
        const frameworkKey = getPackPrimaryFramework(pack);
        if (frameworkKey) {
            merged.set(frameworkKey, pack);
        }
    }
    for (const pack of discoveredPacks) {
        const frameworkKey = getPackPrimaryFramework(pack);
        if (frameworkKey && !merged.has(frameworkKey)) {
            merged.set(frameworkKey, pack);
        }
    }
    return Array.from(merged.values());
}
/**
 * Lookup a built-in pack by either manifest framework id or adapter name.
 *
 * @param framework - Framework or adapter name to lookup.
 * @returns Matching pack or null if no match is found.
 */
export function getBuiltinPackByFramework(framework) {
    return (BUILTIN_PACKS.find((pack) => getPackFrameworks(pack).includes(framework) || pack.adapter.name === framework) ?? null);
}
/**
 * Detect the best matching pack for a project.
 *
 * @param projectRoot - Path to the project for which to detect a pack.
 * @param packs - Optional list of packs to consider for detection. Defaults to built-in packs.
 * @returns Promise<LumisPackModule | null> - Detected pack or null if no compatible pack is found.
 *
 * Selection strategy:
 * 1) Highest adapter detection confidence wins.
 * 2) If equal confidence, higher pack detection priority wins.
 */
function isLumisCoreCompatible(lumisCoreRange, coreVersion = '1.0.0') {
    if (!lumisCoreRange || lumisCoreRange === '*') {
        return true;
    }
    if (lumisCoreRange.startsWith('^')) {
        const major = lumisCoreRange.slice(1).split('.')[0];
        return coreVersion.startsWith(`${major}.`);
    }
    return lumisCoreRange === coreVersion;
}
export async function detectBestPack(projectRoot, packs) {
    const candidates = packs ?? (await getAvailablePacks(projectRoot));
    let best = null;
    for (const pack of candidates) {
        if (!isLumisCoreCompatible(pack.manifest.lumisCore)) {
            continue;
        }
        const detection = await pack.adapter.detect(projectRoot);
        if (!detection.detected) {
            continue;
        }
        const candidate = {
            pack,
            confidence: detection.confidence,
            priority: pack.manifest.detectionPriority,
        };
        if (!best) {
            best = candidate;
            continue;
        }
        const hasHigherConfidence = candidate.confidence > best.confidence;
        const hasSameConfidenceHigherPriority = candidate.confidence === best.confidence && candidate.priority > best.priority;
        if (hasHigherConfidence || hasSameConfidenceHigherPriority) {
            best = candidate;
        }
    }
    return best?.pack ?? null;
}
/**
 * Adapter-only detection path for backward compatibility.
 *
 * @param projectRoot - Path to the project for which to detect an adapter.
 * @param adapters - Optional list of adapters to consider for detection. Defaults to built-in adapters.
 * @returns Best matching adapter, or null when no compatible adapter is detected.
 *
 * Selection strategy: Highest detection confidence wins.
 */
export async function detectBestAdapter(projectRoot, adapters = getBuiltinAdapters()) {
    let best = null;
    for (const adapter of adapters) {
        const detection = await adapter.detect(projectRoot);
        if (!detection.detected) {
            continue;
        }
        if (!best || detection.confidence > best.confidence) {
            best = { adapter, confidence: detection.confidence };
        }
    }
    return best?.adapter ?? null;
}
/**
 * Resolve a pack for a project.
 *
 * @param projectRoot - Path to the project for which to resolve a pack.
 * @param options.configuredAdapter - Optional adapter/framework hint from configuration. If provided, we will first attempt to match this to a built-in pack before falling back to detection.
 * @param options.packs - Optional list of packs to consider for resolution. Defaults to built-in packs.
 * @returns Resolved pack module, or null when no compatible pack can be resolved.
 *
 * Resolution strategy:
 * If a configured adapter/framework is provided and matches a built-in pack,
 * it is preferred. Otherwise we fall back to automatic detection.
 */
export async function resolvePackForProject(projectRoot, options) {
    const configuredAdapter = options?.configuredAdapter;
    const packs = options?.packs ?? (await getAvailablePacks(projectRoot));
    if (configuredAdapter) {
        const configuredPack = packs.find((pack) => getPackFrameworks(pack).includes(configuredAdapter) || pack.adapter.name === configuredAdapter);
        if (configuredPack) {
            return configuredPack;
        }
    }
    // If package.json declares a pack dependency, honor that preference before detection.
    const packageJsonPath = path.join(projectRoot, 'package.json');
    if (fs.existsSync(packageJsonPath)) {
        try {
            const pkg = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
            const dependencies = {
                ...(pkg.dependencies ?? {}),
                ...(pkg.devDependencies ?? {}),
            };
            for (const depName of Object.keys(dependencies)) {
                const declaredPack = packs.find((pack) => pack.manifest.name === depName);
                if (declaredPack && isLumisCoreCompatible(declaredPack.manifest.lumisCore)) {
                    return declaredPack;
                }
            }
        }
        catch {
            // ignore parse errors and continue to detection
        }
    }
    return detectBestPack(projectRoot, packs);
}
/**
 * Resolve adapter for a project through pack resolution.
 *
 * @param projectRoot - Path to the project for which to resolve an adapter.
 * @param options.configuredAdapter - Optional adapter/framework hint from configuration. If provided, we will first attempt to match this to a built-in pack before falling back to detection.
 * @param options.packs - Optional list of packs to consider for resolution. Defaults to built-in packs.
 * @returns Resolved adapter, or null when no compatible pack/adapter can be resolved.
 *
 * Resolution strategy:
 * If a configured adapter/framework is provided and matches a built-in pack,
 * it is preferred. Otherwise we fall back to automatic detection.
 */
export async function resolveAdapterForProject(projectRoot, options) {
    const pack = await resolvePackForProject(projectRoot, options);
    return pack?.adapter ?? null;
}
