// @lumiarq/lumis@1.2.0/dist/cli.js
/**
 * Host CLI entrypoint and command dispatcher.
 *
 * Commands are declared in a registry and resolved by longest prefix match,
 * enabling subcommands with free-form trailing arguments (e.g. `intent <text>`).
 */
import { executeIntentCommand } from './commands/intent.command.js';
import { executeIrRebuildCommand } from './commands/ir-rebuild.command.js';
import { executeIrShowCommand } from './commands/ir-show.command.js';
import { executeMakeCommand } from './commands/make.command.js';
import { executeCommandsListCommand, executeCommandsRunCommand } from './commands/commands.command.js';
import { executeDoctorCommand } from './commands/doctor.command.js';
import { executeInitCommand } from './commands/init.command.js';
import { executeUseCommand } from './commands/use.command.js';
import { ansi } from './output/ansi.js';
import { executePackDetectCommand, executePackListCommand } from './commands/pack.command.js';
import { executeRuntimeRunCommand, executeRuntimeTranscriptsListCommand, executeRuntimeTranscriptsPruneCommand, executeRuntimeTranscriptsShowCommand, } from './commands/runtime.command.js';
import { createCommandContext } from './commands/runtime.js';
import { executeTinkerCommand } from './commands/tinker.command.js';
import { executeTinkerEvalCommand } from './commands/tinker-eval.command.js';
import { runInteractiveLauncher } from './launcher.js';
import { pathToFileURL } from 'node:url';
const COMMANDS = [
    { path: ['use'], description: 'Set preferred AI runtime', handler: executeUseCommand },
    { path: ['pack', 'list'], description: 'List built-in packs', handler: executePackListCommand },
    { path: ['pack', 'detect'], description: 'Detect best pack for project', handler: executePackDetectCommand },
    { path: ['commands', 'list'], description: 'List pack and project commands', handler: executeCommandsListCommand },
    { path: ['commands', 'run'], description: 'Run a pack or project command', handler: executeCommandsRunCommand },
    { path: ['doctor'], description: 'Run Lumis diagnostics for this project', handler: executeDoctorCommand },
    { path: ['init'], description: 'Initialize Lumis in this project', handler: executeInitCommand },
    { path: ['make'], description: 'Create artifacts from intent', handler: executeMakeCommand },
    { path: ['intent'], description: 'Parse and inspect intent', handler: executeIntentCommand },
    { path: ['ir', 'show'], description: 'Show current IR', handler: executeIrShowCommand },
    { path: ['ir', 'rebuild'], description: 'Rebuild project IR', handler: executeIrRebuildCommand },
    { path: ['runtime', 'run'], description: 'Run runtime profile tasks', handler: executeRuntimeRunCommand },
    { path: ['runtime', 'transcripts', 'list'], description: 'List runtime transcripts', handler: executeRuntimeTranscriptsListCommand },
    { path: ['runtime', 'transcripts', 'show'], description: 'Show runtime transcript details', handler: executeRuntimeTranscriptsShowCommand },
    { path: ['runtime', 'transcripts', 'prune'], description: 'Prune runtime transcripts', handler: executeRuntimeTranscriptsPruneCommand },
    { path: ['tinker', 'eval'], description: 'Run Tinker evaluation harness', handler: executeTinkerEvalCommand },
    { path: ['tinker'], description: 'Run interactive tinker mode', handler: executeTinkerCommand },
];
function createTraceListener() {
    if (process.env.LUMIS_TRACE !== '1') {
        return undefined;
    }
    return async (event) => {
        const payload = {
            type: event.type,
            name: event.name,
            status: event.status,
            ...(event.message ? { message: event.message } : {}),
            ...(event.context ? { context: event.context } : {}),
        };
        process.stderr.write(`[lumis-trace] ${JSON.stringify(payload)}\n`);
    };
}
async function emitTrace(onTrace, event) {
    if (!onTrace)
        return;
    await onTrace(event);
}
function formatUsage() {
    const lines = [
        ansi.label('Core CLI'),
        `  ${ansi.dim('Prompt -> Plan -> Confirm -> Execute -> Verified Result')}`,
        '',
        `${ansi.bold('Usage')}:`,
        `  ${ansi.cyan('lumis --help')}`,
    ];
    for (const command of COMMANDS) {
        lines.push(`  ${ansi.cyan(`lumis ${command.path.join(' ')}`)}  ${ansi.dim(`# ${command.description}`)}`);
    }
    return `${lines.join('\n')}\n`;
}
/**
 * Resolve command using longest prefix semantics.
 *
 * Example:
 * - argv: ['intent', 'create', 'record']
 * - match: ['intent'] with args ['create', 'record']
 *
 * @param argv - Raw command tokens after the executable name.
 * @returns Matched command and trailing arguments, or null when no command matches.
 */
function matchCommand(argv) {
    const matches = COMMANDS.filter((command) => {
        if (command.path.length > argv.length) {
            return false;
        }
        return command.path.every((segment, index) => segment === argv[index]);
    }).sort((left, right) => right.path.length - left.path.length);
    const command = matches[0];
    if (!command) {
        return null;
    }
    return {
        command,
        args: argv.slice(command.path.length),
    };
}
/**
 * Execute the host CLI command flow.
 *
 * @param argv - CLI arguments excluding the node/script prefix.
 * @param options - Optional execution overrides for project root and output writers.
 * @returns Exit code (0 success, non-zero failure).
 */
export async function runCli(argv, options = {}) {
    const write = options.write ?? ((message) => process.stdout.write(message));
    const writeError = options.writeError ?? ((message) => process.stderr.write(message));
    const onTrace = createTraceListener();
    const stdin = options.stdin ?? process.stdin;
    const stdout = options.stdout ?? process.stdout;
    const interactiveMode = options.interactiveMode ?? 'auto';
    const shouldRunInteractiveMenu = argv.length === 0 &&
        !argv.includes('--help') &&
        (interactiveMode === 'on' ||
            (interactiveMode === 'auto' && process.env.CI !== 'true' && process.env.LUMIS_NO_TUI !== '1' && stdin.isTTY && stdout.isTTY));
    if (shouldRunInteractiveMenu) {
        await emitTrace(onTrace, {
            type: 'command-dispatch',
            status: 'start',
            name: 'interactive-launcher',
            context: { mode: interactiveMode },
        });
        const launcherResult = await runInteractiveLauncher({
            projectRoot: options.projectRoot ?? process.cwd(),
            write,
            stdin,
            stdout,
        });
        if (launcherResult.kind === 'exit') {
            await emitTrace(onTrace, {
                type: 'command-dispatch',
                status: launcherResult.code === 0 ? 'success' : 'failure',
                name: 'interactive-launcher',
                context: { exitCode: launcherResult.code },
            });
            return launcherResult.code;
        }
        argv = launcherResult.argv;
    }
    if (argv.includes('--help') || argv.length === 0) {
        await emitTrace(onTrace, {
            type: 'command-dispatch',
            status: 'info',
            name: 'help',
            context: { argvLength: argv.length },
        });
        write(formatUsage());
        return 0;
    }
    const matched = matchCommand(argv);
    if (!matched) {
        await emitTrace(onTrace, {
            type: 'command-dispatch',
            status: 'failure',
            name: 'unknown-command',
            message: argv.join(' '),
        });
        writeError(`${ansi.fail(`Unknown command: ${argv.join(' ')}`)}\n`);
        write(formatUsage());
        return 1;
    }
    await emitTrace(onTrace, {
        type: 'command-dispatch',
        status: 'start',
        name: matched.command.path.join(' '),
        context: { args: matched.args },
    });
    const context = await createCommandContext({
        argv,
        commandPath: matched.command.path,
        commandArgs: matched.args,
        projectRoot: options.projectRoot ?? process.cwd(),
        write,
        writeError,
        ...(onTrace ? { onTrace } : {}),
    });
    return matched.command.handler(context);
}
function isMainModule() {
    const argv1 = process.argv[1];
    if (!argv1) {
        return false;
    }
    return import.meta.url === pathToFileURL(argv1).href;
}
if (isMainModule()) {
    void runCli(process.argv.slice(2));
}
