// @lumiarq/lumis@1.2.0/dist/config-loader.js
/**
 * Host config loading for `lumis` CLI/runtime.
 *
 * The config is optional; when absent, runtime falls back to defaults and
 * detection-based pack resolution.
 */
import fs from 'node:fs';
import path from 'node:path';
async function emitTrace(onTrace, event) {
    if (!onTrace)
        return;
    await onTrace(event);
}
/**
 * Load `lumis.config.ts` from project root when present.
 *
 * @param projectRoot - Absolute path of the target project.
 * @returns Parsed Lumis config object, or empty defaults when config is absent.
 */
export async function loadLumisConfig(projectRoot, options = {}) {
    await emitTrace(options.onTrace, {
        type: 'config-load',
        status: 'start',
        name: 'lumis.config',
        context: { projectRoot },
    });
    const configPathTs = path.join(projectRoot, 'lumis.config.ts');
    const configPathJson = path.join(projectRoot, 'lumis.config.json');
    if (fs.existsSync(configPathTs)) {
        const module = await import(configPathTs);
        await emitTrace(options.onTrace, {
            type: 'config-load',
            status: 'success',
            name: 'lumis.config.ts',
            context: { configPath: configPathTs },
        });
        return (module.default ?? {});
    }
    if (fs.existsSync(configPathJson)) {
        const raw = fs.readFileSync(configPathJson, 'utf8');
        try {
            const parsed = JSON.parse(raw);
            await emitTrace(options.onTrace, {
                type: 'config-load',
                status: 'success',
                name: 'lumis.config.json',
                context: { configPath: configPathJson },
            });
            return parsed;
        }
        catch (error) {
            await emitTrace(options.onTrace, {
                type: 'config-load',
                status: 'failure',
                name: 'lumis.config.json',
                message: error instanceof Error ? error.message : String(error),
                context: { configPath: configPathJson },
            });
            return {};
        }
    }
    await emitTrace(options.onTrace, {
        type: 'config-load',
        status: 'info',
        name: 'lumis.config',
        message: 'No config file found. Using defaults.',
        context: { projectRoot },
    });
    return {};
}
/**
 * Save `lumis.config.json` at project root.
 *
 * @param projectRoot - Absolute path of the target project.
 * @param config - Lumis configuration to persist.
 */
export function saveLumisConfig(projectRoot, config) {
    const configPath = path.join(projectRoot, 'lumis.config.json');
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2), 'utf8');
}
