// @lumiarq/lumis@1.2.0/dist/launcher.js
import { spawnSync } from 'node:child_process';
import readline from 'node:readline';
import { loadLumisConfig, saveLumisConfig } from './config-loader.js';
import { SUPPORTED_RUNTIMES } from './commands/use.command.js';
import { ansi } from './output/ansi.js';
const RUNTIME_KEY_ENV = {
    anthropic: 'ANTHROPIC_API_KEY',
    openai: 'OPENAI_API_KEY',
    openrouter: 'OPENROUTER_API_KEY',
    groq: 'GROQ_API_KEY',
    deepseek: 'DEEPSEEK_API_KEY',
    mistral: 'MISTRAL_API_KEY',
    gemini: 'GEMINI_API_KEY',
};
const RUNTIME_GROUPS = [
    { label: 'Agent CLIs', ids: ['claude-code', 'copilot'] },
    { label: 'Model Providers', ids: ['ollama', 'anthropic', 'openai', 'openrouter', 'groq', 'deepseek', 'mistral', 'gemini'] },
    { label: 'Fallback', ids: ['local'] },
];
function commandExists(command) {
    const isWindows = process.platform === 'win32';
    const lookup = isWindows ? 'where' : 'which';
    const result = spawnSync(lookup, [command], {
        stdio: 'ignore',
        env: process.env,
    });
    return result.status === 0;
}
async function checkOllamaAvailability(baseUrl, timeoutMs = 1200) {
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeoutMs);
        const response = await fetch(baseUrl, {
            method: 'HEAD',
            signal: controller.signal,
        });
        clearTimeout(timeoutId);
        return response.status === 404 || response.status === 200;
    }
    catch {
        return false;
    }
}
async function detectRuntimeScores(preferredRuntime) {
    const preferred = preferredRuntime?.toLowerCase();
    const env = process.env;
    const scores = [];
    const ollamaUrl = env['OLLAMA_BASE_URL'] ?? 'http://localhost:11434/v1';
    const ollamaAvailable = await checkOllamaAvailability(ollamaUrl);
    scores.push({
        id: 'ollama',
        score: ollamaAvailable ? (preferred === 'ollama' ? 100 : 55) : 0,
        reason: ollamaAvailable ? `Running at ${ollamaUrl}` : 'Not running',
        preferred: preferred === 'ollama',
        available: ollamaAvailable,
    });
    const providers = [
        { id: 'groq', key: 'GROQ_API_KEY', score: 90 },
        { id: 'openrouter', key: 'OPENROUTER_API_KEY', score: 85 },
        { id: 'anthropic', key: 'ANTHROPIC_API_KEY', score: 80 },
        { id: 'openai', key: 'OPENAI_API_KEY', score: 75 },
        { id: 'deepseek', key: 'DEEPSEEK_API_KEY', score: 70 },
        { id: 'mistral', key: 'MISTRAL_API_KEY', score: 65 },
        { id: 'gemini', key: 'GEMINI_API_KEY', score: 60 },
    ];
    for (const provider of providers) {
        const configured = (env[provider.key] ?? '').trim().length > 0;
        if (!configured)
            continue;
        const isPreferred = preferred === provider.id;
        scores.push({
            id: provider.id,
            score: isPreferred ? 100 : provider.score,
            reason: `Configured via ${provider.key}`,
            preferred: isPreferred,
            available: true,
        });
    }
    if (preferred === 'claude-code') {
        scores.push({ id: 'claude-code', score: 100, reason: 'Preferred runtime requested', preferred: true, available: true });
    }
    if (preferred === 'copilot') {
        scores.push({ id: 'copilot', score: 100, reason: 'Preferred runtime requested', preferred: true, available: true });
    }
    scores.push({
        id: 'local',
        score: preferred === 'local' ? 100 : 5,
        reason: 'Deterministic fallback',
        preferred: preferred === 'local',
        available: true,
    });
    scores.sort((a, b) => b.score - a.score);
    return scores;
}
function runtimeStatusLabel(runtimeId, score) {
    if (runtimeId === 'claude-code') {
        if (commandExists('claude'))
            return ansi.green('installed');
        if (score?.preferred)
            return ansi.yellow('preferred (not detected)');
        return ansi.dim('not detected');
    }
    if (runtimeId === 'copilot') {
        if (commandExists('github-copilot'))
            return ansi.green('installed');
        if (score?.preferred)
            return ansi.yellow('preferred (not detected)');
        return ansi.dim('not detected');
    }
    if (score?.available) {
        return ansi.green('available');
    }
    const keyEnv = RUNTIME_KEY_ENV[runtimeId];
    if (keyEnv) {
        return ansi.dim(`missing ${keyEnv}`);
    }
    if (runtimeId === 'ollama') {
        return ansi.dim('not running');
    }
    if (runtimeId === 'local') {
        return ansi.green('always available');
    }
    return ansi.dim('unavailable');
}
function renderLauncher(args) {
    const runtimeScores = new Map(args.runtimes.map((runtime) => [runtime.id, runtime]));
    const runtimeIndex = new Map(SUPPORTED_RUNTIMES.map((runtimeId, index) => [runtimeId, index]));
    const lines = [];
    lines.push('\u001Bc');
    lines.push(ansi.label('Interactive Launcher'));
    lines.push('');
    lines.push(ansi.bold('AI Runtime'));
    for (const group of RUNTIME_GROUPS) {
        lines.push(` ${ansi.dim(group.label)}`);
        for (const runtimeId of group.ids) {
            const index = runtimeIndex.get(runtimeId);
            if (index === undefined) {
                continue;
            }
            const score = runtimeScores.get(runtimeId);
            const selected = args.mode === 'runtime' && args.selectedIndex === index;
            const pointer = selected ? ansi.cyan('▸') : ' ';
            const preferred = args.preferredRuntime === runtimeId ? ` ${ansi.yellow('★')}` : '';
            const status = runtimeStatusLabel(runtimeId, score);
            const scoreText = score?.available ? ansi.dim(`${score.score}%`) : ansi.dim('--');
            lines.push(` ${pointer} ${runtimeId.padEnd(12)} ${status} ${scoreText}${preferred}`);
        }
        lines.push('');
    }
    lines.push(ansi.bold('Quick Actions'));
    args.actions.forEach((action, index) => {
        const selected = args.mode === 'action' && args.selectedIndex === index;
        const pointer = selected ? ansi.cyan('▸') : ' ';
        lines.push(` ${pointer} ${action.label}`);
        lines.push(`     ${ansi.dim(action.description)}`);
    });
    lines.push('');
    lines.push(ansi.dim(args.mode === 'runtime'
        ? '↑/↓ navigate • tab switch sections • enter → set runtime & start session • esc quit'
        : '↑/↓ navigate • tab switch sections • enter → run action • esc quit'));
    return `${lines.join('\n')}\n`;
}
export async function runInteractiveLauncher(options) {
    const stdin = options.stdin ?? process.stdin;
    const stdout = options.stdout ?? process.stdout;
    if (!stdin.isTTY || !stdout.isTTY || typeof stdin.setRawMode !== 'function') {
        return { kind: 'exit', code: 0 };
    }
    const config = await loadLumisConfig(options.projectRoot);
    const runtimes = await detectRuntimeScores(config.preferredRuntime);
    const actions = [
        {
            label: 'Start AI session',
            description: 'Open interactive AI session using the configured runtime (tinker mode)',
            argv: ['tinker'],
        },
        {
            label: 'Run doctor',
            description: 'Diagnostics for environment and pack routing',
            argv: ['doctor'],
        },
        {
            label: 'Detect pack',
            description: 'Probe this project and select the best adapter',
            argv: ['pack', 'detect'],
        },
        {
            label: 'List commands',
            description: 'Show available pack/project commands',
            argv: ['commands', 'list'],
        },
    ];
    let preferredRuntime = config.preferredRuntime;
    let mode = 'runtime';
    let selectedIndex = 0;
    const maxIndex = () => (mode === 'runtime' ? SUPPORTED_RUNTIMES.length - 1 : actions.length - 1);
    const render = () => {
        options.write(renderLauncher({
            mode,
            runtimes,
            preferredRuntime,
            actions,
            selectedIndex,
        }));
    };
    const setPreferredRuntime = (runtimeId) => {
        saveLumisConfig(options.projectRoot, {
            ...config,
            preferredRuntime: runtimeId,
        });
        // Update local state so next render() shows the ★ on the new selection
        preferredRuntime = runtimeId;
    };
    render();
    return new Promise((resolve) => {
        readline.emitKeypressEvents(stdin);
        stdin.setRawMode(true);
        stdin.resume();
        const cleanup = () => {
            stdin.setRawMode(false);
            stdin.pause();
            stdin.removeListener('keypress', onKeyPress);
        };
        const finish = (result) => {
            cleanup();
            resolve(result);
        };
        const onKeyPress = (_str, key) => {
            if (key.name === 'escape') {
                finish({ kind: 'exit', code: 0 });
                return;
            }
            if (key.ctrl && key.name === 'c') {
                finish({ kind: 'exit', code: 130 });
                return;
            }
            if (key.name === 'tab') {
                mode = mode === 'runtime' ? 'action' : 'runtime';
                selectedIndex = Math.min(selectedIndex, maxIndex());
                render();
                return;
            }
            if (key.name === 'up') {
                selectedIndex = selectedIndex > 0 ? selectedIndex - 1 : maxIndex();
                render();
                return;
            }
            if (key.name === 'down') {
                selectedIndex = selectedIndex < maxIndex() ? selectedIndex + 1 : 0;
                render();
                return;
            }
            if (key.name === 'return' || key.name === 'right') {
                if (mode === 'runtime') {
                    const runtimeId = SUPPORTED_RUNTIMES[selectedIndex];
                    if (!runtimeId) {
                        finish({ kind: 'exit', code: 1 });
                        return;
                    }
                    // 1. Persist preference — writes lumis.config.json
                    setPreferredRuntime(runtimeId);
                    // 2. Re-render so the ★ appears on the saved selection before proceeding
                    render();
                    // 3. Native CLI tools that are full replacements — spawn them directly
                    if (runtimeId === 'claude-code' && commandExists('claude')) {
                        cleanup();
                        const result = spawnSync('claude', [], { stdio: 'inherit', env: process.env });
                        resolve({ kind: 'exit', code: result.status ?? 0 });
                        return;
                    }
                    if (runtimeId === 'copilot' && commandExists('gh')) {
                        cleanup();
                        const result = spawnSync('gh', ['copilot', 'suggest', '--target', 'shell'], {
                            stdio: 'inherit',
                            env: process.env,
                        });
                        resolve({ kind: 'exit', code: result.status ?? 0 });
                        return;
                    }
                    if (runtimeId === 'ollama' && commandExists('ollama')) {
                        cleanup();
                        const result = spawnSync('ollama', ['run', 'llama3'], { stdio: 'inherit', env: process.env });
                        resolve({ kind: 'exit', code: result.status ?? 0 });
                        return;
                    }
                    // 4. API-based runtimes and local — dispatch to tinker so the AI session starts
                    finish({ kind: 'dispatch', argv: ['tinker'] });
                    return;
                }
                const action = actions[selectedIndex];
                if (!action) {
                    finish({ kind: 'exit', code: 1 });
                    return;
                }
                finish({ kind: 'dispatch', argv: action.argv });
            }
        };
        stdin.on('keypress', onKeyPress);
    });
}
