// @lumiarq/rag@0.1.0/dist/stores/memory/index.js
function cosine(a, b) {
    let dot = 0;
    let normA = 0;
    let normB = 0;
    for (let i = 0; i < a.length; i++) {
        dot += a[i] * b[i];
        normA += a[i] * a[i];
        normB += b[i] * b[i];
    }
    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    return denom === 0 ? 0 : dot / denom;
}
/**
 * In-memory VectorStore backed by brute-force cosine similarity.
 *
 * Performance: O(n) per query. Suitable for up to ~50k chunks in development
 * and CI. Use a production adapter (pgvector, Qdrant) for larger corpora.
 */
export class MemoryVectorStore {
    entries = new Map();
    /** documentId → latest upsert timestamp */
    docTimestamps = new Map();
    /** documentId → content hash */
    docContentHashes = new Map();
    async upsert(docs) {
        const now = new Date();
        for (const doc of docs) {
            const chunk = {
                id: doc.id,
                documentId: doc.documentId,
                text: doc.text,
                index: doc.index,
                startChar: doc.startChar,
                endChar: doc.endChar,
                ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
            };
            this.entries.set(doc.id, {
                chunk,
                vector: new Float32Array(doc.vector),
                documentId: doc.documentId,
                contentHash: doc.contentHash ?? null,
                lastIndexedAt: now,
            });
            this.docTimestamps.set(doc.documentId, now);
            this.docContentHashes.set(doc.documentId, doc.contentHash ?? null);
        }
    }
    async delete(ids) {
        for (const id of ids) {
            const entry = this.entries.get(id);
            if (entry) {
                this.entries.delete(id);
                // Clean up doc timestamp only if no more chunks for that document
                const stillHasChunks = Array.from(this.entries.values()).some((e) => e.documentId === entry.documentId);
                if (!stillHasChunks) {
                    this.docTimestamps.delete(entry.documentId);
                    this.docContentHashes.delete(entry.documentId);
                }
            }
        }
    }
    async search(vector, opts) {
        const topK = opts?.topK ?? 10;
        const minScore = opts?.minScore ?? 0;
        const results = [];
        for (const entry of this.entries.values()) {
            const score = cosine(vector, Array.from(entry.vector));
            if (score < minScore)
                continue;
            if (opts?.filter && entry.chunk.metadata && !opts.filter(entry.chunk.metadata))
                continue;
            results.push({ chunk: entry.chunk, score });
        }
        return results.sort((a, b) => b.score - a.score).slice(0, topK);
    }
    async getIndexedIds() {
        return Array.from(this.entries.keys());
    }
    async lastIndexedAt(documentId) {
        return this.docTimestamps.get(documentId) ?? null;
    }
    async getByIds(ids) {
        const result = [];
        for (const id of ids) {
            const entry = this.entries.get(id);
            if (entry)
                result.push(entry.chunk);
        }
        return result;
    }
    async getContentHash(documentId) {
        const hash = this.docContentHashes.get(documentId);
        return hash ?? null;
    }
    /** Number of indexed chunks (useful for tests). */
    get size() {
        return this.entries.size;
    }
}
