// @illumiarq/http@1.0.0/dist/middleware/stack.js
/**
 * Middleware stack — manages an ordered collection of named middleware.
 *
 * Middleware is sorted by priority (descending — higher priority runs first).
 * Supports adding, removing, and retrieving middleware by name.
 */
export class MiddlewareStack {
    _entries = [];
    /** Add a middleware definition to the stack. */
    add(definition) {
        this._entries.push(definition);
        return this;
    }
    /** Remove a middleware by name. Returns true if removed. */
    remove(name) {
        const idx = this._entries.findIndex((e) => e.name === name);
        if (idx === -1)
            return false;
        this._entries.splice(idx, 1);
        return true;
    }
    /** Check if a middleware with the given name exists. */
    has(name) {
        return this._entries.some((e) => e.name === name);
    }
    /** Get all entries, sorted by priority descending (higher priority first). */
    entries() {
        return [...this._entries].sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0));
    }
    /**
     * Returns all middleware handlers in priority order,
     * ready to pass into composeMiddleware().
     */
    handlers() {
        return this.entries().map((e) => e.handler);
    }
    /** Number of middleware in the stack. */
    get size() {
        return this._entries.length;
    }
}
