// @lumiarq/pack-nextjs@1.0.1/dist/adapter.js
import fs from 'node:fs';
import path from 'node:path';
import { CURRENT_IR_VERSION, } from '@lumiarq/context';
import { loadNextjsTemplate, resolveNextjsTemplateName } from './templates/index.js';
const SUPPORTED_ARTIFACT_KINDS = new Set(['page', 'api', 'action', 'service']);
/**
 * Check whether a normalized artifact kind is supported by the Next.js pack.
 *
 * @param kind - Normalized artifact kind token.
 * @returns True when the kind has a template mapping in this pack.
 */
function isSupportedArtifactKind(kind) {
    return SUPPORTED_ARTIFACT_KINDS.has(kind);
}
function toKebabCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
        .replace(/[^a-zA-Z0-9]+/g, '-')
        .replace(/^-+|-+$/g, '')
        .toLowerCase();
}
function getRouterMode(projectRoot) {
    const appPaths = [path.join(projectRoot, 'app'), path.join(projectRoot, 'src', 'app')];
    const pagesPaths = [path.join(projectRoot, 'pages'), path.join(projectRoot, 'src', 'pages')];
    const hasApp = appPaths.some((candidate) => fs.existsSync(candidate));
    const hasPages = pagesPaths.some((candidate) => fs.existsSync(candidate));
    if (hasApp && hasPages) {
        return 'mixed';
    }
    if (hasApp) {
        return 'app';
    }
    if (hasPages) {
        return 'pages';
    }
    return 'unknown';
}
function resolveRouterMode(projectRoot) {
    const mode = getRouterMode(projectRoot);
    if (mode === 'mixed') {
        // Deterministic default for mixed projects in v1.
        return 'app';
    }
    return mode === 'pages' ? 'pages' : 'app';
}
function normalizeKind(rawKind) {
    const lowered = rawKind.trim().toLowerCase();
    if (lowered === 'route' || lowered === 'api-route') {
        return 'api';
    }
    if (lowered === 'server-action') {
        return 'action';
    }
    if (lowered === 'domain-service') {
        return 'service';
    }
    return lowered;
}
function buildTargetFile(projectRoot, router, kind, name, module) {
    const slug = toKebabCase(name || 'new-artifact');
    if (kind === 'page') {
        if (router === 'pages') {
            return path.join(projectRoot, 'src', 'pages', `${slug}.tsx`);
        }
        return path.join(projectRoot, 'src', 'app', slug, 'page.tsx');
    }
    if (kind === 'api') {
        if (router === 'pages') {
            return path.join(projectRoot, 'src', 'pages', 'api', `${slug}.ts`);
        }
        return path.join(projectRoot, 'src', 'app', 'api', slug, 'route.ts');
    }
    if (kind === 'action') {
        return path.join(projectRoot, 'src', 'app', 'actions', `${slug}.ts`);
    }
    return path.join(projectRoot, 'src', 'core', toKebabCase(module || 'core'), 'services', `${slug}.ts`);
}
function normalizeToPosix(value) {
    return value.replaceAll('\\', '/').replace(/\/+/g, '/');
}
function isWithinPath(targetFile, candidateBase) {
    const target = normalizeToPosix(path.resolve(targetFile));
    const base = normalizeToPosix(path.resolve(candidateBase));
    return target === base || target.startsWith(`${base}/`);
}
function pathBelongsToAppRouter(projectRoot, targetFile) {
    return (isWithinPath(targetFile, path.join(projectRoot, 'src', 'app')) ||
        isWithinPath(targetFile, path.join(projectRoot, 'app')));
}
function pathBelongsToPagesRouter(projectRoot, targetFile) {
    return (isWithinPath(targetFile, path.join(projectRoot, 'src', 'pages')) ||
        isWithinPath(targetFile, path.join(projectRoot, 'pages')));
}
function resolveRouterPreferenceFromIR(ir) {
    const routerOverride = ir.metadata?.routerOverride;
    if (routerOverride === 'app' || routerOverride === 'pages') {
        return routerOverride;
    }
    const routerPreference = ir.metadata?.routerPreference;
    if (routerPreference === 'app' || routerPreference === 'pages') {
        return routerPreference;
    }
    return resolveRouterMode(ir.projectRoot);
}
const NEXTJS_VERIFICATION_HOOKS = [
    {
        name: 'nextjs.prevent-create-overwrite',
        when: 'before-write',
        run: async (ctx) => {
            if (ctx.step.operation !== 'create') {
                return { passed: true, guarantee: 'nextjs.create-write-safety' };
            }
            if (!fs.existsSync(ctx.step.targetFile)) {
                return { passed: true, guarantee: 'nextjs.create-write-safety' };
            }
            return {
                passed: false,
                guarantee: 'nextjs.create-write-safety',
                message: `Refusing to overwrite existing file in create step: ${ctx.step.targetFile}`,
            };
        },
    },
    {
        name: 'nextjs.enforce-router-ownership',
        when: 'before-write',
        run: async (ctx) => {
            const kind = ctx.step.artifactKind;
            if (kind !== 'page' && kind !== 'api') {
                return { passed: true, guarantee: 'nextjs.router-ownership' };
            }
            const routerPreference = resolveRouterPreferenceFromIR(ctx.ir);
            const writesToApp = pathBelongsToAppRouter(ctx.projectRoot, ctx.step.targetFile);
            const writesToPages = pathBelongsToPagesRouter(ctx.projectRoot, ctx.step.targetFile);
            if (!writesToApp && !writesToPages) {
                return {
                    passed: false,
                    guarantee: 'nextjs.router-ownership',
                    message: `Router-owned artifact must be written under app/pages router roots: ${ctx.step.targetFile}`,
                };
            }
            if (routerPreference === 'app' && !writesToApp) {
                return {
                    passed: false,
                    guarantee: 'nextjs.router-ownership',
                    message: `Cross-router write blocked. routerPreference=app, attempted target: ${ctx.step.targetFile}`,
                };
            }
            if (routerPreference === 'pages' && !writesToPages) {
                return {
                    passed: false,
                    guarantee: 'nextjs.router-ownership',
                    message: `Cross-router write blocked. routerPreference=pages, attempted target: ${ctx.step.targetFile}`,
                };
            }
            return { passed: true, guarantee: 'nextjs.router-ownership' };
        },
    },
];
const NEXTJS_PACK_COMMANDS = [
    {
        name: 'nextjs.typecheck',
        description: 'Run Next.js project typecheck',
        args: [],
        run: async (_args, projectRoot) => {
            const { spawnSync } = await import('node:child_process');
            spawnSync('pnpm', ['typecheck'], { cwd: projectRoot, stdio: 'inherit' });
        },
    },
    {
        name: 'nextjs.build',
        description: 'Run Next.js production build',
        args: [],
        run: async (_args, projectRoot) => {
            const { spawnSync } = await import('node:child_process');
            spawnSync('pnpm', ['build'], { cwd: projectRoot, stdio: 'inherit' });
        },
    },
];
export const NextjsAdapter = {
    name: 'nextjs',
    version: '1.0.0',
    language: 'typescript',
    detect: async (projectRoot) => {
        const packageJsonPath = path.join(projectRoot, 'package.json');
        let hasNextDependency = false;
        if (fs.existsSync(packageJsonPath)) {
            try {
                const pkg = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
                hasNextDependency = !!pkg.dependencies?.next || !!pkg.devDependencies?.next;
            }
            catch {
                hasNextDependency = false;
            }
        }
        const hasNextConfig = ['next.config.js', 'next.config.mjs', 'next.config.ts'].some((fileName) => fs.existsSync(path.join(projectRoot, fileName)));
        const routerMode = getRouterMode(projectRoot);
        const detected = hasNextDependency || hasNextConfig || routerMode !== 'unknown';
        const confidence = hasNextDependency && hasNextConfig ? 0.97 : detected ? 0.9 : 0;
        return {
            detected,
            confidence,
            reason: detected
                ? routerMode === 'mixed'
                    ? 'nextjs detected (mixed router; defaulting to app router)'
                    : `nextjs detected (${routerMode} router)`
                : 'nextjs not detected',
        };
    },
    buildIR: async (projectRoot) => {
        const routerMode = getRouterMode(projectRoot);
        const routerPreference = resolveRouterMode(projectRoot);
        return {
            irVersion: CURRENT_IR_VERSION,
            version: '1.0',
            projectRoot,
            framework: 'nextjs',
            builtAt: new Date(),
            modules: [],
            dependencies: [],
            violations: [],
            metadata: {
                routerMode,
                routerPreference,
            },
        };
    },
    updateIR: async (ir) => ir,
    placementRules: [],
    dependencyRules: [],
    namingRules: [],
    resolveIntent: async (intent, ir) => {
        if (intent.action !== 'create' || intent.subjects.length === 0) {
            return [];
        }
        const subject = intent.subjects[0];
        if (!subject) {
            return [];
        }
        const kind = normalizeKind(subject.kind);
        if (!isSupportedArtifactKind(kind)) {
            return [];
        }
        const moduleName = typeof subject.attributes?.module === 'string' && subject.attributes.module.length > 0
            ? subject.attributes.module
            : 'Core';
        const name = subject.name ?? `new-${kind}`;
        const routerPreference = resolveRouterPreferenceFromIR(ir);
        const targetFile = buildTargetFile(ir.projectRoot, routerPreference, kind, name, moduleName);
        const templateName = resolveNextjsTemplateName(kind);
        return [
            {
                id: `create:${kind}:${toKebabCase(name)}`,
                operation: 'create',
                artifactKind: kind,
                module: moduleName,
                name,
                targetFile,
                templateName,
                tokens: {
                    Name: name,
                    ModuleName: moduleName,
                    RoutePath: toKebabCase(name),
                    FunctionName: toKebabCase(name).replace(/-([a-z])/g, (_, c) => c.toUpperCase()),
                },
                dependsOn: [],
                requiresConfirmation: false,
                rollback: { operation: 'delete', targetFile },
                rationale: `${intent.action} ${kind} "${name}" in module ${moduleName}`,
            },
        ];
    },
    resolveTemplate: (templateName) => loadNextjsTemplate(templateName),
    verificationHooks: NEXTJS_VERIFICATION_HOOKS,
    buildCommands: {
        typecheck: 'pnpm typecheck',
        build: 'pnpm build',
    },
    tinkerContext: async (projectRoot) => ({ framework: 'nextjs', routerMode: getRouterMode(projectRoot) }),
    commands: async () => NEXTJS_PACK_COMMANDS,
};
