// @lumiarq/ai@1.2.0/dist/gateway/class.js
import { AIAllProvidersFailedError, AIMiddlewareError, AINoEmbeddingProviderError, AINoProvidersError, AITimeoutError, } from '../config/errors.js';
import { parseModelRef, resolveModelForTask, supportsRequest } from './model-selection.js';
export class AIGateway {
    providers = new Map();
    /** Ordered list of provider IDs as registered — used for failover sequencing. */
    providerOrder = [];
    defaultProviderId;
    modelPolicy;
    timeoutMs;
    middleware;
    retryPolicy;
    healthMap = new Map();
    logger;
    constructor(options = {}) {
        for (const provider of options.providers ?? []) {
            this.providers.set(provider.id, provider);
            this.providerOrder.push(provider.id);
        }
        this.defaultProviderId = options.defaultProviderId;
        this.modelPolicy = options.modelPolicy ?? {};
        this.timeoutMs = options.timeoutMs ?? 30000;
        this.middleware = options.middleware ?? [];
        this.logger = options.logger;
        this.retryPolicy = {
            maxConsecutiveFailures: options.retryPolicy?.maxConsecutiveFailures ?? 3,
            cooldownMs: options.retryPolicy?.cooldownMs ?? 60_000,
            failover: options.retryPolicy?.failover ?? true,
        };
    }
    /**
     * Registers a provider after gateway construction.
     * If the provider is new, it is appended to the failover order.
     */
    registerProvider(provider) {
        if (!this.providers.has(provider.id)) {
            this.providerOrder.push(provider.id);
        }
        this.providers.set(provider.id, provider);
    }
    /**
     * Resolves the provider/model pair that should serve the given task.
     * Bare model names route to the default provider or the first registered provider.
     */
    select(task, requestedModel) {
        const modelRef = resolveModelForTask(task, this.modelPolicy, requestedModel);
        const parsed = parseModelRef(modelRef);
        if (parsed.providerId) {
            return {
                providerId: parsed.providerId,
                model: parsed.model,
                task,
            };
        }
        if (this.defaultProviderId) {
            return {
                providerId: this.defaultProviderId,
                model: parsed.model,
                task,
            };
        }
        const firstProvider = this.providers.keys().next().value;
        if (typeof firstProvider === 'string') {
            return {
                providerId: firstProvider,
                model: parsed.model,
                task,
            };
        }
        throw new AINoProvidersError();
    }
    /**
     * Runs a one-shot text completion through middleware, timeout handling,
     * provider health checks, and optional failover.
     */
    async complete(request) {
        let req = request;
        for (let i = 0; i < this.middleware.length; i++) {
            const mw = this.middleware[i];
            if (mw.before) {
                try {
                    req = await mw.before(req);
                }
                catch (err) {
                    throw new AIMiddlewareError(i, 'before', err);
                }
            }
        }
        const primarySelection = this.select(req.task, req.model);
        const candidateIds = this.retryPolicy.failover
            ? this.getFailoverChain(primarySelection.providerId)
            : [primarySelection.providerId];
        const attempted = [];
        let lastError;
        for (const candidateId of candidateIds) {
            const provider = this.providers.get(candidateId);
            if (!provider)
                continue;
            if (!this.isProviderHealthy(candidateId))
                continue;
            const supportRequest = primarySelection.model
                ? { task: req.task, model: primarySelection.model }
                : { task: req.task };
            if (!supportsRequest(provider, supportRequest))
                continue;
            attempted.push(candidateId);
            try {
                const controller = new AbortController();
                const providerRequest = {
                    ...req,
                    model: primarySelection.model,
                    signal: controller.signal,
                };
                const result = await this.withTimeout(provider.complete(providerRequest), this.timeoutMs, candidateId, controller);
                this.markSuccess(candidateId);
                this.logger?.debug?.('ai.complete.success', { provider: candidateId, task: req.task });
                let final = result;
                for (let i = 0; i < this.middleware.length; i++) {
                    const mw = this.middleware[i];
                    if (mw.after) {
                        try {
                            final = await mw.after(final, req);
                        }
                        catch (err) {
                            throw new AIMiddlewareError(i, 'after', err);
                        }
                    }
                }
                return final;
            }
            catch (err) {
                if (err instanceof AIMiddlewareError)
                    throw err;
                this.markFailure(candidateId);
                this.logger?.warn?.('ai.complete.failure', {
                    provider: candidateId,
                    task: req.task,
                    error: err instanceof Error ? err.message : String(err),
                });
                lastError = err;
                if (!this.retryPolicy.failover)
                    throw err;
            }
        }
        const finalErr = attempted.length > 0
            ? new AIAllProvidersFailedError(attempted)
            : (lastError ?? new AIAllProvidersFailedError([]));
        this.logger?.error?.('ai.complete.all_failed', {
            attempted,
            error: finalErr instanceof Error ? finalErr.message : String(finalErr),
        });
        throw finalErr;
    }
    /**
     * Returns all registered provider IDs starting from `primaryId`, wrapping around.
     * Used to build the failover candidate list.
     */
    getFailoverChain(primaryId) {
        const idx = this.providerOrder.indexOf(primaryId);
        if (idx === -1)
            return [primaryId];
        return [
            ...this.providerOrder.slice(idx),
            ...this.providerOrder.slice(0, idx),
        ];
    }
    /**
     * Returns mutable health state for a provider, initializing it on first use.
     */
    getHealth(providerId) {
        let h = this.healthMap.get(providerId);
        if (!h) {
            h = { providerId, consecutiveFailures: 0, isHealthy: true };
            this.healthMap.set(providerId, h);
        }
        return h;
    }
    /**
     * Returns whether a provider should currently receive traffic.
     * Unhealthy providers can recover after the configured cooldown window.
     */
    isProviderHealthy(providerId) {
        const h = this.healthMap.get(providerId);
        if (!h)
            return true;
        if (h.isHealthy)
            return true;
        if (h.lastFailureAt !== undefined) {
            const elapsed = Date.now() - h.lastFailureAt;
            if (elapsed >= this.retryPolicy.cooldownMs)
                return true;
        }
        return false;
    }
    /**
     * Marks a provider request as successful and clears its failure streak.
     */
    markSuccess(providerId) {
        const h = this.getHealth(providerId);
        h.consecutiveFailures = 0;
        h.isHealthy = true;
    }
    /**
     * Records a provider failure and opens the circuit when the threshold is exceeded.
     */
    markFailure(providerId) {
        const h = this.getHealth(providerId);
        h.consecutiveFailures += 1;
        h.lastFailureAt = Date.now();
        if (h.consecutiveFailures >= this.retryPolicy.maxConsecutiveFailures) {
            h.isHealthy = false;
        }
    }
    /**
     * Returns a snapshot of health state for all known providers.
     * Providers that have never been called are reported as healthy.
     */
    health() {
        return this.providerOrder.map((id) => {
            const h = this.healthMap.get(id);
            if (!h)
                return { providerId: id, consecutiveFailures: 0, isHealthy: true };
            const isHealthy = this.isProviderHealthy(id);
            return { ...h, isHealthy };
        });
    }
    /**
     * Streams token deltas from the selected provider while preserving the same
     * middleware, timeout, circuit-breaker, and failover behavior as `complete()`.
     */
    async *completeStream(request) {
        let req = request;
        for (let i = 0; i < this.middleware.length; i++) {
            const mw = this.middleware[i];
            if (mw.before) {
                try {
                    req = await mw.before(req);
                }
                catch (err) {
                    throw new AIMiddlewareError(i, 'before', err);
                }
            }
        }
        const primarySelection = this.select(req.task, req.model);
        const candidateIds = this.retryPolicy.failover
            ? this.getFailoverChain(primarySelection.providerId)
            : [primarySelection.providerId];
        const attempted = [];
        let lastError;
        for (const candidateId of candidateIds) {
            const provider = this.providers.get(candidateId);
            if (!provider)
                continue;
            if (!this.isProviderHealthy(candidateId))
                continue;
            const supportRequest = primarySelection.model
                ? { task: req.task, model: primarySelection.model }
                : { task: req.task };
            if (!supportsRequest(provider, supportRequest))
                continue;
            attempted.push(candidateId);
            const controller = new AbortController();
            const providerRequest = {
                ...req,
                model: primarySelection.model,
                signal: controller.signal,
            };
            try {
                let streamResult;
                if (provider.completeStream) {
                    const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);
                    const parts = [];
                    try {
                        for await (const delta of provider.completeStream(providerRequest)) {
                            parts.push(delta);
                            yield delta;
                        }
                    }
                    finally {
                        clearTimeout(timeoutId);
                    }
                    streamResult = {
                        text: parts.join(''),
                        provider: candidateId,
                        model: primarySelection.model ?? '',
                        finishReason: 'stop',
                    };
                }
                else {
                    const result = await this.withTimeout(provider.complete(providerRequest), this.timeoutMs, candidateId, controller);
                    yield result.text;
                    streamResult = result;
                }
                this.markSuccess(candidateId);
                this.logger?.debug?.('ai.stream.success', { provider: candidateId, task: req.task });
                for (let i = 0; i < this.middleware.length; i++) {
                    const mw = this.middleware[i];
                    if (mw.after) {
                        try {
                            await mw.after(streamResult, req);
                        }
                        catch (err) {
                            throw new AIMiddlewareError(i, 'after', err);
                        }
                    }
                }
                return;
            }
            catch (err) {
                if (err instanceof AIMiddlewareError)
                    throw err;
                controller.abort();
                this.markFailure(candidateId);
                this.logger?.warn?.('ai.stream.failure', {
                    provider: candidateId,
                    task: req.task,
                    error: err instanceof Error ? err.message : String(err),
                });
                lastError = err;
                if (!this.retryPolicy.failover)
                    throw err;
            }
        }
        const finalErr = attempted.length > 0
            ? new AIAllProvidersFailedError(attempted)
            : (lastError ?? new AIAllProvidersFailedError([]));
        throw finalErr;
    }
    /**
     * Requests JSON-shaped output from the model and parses it with the caller's
     * validation function.
     */
    async completeStructured(request) {
        const jsonInstruction = 'Respond with valid JSON only. No markdown, no explanation, no code fences.';
        const systemPrompt = request.systemPrompt
            ? `${request.systemPrompt}\n\n${jsonInstruction}`
            : jsonInstruction;
        const result = await this.complete({ ...request, systemPrompt });
        const raw = result.text.trim().replace(/^```(?:json)?\n?|\n?```$/g, '').trim();
        return request.parse(JSON.parse(raw));
    }
    /**
     * Generates embeddings through the first matching embedding provider and applies
     * the same timeout, health-check, and failover behavior as text completion.
     */
    async embed(request) {
        const parsed = request.model ? parseModelRef(request.model) : undefined;
        const targetId = parsed?.providerId ?? this.defaultProviderId;
        let primaryProviderId;
        if (targetId) {
            const candidate = this.providers.get(targetId);
            if (candidate?.embed)
                primaryProviderId = targetId;
        }
        if (!primaryProviderId) {
            for (const providerId of this.providerOrder) {
                const provider = this.providers.get(providerId);
                if (provider?.embed) {
                    primaryProviderId = providerId;
                    break;
                }
            }
        }
        if (!primaryProviderId) {
            throw new AINoEmbeddingProviderError();
        }
        const candidateIds = this.retryPolicy.failover
            ? this.getFailoverChain(primaryProviderId)
            : [primaryProviderId];
        const attempted = [];
        let lastError;
        for (const candidateId of candidateIds) {
            const provider = this.providers.get(candidateId);
            if (!provider?.embed)
                continue;
            if (!this.isProviderHealthy(candidateId))
                continue;
            attempted.push(candidateId);
            try {
                const controller = new AbortController();
                const providerRequest = {
                    ...request,
                    ...(parsed?.model !== undefined ? { model: parsed.model } : {}),
                    signal: controller.signal,
                };
                const result = await this.withTimeout(provider.embed(providerRequest), this.timeoutMs, candidateId, controller);
                this.markSuccess(candidateId);
                this.logger?.debug?.('ai.embed.success', { provider: candidateId });
                return result;
            }
            catch (err) {
                this.markFailure(candidateId);
                this.logger?.warn?.('ai.embed.failure', {
                    provider: candidateId,
                    error: err instanceof Error ? err.message : String(err),
                });
                lastError = err;
                if (!this.retryPolicy.failover)
                    throw err;
            }
        }
        const finalErr = attempted.length > 0
            ? new AIAllProvidersFailedError(attempted)
            : (lastError ?? new AIAllProvidersFailedError([]));
        this.logger?.error?.('ai.embed.all_failed', {
            attempted,
            error: finalErr instanceof Error ? finalErr.message : String(finalErr),
        });
        throw finalErr;
    }
    /**
     * Convenience wrapper around `embed()` that accepts only an array and returns
     * the embedding matrix directly.
     */
    async embedMany(inputs, options) {
        const req = {
            input: inputs,
            ...(options?.model !== undefined ? { model: options.model } : {}),
        };
        const result = await this.embed(req);
        return result.embeddings;
    }
    /**
     * Rejects requests that exceed the configured timeout and propagates abort
     * signals to providers that support cancellation.
     */
    withTimeout(promise, ms, providerId = 'unknown', controller) {
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                controller?.abort();
                reject(new AITimeoutError(providerId, ms));
            }, ms);
            promise.then((val) => { clearTimeout(timer); resolve(val); }, (err) => { clearTimeout(timer); reject(err); });
        });
    }
}
/**
 * Creates an `AIGateway` instance from the supplied options.
 */
export function createAIGateway(options = {}) {
    return new AIGateway(options);
}
