// @lumiarq/ai@1.2.0/dist/gateway/convenience.js
import { AIGateway } from './class.js';
/**
 * One-shot text completion without constructing a gateway or request object.
 *
 * ```ts
 * import { complete } from '@lumiarq/ai'
 *
 * const text = await complete('List 5 domain name ideas for a hosting brand.')
 * const concise = await complete('Suggest a tagline.', { system: 'You are a copywriter.', temperature: 0.9 })
 * ```
 */
export async function complete(prompt, options = {}) {
    let gateway = options.gateway;
    if (!gateway) {
        const { createGatewayFromEnv } = await import('../providers/index.js');
        gateway = createGatewayFromEnv();
    }
    const result = await gateway.complete({
        task: 'token-suggest',
        prompt,
        ...(options.system ? { systemPrompt: options.system } : {}),
        ...(options.model ? { model: options.model } : {}),
        ...(options.temperature !== undefined ? { temperature: options.temperature } : {}),
    });
    return result.text;
}
/**
 * Streaming text completion that returns both a token stream and a promise for
 * the fully accumulated output.
 */
export async function streamText(prompt, options = {}) {
    let gateway = options.gateway;
    if (!gateway) {
        const { createGatewayFromEnv } = await import('../providers/index.js');
        gateway = createGatewayFromEnv();
    }
    const request = {
        task: 'token-suggest',
        prompt,
        ...(options.system ? { systemPrompt: options.system } : {}),
        ...(options.model ? { model: options.model } : {}),
        ...(options.temperature !== undefined ? { temperature: options.temperature } : {}),
    };
    const chunks = [];
    let textResolve;
    let textReject;
    const text = new Promise((resolve, reject) => {
        textResolve = resolve;
        textReject = reject;
    });
    const source = gateway.completeStream(request);
    async function* teedStream() {
        try {
            for await (const chunk of source) {
                chunks.push(chunk);
                yield chunk;
            }
            textResolve(chunks.join(''));
        }
        catch (err) {
            textReject(err);
            throw err;
        }
    }
    return { stream: teedStream(), text };
}
/**
 * Generates a single embedding vector for the supplied input string.
 */
export async function embed(input, options = {}) {
    let gateway = options.gateway;
    if (!gateway) {
        const { createGatewayFromEnv } = await import('../providers/index.js');
        gateway = createGatewayFromEnv();
    }
    const result = await gateway.embed({
        input,
        ...(options.model !== undefined ? { model: options.model } : {}),
    });
    return result.embeddings[0] ?? [];
}
/**
 * Generates embedding vectors for multiple inputs in one call.
 */
export async function embedMany(inputs, options = {}) {
    let gateway = options.gateway;
    if (!gateway) {
        const { createGatewayFromEnv } = await import('../providers/index.js');
        gateway = createGatewayFromEnv();
    }
    const result = await gateway.embed({
        input: inputs,
        ...(options.model !== undefined ? { model: options.model } : {}),
    });
    return result.embeddings;
}
/**
 * Requests structured JSON output and validates it with the supplied `parse`
 * function before returning the typed value.
 */
export async function generateObject(prompt, parse, options = {}) {
    let gateway = options.gateway;
    if (!gateway) {
        const { createGatewayFromEnv } = await import('../providers/index.js');
        gateway = createGatewayFromEnv();
    }
    return gateway.completeStructured({
        task: 'token-suggest',
        prompt,
        parse,
        ...(options.system ? { systemPrompt: options.system } : {}),
        ...(options.model ? { model: options.model } : {}),
        ...(options.temperature !== undefined ? { temperature: options.temperature } : {}),
    });
}
export { AIGateway };
