// @illumiarq/http@1.1.0/dist/errors/http-errors.js
/** Base HTTP/package error. */
export class HttpError extends Error {
    statusCode;
    constructor(message, statusCode = 500) {
        super(message);
        this.name = 'HttpError';
        this.statusCode = statusCode;
    }
}
export class MissingRenderStrategyError extends HttpError {
    constructor(path) {
        super(`Web route \"${path}\" is missing a required render declaration ('static' | 'dynamic').`, 500);
        this.name = 'MissingRenderStrategyError';
    }
}
export class InvalidApiRouteError extends HttpError {
    constructor(path, field) {
        super(`API route \"${path}\" must not declare \"${field}\" — that field is for web routes only.`, 500);
        this.name = 'InvalidApiRouteError';
    }
}
export class DuplicateRouteError extends HttpError {
    constructor(method, path) {
        super(`Duplicate route: ${method} ${path} is registered more than once.`, 500);
        this.name = 'DuplicateRouteError';
    }
}
export class MissingMetaError extends HttpError {
    constructor(path) {
        super(`Public web route \"${path}\" is missing a required meta() function.`, 500);
        this.name = 'MissingMetaError';
    }
}
export class RouteNotFoundError extends Error {
    constructor(name) {
        super(`No route registered with name \"${name}\".`);
        this.name = 'RouteNotFoundError';
    }
}
