// @illumiarq/core@1.1.0/dist/http/resource.js
/**
 * defineResource() — API response transformer factory.
 *
 * Creates a typed resource transformer with three methods:
 *   `make(item)`        — transforms a single item
 *   `collection(items)` — transforms an array of items
 *   `paginated(result)` — wraps a PaginatedResult with `{ data, meta }` envelope
 *
 * @example
 * // src/modules/Billing/http/resources/invoice.resource.ts
 * import { defineResource } from '@lumiarq/framework';
 *
 * export const InvoiceResource = defineResource((invoice: Invoice) => ({
 *   id: invoice.id,
 *   amount: invoice.amount,
 *   currency: invoice.currency,
 *   status: invoice.status,
 *   createdAt: invoice.createdAt.toISOString(),
 * }));
 *
 * // In a handler:
 * return json(InvoiceResource.make(invoice));
 * return json(InvoiceResource.collection(invoices));
 * return json(InvoiceResource.paginated(paginatedResult));
 */
/**
 * Define a typed API resource transformer.
 *
 * @param transform - A function that maps one domain object to its API representation.
 * @returns A `ResourceDefinition` with `make()`, `collection()`, and `paginated()`.
 */
export function defineResource(transform) {
    return {
        make(item) {
            return transform(item);
        },
        collection(items) {
            return items.map(transform);
        },
        paginated(result) {
            return {
                data: result.data.map(transform),
                meta: {
                    page: result.page,
                    perPage: result.perPage,
                    total: result.total,
                    lastPage: result.lastPage,
                    from: result.from,
                    to: result.to,
                },
            };
        },
    };
}
