// @illumiarq/runtime@1.1.2/dist/notification/notification-service.js
export class NotificationService {
    mailer;
    queueDriver;
    from;
    constructor(opts) {
        this.mailer = opts.mailer;
        this.queueDriver = opts.queue ?? undefined;
        this.from = opts.from ?? 'notifications@example.com';
    }
    async send(notifiable, notification) {
        if (!hasToMail(notification)) {
            throw new Error('[NotificationService] notification must implement toMail()');
        }
        if (!notifiable.email) {
            throw new Error(`[NotificationService] notifiable "${notifiable.id}" has no email address`);
        }
        const message = {
            ...notification.toMail(),
            to: notifiable.email,
            from: notification.toMail().from ?? this.from,
        };
        await this.mailer.send(message);
    }
    async queue(notifiable, notification) {
        if (!hasToMail(notification)) {
            throw new Error('[NotificationService] notification must implement toMail()');
        }
        if (!notifiable.email) {
            throw new Error(`[NotificationService] notifiable "${notifiable.id}" has no email address`);
        }
        const message = {
            ...notification.toMail(),
            to: notifiable.email,
            from: notification.toMail().from ?? this.from,
        };
        if (this.queueDriver) {
            await this.queueDriver.dispatch({ name: 'mail.send', data: message }, { queue: 'mail' });
        }
        else {
            // No queue configured — send immediately
            await this.mailer.send(message);
        }
    }
}
function hasToMail(notification) {
    return (notification !== null &&
        typeof notification === 'object' &&
        'toMail' in notification &&
        typeof notification.toMail === 'function');
}
