// @illumiarq/lumis@1.2.8/dist/commands/app-commands.js
import { existsSync, writeFileSync } from 'node:fs';
import { join, relative } from 'node:path';
import { generateAuthModule, generateAuthUI, generateIAMModule, generateLangFile, generateUserModule, publishAuthStubs, publishIAMStubs, } from '@illumiarq/auth-starter';
import { ui, writeError, writeLine } from '../console.js';
import { runBridge } from './app/command-runtime.js';
import { cacheSearchIndex, cacheViews, clearOptimizationCaches, clearSearchIndex, clearViews, optimizeForProduction, } from './app/cache-workflows.js';
import { cacheConfig, clearConfigCache } from './app/config-cache.js';
import { pingDatabaseConnection, runDatabaseCommand } from './app/database.js';
import { publishConfig } from './app/config-publish.js';
import { cacheRoutes, checkRoutes, clearRouteCache, listRoutes } from './app/routes.js';
import { ensureParentDir, resolvePathWithinRoot } from './app/security.js';
import { scheduleList, scheduleRun, workerList, workerStart } from './app/worker-schedule.js';
export { cacheConfig, cacheRoutes, cacheSearchIndex, cacheViews, clearConfigCache, clearOptimizationCaches, checkRoutes, clearRouteCache, clearSearchIndex, clearViews, listRoutes, optimizeForProduction, pingDatabaseConnection, publishConfig, runDatabaseCommand, scheduleList, scheduleRun, workerList, workerStart, };
function writeGeneratedFiles(cwd, files) {
    let created = 0;
    let skipped = 0;
    for (const file of files) {
        const target = resolvePathWithinRoot(cwd, file.path);
        ensureParentDir(target);
        if (existsSync(target) && file.path !== 'lang/en.json') {
            skipped += 1;
            continue;
        }
        writeFileSync(target, file.content, 'utf8');
        created += 1;
    }
    return { created, skipped };
}
export function showResolvedConfig(configName, cwd = process.cwd()) {
    const result = runBridge(['config:show', cwd, configName], cwd);
    if (result.status !== 0) {
        writeError(ui.fail(result.stderr.trim() || `Unable to load config/${configName}.ts`));
        return result.status;
    }
    const parsed = JSON.parse(result.stdout);
    writeLine();
    writeLine(ui.section(`Config ${configName}`));
    writeLine(JSON.stringify(parsed.config, null, 2));
    writeLine();
    return 0;
}
export function publishStubs(options = {}, cwd = process.cwd()) {
    if (options.all || !options.iam) {
        publishAuthStubs(cwd);
    }
    if (options.all || options.iam) {
        publishIAMStubs(cwd);
    }
    writeLine();
    writeLine(ui.section('Stub Publish'));
    writeLine(`  ${ui.ok(`Published stubs into ${relative(cwd, join(cwd, 'stubs')) || 'stubs'}`)}`);
    writeLine();
    return 0;
}
export function installAuth(options = {}, cwd = process.cwd()) {
    const generated = options.iam
        ? generateIAMModule()
        : [...generateAuthModule(), ...generateUserModule()];
    const uiFiles = options.ui ? generateAuthUI(options.ui) : [];
    const langFile = generateLangFile(cwd);
    const result = writeGeneratedFiles(cwd, [...generated, ...uiFiles, langFile]);
    writeLine();
    writeLine(ui.section('Auth Install'));
    writeLine(`  ${ui.ok(`Created ${result.created} files`)}`);
    if (result.skipped > 0) {
        writeLine(`  ${ui.warn(`Skipped ${result.skipped} existing files`)}`);
    }
    if (options.ui) {
        writeLine(`  ${ui.bullet(`UI scaffold: ${options.ui}`)}`);
    }
    if (options.iam) {
        writeLine(`  ${ui.bullet('Installed IAM module scaffold')}`);
    }
    writeLine();
    return 0;
}
