// @lumiarq/rag@0.1.0/dist/pipeline/index.js
import { VectorRetriever } from '../retriever/index.js';
/**
 * Create a fully wired RAG pipeline.
 *
 * ```ts
 * const rag = createRAGPipeline({
 *   chunker: new SlidingWindowChunker(),
 *   embedder: myEmbedder,
 *   store: new MemoryVectorStore(),
 *   generator: myGenerator,
 * })
 *
 * await rag.ingest([{ id: 'doc1', content: '...' }])
 * const answer = await rag.answer('What is X?')
 * ```
 */
export function createRAGPipeline(opts) {
    const topK = opts.topK ?? 10;
    const topN = opts.topN ?? 5;
    const retriever = opts.retriever ?? new VectorRetriever(opts.embedder, opts.store);
    const evalHooks = opts.evalHooks ?? [];
    const log = opts.logger;
    async function ingest(docs) {
        for (const doc of docs) {
            // Incremental indexing: skip if the stored content hash matches.
            if (doc.contentHash) {
                const storedHash = await opts.store.getContentHash(doc.id);
                if (storedHash === doc.contentHash) {
                    // Hash matches — document unchanged, skip re-indexing.
                    await log?.debug?.('rag:ingest:skip', { docId: doc.id, reason: 'hash-match' });
                    continue;
                }
                // Hash differs (or never stored) — fall through to re-index.
            }
            const chunks = await opts.chunker.chunk(doc);
            if (chunks.length === 0) {
                await log?.debug?.('rag:ingest:skip', { docId: doc.id, reason: 'no-chunks' });
                continue;
            }
            const texts = chunks.map((c) => c.text);
            const vectors = await opts.embedder.embed(texts);
            const indexed = chunks.map((chunk, i) => ({
                ...chunk,
                vector: vectors[i],
                ...(doc.contentHash !== undefined ? { contentHash: doc.contentHash } : {}),
            }));
            await opts.store.upsert(indexed);
            // If the configured retriever supports BM25 (HybridRetriever), keep it in sync.
            if ('syncChunks' in retriever && typeof retriever.syncChunks === 'function') {
                ;
                retriever.syncChunks(indexed);
            }
            await log?.debug?.('rag:ingest:upsert', { docId: doc.id, chunks: chunks.length });
        }
        await log?.info?.('rag:ingest:complete', { docs: docs.length });
    }
    async function retrieve(query, retrieveOpts) {
        const start = Date.now();
        const results = await retriever.retrieve(query, { topK, ...retrieveOpts });
        await log?.debug?.('rag:retrieve', { retrieved: results.length, latencyMs: Date.now() - start });
        return results;
    }
    async function answer(query, retrieveOpts) {
        const start = Date.now();
        let chunks = await retrieve(query, retrieveOpts);
        if (opts.reranker) {
            chunks = await opts.reranker.rerank(query, chunks, topN);
        }
        else {
            chunks = chunks.slice(0, topN);
        }
        // Embed query tokens for usage accounting (approximation)
        const queryEmbedTokens = Math.ceil(query.length / 4);
        const result = await opts.generator.generate(query, chunks);
        // Patch usage with pipeline-level accounting
        const patchedResult = {
            ...result,
            usage: {
                ...result.usage,
                embedTokens: queryEmbedTokens + result.usage.embedTokens,
                latencyMs: Date.now() - start,
            },
        };
        await log?.info?.('rag:answer', {
            chunksUsed: chunks.length,
            citations: patchedResult.citations.length,
            latencyMs: patchedResult.usage.latencyMs,
            generationTokens: patchedResult.usage.generationTokens,
        });
        // Fire eval hooks (non-blocking, errors swallowed)
        for (const hook of evalHooks) {
            Promise.resolve(hook.onAnswer(query, patchedResult, chunks)).catch(() => undefined);
        }
        return patchedResult;
    }
    async function* answerStream(query, retrieveOpts) {
        const start = Date.now();
        let chunks = await retrieve(query, retrieveOpts);
        if (opts.reranker) {
            chunks = await opts.reranker.rerank(query, chunks, topN);
        }
        else {
            chunks = chunks.slice(0, topN);
        }
        await log?.debug?.('rag:answerStream:start', { chunksUsed: chunks.length });
        if (opts.generator.generateStream) {
            const collectedChunks = [];
            for await (const delta of opts.generator.generateStream(query, chunks)) {
                collectedChunks.push(delta);
                yield delta;
            }
            // Fire eval hooks after stream completes.
            const fullText = collectedChunks.join('');
            const queryEmbedTokens = Math.ceil(query.length / 4);
            const streamAnswer = {
                text: fullText,
                citations: [],
                chunks,
                usage: {
                    embedTokens: queryEmbedTokens,
                    generationTokens: 0,
                    chunksRetrieved: chunks.length,
                    chunksAfterRerank: chunks.length,
                    latencyMs: Date.now() - start,
                },
            };
            await log?.info?.('rag:answerStream:complete', {
                chunksUsed: chunks.length,
                latencyMs: streamAnswer.usage.latencyMs,
            });
            if (evalHooks.length > 0) {
                for (const hook of evalHooks) {
                    Promise.resolve(hook.onAnswer(query, streamAnswer, chunks)).catch(() => undefined);
                }
            }
        }
        else {
            const result = await opts.generator.generate(query, chunks);
            yield result.text;
            // Fire eval hooks for non-streaming fallback.
            const queryEmbedTokens = Math.ceil(query.length / 4);
            const streamAnswer = {
                ...result,
                usage: {
                    ...result.usage,
                    embedTokens: queryEmbedTokens + result.usage.embedTokens,
                    latencyMs: Date.now() - start,
                },
            };
            await log?.info?.('rag:answerStream:complete', {
                chunksUsed: chunks.length,
                latencyMs: streamAnswer.usage.latencyMs,
            });
            if (evalHooks.length > 0) {
                for (const hook of evalHooks) {
                    Promise.resolve(hook.onAnswer(query, streamAnswer, chunks)).catch(() => undefined);
                }
            }
        }
    }
    return { ingest, retrieve, answer, answerStream };
}
