// @lumiarq/rag@0.1.0/dist/eval/trazze.js
/**
 * Creates a `RAGEvalHook` that pushes a structured `rag-eval` event to Trazze
 * (or any compatible structured logger) after every answer is generated.
 *
 * Usage with `@trazze/sdk`:
 * ```ts
 * import { createLogger } from '@trazze/sdk'
 * import { createTrazzeRAGEvalHook } from '@lumiarq/rag'
 *
 * const logger = createLogger({ channels: [myChannel] })
 * const hook = createTrazzeRAGEvalHook({ logger })
 *
 * const rag = createRAGPipeline({ ..., evalHooks: [hook] })
 * ```
 */
export function createTrazzeRAGEvalHook(options) {
    const { logger, context: baseContext = {}, includeChunks = false, faithfulnessScore } = options;
    return {
        async onAnswer(query, answer, chunks) {
            const ctx = {
                ...baseContext,
                event: 'rag-eval',
                // Query
                query,
                queryLength: query.length,
                // Answer
                answerLength: answer.text.length,
                citations: answer.citations.length,
                // Usage
                embedTokens: answer.usage.embedTokens,
                generationTokens: answer.usage.generationTokens,
                chunksRetrieved: answer.usage.chunksRetrieved,
                chunksAfterRerank: answer.usage.chunksAfterRerank,
                latencyMs: answer.usage.latencyMs,
            };
            if (faithfulnessScore !== undefined) {
                ctx['faithfulness'] = faithfulnessScore(answer, chunks);
            }
            if (includeChunks) {
                ctx['chunks'] = chunks.map((c) => ({
                    id: c.chunk.id,
                    documentId: c.chunk.documentId,
                    score: c.score,
                    source: c.source,
                }));
            }
            await Promise.resolve(logger.info('rag-eval', ctx));
        },
    };
}
