// @lumiarq/pack-lumiarq@1.0.1/dist/commands/index.js
import { spawn } from 'node:child_process';
import { once } from 'node:events';
export async function getLumiarqCommands() {
    return [
        {
            name: 'lumiarq.boundary-check',
            description: 'Run LumiARQ boundary checks for the current project.',
            args: [],
            run: async (_args, projectRoot) => {
                await runShellCommand('pnpm turbo boundary:check', projectRoot);
            },
        },
        {
            name: 'lumiarq.typecheck',
            description: 'Run TypeScript typecheck in the current project.',
            args: [],
            run: async (_args, projectRoot) => {
                await runShellCommand('pnpm typecheck', projectRoot);
            },
        },
    ];
}
async function runShellCommand(command, projectRoot) {
    const child = spawn(command, {
        cwd: projectRoot,
        shell: true,
        stdio: 'inherit',
    });
    const [exitCode] = (await once(child, 'close'));
    if (exitCode !== 0) {
        throw new Error(`Command failed: ${command}`);
    }
}
