// @lumiarq/lumis@1.2.0/dist/runtime/executor.js
import { spawn } from 'node:child_process';
import { redactRuntimeProfileForLogs } from './profiles.js';
function nowIso() {
    return new Date().toISOString();
}
function shouldRedactEnvKey(key) {
    return /token|secret|password|key/i.test(key);
}
function collectSensitiveValues(request) {
    const values = new Set();
    const processEnv = request.processEnv ?? process.env;
    for (const [key, value] of Object.entries(request.env ?? {})) {
        if (value && shouldRedactEnvKey(key)) {
            values.add(value);
        }
    }
    if (request.profile.kind === 'ssh') {
        const sshProfile = request.profile;
        if (sshProfile.privateKeyEnvVar) {
            const keyValue = processEnv[sshProfile.privateKeyEnvVar];
            if (keyValue) {
                values.add(keyValue);
            }
        }
        if (sshProfile.passwordEnvVar) {
            const passwordValue = processEnv[sshProfile.passwordEnvVar];
            if (passwordValue) {
                values.add(passwordValue);
            }
        }
    }
    return Array.from(values);
}
function redactText(text, sensitiveValues) {
    let redacted = text;
    for (const value of sensitiveValues) {
        if (!value) {
            continue;
        }
        redacted = redacted.split(value).join('[REDACTED]');
    }
    return redacted;
}
function quoteArg(arg) {
    if (/^[A-Za-z0-9_./:-]+$/.test(arg)) {
        return arg;
    }
    return `'${arg.replace(/'/g, `'\\''`)}'`;
}
function buildSshArgs(profile, request) {
    const processEnv = request.processEnv ?? process.env;
    const args = [];
    if (profile.port !== undefined) {
        args.push('-p', String(profile.port));
    }
    if (profile.knownHostsPath) {
        args.push('-o', `UserKnownHostsFile=${profile.knownHostsPath}`);
    }
    if (profile.privateKeyEnvVar) {
        const privateKeyPath = processEnv[profile.privateKeyEnvVar];
        if (!privateKeyPath) {
            return { ok: false, error: `Missing SSH private key env var: ${profile.privateKeyEnvVar}` };
        }
        args.push('-i', privateKeyPath);
    }
    if (profile.passwordEnvVar && !processEnv[profile.passwordEnvVar]) {
        return { ok: false, error: `Missing SSH password env var: ${profile.passwordEnvVar}` };
    }
    args.push(`${profile.user}@${profile.host}`);
    const remoteTokens = [request.command, ...(request.args ?? [])].map((token) => quoteArg(token));
    if (request.cwd && request.cwd.trim().length > 0) {
        remoteTokens.unshift('&&');
        remoteTokens.unshift(quoteArg(request.cwd));
        remoteTokens.unshift('cd');
    }
    args.push(remoteTokens.join(' '));
    return { ok: true, args };
}
function createResultSkeleton(request) {
    const startedAt = nowIso();
    return {
        profileId: request.profile.id,
        profileKind: request.profile.kind,
        command: request.command,
        args: request.args ?? [],
        success: false,
        exitCode: null,
        startedAt,
        endedAt: startedAt,
        durationMs: 0,
        transcript: [],
        stdout: '',
        stderr: '',
        redactedProfile: redactRuntimeProfileForLogs(request.profile),
        diagnostics: [],
    };
}
export async function executeOnRuntimeProfile(request) {
    const result = createResultSkeleton(request);
    const startedAtMs = Date.now();
    const sensitiveValues = collectSensitiveValues(request);
    const writeTranscript = (stream, text) => {
        const redacted = redactText(text, sensitiveValues);
        if (!redacted) {
            return;
        }
        result.transcript.push({
            stream,
            text: redacted,
            timestamp: nowIso(),
        });
        if (stream === 'stdout') {
            result.stdout += redacted;
        }
        if (stream === 'stderr' || stream === 'system') {
            result.stderr += redacted;
        }
    };
    const env = {
        ...(request.processEnv ?? process.env),
        ...(request.env ?? {}),
    };
    let executable = request.command;
    let executableArgs = request.args ?? [];
    if (request.profile.kind === 'ssh') {
        executable = 'ssh';
        const sshArgs = buildSshArgs(request.profile, request);
        if (!sshArgs.ok) {
            result.diagnostics.push(sshArgs.error);
            writeTranscript('system', `${sshArgs.error}\n`);
            result.endedAt = nowIso();
            result.durationMs = Date.now() - startedAtMs;
            return result;
        }
        executableArgs = sshArgs.args;
    }
    try {
        const child = spawn(executable, executableArgs, {
            cwd: request.cwd,
            env,
            shell: false,
        });
        let timedOut = false;
        let timeoutHandle;
        if (request.timeoutMs && request.timeoutMs > 0) {
            timeoutHandle = setTimeout(() => {
                timedOut = true;
                child.kill('SIGTERM');
            }, request.timeoutMs);
        }
        child.stdout.on('data', (chunk) => {
            writeTranscript('stdout', String(chunk));
        });
        child.stderr.on('data', (chunk) => {
            writeTranscript('stderr', String(chunk));
        });
        const exitCode = await new Promise((resolve, reject) => {
            child.on('error', reject);
            child.on('close', (code) => resolve(code));
        });
        if (timeoutHandle) {
            clearTimeout(timeoutHandle);
        }
        if (timedOut) {
            result.diagnostics.push(`Execution timed out after ${request.timeoutMs}ms`);
        }
        result.exitCode = exitCode;
        result.success = exitCode === 0 && !timedOut;
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        result.diagnostics.push(message);
        writeTranscript('system', `${message}\n`);
        result.success = false;
    }
    result.endedAt = nowIso();
    result.durationMs = Date.now() - startedAtMs;
    return result;
}
