// @lumiarq/lumis@1.2.0/dist/runtime/orchestration.js
import path from 'node:path';
import { executeOnRuntimeProfile } from './executor.js';
import { evaluateRuntimePolicy } from './policy.js';
import { getRuntimeProfile, loadRuntimeProfileStore } from './profiles.js';
import { persistRuntimeTranscript } from './transcripts.js';
function resolveRuntimeTasks(config, adapter) {
    const requestedTasks = config.runtime?.postMakeTasks ?? ['typecheck', 'boundaryCheck'];
    const resolved = [];
    for (const task of requestedTasks) {
        const override = config.runtime?.taskCommandOverrides?.[task];
        const command = override ?? adapter.buildCommands[task];
        if (typeof command === 'string' && command.trim().length > 0) {
            resolved.push({ task, command: command.trim() });
        }
    }
    return resolved;
}
function resolveRuntimeTasksWithSelection(config, adapter, requestedTasks) {
    if (!requestedTasks || requestedTasks.length === 0) {
        return resolveRuntimeTasks(config, adapter);
    }
    return requestedTasks
        .map((task) => {
        const override = config.runtime?.taskCommandOverrides?.[task];
        const command = override ?? adapter.buildCommands[task];
        if (!command || command.trim().length === 0) {
            return null;
        }
        return { task, command: command.trim() };
    })
        .filter((entry) => entry !== null);
}
export async function executeRuntimeTasks(input) {
    const emitTrace = async (event) => {
        if (!input.onTrace)
            return;
        await input.onTrace(event);
    };
    await emitTrace({
        type: 'runtime-task',
        status: 'start',
        name: 'runtime-orchestration',
        context: { enabled: input.enabled },
    });
    if (!input.enabled) {
        await emitTrace({
            type: 'runtime-task',
            status: 'info',
            name: 'runtime-disabled',
        });
        return {
            enabled: false,
            steps: [],
            success: true,
            errors: [],
        };
    }
    const store = loadRuntimeProfileStore(input.projectRoot);
    const profile = getRuntimeProfile(store, input.profileId ?? input.config.runtime?.profileId);
    if (!profile) {
        const error = 'Runtime profile not found. Configure runtime.profileId or defaultProfileId in .lumis/runtime-profiles.json.';
        await emitTrace({
            type: 'runtime-task',
            status: 'failure',
            name: 'runtime-profile-not-found',
            message: error,
        });
        input.writeError(`${error}\n`);
        return {
            enabled: true,
            steps: [],
            success: false,
            errors: [error],
        };
    }
    const policyDecision = evaluateRuntimePolicy({
        profile,
        config: input.config,
        ...(typeof input.approveRemote === 'boolean'
            ? { approveRemote: input.approveRemote }
            : {}),
        ...(input.processEnv ? { processEnv: input.processEnv } : {}),
    });
    if (!policyDecision.allowExecution) {
        const blockingReason = policyDecision.blockingReason ?? 'Runtime policy blocked execution.';
        await emitTrace({
            type: 'runtime-task',
            status: 'failure',
            name: 'runtime-policy-blocked',
            message: blockingReason,
            context: { profileId: profile.id, policyMode: policyDecision.mode },
        });
        input.writeError(`${blockingReason}\n`);
        return {
            enabled: true,
            profileId: profile.id,
            profileKind: profile.kind,
            steps: [],
            success: false,
            errors: [blockingReason],
        };
    }
    if (policyDecision.requiresApproval && policyDecision.approved) {
        input.write('Runtime policy: remote execution approved.\n');
    }
    const commands = resolveRuntimeTasksWithSelection(input.config, input.adapter, input.requestedTasks);
    if (commands.length === 0) {
        await emitTrace({
            type: 'runtime-task',
            status: 'info',
            name: 'runtime-no-tasks',
            context: { profileId: profile.id },
        });
        input.write('Runtime orchestration enabled, but no buildCommands matched configured tasks.\n');
        return {
            enabled: true,
            profileId: profile.id,
            profileKind: profile.kind,
            steps: [],
            success: true,
            errors: [],
        };
    }
    input.write(`Runtime profile: ${profile.name} (${profile.kind})\n`);
    const steps = [];
    const errors = [];
    const failOnTaskError = input.failOnTaskError ?? input.config.runtime?.failOnTaskError !== false;
    const labelPrefix = input.labelPrefix ?? 'runtime';
    for (const runtimeTask of commands) {
        await emitTrace({
            type: 'runtime-task',
            status: 'start',
            name: runtimeTask.task,
            context: { command: runtimeTask.command, profileId: profile.id },
        });
        input.write(`Runtime task [${runtimeTask.task}]: ${runtimeTask.command}\n`);
        const result = await executeOnRuntimeProfile({
            profile,
            command: process.platform === 'win32' ? 'cmd' : 'sh',
            args: process.platform === 'win32' ? ['/c', runtimeTask.command] : ['-lc', runtimeTask.command],
            cwd: input.projectRoot,
        });
        const transcriptPath = persistRuntimeTranscript(input.projectRoot, `${labelPrefix}-${runtimeTask.task}`, result);
        const relativePath = path.relative(input.projectRoot, transcriptPath);
        steps.push({
            task: runtimeTask.task,
            command: runtimeTask.command,
            result,
            transcriptPath,
        });
        if (result.success) {
            await emitTrace({
                type: 'runtime-task',
                status: 'success',
                name: runtimeTask.task,
                context: { transcriptPath: relativePath },
            });
            input.write(`  [ok] ${runtimeTask.task} (transcript: ${relativePath})\n`);
            continue;
        }
        const error = `${runtimeTask.task} failed with exit code ${result.exitCode ?? 'null'}`;
        errors.push(error);
        await emitTrace({
            type: 'runtime-task',
            status: 'failure',
            name: runtimeTask.task,
            message: error,
            context: { transcriptPath: relativePath },
        });
        input.writeError(`  [fail] ${error} (transcript: ${relativePath})\n`);
        for (const diagnostic of result.diagnostics) {
            input.writeError(`  [diag] ${diagnostic}\n`);
        }
        if (failOnTaskError) {
            break;
        }
    }
    const advisoryMode = policyDecision.mode === 'advisory';
    if (advisoryMode && errors.length > 0) {
        input.writeError('Runtime policy mode is advisory: task failures recorded but execution continues.\n');
    }
    await emitTrace({
        type: 'runtime-task',
        status: errors.length === 0 || advisoryMode ? 'success' : 'failure',
        name: 'runtime-orchestration-complete',
        context: { errors: errors.length, advisoryMode, profileId: profile.id },
    });
    return {
        enabled: true,
        profileId: profile.id,
        profileKind: profile.kind,
        steps,
        success: advisoryMode ? true : errors.length === 0,
        errors,
    };
}
export async function runPostMakeRuntimeTasks(input) {
    return executeRuntimeTasks({
        projectRoot: input.projectRoot,
        adapter: input.adapter,
        config: input.config,
        write: input.write,
        writeError: input.writeError,
        ...(input.onTrace ? { onTrace: input.onTrace } : {}),
        enabled: input.config.runtime?.enabled === true,
        labelPrefix: 'make',
    });
}
