// @lumiarq/lumis@1.2.0/dist/runtime/policy.js
function isTruthy(value) {
    if (!value) {
        return false;
    }
    const normalized = value.trim().toLowerCase();
    return normalized === '1' || normalized === 'true' || normalized === 'yes' || normalized === 'y';
}
export function resolveRuntimePolicy(config) {
    const mode = config.runtime?.policy?.mode ?? 'strict';
    const requireApprovalForRemote = config.runtime?.policy?.requireApprovalForRemote !== false;
    const approvalEnvVar = config.runtime?.policy?.approvalEnvVar ?? 'LUMIS_RUNTIME_APPROVED';
    return {
        mode,
        requireApprovalForRemote,
        approvalEnvVar,
    };
}
export function evaluateRuntimePolicy(input) {
    const policy = resolveRuntimePolicy(input.config);
    const isRemoteProfile = input.profile.kind !== 'local';
    const requiresApproval = isRemoteProfile && policy.requireApprovalForRemote;
    const env = input.processEnv ?? process.env;
    const approvedByEnv = isTruthy(env[policy.approvalEnvVar]);
    const approvedByFlag = input.approveRemote === true;
    const approved = !requiresApproval || approvedByEnv || approvedByFlag;
    if (!approved) {
        return {
            mode: policy.mode,
            requiresApproval,
            approved,
            allowExecution: false,
            blockingReason: `Runtime profile ${input.profile.id} (${input.profile.kind}) requires explicit approval. ` +
                `Set ${policy.approvalEnvVar}=true or pass --approve-remote.`,
        };
    }
    return {
        mode: policy.mode,
        requiresApproval,
        approved,
        allowExecution: true,
    };
}
