// @lumiarq/lumis@1.2.0/dist/runtime/profiles.js
import fs from 'node:fs';
import path from 'node:path';
const PROFILE_STORE_RELATIVE_PATH = path.join('.lumis', 'runtime-profiles.json');
function isObject(value) {
    return typeof value === 'object' && value !== null;
}
function asString(value) {
    return typeof value === 'string' && value.trim().length > 0 ? value.trim() : null;
}
function validateProfile(profile, index) {
    if (!isObject(profile)) {
        return [`profiles[${index}] must be an object`];
    }
    const id = asString(profile.id);
    const name = asString(profile.name);
    const kind = asString(profile.kind);
    const errors = [];
    if (!id)
        errors.push(`profiles[${index}].id must be a non-empty string`);
    if (!name)
        errors.push(`profiles[${index}].name must be a non-empty string`);
    if (!kind)
        errors.push(`profiles[${index}].kind must be a non-empty string`);
    if (kind === 'ssh') {
        if (!asString(profile.host)) {
            errors.push(`profiles[${index}].host is required for ssh profile`);
        }
        if (!asString(profile.user)) {
            errors.push(`profiles[${index}].user is required for ssh profile`);
        }
    }
    if (kind === 'container') {
        if (!asString(profile.image)) {
            errors.push(`profiles[${index}].image is required for container profile`);
        }
        if (!asString(profile.workspacePath)) {
            errors.push(`profiles[${index}].workspacePath is required for container profile`);
        }
    }
    if (kind === 'ci' && !asString(profile.provider)) {
        errors.push(`profiles[${index}].provider is required for ci profile`);
    }
    return errors;
}
export function validateRuntimeProfileStore(input) {
    if (!isObject(input)) {
        return { ok: false, errors: ['profile store must be an object'] };
    }
    const version = asString(input.version);
    const profiles = input.profiles;
    const defaultProfileId = input.defaultProfileId;
    const errors = [];
    if (version !== '1.0') {
        errors.push('profile store version must be "1.0"');
    }
    if (!Array.isArray(profiles)) {
        errors.push('profiles must be an array');
    }
    else {
        for (let index = 0; index < profiles.length; index += 1) {
            errors.push(...validateProfile(profiles[index], index));
        }
    }
    if (defaultProfileId !== undefined && asString(defaultProfileId) === null) {
        errors.push('defaultProfileId must be a non-empty string when provided');
    }
    if (errors.length > 0) {
        return { ok: false, errors };
    }
    return {
        ok: true,
        value: input,
    };
}
export function createDefaultRuntimeProfileStore(projectRoot) {
    return {
        version: '1.0',
        defaultProfileId: 'local',
        profiles: [
            {
                id: 'local',
                name: 'Local shell',
                kind: 'local',
                workingDirectory: projectRoot,
            },
        ],
    };
}
export function loadRuntimeProfileStore(projectRoot) {
    const storePath = path.join(projectRoot, PROFILE_STORE_RELATIVE_PATH);
    if (!fs.existsSync(storePath)) {
        return createDefaultRuntimeProfileStore(projectRoot);
    }
    const raw = fs.readFileSync(storePath, 'utf8');
    let parsed;
    try {
        parsed = JSON.parse(raw);
    }
    catch {
        return createDefaultRuntimeProfileStore(projectRoot);
    }
    const validated = validateRuntimeProfileStore(parsed);
    if (!validated.ok) {
        return createDefaultRuntimeProfileStore(projectRoot);
    }
    return validated.value;
}
export function saveRuntimeProfileStore(projectRoot, store) {
    const storePath = path.join(projectRoot, PROFILE_STORE_RELATIVE_PATH);
    fs.mkdirSync(path.dirname(storePath), { recursive: true });
    fs.writeFileSync(storePath, JSON.stringify(store, null, 2), 'utf8');
}
export function getRuntimeProfile(store, profileId) {
    const id = profileId ?? store.defaultProfileId;
    if (!id) {
        return null;
    }
    return store.profiles.find((profile) => profile.id === id) ?? null;
}
export function redactRuntimeProfileForLogs(profile) {
    if (profile.kind !== 'ssh') {
        return profile;
    }
    const sshProfile = profile;
    return {
        ...sshProfile,
        ...(sshProfile.privateKeyEnvVar ? { privateKeyEnvVar: '[REDACTED_ENV_VAR]' } : {}),
        ...(sshProfile.passwordEnvVar ? { passwordEnvVar: '[REDACTED_ENV_VAR]' } : {}),
    };
}
