// @lumiarq/lumis@1.2.0/dist/runtime/transcripts.js
import fs from 'node:fs';
import path from 'node:path';
const TRANSCRIPTS_RELATIVE_DIR = path.join('.lumis', 'transcripts');
const TRANSCRIPT_INDEX_FILE = 'index.json';
function getTranscriptPaths(projectRoot) {
    const dir = path.join(projectRoot, TRANSCRIPTS_RELATIVE_DIR);
    return {
        dir,
        indexPath: path.join(dir, TRANSCRIPT_INDEX_FILE),
    };
}
function sanitizeFileToken(token) {
    return token.replace(/[^a-zA-Z0-9_-]/g, '-').slice(0, 40);
}
export function persistRuntimeTranscript(projectRoot, label, result) {
    const transcriptsDir = path.join(projectRoot, TRANSCRIPTS_RELATIVE_DIR);
    fs.mkdirSync(transcriptsDir, { recursive: true });
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const fileName = `${timestamp}-${sanitizeFileToken(label)}-${sanitizeFileToken(result.profileId)}.json`;
    const filePath = path.join(transcriptsDir, fileName);
    const payload = {
        version: '1.0',
        createdAt: new Date().toISOString(),
        command: result.command,
        args: result.args,
        profileId: result.profileId,
        profileKind: result.profileKind,
        success: result.success,
        exitCode: result.exitCode,
        durationMs: result.durationMs,
        diagnostics: result.diagnostics,
        stdout: result.stdout,
        stderr: result.stderr,
        transcript: result.transcript,
        redactedProfile: result.redactedProfile,
    };
    fs.writeFileSync(filePath, JSON.stringify(payload, null, 2), 'utf8');
    appendRuntimeTranscriptIndex(projectRoot, {
        id: `${Date.now()}-${Math.random().toString(16).slice(2, 10)}`,
        createdAt: payload.createdAt,
        label,
        profileId: payload.profileId,
        profileKind: payload.profileKind,
        transcriptFile: path.relative(projectRoot, filePath).replaceAll('\\', '/'),
        success: payload.success,
        exitCode: payload.exitCode,
        durationMs: payload.durationMs,
    });
    return filePath;
}
export function loadRuntimeTranscriptIndex(projectRoot) {
    const { indexPath } = getTranscriptPaths(projectRoot);
    if (!fs.existsSync(indexPath)) {
        return {
            version: '1.0',
            updatedAt: new Date().toISOString(),
            entries: [],
        };
    }
    try {
        const raw = fs.readFileSync(indexPath, 'utf8');
        const parsed = JSON.parse(raw);
        if (parsed.version !== '1.0' || !Array.isArray(parsed.entries)) {
            return {
                version: '1.0',
                updatedAt: new Date().toISOString(),
                entries: [],
            };
        }
        return parsed;
    }
    catch {
        return {
            version: '1.0',
            updatedAt: new Date().toISOString(),
            entries: [],
        };
    }
}
export function saveRuntimeTranscriptIndex(projectRoot, index) {
    const { dir: transcriptsDir, indexPath } = getTranscriptPaths(projectRoot);
    fs.mkdirSync(transcriptsDir, { recursive: true });
    fs.writeFileSync(indexPath, JSON.stringify(index, null, 2), 'utf8');
}
export function appendRuntimeTranscriptIndex(projectRoot, entry) {
    const index = loadRuntimeTranscriptIndex(projectRoot);
    index.entries.unshift(entry);
    index.updatedAt = new Date().toISOString();
    if (index.entries.length > 200) {
        index.entries = index.entries.slice(0, 200);
    }
    saveRuntimeTranscriptIndex(projectRoot, index);
}
export function summarizeRuntimeTranscripts(index) {
    const successful = index.entries.filter((entry) => entry.success).length;
    const failed = index.entries.length - successful;
    return {
        total: index.entries.length,
        successful,
        failed,
        latestEntry: index.entries[0] ?? null,
    };
}
function shouldKeepEntry(entry, options, nowMs) {
    if (options.profileId && entry.profileId !== options.profileId) {
        return true;
    }
    if (options.label && !entry.label.startsWith(options.label)) {
        return true;
    }
    if (options.maxAgeDays !== undefined) {
        const createdAtMs = Date.parse(entry.createdAt);
        const maxAgeMs = options.maxAgeDays * 24 * 60 * 60 * 1000;
        if (!Number.isNaN(createdAtMs) && nowMs - createdAtMs <= maxAgeMs) {
            return true;
        }
    }
    return false;
}
export function pruneRuntimeTranscripts(projectRoot, options = {}) {
    const index = loadRuntimeTranscriptIndex(projectRoot);
    const { indexPath } = getTranscriptPaths(projectRoot);
    const nowMs = Date.now();
    const keepCount = options.keep ?? 20;
    const filtered = index.entries.filter((entry) => {
        if (options.profileId && entry.profileId !== options.profileId) {
            return false;
        }
        if (options.label && !entry.label.startsWith(options.label)) {
            return false;
        }
        return true;
    });
    const guaranteedKeepIds = new Set(filtered.slice(0, keepCount).map((entry) => entry.id));
    const removedEntries = [];
    const keptEntries = [];
    let missingFilesRemoved = 0;
    for (const entry of index.entries) {
        const transcriptPath = path.join(projectRoot, entry.transcriptFile);
        const fileExists = fs.existsSync(transcriptPath);
        const matchesFilter = filtered.some((candidate) => candidate.id === entry.id);
        const keepBecauseRecent = guaranteedKeepIds.has(entry.id);
        const keepBecausePolicy = matchesFilter ? shouldKeepEntry(entry, options, nowMs) : true;
        if (keepBecauseRecent || keepBecausePolicy) {
            keptEntries.push(entry);
            continue;
        }
        removedEntries.push(entry);
        if (!options.dryRun) {
            if (fileExists) {
                fs.rmSync(transcriptPath, { force: true });
            }
            else {
                missingFilesRemoved += 1;
            }
        }
    }
    if (!options.dryRun) {
        saveRuntimeTranscriptIndex(projectRoot, {
            version: '1.0',
            updatedAt: new Date().toISOString(),
            entries: keptEntries,
        });
    }
    return {
        removedEntries,
        keptEntries,
        missingFilesRemoved,
        indexPath,
    };
}
