// @lumiarq/rag@0.1.0/dist/bm25/index.js
/**
 * BM25 (Okapi BM25) sparse retrieval implementation.
 * No external dependencies — pure TypeScript.
 *
 * Tuning parameters:
 *   k1 = 1.5  (term frequency saturation)
 *   b  = 0.75 (document length normalisation)
 */
function tokenize(text) {
    return text
        .toLowerCase()
        .replace(/[^\w\s]/g, ' ')
        .split(/\s+/)
        .filter((t) => t.length > 1);
}
export class BM25Index {
    k1;
    b;
    docs = new Map();
    df = new Map(); // term → document frequency
    avgDocLength = 0;
    constructor(k1 = 1.5, b = 0.75) {
        this.k1 = k1;
        this.b = b;
    }
    /** Add or replace a document in the index. */
    upsert(id, text) {
        // Remove existing document contribution to df
        const existing = this.docs.get(id);
        if (existing) {
            for (const term of existing.terms.keys()) {
                const count = (this.df.get(term) ?? 0) - 1;
                if (count <= 0) {
                    this.df.delete(term);
                }
                else {
                    this.df.set(term, count);
                }
            }
        }
        const tokens = tokenize(text);
        const tf = new Map();
        for (const token of tokens) {
            tf.set(token, (tf.get(token) ?? 0) + 1);
        }
        this.docs.set(id, { id, terms: tf, length: tokens.length });
        for (const term of tf.keys()) {
            this.df.set(term, (this.df.get(term) ?? 0) + 1);
        }
        this.recomputeAvg();
    }
    /** Remove a document from the index. */
    delete(id) {
        const doc = this.docs.get(id);
        if (!doc)
            return;
        for (const term of doc.terms.keys()) {
            const count = (this.df.get(term) ?? 0) - 1;
            if (count <= 0) {
                this.df.delete(term);
            }
            else {
                this.df.set(term, count);
            }
        }
        this.docs.delete(id);
        this.recomputeAvg();
    }
    /** Score all documents against `query`, return top-K sorted descending. */
    search(query, topK = 10) {
        const N = this.docs.size;
        if (N === 0)
            return [];
        const queryTerms = tokenize(query);
        const scores = new Map();
        for (const term of queryTerms) {
            const df = this.df.get(term) ?? 0;
            if (df === 0)
                continue;
            const idf = Math.log((N - df + 0.5) / (df + 0.5) + 1);
            for (const doc of this.docs.values()) {
                const tf = doc.terms.get(term) ?? 0;
                if (tf === 0)
                    continue;
                const norm = 1 - this.b + this.b * (doc.length / this.avgDocLength);
                const score = idf * ((tf * (this.k1 + 1)) / (tf + this.k1 * norm));
                scores.set(doc.id, (scores.get(doc.id) ?? 0) + score);
            }
        }
        return Array.from(scores.entries())
            .map(([id, score]) => ({ id, score }))
            .sort((a, b) => b.score - a.score)
            .slice(0, topK);
    }
    get size() {
        return this.docs.size;
    }
    recomputeAvg() {
        if (this.docs.size === 0) {
            this.avgDocLength = 0;
            return;
        }
        let total = 0;
        for (const doc of this.docs.values())
            total += doc.length;
        this.avgDocLength = total / this.docs.size;
    }
}
