// @lumiarq/pack-lumiarq@1.0.1/dist/rules/index.js
export const lumiarqPlacementRules = [
    {
        artifactKind: 'action',
        directoryPattern: 'logic/actions/',
        filePattern: '{verb}-{entity}.action.ts',
        classPattern: '{Verb}{Entity}Action',
    },
    {
        artifactKind: 'task',
        directoryPattern: 'logic/tasks/',
        filePattern: '{verb}-{entity}.task.ts',
        classPattern: '{Verb}{Entity}Task',
    },
    {
        artifactKind: 'query',
        directoryPattern: 'logic/queries/',
        filePattern: '{verb}-{entity}.query.ts',
        classPattern: '{Verb}{Entity}Query',
    },
    {
        artifactKind: 'handler',
        directoryPattern: 'http/handlers/',
        filePattern: '{verb}-{entity}.handler.ts',
        classPattern: '{Verb}{Entity}Handler',
    },
    {
        artifactKind: 'schema',
        directoryPattern: 'database/schemas/',
        filePattern: '{entity}.schema.ts',
        classPattern: '{entity}s',
    },
    {
        artifactKind: 'repository',
        directoryPattern: 'database/repositories/',
        filePattern: '{entity}.repository.ts',
        classPattern: '{Entity}Repository',
    },
    {
        artifactKind: 'event',
        directoryPattern: 'events/definitions/',
        filePattern: '{entity}-{pastTense}.event.ts',
        classPattern: '{Entity}{PastTense}',
    },
    {
        artifactKind: 'dto',
        directoryPattern: 'contracts/types/',
        filePattern: '{verb}-{entity}.dto.ts',
        classPattern: '{Verb}{Entity}DTO',
    },
];
export const lumiarqDependencyRules = [
    {
        fromKind: 'handler',
        toKind: 'task',
        allowed: false,
        message: 'Handlers must not call Tasks directly — use an Action',
    },
    {
        fromKind: 'action',
        toKind: 'schema',
        allowed: false,
        message: 'Actions must not access schemas directly — use a Task + Repository',
    },
    {
        fromKind: 'loader',
        toKind: 'action',
        allowed: false,
        message: 'Loaders must not import Actions — Loaders are read-only',
    },
];
export const lumiarqNamingRules = [
    {
        artifactKind: 'action',
        filePattern: /^[a-z]+(-[a-z]+)*\.action\.ts$/,
        classPattern: /^[A-Z][a-zA-Z]*Action$/,
        message: 'Actions must follow kebab-case filenames and {Verb}{Entity}Action class names',
    },
    {
        artifactKind: 'task',
        filePattern: /^[a-z]+(-[a-z]+)*\.task\.ts$/,
        classPattern: /^[A-Z][a-zA-Z]*Task$/,
        message: 'Tasks must follow kebab-case filenames and {Verb}{Entity}Task class names',
    },
    {
        artifactKind: 'schema',
        filePattern: /^[a-z]+(-[a-z]+)*\.schema\.ts$/,
        classPattern: /^[a-z]+s$/,
        message: 'Schemas must use kebab-case filenames and lowercase table names',
    },
];
