// @lumiarq/rag@0.1.0/dist/adapters/embedder/base.js
function l2Normalize(vec) {
    let norm = 0;
    for (const v of vec)
        norm += v * v;
    norm = Math.sqrt(norm);
    if (norm === 0)
        return vec;
    return vec.map((v) => v / norm);
}
export class BaseEmbedder {
    maxBatchSize;
    shouldNormalize;
    /** Wall-clock milliseconds of the most recent `embed()` call. */
    lastLatencyMs = 0;
    constructor(opts = {}) {
        this.maxBatchSize = opts.maxBatchSize ?? 512;
        this.shouldNormalize = opts.normalize ?? false;
    }
    async embed(texts) {
        if (texts.length === 0)
            return [];
        const t0 = Date.now();
        // Deduplicate within the call to avoid paying twice for repeated strings.
        const unique = Array.from(new Set(texts));
        const indexMap = new Map();
        for (let i = 0; i < unique.length; i++)
            indexMap.set(unique[i], i);
        // Chunk into maxBatchSize slices.
        const allVecs = new Array(unique.length);
        for (let offset = 0; offset < unique.length; offset += this.maxBatchSize) {
            const slice = unique.slice(offset, offset + this.maxBatchSize);
            const vecs = await this.embedBatch(slice);
            for (let j = 0; j < vecs.length; j++) {
                const vec = vecs[j];
                allVecs[offset + j] = this.shouldNormalize ? l2Normalize(vec) : vec;
            }
        }
        // Reconstruct output in original input order (including duplicates).
        const result = texts.map((t) => allVecs[indexMap.get(t)]);
        this.lastLatencyMs = Date.now() - t0;
        return result;
    }
}
