// @lumiarq/rag@0.1.0/dist/adapters/embedder/cached.js
export class CachedEmbedder {
    dimensions;
    model;
    inner;
    maxSize;
    ttlMs;
    /** Ordered map — iteration order tracks LRU (most-recently-used is last). */
    cache = new Map();
    /** Number of cache hits on the last `embed()` call. */
    lastHits = 0;
    /** Number of cache misses (API calls needed) on the last `embed()` call. */
    lastMisses = 0;
    constructor(inner, opts = {}) {
        this.inner = inner;
        this.dimensions = inner.dimensions;
        this.model = inner.model;
        this.maxSize = opts.maxSize ?? 10_000;
        this.ttlMs = opts.ttlMs ?? Infinity;
    }
    async embed(texts) {
        if (texts.length === 0)
            return [];
        const now = Date.now();
        const result = new Array(texts.length).fill(null);
        const missIndices = [];
        const missTexts = [];
        this.lastHits = 0;
        this.lastMisses = 0;
        // First pass: serve from cache where possible.
        for (let i = 0; i < texts.length; i++) {
            const key = `${this.model}:${texts[i]}`;
            const entry = this.cache.get(key);
            if (entry && now - entry.insertedAt < this.ttlMs) {
                // Refresh LRU position.
                this.cache.delete(key);
                this.cache.set(key, entry);
                result[i] = entry.vector;
                this.lastHits++;
            }
            else {
                if (entry)
                    this.cache.delete(key); // evict stale
                missIndices.push(i);
                missTexts.push(texts[i]);
                this.lastMisses++;
            }
        }
        // Second pass: embed misses via the inner embedder.
        if (missTexts.length > 0) {
            const fresh = await this.inner.embed(missTexts);
            for (let j = 0; j < missIndices.length; j++) {
                const idx = missIndices[j];
                const vec = fresh[j];
                result[idx] = vec;
                const key = `${this.model}:${texts[idx]}`;
                this.evictIfFull();
                this.cache.set(key, { vector: vec, insertedAt: now });
            }
        }
        return result;
    }
    /** Evict the least-recently-used entry when at capacity. */
    evictIfFull() {
        if (this.cache.size < this.maxSize)
            return;
        // Map iterator preserves insertion order; first key is LRU.
        const lruKey = this.cache.keys().next().value;
        if (typeof lruKey === 'string')
            this.cache.delete(lruKey);
    }
    /** Remove all cached entries. */
    clear() {
        this.cache.clear();
    }
    /** Current number of cached entries. */
    get size() {
        return this.cache.size;
    }
}
