// @lumiarq/rag@0.1.0/dist/adapters/embedder/gateway.js
/**
 * AIGatewayEmbedder — bridges `@lumiarq/ai` into the RAG `Embedder` interface.
 *
 * This is the primary embedder for any app already using `@lumiarq/ai`. It:
 *  • Delegates to `gateway.embedMany()` which handles provider selection,
 *    failover, and usage tracking automatically.
 *  • Accepts an optional model string in provider-prefixed format
 *    (`openai:text-embedding-3-large`, `gemini:text-embedding-004`).
 *  • Inherits batching and normalisation from BaseEmbedder.
 *
 * Usage:
 *   const embedder = new AIGatewayEmbedder(gateway, { model: 'openai:text-embedding-3-small', dimensions: 1536 })
 *   const pipeline = createRAGPipeline({ embedder, store })
 */
import { BaseEmbedder } from './base.js';
export class AIGatewayEmbedder extends BaseEmbedder {
    dimensions;
    model;
    gateway;
    modelRef;
    constructor(gateway, opts = {}) {
        super(opts);
        this.gateway = gateway;
        this.modelRef = opts.model;
        this.dimensions = opts.dimensions ?? 1536;
        this.model = opts.model ?? 'auto';
    }
    async embedBatch(texts) {
        if (typeof this.gateway.embedMany === 'function') {
            if (this.modelRef !== undefined) {
                return this.gateway.embedMany(texts, { model: this.modelRef });
            }
            return this.gateway.embedMany(texts);
        }
        const request = this.modelRef !== undefined
            ? { input: texts, model: this.modelRef }
            : { input: texts };
        const result = await this.gateway.embed(request);
        return result.embeddings;
    }
}
