// @trazze/ignite@0.1.0/dist/index.js
// src/parser.ts
var FRAME_RE = /^\s*at\s+(?:(.+?)\s+\()?(?:file:\/\/)?(\/[^)]+?)(?::(\d+))?(?::(\d+))?\)?$/;
function parseFrame(line) {
  const m = FRAME_RE.exec(line);
  if (!m) return null;
  const fn = m[1]?.trim() || void 0;
  const file = m[2] || void 0;
  const lineNo = m[3] ? parseInt(m[3], 10) : void 0;
  const col = m[4] ? parseInt(m[4], 10) : void 0;
  return {
    fn,
    file,
    line: lineNo,
    column: col,
    isVendor: file ? file.includes("node_modules") : false
  };
}
function parseStack(error) {
  if (!error.stack) return [];
  const lines = error.stack.split("\n");
  const frames = [];
  for (const line of lines) {
    const frame = parseFrame(line);
    if (frame) frames.push(frame);
  }
  return frames;
}

// src/reader.ts
import { readFileSync } from "node:fs";
function readContext(filePath, line, radius = 6) {
  try {
    const contents = readFileSync(filePath, "utf-8");
    const lines = contents.split("\n");
    const start = Math.max(0, line - 1 - radius);
    const end = Math.min(lines.length - 1, line - 1 + radius);
    const result = [];
    for (let i = start; i <= end; i++) {
      result.push({
        lineNo: i + 1,
        code: lines[i] ?? "",
        isFocus: i + 1 === line
      });
    }
    return result;
  } catch {
    return [];
  }
}

// src/editors.ts
function resolveEditorScheme() {
  const raw = (process.env["TRAZZE_EDITOR"] ?? process.env["EDITOR"] ?? "").toLowerCase();
  if (raw.includes("cursor")) return "cursor";
  if (raw.includes("code")) return "vscode";
  if (raw.includes("phpstorm")) return "phpstorm";
  if (raw.includes("webstorm")) return "webstorm";
  if (raw.includes("sublime") || raw.includes("subl")) return "sublime";
  if (raw.includes("textmate") || raw.includes("mate")) return "textmate";
  if (raw.includes("atom")) return "atom";
  const explicit = (process.env["TRAZZE_EDITOR"] ?? "").toLowerCase();
  const known = ["vscode", "cursor", "phpstorm", "webstorm", "sublime", "textmate", "atom"];
  if (known.includes(explicit)) return explicit;
  return null;
}
function buildEditorUrl(scheme, file, line, column = 1) {
  const enc = encodeURIComponent(file);
  switch (scheme) {
    case "vscode":
      return `vscode://file/${file}:${line}:${column}`;
    case "cursor":
      return `cursor://file/${file}:${line}:${column}`;
    case "phpstorm":
      return `phpstorm://open?file=${enc}&line=${line}`;
    case "webstorm":
      return `webstorm://open?file=${enc}&line=${line}`;
    case "sublime":
      return `subl://${file}:${line}`;
    case "textmate":
      return `txmt://open?url=file://${enc}&line=${line}`;
    case "atom":
      return `atom://core/open/file?filename=${enc}&line=${line}`;
  }
}
function editorLabel(scheme) {
  const labels = {
    vscode: "VS Code",
    cursor: "Cursor",
    phpstorm: "PhpStorm",
    webstorm: "WebStorm",
    sublime: "Sublime",
    textmate: "TextMate",
    atom: "Atom"
  };
  return labels[scheme];
}

// src/renderer.ts
function esc(s) {
  return s.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
}
function suggestFix(error) {
  const name = error.name ?? "";
  const msg = error.message ?? "";
  const hints = [];
  if (name === "TypeError") {
    if (/cannot read propert(y|ies) of (null|undefined)/i.test(msg))
      hints.push("A value is <code>null</code> or <code>undefined</code> where an object was expected. Add a null-check or optional chaining (<code>?.</code>) before accessing the property.");
    else if (/is not a function/i.test(msg))
      hints.push("The value you are calling is not a function. Check that the import is correct and the export name matches.");
    else if (/cannot set propert/i.test(msg))
      hints.push("Attempted to assign to a non-writable or frozen property. Verify the object structure.");
    else
      hints.push("A type mismatch occurred. Confirm the shape of data returned from your database or API before using it.");
  }
  if (name === "ReferenceError")
    hints.push("A variable or binding is referenced before it is declared. Check for circular imports or a missing <code>await</code> on a dynamic import.");
  if (name === "SyntaxError")
    hints.push("JavaScript syntax is invalid. This often means a build step failed or a JSON payload is malformed.");
  if (/ENOENT|no such file/i.test(msg))
    hints.push("A file or directory does not exist at the path shown in the stack. Check your config paths and ensure the file was deployed.");
  if (/EACCES|permission denied/i.test(msg))
    hints.push("The process lacks permission to read/write the path. Check file ownership and environment capabilities.");
  if (/ECONNREFUSED|connect ECONNREFUSED/i.test(msg))
    hints.push("A downstream service refused the TCP connection. Make sure the database, cache, or API is running and the port/host in your <code>.env</code> is correct.");
  if (/timeout|timed out/i.test(msg))
    hints.push("A network call or async operation exceeded its timeout. Check service health and consider increasing the timeout threshold.");
  if (/invalid.*token|unexpected token/i.test(msg))
    hints.push("A JSON or expression parse failed. Inspect the raw response body for HTML error pages returned by downstream services.");
  if (hints.length === 0)
    hints.push("Inspect the first application frame below. The highlighted line is where the error originated.");
  return hints;
}
function renderHints(error) {
  const hints = suggestFix(error);
  return `
    <div class="hints">
      <div class="hints-title">
        <svg width="14" height="14" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><circle cx="12" cy="12" r="10"/><line x1="12" y1="8" x2="12" y2="12"/><line x1="12" y1="16" x2="12.01" y2="16"/></svg>
        Possible cause
      </div>
      <ul class="hints-list">
        ${hints.map((h) => `<li>${h}</li>`).join("")}
      </ul>
    </div>`;
}
function renderFrame(frame, idx, editorScheme) {
  const file = frame.file ?? "(unknown)";
  const short = file.split("/").slice(-3).join("/");
  const loc = frame.line != null ? `:${frame.line}` : "";
  const fn = esc(frame.fn ?? "(anonymous)");
  const editorLink = editorScheme && frame.file && frame.line ? buildEditorUrl(editorScheme, frame.file, frame.line, frame.column ?? 1) : null;
  const openBtn = editorLink ? `<a class="open-editor-btn" href="${esc(editorLink)}" title="Open in ${editorLabel(editorScheme)}">
        <svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><path d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"/><polyline points="15 3 21 3 21 9"/><line x1="10" y1="14" x2="21" y2="3"/></svg>
        Open in ${esc(editorLabel(editorScheme))}
      </a>` : "";
  const context = frame.file && frame.line ? readContext(frame.file, frame.line) : [];
  const contextHtml = context.length > 0 ? `<div class="code-block">
        <table class="code-table">
          ${context.map((cl) => `
            <tr class="${cl.isFocus ? "focus-line" : ""}">
              <td class="line-no">${cl.lineNo}</td>
              <td class="line-code"><code>${esc(cl.code)}</code></td>
            </tr>`).join("")}
        </table>
      </div>` : `<div class="code-block code-block--empty"><span class="muted">Source not available for this frame</span></div>`;
  return `
    <div class="frame ${idx === 0 ? "frame--first" : ""} ${frame.isVendor ? "frame--vendor" : ""}" data-frame-idx="${idx}">
      <div class="frame-header">
        <span class="frame-fn">${fn}</span>
        <span class="frame-loc">${esc(short)}${esc(loc)}</span>
        ${openBtn}
      </div>
      ${contextHtml}
    </div>`;
}
function renderRequestSection(req) {
  const url = new URL(req.url);
  const headers = [];
  req.headers.forEach((val, key) => {
    if (key.toLowerCase() !== "cookie" && key.toLowerCase() !== "authorization")
      headers.push([key, val]);
  });
  const headersHtml = headers.length > 0 ? `<table class="kv-table">
        ${headers.map(([k, v]) => `
          <tr>
            <td class="kv-key">${esc(k)}</td>
            <td class="kv-val">${esc(v)}</td>
          </tr>`).join("")}
      </table>` : `<span class="muted">No request headers available</span>`;
  return `
    <details class="collapsible" open>
      <summary class="collapsible-summary">
        <svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><polyline points="6 9 12 15 18 9"/></svg>
        Request
        <span class="collapsible-meta">${esc(req.method)} ${esc(url.pathname)}</span>
      </summary>
      <div class="collapsible-body">
        <div class="kv-group">
          <div class="kv-label">URL</div>
          <div class="kv-mono">${esc(url.toString())}</div>
        </div>
        <div class="kv-group">
          <div class="kv-label">Headers</div>
          ${headersHtml}
        </div>
      </div>
    </details>`;
}
function renderContextSection(env) {
  const keys = [
    ["APP_ENV", env["APP_ENV"] ?? "\u2014"],
    ["NODE_ENV", env["NODE_ENV"] ?? "\u2014"],
    ["NODE", process.version],
    ["Platform", process.platform]
  ];
  return `
    <details class="collapsible">
      <summary class="collapsible-summary">
        <svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><polyline points="6 9 12 15 18 9"/></svg>
        Environment
      </summary>
      <div class="collapsible-body">
        <table class="kv-table">
          ${keys.map(([k, v]) => `
            <tr>
              <td class="kv-key">${esc(k)}</td>
              <td class="kv-val">${esc(String(v))}</td>
            </tr>`).join("")}
        </table>
      </div>
    </details>`;
}
function renderIgnitionPage(error, req) {
  const frames = parseStack(error);
  const title = esc(error.name ?? "Error");
  const msg = esc(error.message ?? "");
  const url = new URL(req.url);
  const ts = /* @__PURE__ */ new Date();
  const editorScheme = resolveEditorScheme();
  const appFrames = frames.filter((f) => !f.isVendor);
  const vendorFrames = frames.filter((f) => f.isVendor);
  const allFrames = [...appFrames, ...vendorFrames];
  const tracePayload = JSON.stringify({
    error: { name: error.name, message: error.message, stack: error.stack },
    request: { method: req.method, url: req.url },
    timestamp: ts.toISOString(),
    frames: allFrames.map((f) => ({
      fn: f.fn,
      file: f.file,
      line: f.line,
      column: f.column,
      isVendor: f.isVendor
    }))
  }, null, 2);
  return `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <title>${title}: ${msg}</title>
  <style>
    *, *::before, *::after { box-sizing: border-box; margin: 0; padding: 0; }
    :root {
      --bg: #09090b;
      --surface: #18181b;
      --surface2: #27272a;
      --border: #3f3f46;
      --text: #f4f4f5;
      --muted: #71717a;
      --red: #f87171;
      --red-dim: rgba(239,68,68,0.12);
      --amber: #fbbf24;
      --green: #4ade80;
      --focus-line: rgba(248,113,113,0.10);
      --focus-border: #ef4444;
      font-family: ui-sans-serif, system-ui, sans-serif;
      font-size: 14px;
      color-scheme: dark;
    }
    body { background: var(--bg); color: var(--text); min-height: 100vh; }
    .muted { color: var(--muted); }

    /* \u2500\u2500 Header \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .header {
      background: var(--red-dim);
      border-bottom: 1px solid rgba(239,68,68,0.25);
      padding: 20px 24px;
    }
    .header-inner { max-width: 1200px; margin: 0 auto; }
    .badge {
      display: inline-flex; align-items: center; gap: 5px;
      padding: 3px 10px; border-radius: 6px;
      background: var(--red-dim); border: 1px solid rgba(239,68,68,0.4);
      color: var(--red); font-size: 11px; font-weight: 600;
      letter-spacing: .06em; text-transform: uppercase; margin-bottom: 10px;
    }
    .badge svg { opacity: .8; }
    h1 { font-size: clamp(16px, 2.5vw, 22px); font-weight: 700; color: var(--red); word-break: break-word; }
    .msg { margin-top: 6px; font-size: clamp(13px, 1.5vw, 15px); color: var(--text); opacity: .85; word-break: break-word; }
    .meta {
      margin-top: 12px; display: flex; flex-wrap: wrap; align-items: center; gap: 8px 16px;
      font-size: 12px; color: var(--muted); font-family: ui-monospace, monospace;
    }
    .meta-pill {
      display: inline-flex; align-items: center; gap: 4px;
      background: var(--surface2); border: 1px solid var(--border);
      border-radius: 4px; padding: 2px 8px;
    }
    .copy-btn {
      margin-left: auto; padding: 4px 12px; border-radius: 6px;
      background: var(--surface2); border: 1px solid var(--border);
      color: var(--muted); font-size: 11px; cursor: pointer;
      transition: color .15s, border-color .15s;
    }
    .copy-btn:hover { color: var(--text); border-color: var(--muted); }
    .copy-btn.copied { color: var(--green); border-color: var(--green); }

    /* \u2500\u2500 Root layout \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .layout {
      display: grid;
      grid-template-columns: 300px 1fr;
      max-width: 1200px;
      margin: 0 auto;
    }

    /* \u2500\u2500 Sidebar \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .sidebar {
      border-right: 1px solid var(--border);
      height: calc(100dvh - var(--header-h, 130px));
      overflow-y: auto;
      position: sticky;
      top: 0;
    }
    .sidebar-section-label {
      padding: 8px 14px; font-size: 10px; font-weight: 700;
      letter-spacing: .08em; text-transform: uppercase; color: var(--muted);
      border-bottom: 1px solid var(--border);
    }
    .frame-item {
      padding: 9px 14px; border-bottom: 1px solid var(--border);
      cursor: pointer; transition: background .1s;
      border-left: 3px solid transparent;
    }
    .frame-item:hover { background: var(--surface2); }
    .frame-item--active { background: var(--surface); border-left-color: var(--red); }
    .frame-item--vendor { opacity: .4; }
    .frame-item--vendor:hover { opacity: .7; }
    .fi-fn { display: block; font-weight: 600; font-size: 12px; margin-bottom: 2px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; }
    .fi-path { font-size: 10px; color: var(--muted); font-family: ui-monospace, monospace; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; }

    /* \u2500\u2500 Main content \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .content { padding: 20px 24px; overflow-x: hidden; }

    /* \u2500\u2500 Hints \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .hints {
      background: rgba(251,191,36,0.06); border: 1px solid rgba(251,191,36,0.25);
      border-radius: 8px; padding: 14px 16px; margin-bottom: 20px;
    }
    .hints-title {
      display: flex; align-items: center; gap: 6px;
      font-size: 12px; font-weight: 700; letter-spacing: .04em; text-transform: uppercase;
      color: var(--amber); margin-bottom: 10px;
    }
    .hints-list { padding-left: 18px; display: flex; flex-direction: column; gap: 5px; }
    .hints-list li { font-size: 13px; color: var(--text); line-height: 1.55; }
    .hints-list code { font-family: ui-monospace, monospace; font-size: 12px; background: var(--surface2); padding: 1px 5px; border-radius: 3px; }

    /* \u2500\u2500 Frame detail \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .frame { margin-bottom: 14px; }
    .frame--vendor { opacity: .5; }
    .frame-header {
      display: flex; align-items: center; justify-content: space-between; gap: 8px;
      padding: 9px 14px; background: var(--surface);
      border: 1px solid var(--border); border-radius: 8px 8px 0 0;
    }
    .frame-fn  { font-weight: 600; font-size: 12px; flex: 1; min-width: 0; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; }
    .frame-loc { font-size: 10px; color: var(--muted); font-family: ui-monospace, monospace; flex-shrink: 0; }
    .open-editor-btn {
      display: inline-flex; align-items: center; gap: 4px; flex-shrink: 0;
      padding: 2px 8px; border-radius: 4px; font-size: 10px; font-weight: 500;
      background: var(--surface2); border: 1px solid var(--border);
      color: var(--muted); text-decoration: none; white-space: nowrap;
      transition: color .15s, border-color .15s, background .15s;
    }
    .open-editor-btn:hover { color: var(--text); border-color: var(--muted); background: var(--border); }
    .code-block {
      background: var(--surface); border: 1px solid var(--border);
      border-top: none; border-radius: 0 0 8px 8px; overflow: auto;
    }
    .code-block--empty { padding: 10px 14px; font-size: 12px; }
    .code-table { width: 100%; border-collapse: collapse; }
    .code-table tr { border-bottom: none; }
    .focus-line { background: var(--focus-line); border-left: 3px solid var(--focus-border); }
    .line-no {
      width: 44px; text-align: right; padding: 2px 10px 2px 8px;
      font-family: ui-monospace, monospace; font-size: 11px; color: var(--muted);
      user-select: none; vertical-align: top;
    }
    .line-code { padding: 2px 16px 2px 0; white-space: pre; font-family: ui-monospace, monospace; font-size: 11px; }

    /* \u2500\u2500 Vendor toggle \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .vendor-toggle {
      margin-top: 16px; border: 1px solid var(--border); border-radius: 8px;
      overflow: hidden;
    }

    /* \u2500\u2500 Collapsible sections \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .collapsible { border: 1px solid var(--border); border-radius: 8px; overflow: hidden; margin-bottom: 12px; }
    .collapsible-summary {
      display: flex; align-items: center; gap: 6px;
      padding: 10px 14px; cursor: pointer; background: var(--surface);
      font-size: 12px; font-weight: 600; color: var(--text); list-style: none;
      user-select: none;
    }
    .collapsible-summary::-webkit-details-marker { display: none; }
    .collapsible-summary::marker { display: none; }
    .collapsible-summary svg { transition: transform .2s; flex-shrink: 0; color: var(--muted); }
    details[open] .collapsible-summary svg { transform: rotate(0deg); }
    .collapsible-summary svg { transform: rotate(-90deg); }
    details[open] .collapsible-summary svg { transform: rotate(0deg); }
    .collapsible-meta { margin-left: auto; font-weight: 400; color: var(--muted); font-family: ui-monospace, monospace; font-size: 11px; }
    .collapsible-body { padding: 14px 16px; border-top: 1px solid var(--border); display: flex; flex-direction: column; gap: 14px; }

    /* \u2500\u2500 Key-value tables \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .kv-group { display: flex; flex-direction: column; gap: 5px; }
    .kv-label { font-size: 10px; font-weight: 700; letter-spacing: .06em; text-transform: uppercase; color: var(--muted); }
    .kv-mono  { font-family: ui-monospace, monospace; font-size: 12px; word-break: break-all; }
    .kv-table { width: 100%; border-collapse: collapse; font-size: 12px; }
    .kv-table tr { border-bottom: 1px solid var(--border); }
    .kv-table tr:last-child { border-bottom: none; }
    .kv-key { padding: 5px 12px 5px 0; font-family: ui-monospace, monospace; color: var(--muted); white-space: nowrap; vertical-align: top; width: 40%; }
    .kv-val { padding: 5px 0; font-family: ui-monospace, monospace; word-break: break-all; }

    /* \u2500\u2500 Trazze JSON section \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    .trazze-section {
      margin-top: 20px; border: 1px solid var(--border); border-radius: 8px; overflow: hidden;
    }
    .trazze-header {
      display: flex; align-items: center; justify-content: space-between;
      padding: 10px 14px; background: var(--surface);
      font-size: 12px; font-weight: 600; border-bottom: 1px solid var(--border);
    }
    .trazze-badge {
      display: inline-flex; align-items: center; gap: 4px;
      padding: 1px 8px; border-radius: 4px; font-size: 10px; font-weight: 700;
      letter-spacing: .06em; text-transform: uppercase;
      background: rgba(74,222,128,0.1); border: 1px solid rgba(74,222,128,0.3); color: var(--green);
    }
    .trazze-json {
      padding: 12px 16px; background: var(--bg);
      font-family: ui-monospace, monospace; font-size: 11px; line-height: 1.6;
      overflow-x: auto; max-height: 280px; white-space: pre;
    }

    /* \u2500\u2500 Responsive \u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500 */
    @media (max-width: 800px) {
      .layout {
        grid-template-columns: 1fr;
      }
      .sidebar {
        height: auto;
        max-height: 220px;
        position: static;
        border-right: none;
        border-bottom: 1px solid var(--border);
        overflow-y: auto;
      }
      .content { padding: 16px; }
      .header { padding: 16px; }
    }
    @media (max-width: 500px) {
      .meta-pill { display: none; }
      .frame-header { flex-direction: column; align-items: flex-start; gap: 3px; }
    }
  </style>
</head>
<body>
  <div class="header" id="__ignite_header">
    <div class="header-inner">
      <span class="badge">
        <svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><path d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"/><line x1="12" y1="9" x2="12" y2="13"/><line x1="12" y1="17" x2="12.01" y2="17"/></svg>
        Server Error
      </span>
      <h1>${title}</h1>
      <p class="msg">${msg}</p>
      <div class="meta">
        <span class="meta-pill">${esc(req.method)}</span>
        <span>${esc(url.pathname)}</span>
        <span class="meta-pill">${ts.toLocaleTimeString()}</span>
        <button class="copy-btn" id="__ignite_copy" type="button">Copy trace JSON</button>
      </div>
    </div>
  </div>

  <div class="layout">
    <aside class="sidebar" id="__ignite_sidebar">
      ${appFrames.length > 0 ? `<div class="sidebar-section-label">App frames</div>` : ""}
      ${appFrames.map((f, i) => `
        <div class="frame-item ${i === 0 ? "frame-item--active" : ""}" data-frame="${i}" role="button" tabindex="0">
          <span class="fi-fn">${esc(f.fn ?? "(anonymous)")}</span>
          <span class="fi-path">${esc((f.file ?? "").split("/").slice(-3).join("/"))}${f.line ? `:${f.line}` : ""}</span>
        </div>`).join("")}
      ${vendorFrames.length > 0 ? `
        <div class="sidebar-section-label">Vendor frames</div>
        ${vendorFrames.map((f, i) => `
          <div class="frame-item frame-item--vendor" data-frame="${appFrames.length + i}" role="button" tabindex="0">
            <span class="fi-fn">${esc(f.fn ?? "(anonymous)")}</span>
            <span class="fi-path">${esc((f.file ?? "").split("/").slice(-3).join("/"))}${f.line ? `:${f.line}` : ""}</span>
          </div>`).join("")}
      ` : ""}
    </aside>

    <main class="content" id="__ignite_content">
      ${renderHints(error)}

      <div id="__ignite_frames">
        ${appFrames.length > 0 ? appFrames.slice(0, 8).map((f, i) => renderFrame(f, i, editorScheme)).join("") : allFrames.slice(0, 4).map((f, i) => renderFrame(f, i, editorScheme)).join("")}
      </div>

      ${vendorFrames.length > 0 ? `
        <details class="vendor-toggle collapsible">
          <summary class="collapsible-summary">
            <svg width="12" height="12" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2.5" stroke-linecap="round" stroke-linejoin="round" aria-hidden="true"><polyline points="6 9 12 15 18 9"/></svg>
            ${vendorFrames.length} vendor frame${vendorFrames.length !== 1 ? "s" : ""}
          </summary>
          <div class="collapsible-body" style="padding:0">
            ${vendorFrames.slice(0, 10).map((f, i) => renderFrame(f, appFrames.length + i, editorScheme)).join("")}
          </div>
        </details>` : ""}

      ${renderRequestSection(req)}
      ${renderContextSection(process.env)}

      <div class="trazze-section">
        <div class="trazze-header">
          <span>Trace payload</span>
          <span class="trazze-badge">Trazze-ready</span>
        </div>
        <pre class="trazze-json" id="__ignite_json">${esc(tracePayload)}</pre>
      </div>
    </main>
  </div>

  <script>
    (function () {
      // Measure header height for sticky sidebar offset
      const header = document.getElementById('__ignite_header');
      if (header) {
        document.documentElement.style.setProperty('--header-h', header.offsetHeight + 'px');
      }

      // Sidebar frame navigation \u2014 scroll to and highlight the selected frame
      const sidebar = document.getElementById('__ignite_sidebar');
      const framesEl = document.getElementById('__ignite_frames');

      if (sidebar && framesEl) {
        sidebar.addEventListener('click', function (e) {
          const item = e.target.closest('[data-frame]');
          if (!item) return;

          const idx = parseInt(item.dataset.frame, 10);

          // Update active state
          sidebar.querySelectorAll('.frame-item').forEach(function (el) {
            el.classList.toggle('frame-item--active', el === item);
          });

          // Scroll the matching frame into view in the content area
          const target = framesEl.querySelector('[data-frame-idx="' + idx + '"]')
            || document.querySelector('[data-frame-idx="' + idx + '"]');
          if (target) target.scrollIntoView({ behavior: 'smooth', block: 'start' });
        });
      }

      // Copy trace JSON to clipboard
      const copyBtn = document.getElementById('__ignite_copy');
      const jsonEl  = document.getElementById('__ignite_json');
      if (copyBtn && jsonEl) {
        copyBtn.addEventListener('click', function () {
          navigator.clipboard.writeText(jsonEl.textContent || '').then(function () {
            copyBtn.textContent = 'Copied!';
            copyBtn.classList.add('copied');
            setTimeout(function () {
              copyBtn.textContent = 'Copy trace JSON';
              copyBtn.classList.remove('copied');
            }, 2000);
          });
        });
      }
    })();
  </script>
</body>
</html>`;
}

// src/push.ts
async function pushToTrazze(err, req) {
  const baseUrl = process.env["TRAZZE_URL"];
  const apiKey = process.env["TRAZZE_API_KEY"];
  if (!baseUrl || !apiKey) return;
  const env = process.env["APP_ENV"] ?? process.env["NODE_ENV"] ?? "unknown";
  const project = process.env["TRAZZE_PROJECT"] ?? "default";
  const frames = parseStack(err);
  const ts = (/* @__PURE__ */ new Date()).toISOString();
  const payload = {
    name: err.name,
    message: err.message,
    stack: err.stack,
    frames: frames.slice(0, 20),
    method: req.method,
    url: req.url,
    timestamp: ts
  };
  const envelope = {
    type: "error",
    timestamp: ts,
    project,
    environment: env,
    sdk: { name: "@trazze/ignite", version: "0.1.0" },
    data: payload
  };
  try {
    await fetch(`${baseUrl.replace(/\/$/, "")}/v1/ingest`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "x-api-key": apiKey
      },
      body: JSON.stringify(envelope)
    });
  } catch {
  }
}

// src/index.ts
function emitToTrazze(err, req) {
  const frames = parseStack(err);
  const entry = {
    level: "error",
    message: `[ignite] ${err.name}: ${err.message}`,
    timestamp: (/* @__PURE__ */ new Date()).toISOString(),
    context: {
      error: {
        name: err.name,
        message: err.message,
        stack: err.stack
      },
      request: {
        method: req.method,
        url: req.url
      },
      frames: frames.slice(0, 10).map((f) => ({
        fn: f.fn,
        file: f.file,
        line: f.line,
        column: f.column,
        isVendor: f.isVendor
      }))
    }
  };
  console.error(JSON.stringify(entry));
}
async function handleIgnitionError(err, req) {
  const isDev = process.env.NODE_ENV === "development" || process.env.APP_ENV === "development" || process.env.APP_ENV === "local";
  emitToTrazze(err, req);
  await pushToTrazze(err, req);
  if (!isDev) {
    return new Response("Internal Server Error", { status: 500 });
  }
  const html = renderIgnitionPage(err, req);
  return new Response(html, {
    status: 500,
    headers: { "Content-Type": "text/html; charset=utf-8" }
  });
}
export {
  buildEditorUrl,
  editorLabel,
  handleIgnitionError,
  pushToTrazze,
  resolveEditorScheme
};
