// @lumiarq/pack-laravel@1.0.1/dist/adapter.js
import fs from 'node:fs';
import path from 'node:path';
import { CURRENT_IR_VERSION } from '@lumiarq/context';
import { loadLaravelTemplate, resolveLaravelTemplateName } from './templates/index.js';
const LARAVEL_MIGRATION_INTENT_KINDS = new Set(['schema', 'migration', 'table']);
function toKebabCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
        .replace(/[^a-zA-Z0-9]+/g, '-')
        .replace(/^-+|-+$/g, '')
        .toLowerCase();
}
function getMigrationPostStepPath(projectRoot) {
    return path.join(projectRoot, '.lumis', 'poststeps', 'laravel-migrations.md');
}
function buildMigrationPostStep(intent, ir) {
    const subject = intent.subjects[0];
    const modelOrTableName = subject?.name ?? 'example';
    const resourceName = toKebabCase(modelOrTableName);
    const migrationName = `add_${resourceName}_changes`;
    const targetFile = getMigrationPostStepPath(ir.projectRoot);
    return {
        id: `poststep:migration:${resourceName}`,
        operation: 'append',
        artifactKind: 'migration-post-step',
        module: 'database',
        name: migrationName,
        targetFile,
        templateName: resolveLaravelTemplateName('migration'),
        tokens: {
            MigrationName: migrationName,
            ResourceName: resourceName,
        },
        dependsOn: [],
        requiresConfirmation: false,
        rollback: {
            operation: 'restore',
            targetFile,
        },
        rationale: `${intent.action} migration post-step for "${modelOrTableName}"`,
    };
}
function isMigrationKind(kind) {
    return LARAVEL_MIGRATION_INTENT_KINDS.has(kind);
}
function normalizePath(value) {
    return value.replaceAll('\\', '/').replace(/\/+/g, '/');
}
function isLaravelMigrationFilePath(targetFile) {
    const normalized = normalizePath(targetFile);
    return normalized.includes('/database/migrations/') && normalized.endsWith('.php');
}
const LARAVEL_VERIFICATION_HOOKS = [
    {
        name: 'laravel.prevent-generated-migrations',
        when: 'before-write',
        run: async (ctx) => {
            if (!isLaravelMigrationFilePath(ctx.step.targetFile)) {
                return { passed: true, guarantee: 'laravel.migration-contract' };
            }
            return {
                passed: false,
                guarantee: 'laravel.migration-contract',
                message: 'Laravel migration generation is blocked by contract. Use the generated post-step guide and run artisan make:migration manually.',
            };
        },
    },
];
export const LaravelAdapter = {
    name: 'laravel',
    version: '1.0.0',
    language: 'php',
    detect: async (projectRoot) => {
        const artisanPath = path.join(projectRoot, 'artisan');
        const composerPath = path.join(projectRoot, 'composer.json');
        const detected = fs.existsSync(artisanPath) || fs.existsSync(composerPath);
        return {
            detected,
            confidence: detected ? 0.9 : 0,
            reason: detected ? 'Laravel markers detected (artisan/composer.json)' : 'No Laravel markers detected',
        };
    },
    buildIR: async (projectRoot) => ({
        irVersion: CURRENT_IR_VERSION,
        version: '1.0',
        projectRoot,
        framework: 'laravel',
        builtAt: new Date(),
        modules: [],
        dependencies: [],
        violations: [],
        metadata: {
            migrationPolicy: 'manual-post-steps-only',
        },
    }),
    updateIR: async (ir) => ir,
    placementRules: [],
    dependencyRules: [],
    namingRules: [],
    resolveIntent: async (intent, ir) => {
        if (intent.action !== 'create' || intent.subjects.length === 0) {
            return [];
        }
        const kind = intent.subjects[0]?.kind.trim().toLowerCase();
        if (!kind || !isMigrationKind(kind)) {
            return [];
        }
        return [buildMigrationPostStep(intent, ir)];
    },
    resolveTemplate: (templateName) => loadLaravelTemplate(templateName),
    verificationHooks: LARAVEL_VERIFICATION_HOOKS,
    buildCommands: {
        typecheck: 'php artisan test --testsuite=Unit',
        build: 'php artisan optimize',
        test: 'php artisan test',
    },
    tinkerContext: async () => ({ framework: 'laravel', migrationPolicy: 'manual-post-steps-only' }),
};
