// @illumiarq/runtime@1.1.2/dist/storage/local-storage.js
import { createWriteStream } from 'node:fs';
import { mkdir, unlink, access, readFile } from 'node:fs/promises';
import { join, dirname } from 'node:path';
import { pipeline } from 'node:stream/promises';
import { Readable } from 'node:stream';
export class LocalStorage {
    root;
    baseUrl;
    constructor(opts) {
        this.root = opts.root;
        this.baseUrl = opts.baseUrl?.replace(/\/$/, '') ?? `file://${opts.root}`;
    }
    async put(path, file, options) {
        const full = join(this.root, path);
        await mkdir(dirname(full), { recursive: true });
        const readable = file instanceof Buffer
            ? Readable.from(file)
            : Readable.fromWeb(file);
        await pipeline(readable, createWriteStream(full));
        const size = file instanceof Buffer ? file.length : 0;
        const storedFile = {
            path,
            url: this.url(path),
            size,
            mimeType: options?.contentType ?? 'application/octet-stream',
        };
        if (options?.metadata !== undefined) {
            storedFile.metadata = options.metadata;
        }
        return storedFile;
    }
    async get(path) {
        try {
            return await readFile(join(this.root, path));
        }
        catch {
            return null;
        }
    }
    async delete(path) {
        try {
            await unlink(join(this.root, path));
        }
        catch {
            /* file may not exist */
        }
    }
    async exists(path) {
        try {
            await access(join(this.root, path));
            return true;
        }
        catch {
            return false;
        }
    }
    url(path) {
        return `${this.baseUrl}/${path.replace(/^\//, '')}`;
    }
    async temporaryUrl(path, _expiry) {
        // Local driver does not support signed URLs — return permanent URL.
        return this.url(path);
    }
}
