// @illumiarq/runtime@1.1.2/dist/storage/s3-storage.js
async function loadS3() {
    try {
        return (await import('@aws-sdk/client-s3'));
    }
    catch {
        throw new Error('[S3Storage] @aws-sdk/client-s3 is not installed. Run: pnpm add @aws-sdk/client-s3 @aws-sdk/s3-request-presigner');
    }
}
async function loadPresigner() {
    try {
        return (await import('@aws-sdk/s3-request-presigner'));
    }
    catch {
        throw new Error('[S3Storage] @aws-sdk/s3-request-presigner is not installed. Run: pnpm add @aws-sdk/s3-request-presigner');
    }
}
export class S3Storage {
    opts;
    client = null;
    sdk = null;
    constructor(opts) {
        this.opts = opts;
    }
    async getClient() {
        if (!this.client || !this.sdk) {
            this.sdk = await loadS3();
            this.client = new this.sdk.S3Client({
                region: this.opts.region,
                credentials: this.opts.credentials,
                ...(this.opts.endpoint ? { endpoint: this.opts.endpoint, forcePathStyle: true } : {}),
            });
        }
        return { client: this.client, sdk: this.sdk };
    }
    async put(path, file, options) {
        const { client, sdk } = await this.getClient();
        const body = file instanceof Buffer ? file : await streamToBuffer(file);
        await client.send(new sdk.PutObjectCommand({
            Bucket: this.opts.bucket,
            Key: path,
            Body: body,
            ContentType: options?.contentType ?? 'application/octet-stream',
            ACL: options?.visibility === 'public' ? 'public-read' : 'private',
            Metadata: options?.metadata,
        }));
        return {
            path,
            url: this.url(path),
            size: body.length,
            mimeType: options?.contentType ?? 'application/octet-stream',
            ...(options?.metadata ? { metadata: options.metadata } : {}),
        };
    }
    async get(path) {
        const { client, sdk } = await this.getClient();
        try {
            const result = (await client.send(new sdk.GetObjectCommand({ Bucket: this.opts.bucket, Key: path })));
            if (!result.Body)
                return null;
            const bytes = await result.Body.transformToByteArray();
            return Buffer.from(bytes);
        }
        catch {
            return null;
        }
    }
    async delete(path) {
        const { client, sdk } = await this.getClient();
        await client.send(new sdk.DeleteObjectCommand({ Bucket: this.opts.bucket, Key: path }));
    }
    async exists(path) {
        const { client, sdk } = await this.getClient();
        try {
            await client.send(new sdk.HeadObjectCommand({ Bucket: this.opts.bucket, Key: path }));
            return true;
        }
        catch {
            return false;
        }
    }
    url(path) {
        if (this.opts.publicBaseUrl) {
            return `${this.opts.publicBaseUrl.replace(/\/$/, '')}/${path.replace(/^\//, '')}`;
        }
        return `https://${this.opts.bucket}.s3.${this.opts.region}.amazonaws.com/${path}`;
    }
    async temporaryUrl(path, expiry) {
        const { client, sdk } = await this.getClient();
        const presigner = await loadPresigner();
        const expiresIn = Math.max(1, Math.floor((expiry.getTime() - Date.now()) / 1000));
        return presigner.getSignedUrl(client, new sdk.GetObjectCommand({ Bucket: this.opts.bucket, Key: path }), { expiresIn });
    }
}
async function streamToBuffer(stream) {
    const reader = stream.getReader();
    const chunks = [];
    for (;;) {
        const { done, value } = await reader.read();
        if (done)
            break;
        if (value)
            chunks.push(value);
    }
    return Buffer.concat(chunks);
}
