// @lumiarq/pack-lumiarq@1.0.1/dist/adapter.js
import { CURRENT_IR_VERSION } from '@lumiarq/context';
import path from 'node:path';
import { buildIR } from './ir/builder.js';
import { getLumiarqCommands } from './commands/index.js';
import { LUMIARQ_FRAMEWORK } from './frameworks/lumiarq.js';
import { lumiarqDependencyRules, lumiarqNamingRules, lumiarqPlacementRules } from './rules/index.js';
import { loadLumiarqTemplate } from './templates/index.js';
import { getLumiarqTinkerContext } from './tinker/index.js';
import { lumiarqVerificationHooks } from './verify/index.js';
const SUPPORTED_INTENT_KINDS = new Set([
    'action',
    'task',
    'query',
    'policy',
    'schema',
    'repository',
    'handler',
    'loader',
    'event',
    'listener',
    'validator',
    'dto',
]);
const KIND_CLASS_SUFFIX = {
    action: 'Action',
    task: 'Task',
    query: 'Query',
    policy: 'Policy',
    schema: 'Schema',
    repository: 'Repository',
    handler: 'Handler',
    loader: 'Loader',
    event: 'Event',
    listener: 'Listener',
    validator: 'Validator',
    dto: 'DTO',
};
/**
 * LumiARQ Adapter for Lumis — translates LumiARQ Porto architecture
 * into generic Lumis IR, plans, and execution model.
 *
 * This adapter enables Lumis to understand and safely execute changes
 * in LumiARQ projects by mapping Porto layer structure to generic artifacts.
 */
export const LumiarqAdapter = {
    name: 'lumiarq',
    version: '1.0.0',
    language: 'typescript',
    async detect(projectRoot) {
        try {
            const fs = await import('node:fs');
            const path = await import('node:path');
            // Check for LumiARQ markers
            const packageJsonPath = path.join(projectRoot, 'package.json');
            if (!fs.existsSync(packageJsonPath)) {
                return { detected: false, confidence: 0, reason: 'No package.json found' };
            }
            const pkgContent = fs.readFileSync(packageJsonPath, 'utf8');
            const pkg = JSON.parse(pkgContent);
            const hasFramework = !!(pkg?.dependencies ?? {})['@lumiarq/framework'] ||
                !!(pkg?.devDependencies ?? {})['@lumiarq/framework'];
            if (!hasFramework) {
                return {
                    detected: false,
                    confidence: 0,
                    reason: '@lumiarq/framework not in dependencies',
                };
            }
            // Check for Porto module structure
            const modulesPath = path.join(projectRoot, 'src/modules');
            const hasModules = fs.existsSync(modulesPath);
            return {
                detected: hasFramework && hasModules,
                confidence: hasFramework ? (hasModules ? 0.95 : 0.75) : 0,
                reason: 'LumiARQ framework detected with Porto modules',
            };
        }
        catch {
            return { detected: false, confidence: 0, reason: 'Detection failed' };
        }
    },
    async buildIR(projectRoot) {
        // Build the LumiARQ-specific IR
        const lumiarqIR = buildIR(projectRoot, LUMIARQ_FRAMEWORK);
        // Convert to generic Module[] and Artifact[] shape
        const modules = convertLumiarqModules(lumiarqIR);
        const artifacts = extractArtifacts(lumiarqIR);
        const dependencies = convertDependencies(lumiarqIR);
        const violations = dependencies.filter((d) => d.isViolation);
        return {
            irVersion: CURRENT_IR_VERSION,
            version: '1.0',
            projectRoot,
            framework: 'lumiarq',
            builtAt: new Date(),
            modules,
            dependencies,
            violations,
            metadata: {
                builtAt: new Date().toISOString(),
            },
        };
    },
    async updateIR(ir, changedFile) {
        // For now, rebuild full IR — incremental update comes in v1.1
        return this.buildIR(ir.projectRoot);
    },
    placementRules: lumiarqPlacementRules,
    dependencyRules: lumiarqDependencyRules,
    namingRules: lumiarqNamingRules,
    async resolveIntent(intent, ir) {
        if (intent.action !== 'create') {
            return [];
        }
        const primarySubject = intent.subjects[0];
        if (!primarySubject) {
            return [];
        }
        const artifactKind = normalizeArtifactKind(primarySubject.kind);
        if (!artifactKind || !SUPPORTED_INTENT_KINDS.has(artifactKind)) {
            return [];
        }
        const kindPath = LUMIARQ_FRAMEWORK.paths[artifactKind];
        if (!kindPath) {
            return [];
        }
        const moduleName = resolveIntentModule(primarySubject, ir);
        const resolvedName = resolveArtifactName(primarySubject.name, artifactKind);
        const normalizedName = normalizeArtifactDisplayName(resolvedName);
        const fileStem = toKebabCase(stripKindSuffix(resolvedName, artifactKind));
        const relativeTargetFile = `src/modules/${moduleName}/${trimSlashes(kindPath)}/${fileStem}.${artifactKind}.ts`;
        const targetFile = path.join(ir.projectRoot, relativeTargetFile);
        const step = {
            id: `create:${artifactKind}:${moduleName}:${fileStem}`,
            operation: 'create',
            artifactKind,
            module: moduleName,
            name: normalizedName,
            targetFile,
            templateName: artifactKind,
            tokens: {
                Name: normalizedName,
                ModuleName: moduleName,
                ArtifactKind: artifactKind,
                FileStem: fileStem,
                ClassName: buildClassName(resolvedName, artifactKind),
            },
            dependsOn: [],
            requiresConfirmation: false,
            rollback: {
                operation: 'delete',
                targetFile,
            },
            rationale: `${intent.action} ${artifactKind} "${normalizedName}" in module ${moduleName}`,
        };
        return [step];
    },
    resolveTemplate(templateName) {
        return loadLumiarqTemplate(templateName);
    },
    verificationHooks: lumiarqVerificationHooks,
    buildCommands: {
        typecheck: 'pnpm typecheck',
        lint: 'pnpm lint',
        boundaryCheck: 'pnpm turbo boundary:check',
        build: 'pnpm build',
        test: 'pnpm test',
    },
    tinkerContext: getLumiarqTinkerContext,
    commands: getLumiarqCommands,
};
/**
 * Normalize free-form intent subject kind into a canonical LumiARQ artifact kind.
 *
 * @param rawKind - Raw kind token from parsed intent.
 * @returns Canonical kind when recognized, otherwise null.
 */
function normalizeArtifactKind(rawKind) {
    const lowered = rawKind.trim().toLowerCase();
    if (!lowered) {
        return null;
    }
    const aliases = {
        actions: 'action',
        tasks: 'task',
        queries: 'query',
        policies: 'policy',
        schemas: 'schema',
        repositories: 'repository',
        handlers: 'handler',
        loaders: 'loader',
        events: 'event',
        listeners: 'listener',
        validators: 'validator',
        dtos: 'dto',
    };
    return aliases[lowered] ?? lowered;
}
/**
 * Resolve module name from parsed intent and IR fallback context.
 *
 * @param subject - Parsed primary subject from intent.
 * @param ir - Current IR used for fallback module inference.
 * @returns Module name to use for generated step paths.
 */
function resolveIntentModule(subject, ir) {
    const candidate = subject.attributes?.module;
    if (typeof candidate === 'string' && candidate.trim().length > 0) {
        return candidate.trim();
    }
    return ir.modules[0]?.name ?? 'Core';
}
/**
 * Resolve artifact display name from optional subject name and kind defaults.
 *
 * @param rawName - Optional name parsed from the user intent.
 * @param kind - Canonical artifact kind.
 * @returns A non-empty artifact display name.
 */
function resolveArtifactName(rawName, kind) {
    if (rawName && rawName.trim().length > 0) {
        return rawName.trim();
    }
    return `New${toPascalCase(kind)}${KIND_CLASS_SUFFIX[kind] ?? ''}`;
}
/**
 * Normalize a generated artifact name for user-facing plan output.
 *
 * @param rawName - Raw artifact name candidate.
 * @returns Trimmed non-empty display name.
 */
function normalizeArtifactDisplayName(rawName) {
    return rawName.trim() || 'UnnamedArtifact';
}
/**
 * Remove the artifact kind suffix from a class-like artifact name when present.
 *
 * @param rawName - Artifact name potentially containing a kind suffix.
 * @param kind - Canonical artifact kind.
 * @returns Name without trailing kind suffix when matched.
 */
function stripKindSuffix(rawName, kind) {
    const suffix = KIND_CLASS_SUFFIX[kind];
    if (!suffix) {
        return rawName;
    }
    return rawName.endsWith(suffix) ? rawName.slice(0, rawName.length - suffix.length) || rawName : rawName;
}
/**
 * Build a class name token that follows LumiARQ naming style.
 *
 * @param rawName - Source artifact name from intent.
 * @param kind - Canonical artifact kind.
 * @returns Class-like name for templates.
 */
function buildClassName(rawName, kind) {
    const suffix = KIND_CLASS_SUFFIX[kind] ?? '';
    const base = toPascalCase(stripKindSuffix(rawName, kind));
    return `${base}${suffix}`;
}
/**
 * Convert free-form text to kebab-case.
 *
 * @param value - Free-form string.
 * @returns Kebab-cased value.
 */
function toKebabCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
        .replace(/[\s_]+/g, '-')
        .replace(/[^a-zA-Z0-9-]/g, '-')
        .replace(/-+/g, '-')
        .replace(/^-|-$/g, '')
        .toLowerCase();
}
/**
 * Convert free-form text to PascalCase.
 *
 * @param value - Free-form string.
 * @returns PascalCased value.
 */
function toPascalCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1 $2')
        .split(/[^a-zA-Z0-9]+/)
        .filter(Boolean)
        .map((part) => part.charAt(0).toUpperCase() + part.slice(1).toLowerCase())
        .join('');
}
/**
 * Trim leading and trailing slashes for safe path concatenation.
 *
 * @param value - Path segment possibly containing edge slashes.
 * @returns Path segment without leading/trailing slashes.
 */
function trimSlashes(value) {
    return value.replace(/^\/+|\/+$/g, '');
}
// ─────────────────────────────────────────────────────────────────
// Conversion Helpers
// ─────────────────────────────────────────────────────────────────
function convertLumiarqModules(lumiarqIR) {
    return lumiarqIR.modules.map((m) => ({
        name: m.name,
        directory: m.directory,
        artifacts: extractModuleArtifacts(lumiarqIR, m.name),
        publicExports: m.exports.actions
            .concat(m.exports.queries, m.exports.events, m.exports.types, m.exports.models)
            .map((name) => `${m.name}:${name}`),
        metadata: {
            portoModule: true,
            exports: m.exports,
        },
    }));
}
function extractModuleArtifacts(lumiarqIR, moduleName) {
    const moduleRecord = lumiarqIR.modules.find((m) => m.name === moduleName);
    if (!moduleRecord)
        return [];
    const artifacts = [];
    const layers = moduleRecord.layers;
    // HTTP layer
    if (layers.http) {
        for (const h of layers.http.handlers) {
            artifacts.push({
                id: `${moduleName}:handler:${h.name}`,
                kind: 'handler',
                name: h.name,
                file: h.file,
                module: moduleName,
                metadata: { action: h.action },
            });
        }
    }
    // Database layer
    if (layers.database) {
        for (const s of layers.database.schemas) {
            artifacts.push({
                id: `${moduleName}:schema:${s.entity}`,
                kind: 'schema',
                name: s.entity,
                file: s.file,
                module: moduleName,
                metadata: { entity: s.entity },
            });
        }
        for (const r of layers.database.repositories) {
            artifacts.push({
                id: `${moduleName}:repository:${r.name}`,
                kind: 'repository',
                name: r.name,
                file: r.file,
                module: moduleName,
                metadata: { entity: r.entity },
            });
        }
    }
    // Logic layer
    if (layers.logic) {
        for (const a of layers.logic.actions) {
            artifacts.push({
                id: `${moduleName}:action:${a.name}`,
                kind: 'action',
                name: a.name,
                file: a.file,
                module: moduleName,
                metadata: { input: a.input },
            });
        }
        for (const t of layers.logic.tasks) {
            artifacts.push({
                id: `${moduleName}:task:${t.name}`,
                kind: 'task',
                name: t.name,
                file: t.file,
                module: moduleName,
                metadata: { input: t.input },
            });
        }
        for (const q of layers.logic.queries) {
            artifacts.push({
                id: `${moduleName}:query:${q.name}`,
                kind: 'query',
                name: q.name,
                file: q.file,
                module: moduleName,
                metadata: { repository: q.repository },
            });
        }
    }
    // Events layer
    if (layers.events) {
        for (const e of layers.events.definitions) {
            artifacts.push({
                id: `${moduleName}:event:${e.name}`,
                kind: 'event',
                name: e.name,
                file: e.file,
                module: moduleName,
                metadata: { schema: e.schema },
            });
        }
    }
    // Contracts layer
    if (layers.contracts) {
        for (const t of layers.contracts.types) {
            artifacts.push({
                id: `${moduleName}:${t.kind}:${t.name}`,
                kind: t.kind,
                name: t.name,
                file: t.file,
                module: moduleName,
                metadata: {},
            });
        }
    }
    return artifacts;
}
function extractArtifacts(lumiarqIR) {
    const artifacts = [];
    for (const module of lumiarqIR.modules) {
        artifacts.push(...extractModuleArtifacts(lumiarqIR, module.name));
    }
    return artifacts;
}
function convertDependencies(lumiarqIR) {
    const deps = [];
    // Extract from dependency graph
    for (const [moduleName, moduleData] of Object.entries(lumiarqIR.dependencies.modules)) {
        for (const imp of moduleData.imports) {
            deps.push({
                from: `${moduleName}:*`,
                to: imp.module,
                kind: 'import',
                isViolation: false,
            });
        }
    }
    // Add violations
    for (const violation of lumiarqIR.dependencies.violations) {
        deps.push({
            from: violation.from,
            to: violation.to,
            kind: 'violation',
            isViolation: true,
            violatedRule: violation.type,
        });
    }
    return deps;
}
