// @lumiarq/pack-lumiarq@1.0.1/dist/templates/index.js
/**
 * Framework templates for LumiARQ artifacts.
 *
 * Template payloads are organised as stub-style modules to mirror the
 * `framework/packages/lumis/src/stubs/make` pattern while keeping the pack
 * self-contained for standalone builds.
 */
import { createTemplateStubLoader, resolveDirectoryFromImportMeta } from '@lumiarq/context';
const loader = createTemplateStubLoader({
    namespace: 'LumiARQ',
    stubsDir: resolveDirectoryFromImportMeta(import.meta.url, 'stubs'),
    templateToStubFile: {
        action: 'action.stub',
        task: 'task.stub',
        query: 'query.stub',
        policy: 'policy.stub',
        repository: 'repository.stub',
        handler: 'handler.stub',
        loader: 'loader.stub',
        listener: 'listener.stub',
        event: 'event.stub',
        dto: 'dto.stub',
        schema: 'schema.stub',
        validator: 'validator.stub',
    },
});
export function listLumiarqTemplateNames() {
    return loader.listTemplateNames();
}
/**
 * Resolve a LumiARQ template by artifact kind.
 *
 * @param templateName - Canonical artifact kind used as the template lookup key.
 * @returns Source template string for the requested artifact kind.
 * @throws Error when the requested template is not supported by the pack.
 */
export function loadLumiarqTemplate(templateName) {
    return loader.loadTemplate(templateName);
}
