// @lumiarq/rag@0.1.0/dist/retriever/index.js
// ── Vector Retriever ──────────────────────────────────────────────────────────
/**
 * Pure dense vector retrieval. Embeds the query and searches the VectorStore.
 */
export class VectorRetriever {
    embedder;
    store;
    constructor(embedder, store) {
        this.embedder = embedder;
        this.store = store;
    }
    async retrieve(query, opts) {
        const [vector] = await this.embedder.embed([query]);
        if (!vector)
            return [];
        const results = await this.store.search(vector, opts);
        return results.map((r) => ({ ...r, source: 'vector', vectorScore: r.score }));
    }
}
// ── Hybrid Retriever ──────────────────────────────────────────────────────────
/**
 * Combines dense vector retrieval and sparse BM25 retrieval using
 * Reciprocal Rank Fusion (RRF) with a configurable vector alpha weight.
 *
 * `hybridAlpha = 1.0` → pure vector
 * `hybridAlpha = 0.0` → pure BM25
 * `hybridAlpha = 0.6` → default balanced blend
 */
export class HybridRetriever {
    embedder;
    store;
    bm25;
    vectorRetriever;
    constructor(embedder, store, bm25) {
        this.embedder = embedder;
        this.store = store;
        this.bm25 = bm25;
        this.vectorRetriever = new VectorRetriever(embedder, store);
    }
    /** Sync helper to register a chunk in the BM25 index. Call after each upsert. */
    indexChunk(chunk) {
        this.bm25.upsert(chunk.id, chunk.text);
    }
    /** Batch-sync a set of chunks into the BM25 index. Called automatically by the pipeline on ingest. */
    syncChunks(chunks) {
        for (const chunk of chunks) {
            this.bm25.upsert(chunk.id, chunk.text);
        }
    }
    /** Remove a chunk from the BM25 index. Call after each delete. */
    removeChunk(id) {
        this.bm25.delete(id);
    }
    async retrieve(query, opts) {
        const topK = opts?.topK ?? 10;
        const alpha = opts?.hybridAlpha ?? 0.6;
        const fetchK = topK * 3; // fetch more before fusion
        // Run vector and BM25 in parallel
        const [vectorResults, bm25Results] = await Promise.all([
            this.vectorRetriever.retrieve(query, { ...opts, topK: fetchK }),
            Promise.resolve(this.bm25.search(query, fetchK)),
        ]);
        // Build score maps
        const vectorMap = new Map();
        for (const r of vectorResults)
            vectorMap.set(r.chunk.id, r.score);
        const bm25Map = new Map();
        for (const r of bm25Results)
            bm25Map.set(r.id, r.score);
        // Collect all unique chunk IDs seen in either result set
        const allIds = new Set([...vectorMap.keys(), ...bm25Map.keys()]);
        // Normalise BM25 scores to [0, 1]
        const maxBM25 = Math.max(...bm25Map.values(), 1);
        // Build chunk map from vector results
        const chunkMap = new Map();
        for (const r of vectorResults)
            chunkMap.set(r.chunk.id, r.chunk);
        // Fetch any BM25-only hits (not in vector results) from the store
        const missingIds = [...allIds].filter((id) => !chunkMap.has(id));
        if (missingIds.length > 0) {
            const fetched = await this.store.getByIds(missingIds);
            for (const chunk of fetched)
                chunkMap.set(chunk.id, chunk);
        }
        const fused = [];
        for (const id of allIds) {
            const vectorScore = vectorMap.get(id) ?? 0;
            const bm25Score = (bm25Map.get(id) ?? 0) / maxBM25;
            const fusedScore = alpha * vectorScore + (1 - alpha) * bm25Score;
            const chunk = chunkMap.get(id);
            if (!chunk)
                continue; // chunk not found in store (deleted or never stored)
            if (opts?.filter && chunk.metadata && !opts.filter(chunk.metadata))
                continue;
            if (opts?.minScore && fusedScore < opts.minScore)
                continue;
            fused.push({
                chunk,
                score: fusedScore,
                source: 'hybrid',
                vectorScore,
                bm25Score,
            });
        }
        return fused.sort((a, b) => b.score - a.score).slice(0, topK);
    }
}
