// @lumiarq/pack-nextjs@1.0.1/dist/templates/index.js
import { createTemplateStubLoader, resolveDirectoryFromImportMeta } from '@lumiarq/context';
const NEXTJS_COMMAND_TEMPLATE_MAP = {
    page: 'nextjs-page',
    api: 'nextjs-api-route',
    action: 'nextjs-server-action',
    service: 'nextjs-domain-service',
};
const loader = createTemplateStubLoader({
    namespace: 'Nextjs',
    stubsDir: resolveDirectoryFromImportMeta(import.meta.url, 'stubs'),
    projectOverrideDirs: ['stubs/lumis/nextjs', 'stubs/lumis'],
    templateToStubFile: {
        'nextjs-page': 'nextjs-page.stub',
        'nextjs-api-route': 'nextjs-api-route.stub',
        'nextjs-server-action': 'nextjs-server-action.stub',
        'nextjs-domain-service': 'nextjs-domain-service.stub',
    },
});
export function loadNextjsTemplate(templateName) {
    return loader.loadTemplate(templateName);
}
/**
 * Resolve the template name used for a normalized Next.js artifact kind.
 *
 * @param kind - Canonical artifact kind resolved during intent planning.
 * @returns Template catalog key used by `loadNextjsTemplate`.
 */
export function resolveNextjsTemplateName(kind) {
    return NEXTJS_COMMAND_TEMPLATE_MAP[kind];
}
export function listNextjsTemplateNames() {
    return loader.listTemplateNames();
}
