// @lumiarq/context@1.0.1/dist/persistence/ir-version.js
export const CURRENT_IR_VERSION = '1.0.0';
/**
 * Normalize legacy IR version aliases into semver-like form.
 * - "1.0" -> "1.0.0"
 */
function normalizeVersion(version) {
    const parts = version.split('.');
    if (parts.length === 2) {
        return `${version}.0`;
    }
    return version;
}
function majorOf(version) {
    const normalized = normalizeVersion(version);
    const major = Number.parseInt(normalized.split('.')[0] ?? '0', 10);
    return Number.isFinite(major) ? major : 0;
}
/**
 * Ensure persisted IR can be consumed by the current runtime.
 * Returns null for incompatible major versions.
 */
export function migratePersistedIR(raw) {
    const detected = raw.irVersion ?? raw.version;
    if (!detected) {
        return null;
    }
    const normalized = normalizeVersion(detected);
    if (majorOf(normalized) !== majorOf(CURRENT_IR_VERSION)) {
        return null;
    }
    if (raw.irVersion === normalized) {
        return raw;
    }
    return {
        ...raw,
        irVersion: normalized,
        version: normalized,
        metadata: {
            ...raw.metadata,
            migratedFromIrVersion: detected,
        },
    };
}
/**
 * Apply the current canonical IR version on writes.
 */
export function stampCurrentIRVersion(ir) {
    return {
        ...ir,
        irVersion: CURRENT_IR_VERSION,
        // Keep legacy field for backward compatibility with older readers.
        version: CURRENT_IR_VERSION,
    };
}
