// @lumiarq/prettier-plugin-veil@1.0.0/dist/index.js
/**
 * @lumiarq/prettier-plugin-veil
 *
 * A Prettier plugin that understands Veil template syntax (.veil.html).
 * Register in .prettierrc:
 *
 *   { "plugins": ["@lumiarq/prettier-plugin-veil"] }
 *
 * Prettier will automatically apply this plugin to any file whose name ends
 * in `.veil.html` because we register the `veil-html` parser with that glob.
 */
import { parseVeil } from './veil-parser.js';
import { printVeil } from './veil-printer.js';
// ── Parser ────────────────────────────────────────────────────────────────────
const veilParser = {
    astFormat: 'veil-ast',
    locStart: (node) => node.start,
    locEnd: (node) => node.end,
    parse(text) {
        const nodes = parseVeil(text);
        return {
            type: 'veil-root',
            nodes,
            start: 0,
            end: text.length,
        };
    },
};
// ── Printer ───────────────────────────────────────────────────────────────────
const veilPrinter = {
    print(path, options) {
        const node = path.node;
        if (node.type !== 'veil-root')
            return '';
        const formatted = printVeil(node.nodes, {
            tabWidth: options.tabWidth ?? 2,
            useTabs: options.useTabs ?? false,
            printWidth: options.printWidth ?? 120,
        });
        // Return as a Prettier Doc string — for a text-only printer this is valid
        return formatted;
    },
};
// ── Options ───────────────────────────────────────────────────────────────────
const options = {
    veilSortTailwindClasses: {
        type: 'boolean',
        category: 'Veil',
        default: true,
        description: 'Sort Tailwind utility classes in Veil templates. Requires prettier-plugin-tailwindcss to be installed.',
    },
};
// ── Plugin export ─────────────────────────────────────────────────────────────
const plugin = {
    languages: [
        {
            name: 'Veil',
            parsers: ['veil-html'],
            extensions: ['.veil.html'],
            vscodeLanguageIds: ['html'],
        },
    ],
    parsers: {
        'veil-html': veilParser,
    },
    printers: {
        'veil-ast': veilPrinter,
    },
    options,
};
export default plugin;
export { parseVeil, printVeil };
