// @lumiarq/rag@0.1.0/dist/adapters/store/base.js
export class BaseVectorStore {
    upsertBatchSize;
    filterOverfetchMultiplier;
    deleteBatchSize;
    /**
     * Set to true in subclasses that pass the filter predicate natively to the backend.
     * When false, the base applies it in-process after fetching more results.
     */
    nativeFilter = false;
    constructor(opts = {}) {
        this.upsertBatchSize = opts.upsertBatchSize ?? 200;
        this.filterOverfetchMultiplier = opts.filterOverfetchMultiplier ?? 5;
        this.deleteBatchSize = opts.deleteBatchSize ?? 500;
    }
    // ── VectorStore interface ─────────────────────────────────────────────────
    async upsert(docs) {
        for (let offset = 0; offset < docs.length; offset += this.upsertBatchSize) {
            await this.doUpsert(docs.slice(offset, offset + this.upsertBatchSize));
        }
    }
    async delete(ids) {
        for (let offset = 0; offset < ids.length; offset += this.deleteBatchSize) {
            await this.doDelete(ids.slice(offset, offset + this.deleteBatchSize));
        }
    }
    async search(vector, opts) {
        const topK = opts?.topK ?? 10;
        const minScore = opts?.minScore ?? 0;
        const filter = opts?.filter;
        if (!filter || this.nativeFilter) {
            // Backend handles filtering (or there's no filter): direct call.
            return this.doSearch(vector, topK, minScore);
        }
        // In-process filter: over-fetch, filter, trim to topK.
        const overFetch = topK * this.filterOverfetchMultiplier;
        const raw = await this.doSearch(vector, overFetch, minScore);
        const filtered = raw.filter((r) => r.chunk.metadata === undefined || filter(r.chunk.metadata));
        return filtered.slice(0, topK);
    }
    async getIndexedIds() {
        return this.doGetIndexedIds();
    }
    async lastIndexedAt(documentId) {
        return this.doLastIndexedAt(documentId);
    }
    async getByIds(ids) {
        if (ids.length === 0)
            return [];
        return this.doGetByIds(ids);
    }
    async getContentHash(documentId) {
        return this.doGetContentHash(documentId);
    }
}
