// @lumiarq/rag@0.1.0/dist/adapters/store/pg.js
import { BaseVectorStore } from './base.js';
import { buildCreateTableSql, buildDeleteSql, buildGetByIdsSql, buildGetContentHashSql, buildGetIdsSql, buildLastIndexedAtSql, buildSearchSql, buildUpsertSql, resolveConfig, } from './sql-base.js';
export class PgVectorStore extends BaseVectorStore {
    client;
    cfg;
    dimensions;
    constructor(opts) {
        super(opts);
        this.client = opts.client;
        this.cfg = resolveConfig(opts);
        this.dimensions = opts.dimensions;
    }
    /** Create the chunks table and indexes. Idempotent. */
    async createTable() {
        const sql = buildCreateTableSql(this.cfg, this.dimensions);
        // Split on ; to run each statement separately (pg driver handles one at a time).
        for (const stmt of sql.split(';').map((s) => s.trim()).filter(Boolean)) {
            await this.client.query(stmt);
        }
    }
    async doUpsert(docs) {
        const batch = docs.map((doc) => ({
            id: doc.id,
            documentId: doc.documentId,
            text: doc.text,
            chunkIndex: doc.index,
            startChar: doc.startChar,
            endChar: doc.endChar,
            ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
            ...(doc.contentHash !== undefined ? { contentHash: doc.contentHash } : {}),
            vector: doc.vector,
        }));
        const { sql, values } = buildUpsertSql(this.cfg, batch);
        await this.client.query(sql, values);
    }
    async doSearch(vector, topK, minScore) {
        const { sql } = buildSearchSql(this.cfg, topK);
        const vectorLiteral = `[${vector.join(',')}]`;
        const { rows } = await this.client.query(sql, [vectorLiteral, minScore, topK]);
        return rows.map((row) => ({
            chunk: {
                id: row.id,
                documentId: row.document_id,
                text: row.text,
                index: row.chunk_index,
                startChar: row.start_char,
                endChar: row.end_char,
                ...(row.metadata !== null ? { metadata: row.metadata } : {}),
            },
            score: typeof row.score === 'number' ? row.score : Number(row.score),
        }));
    }
    async doDelete(ids) {
        const { sql } = buildDeleteSql(this.cfg, ids.length);
        await this.client.query(sql, ids);
    }
    async doGetIndexedIds() {
        const sql = buildGetIdsSql(this.cfg);
        const { rows } = await this.client.query(sql);
        return rows.map((r) => r.id);
    }
    async doLastIndexedAt(documentId) {
        const sql = buildLastIndexedAtSql(this.cfg);
        const { rows } = await this.client.query(sql, [documentId]);
        const ts = rows[0]?.ts;
        return ts ? new Date(ts) : null;
    }
    async doGetByIds(ids) {
        const { sql } = buildGetByIdsSql(this.cfg, ids.length);
        const { rows } = await this.client.query(sql, ids);
        return rows.map((row) => ({
            id: row.id,
            documentId: row.document_id,
            text: row.text,
            index: row.chunk_index,
            startChar: row.start_char,
            endChar: row.end_char,
            ...(row.metadata !== null ? { metadata: row.metadata } : {}),
        }));
    }
    async doGetContentHash(documentId) {
        const sql = buildGetContentHashSql(this.cfg);
        const { rows } = await this.client.query(sql, [documentId]);
        return rows[0]?.content_hash ?? null;
    }
}
