// @lumiarq/rag@0.1.0/dist/adapters/store/qdrant.js
import { BaseVectorStore } from './base.js';
export class QdrantVectorStore extends BaseVectorStore {
    baseUrl;
    collection;
    dimensions;
    distance;
    headers;
    constructor(opts) {
        super(opts);
        this.baseUrl = (opts.url ?? 'http://localhost:6333').replace(/\/$/, '');
        this.collection = opts.collectionName;
        this.dimensions = opts.dimensions;
        this.distance = opts.distance ?? 'Cosine';
        this.headers = {
            'Content-Type': 'application/json',
            ...(opts.apiKey ? { 'api-key': opts.apiKey } : {}),
        };
    }
    /** Create the Qdrant collection. Idempotent (ignores 409 conflicts). */
    async createCollection() {
        const url = `${this.baseUrl}/collections/${this.collection}`;
        const res = await fetch(url, {
            method: 'PUT',
            headers: this.headers,
            body: JSON.stringify({
                vectors: {
                    size: this.dimensions,
                    distance: this.distance,
                },
            }),
        });
        if (!res.ok && res.status !== 409) {
            throw new Error(`QdrantVectorStore.createCollection: HTTP ${res.status} ${await res.text()}`);
        }
    }
    async fetchJson(method, path, body) {
        const res = await fetch(`${this.baseUrl}${path}`, {
            method,
            headers: this.headers,
            ...(body !== undefined ? { body: JSON.stringify(body) } : {}),
        });
        if (!res.ok) {
            throw new Error(`QdrantVectorStore fetch error: HTTP ${res.status} ${await res.text()}`);
        }
        return res.json();
    }
    async doUpsert(docs) {
        const points = docs.map((doc) => ({
            id: doc.id,
            vector: doc.vector,
            payload: {
                document_id: doc.documentId,
                text: doc.text,
                chunk_index: doc.index,
                start_char: doc.startChar,
                end_char: doc.endChar,
                metadata: doc.metadata ?? null,
                content_hash: doc.contentHash ?? null,
                last_indexed_at: new Date().toISOString(),
            },
        }));
        await this.fetchJson('PUT', `/collections/${this.collection}/points`, {
            points,
            wait: true,
        });
    }
    async doSearch(vector, topK, minScore) {
        const result = await this.fetchJson('POST', `/collections/${this.collection}/points/search`, {
            vector,
            limit: topK,
            score_threshold: minScore,
            with_payload: true,
        });
        return result.result.map((pt) => ({
            chunk: {
                id: pt.id,
                documentId: pt.payload.document_id,
                text: pt.payload.text,
                index: pt.payload.chunk_index,
                startChar: pt.payload.start_char,
                endChar: pt.payload.end_char,
                ...(pt.payload.metadata !== null ? { metadata: pt.payload.metadata } : {}),
            },
            score: pt.score,
        }));
    }
    async doDelete(ids) {
        await this.fetchJson('POST', `/collections/${this.collection}/points/delete`, {
            points: ids,
            wait: true,
        });
    }
    async doGetIndexedIds() {
        // Qdrant scroll API — fetch all IDs without vectors.
        const all = [];
        let nextOffset = undefined;
        do {
            const body = {
                limit: 250,
                with_payload: false,
                with_vector: false,
            };
            if (nextOffset !== undefined)
                body['offset'] = nextOffset;
            const result = await this.fetchJson('POST', `/collections/${this.collection}/points/scroll`, body);
            for (const pt of result.result)
                all.push(pt.id);
            nextOffset = result.next_page_offset;
        } while (nextOffset);
        return all;
    }
    async doLastIndexedAt(documentId) {
        // Qdrant doesn't have a MAX aggregate; scroll with filter and find the latest.
        const result = await this.fetchJson('POST', `/collections/${this.collection}/points/scroll`, {
            filter: {
                must: [{ key: 'document_id', match: { value: documentId } }],
            },
            limit: 10_000,
            with_payload: ['last_indexed_at'],
            with_vector: false,
        });
        const timestamps = result.result
            .map((pt) => pt.payload.last_indexed_at)
            .filter((ts) => typeof ts === 'string')
            .map((ts) => new Date(ts));
        const first = timestamps[0];
        if (!first)
            return null;
        return timestamps.reduce((max, d) => (d > max ? d : max), first);
    }
    async doGetByIds(ids) {
        const result = await this.fetchJson('POST', `/collections/${this.collection}/points`, {
            ids,
            with_payload: true,
            with_vector: false,
        });
        return result.result.map((pt) => ({
            id: pt.id,
            documentId: pt.payload.document_id,
            text: pt.payload.text,
            index: pt.payload.chunk_index,
            startChar: pt.payload.start_char,
            endChar: pt.payload.end_char,
            ...(pt.payload.metadata !== null ? { metadata: pt.payload.metadata } : {}),
        }));
    }
    async doGetContentHash(documentId) {
        // Fetch first chunk for this document and return its content_hash payload field.
        const result = await this.fetchJson('POST', `/collections/${this.collection}/points/scroll`, {
            filter: {
                must: [{ key: 'document_id', match: { value: documentId } }],
            },
            limit: 1,
            with_payload: ['content_hash'],
            with_vector: false,
        });
        const hash = result.result[0]?.payload?.content_hash;
        return typeof hash === 'string' ? hash : null;
    }
}
