// @lumiarq/rag@0.1.0/dist/adapters/store/sql-base.js
/**
 * SqlVectorStoreBase — shared pgvector SQL string builders.
 *
 * This class is NOT a VectorStore itself. It is a pure SQL-builder mixin that
 * both PgVectorStore and SupabaseVectorStore compose. It has zero DB client
 * dependencies — it only generates strings.
 *
 * Why separate from BaseVectorStore:
 *   - PgVectorStore uses the `pg` pool directly
 *   - SupabaseVectorStore uses `@supabase/supabase-js` RPC
 *   Both need the same SQL, but neither should inherit from the other.
 */
/**
 * Generates the DDL statement to create the chunks table.
 * Call once during setup/migration.
 */
export function buildCreateTableSql(cfg, dimensions) {
    return `
CREATE TABLE IF NOT EXISTS ${cfg.schema}.${cfg.table} (
  id             TEXT        PRIMARY KEY,
  document_id    TEXT        NOT NULL,
  text           TEXT        NOT NULL,
  chunk_index    INTEGER     NOT NULL,
  start_char     INTEGER     NOT NULL,
  end_char       INTEGER     NOT NULL,
  metadata       JSONB,
  content_hash   TEXT,
  last_indexed_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
  ${cfg.embeddingColumn} vector(${dimensions}) NOT NULL
);
CREATE INDEX IF NOT EXISTS ${cfg.table}_doc_id_idx ON ${cfg.schema}.${cfg.table} (document_id);
CREATE INDEX IF NOT EXISTS ${cfg.table}_embedding_idx ON ${cfg.schema}.${cfg.table}
  USING hnsw (${cfg.embeddingColumn} vector_cosine_ops);
`.trim();
}
/**
 * Generates a parameterised UPSERT statement for a single batch.
 * Returns `{ sql, values }` ready for a `pg` client's `query(sql, values)`.
 *
 * Uses `ON CONFLICT (id) DO UPDATE` so re-indexing is idempotent.
 */
export function buildUpsertSql(cfg, batch) {
    const values = [];
    const rowPlaceholders = [];
    let p = 1;
    for (const row of batch) {
        rowPlaceholders.push(`($${p++}, $${p++}, $${p++}, $${p++}, $${p++}, $${p++}, $${p++}, $${p++}, NOW(), $${p++}::vector)`);
        values.push(row.id, row.documentId, row.text, row.chunkIndex, row.startChar, row.endChar, row.metadata !== undefined ? JSON.stringify(row.metadata) : null, row.contentHash ?? null, `[${row.vector.join(',')}]`);
    }
    const sql = `
INSERT INTO ${cfg.schema}.${cfg.table}
  (id, document_id, text, chunk_index, start_char, end_char, metadata, content_hash, last_indexed_at, ${cfg.embeddingColumn})
VALUES ${rowPlaceholders.join(', ')}
ON CONFLICT (id) DO UPDATE SET
  document_id     = EXCLUDED.document_id,
  text            = EXCLUDED.text,
  chunk_index     = EXCLUDED.chunk_index,
  start_char      = EXCLUDED.start_char,
  end_char        = EXCLUDED.end_char,
  metadata        = EXCLUDED.metadata,
  content_hash    = EXCLUDED.content_hash,
  last_indexed_at = EXCLUDED.last_indexed_at,
  ${cfg.embeddingColumn} = EXCLUDED.${cfg.embeddingColumn}
`.trim();
    return { sql, values };
}
/**
 * Generates a parameterised cosine-distance search query.
 */
export function buildSearchSql(cfg, topK) {
    // 1 - cosine_distance = cosine_similarity
    const sql = `
SELECT id, document_id, text, chunk_index, start_char, end_char, metadata, last_indexed_at,
       1 - (${cfg.embeddingColumn} <=> $1::vector) AS score
FROM   ${cfg.schema}.${cfg.table}
WHERE  1 - (${cfg.embeddingColumn} <=> $1::vector) >= $2
ORDER  BY ${cfg.embeddingColumn} <=> $1::vector
LIMIT  $3
`.trim();
    return { sql };
}
/**
 * Generates a DELETE statement for a list of chunk IDs.
 */
export function buildDeleteSql(cfg, count) {
    const placeholders = Array.from({ length: count }, (_, i) => `$${i + 1}`).join(', ');
    const sql = `DELETE FROM ${cfg.schema}.${cfg.table} WHERE id IN (${placeholders})`;
    return { sql };
}
/** Generates a SELECT for all stored chunk IDs. */
export function buildGetIdsSql(cfg) {
    return `SELECT id FROM ${cfg.schema}.${cfg.table}`;
}
/** Generates a parameterised SELECT for chunks by their IDs. */
export function buildGetByIdsSql(cfg, count) {
    const placeholders = Array.from({ length: count }, (_, i) => `$${i + 1}`).join(', ');
    const sql = `
SELECT id, document_id, text, chunk_index, start_char, end_char, metadata, content_hash, last_indexed_at
FROM   ${cfg.schema}.${cfg.table}
WHERE  id IN (${placeholders})
`.trim();
    return { sql };
}
/** Generates a SELECT for the content_hash of the most recently indexed chunk of a document. */
export function buildGetContentHashSql(cfg) {
    return `
SELECT content_hash
FROM   ${cfg.schema}.${cfg.table}
WHERE  document_id = $1
LIMIT  1
`.trim();
}
/** Generates a SELECT for the latest indexed timestamp of a document. */
export function buildLastIndexedAtSql(cfg) {
    return `SELECT MAX(last_indexed_at) AS ts FROM ${cfg.schema}.${cfg.table} WHERE document_id = $1`;
}
/** Resolve config with defaults applied. */
export function resolveConfig(cfg) {
    return {
        schema: cfg.schema ?? 'public',
        table: cfg.table ?? 'lumiarq_chunks',
        embeddingColumn: cfg.embeddingColumn ?? 'embedding',
    };
}
