// @lumiarq/rag@0.1.0/dist/adapters/store/supabase.js
import { BaseVectorStore } from './base.js';
import { resolveConfig, } from './sql-base.js';
export class SupabaseVectorStore extends BaseVectorStore {
    client;
    cfg;
    matchFn;
    constructor(opts) {
        super(opts);
        this.client = opts.client;
        this.cfg = resolveConfig(opts);
        this.matchFn = opts.matchFn ?? 'match_chunks';
    }
    async doUpsert(docs) {
        const rows = docs.map((doc) => ({
            id: doc.id,
            document_id: doc.documentId,
            text: doc.text,
            chunk_index: doc.index,
            start_char: doc.startChar,
            end_char: doc.endChar,
            metadata: doc.metadata ?? null,
            content_hash: doc.contentHash ?? null,
            last_indexed_at: new Date().toISOString(),
            [this.cfg.embeddingColumn]: `[${doc.vector.join(',')}]`,
        }));
        const { error } = await this.client.from(this.cfg.table).upsert(rows, { onConflict: 'id' });
        if (error)
            throw new Error(`SupabaseVectorStore.upsert: ${error.message}`);
    }
    async doSearch(vector, topK, minScore) {
        const { data, error } = await this.client.rpc(this.matchFn, {
            query_embedding: `[${vector.join(',')}]`,
            match_threshold: minScore,
            match_count: topK,
            table_name: this.cfg.table,
        });
        if (error)
            throw new Error(`SupabaseVectorStore.search: ${error.message}`);
        return (data ?? []).map((row) => ({
            chunk: {
                id: row.id,
                documentId: row.document_id,
                text: row.text,
                index: row.chunk_index,
                startChar: row.start_char,
                endChar: row.end_char,
                ...(row.metadata !== null ? { metadata: row.metadata } : {}),
            },
            score: row.similarity,
        }));
    }
    async doDelete(ids) {
        const { error } = await this.client.from(this.cfg.table).delete().in('id', ids);
        if (error)
            throw new Error(`SupabaseVectorStore.delete: ${error.message}`);
    }
    async doGetIndexedIds() {
        const { data, error } = await this.client.from(this.cfg.table).select('id');
        if (error)
            throw new Error(`SupabaseVectorStore.getIndexedIds: ${error.message}`);
        return (data ?? []).map((r) => r['id']);
    }
    async doLastIndexedAt(documentId) {
        const { data, error } = await this.client
            .from(this.cfg.table)
            .select('last_indexed_at')
            .eq('document_id', documentId);
        if (error)
            throw new Error(`SupabaseVectorStore.lastIndexedAt: ${error.message}`);
        if (!data || data.length === 0)
            return null;
        const dates = data
            .map((r) => r['last_indexed_at'])
            .filter((ts) => typeof ts === 'string')
            .map((ts) => new Date(ts));
        const first = dates[0];
        if (!first)
            return null;
        return dates.reduce((max, d) => (d > max ? d : max), first);
    }
    async doGetByIds(ids) {
        const { data, error } = await this.client
            .from(this.cfg.table)
            .select('id, document_id, text, chunk_index, start_char, end_char, metadata')
            .in('id', ids);
        if (error)
            throw new Error(`SupabaseVectorStore.getByIds: ${error.message}`);
        return (data ?? []).map((r) => ({
            id: r['id'],
            documentId: r['document_id'],
            text: r['text'],
            index: r['chunk_index'],
            startChar: r['start_char'],
            endChar: r['end_char'],
            ...(r['metadata'] !== null
                ? { metadata: r['metadata'] }
                : {}),
        }));
    }
    async doGetContentHash(documentId) {
        const { data, error } = await this.client
            .from(this.cfg.table)
            .select('content_hash')
            .eq('document_id', documentId);
        if (error)
            throw new Error(`SupabaseVectorStore.getContentHash: ${error.message}`);
        const first = data?.[0];
        if (!first)
            return null;
        const hash = first['content_hash'];
        return typeof hash === 'string' ? hash : null;
    }
}
