// @lumiarq/ai@1.2.0/dist/deterministic.js
function parseDeterministicIntent(text) {
    const stripped = text.toLowerCase().trim();
    if (stripped.includes('create') || stripped.includes('add') || stripped.includes('new')) {
        return 'create';
    }
    if (stripped.includes('delete') || stripped.includes('remove')) {
        return 'delete';
    }
    if (stripped.includes('update') || stripped.includes('modify') || stripped.includes('change')) {
        return 'update';
    }
    if (stripped.includes('read') || stripped.includes('show') || stripped.includes('get') || stripped.includes('list')) {
        return 'read';
    }
    return 'unknown';
}
function ensureRange(value, min, max) {
    if (Number.isNaN(value)) {
        return min;
    }
    return Math.max(min, Math.min(max, value));
}
/**
 * Deterministic fallback provider.
 * Used when no AI providers are available or configured.
 * Supports only `intent-parse` via keyword recognition; always available, zero latency.
 * Returns explicit error for unsupported tasks instead of silent failure.
 */
export function createDeterministicProvider() {
    return {
        id: 'local',
        supports: ({ task, model }) => task === 'intent-parse' && (!model || model === 'deterministic'),
        complete: async (request) => {
            // Reject unsupported tasks with a clear error message.
            if (request.task !== 'intent-parse') {
                throw new Error(`Deterministic provider does not support task '${request.task}'. ` +
                    `Only 'intent-parse' is supported. Please configure an AI provider (Groq, OpenAI, Anthropic, etc.).`);
            }
            const intent = parseDeterministicIntent(request.prompt);
            const response = {
                intent,
                confidence: 0.55,
                ambiguityReasons: ['deterministic fallback parser used; AI provider unavailable'],
            };
            return {
                provider: 'local',
                model: 'deterministic',
                text: JSON.stringify(response),
            };
        },
    };
}
