// @lumiarq/ai@1.2.0/dist/index.js
// ── AI Gateway (provider router) ──────────────────────────────────────────────
export { AIGateway, createAIGateway, complete, streamText, generateObject, embed, embedMany } from './gateway.js';
// ── Model registry ────────────────────────────────────────────────────────────
export { PROVIDER_DEFAULT_MODELS, KNOWN_MODELS, mergeModelRegistry } from './config/models.js';
// ── Provider paths ────────────────────────────────────────────────────────────
export { AI_PROVIDER_URL_PATH, AI_PROVIDER_ENDPOINT_PATH } from './config/paths.js';
// ── Provider adapters ─────────────────────────────────────────────────────────
export { createOpenAIProvider } from './providers/openai.js';
export { createAnthropicProvider } from './providers/anthropic.js';
export { createGeminiProvider } from './providers/gemini.js';
export { createOllamaProvider } from './providers/ollama.js';
export { createOpenRouterProvider } from './providers/openrouter.js';
export { createDeepSeekProvider } from './providers/deepseek.js';
export { createMistralProvider } from './providers/mistral.js';
// ── Provider factory & gateway builder ────────────────────────────────────────
export { createGatewayFromEnv, createProvidersFromEnv, inferDefaultProvider, AUTO_POLICY_TASK_MODELS, } from './providers/index.js';
// ── Deterministic fallback provider ───────────────────────────────────────────
export { createDeterministicProvider } from './deterministic.js';
// ── Structured errors ─────────────────────────────────────────────────────────
export { AITimeoutError, AIProviderError, AIAllProvidersFailedError, AINoEmbeddingProviderError, AINoProvidersError, AIMiddlewareError, AINoModelForTaskError, } from './config/errors.js';
// ── Auto-policy helpers ───────────────────────────────────────────────────────
export { createAutoPolicy, autoSelectModelForTask } from './auto-policy.js';
