// @lumiarq/rag@0.1.0/dist/chunker/index.js
function chunkId(documentId, index) {
    return `${documentId}__chunk_${index}`;
}
// ── Sliding Window Chunker ────────────────────────────────────────────────────
/**
 * Splits text into overlapping fixed-size character windows.
 * Simple and predictable — good for dense prose.
 */
export class SlidingWindowChunker {
    async chunk(doc, opts) {
        const size = opts?.chunkSize ?? 512;
        const overlap = opts?.overlap ?? 64;
        const stride = size - overlap;
        if (stride <= 0) {
            throw new Error(`SlidingWindowChunker: overlap (${overlap}) must be less than chunkSize (${size})`);
        }
        const text = doc.content;
        const chunks = [];
        let startChar = 0;
        let index = 0;
        while (startChar < text.length) {
            const endChar = Math.min(startChar + size, text.length);
            const chunk = {
                id: chunkId(doc.id, index),
                documentId: doc.id,
                text: text.slice(startChar, endChar),
                index,
                startChar,
                endChar,
                ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
            };
            chunks.push(chunk);
            if (endChar === text.length)
                break;
            startChar += stride;
            index++;
        }
        return chunks;
    }
}
// ── Sentence Chunker ──────────────────────────────────────────────────────────
/**
 * Groups sentences into chunks that stay under the target size.
 * Preserves sentence boundaries — better for Q&A over prose.
 */
export class SentenceChunker {
    async chunk(doc, opts) {
        const size = opts?.chunkSize ?? 512;
        // Split on sentence-ending punctuation followed by whitespace/EOL.
        const sentenceRegex = /[^.!?]*[.!?]+(?:\s|$)/g;
        const sentences = [];
        let match;
        let lastEnd = 0;
        while ((match = sentenceRegex.exec(doc.content)) !== null) {
            sentences.push({ text: match[0], start: match.index });
            lastEnd = match.index + match[0].length;
        }
        // Capture any trailing text without ending punctuation
        if (lastEnd < doc.content.length) {
            sentences.push({ text: doc.content.slice(lastEnd), start: lastEnd });
        }
        const chunks = [];
        let currentText = '';
        let currentStart = 0;
        let chunkIndex = 0;
        for (const sentence of sentences) {
            const candidate = currentText + sentence.text;
            if (candidate.length > size && currentText.length > 0) {
                // Flush current chunk
                const flushChunk = {
                    id: chunkId(doc.id, chunkIndex),
                    documentId: doc.id,
                    text: currentText.trim(),
                    index: chunkIndex,
                    startChar: currentStart,
                    endChar: currentStart + currentText.length,
                    ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
                };
                chunks.push(flushChunk);
                chunkIndex++;
                currentText = sentence.text;
                currentStart = sentence.start;
            }
            else {
                if (currentText.length === 0)
                    currentStart = sentence.start;
                currentText += sentence.text;
            }
        }
        if (currentText.trim().length > 0) {
            const lastChunk = {
                id: chunkId(doc.id, chunkIndex),
                documentId: doc.id,
                text: currentText.trim(),
                index: chunkIndex,
                startChar: currentStart,
                endChar: currentStart + currentText.length,
                ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
            };
            chunks.push(lastChunk);
        }
        return chunks;
    }
}
// ── Markdown Chunker ──────────────────────────────────────────────────────────
/**
 * Splits markdown content at heading boundaries (H1–H6).
 * Each heading is joined with its following body text into one section,
 * ensuring no heading-only chunks are produced.
 * Falls back to SlidingWindowChunker for sections that exceed `chunkSize`.
 */
export class MarkdownChunker {
    fallback = new SlidingWindowChunker();
    async chunk(doc, opts) {
        const size = opts?.chunkSize ?? 1024;
        // Split on markdown headings (capturing group keeps the heading in the result array).
        // The resulting array alternates: [preamble?, heading, body, heading, body, ...]
        const headingRegex = /^(#{1,6}\s.+)$/m;
        const parts = doc.content.split(headingRegex);
        // Re-join heading+body pairs so we never emit a heading-only chunk.
        const sections = [];
        let pos = 0;
        // parts[0] is pre-heading preamble (may be empty)
        const preamble = parts[0] ?? '';
        if (preamble.trim().length > 0) {
            sections.push({ text: preamble, start: 0 });
        }
        pos += preamble.length;
        // Walk the remaining parts in heading+body pairs
        for (let i = 1; i < parts.length; i += 2) {
            const heading = parts[i] ?? '';
            const body = parts[i + 1] ?? '';
            const combined = heading + body;
            if (combined.trim().length > 0) {
                sections.push({ text: combined, start: pos });
            }
            pos += combined.length;
        }
        const chunks = [];
        let globalIndex = 0;
        for (const section of sections) {
            if (section.text.length <= size) {
                const chunk = {
                    id: chunkId(doc.id, globalIndex),
                    documentId: doc.id,
                    text: section.text.trim(),
                    index: globalIndex,
                    startChar: section.start,
                    endChar: section.start + section.text.length,
                    ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
                };
                chunks.push(chunk);
                globalIndex++;
            }
            else {
                // Section too large — split with sliding window
                const subDoc = {
                    id: doc.id,
                    content: section.text,
                    ...(doc.metadata !== undefined ? { metadata: doc.metadata } : {}),
                };
                const subChunks = await this.fallback.chunk(subDoc, opts);
                for (const sub of subChunks) {
                    chunks.push({
                        ...sub,
                        id: chunkId(doc.id, globalIndex),
                        index: globalIndex,
                        startChar: section.start + sub.startChar,
                        endChar: section.start + sub.endChar,
                    });
                    globalIndex++;
                }
            }
        }
        return chunks;
    }
}
