// @illumiarq/runtime@1.1.2/dist/queue/bullmq-queue.js
async function loadQueue() {
    try {
        const bullmq = (await import('bullmq'));
        return bullmq.Queue;
    }
    catch {
        throw new Error('[BullMQQueue] bullmq is not installed. Run: pnpm add bullmq ioredis');
    }
}
export class BullMQQueue {
    opts;
    queues = new Map();
    QueueCtor = null;
    constructor(opts) {
        this.opts = opts;
    }
    async getQueue(name) {
        if (!this.queues.has(name)) {
            if (!this.QueueCtor) {
                this.QueueCtor = await loadQueue();
            }
            const q = new this.QueueCtor(name, { connection: this.opts.connection });
            this.queues.set(name, q);
        }
        return this.queues.get(name);
    }
    async dispatch(job, options) {
        const queueName = options?.queue ?? this.opts.defaultQueue ?? 'default';
        const q = await this.getQueue(queueName);
        await q.add(job.name, job.data, {
            delay: options?.delay ? options.delay * 1000 : undefined,
            attempts: options?.tries ?? 3,
            backoff: options?.backoff
                ? { type: 'exponential', delay: options.backoff * 1000 }
                : { type: 'exponential', delay: 5000 },
        });
    }
    async later(job, seconds, options) {
        return this.dispatch(job, { ...options, delay: seconds });
    }
    /** Close all queue connections. Call on graceful shutdown. */
    async close() {
        await Promise.all([...this.queues.values()].map((q) => q.close()));
        this.queues.clear();
    }
}
export class BullMQWorker {
    opts;
    workers = [];
    constructor(opts) {
        this.opts = opts;
    }
    async start() {
        let Worker;
        try {
            const bullmq = (await import('bullmq'));
            Worker = bullmq.Worker;
        }
        catch {
            throw new Error('[BullMQWorker] bullmq is not installed. Run: pnpm add bullmq ioredis');
        }
        for (const queueName of this.opts.queues) {
            const worker = new Worker(queueName, async (job) => {
                const handler = this.opts.handlers?.get(job.name);
                if (handler) {
                    await handler(job.name, job.data);
                }
                else {
                    console.warn(`[BullMQWorker] No handler registered for job "${job.name}" on queue "${queueName}"`);
                }
            }, { connection: this.opts.connection });
            this.workers.push(worker);
        }
    }
    async stop() {
        await Promise.all(this.workers.map((w) => w.close()));
        this.workers = [];
    }
}
