// @illumiarq/runtime@1.1.2/dist/audit/database-audit.js
import { getContext } from '@illumiarq/context';
function generateId() {
    return `audit-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
}
export class DatabaseAudit {
    db;
    constructor(opts) {
        this.db = opts.db;
    }
    async recordAttempt(entry) {
        const id = generateId();
        let requestId = entry.requestId;
        try {
            const ctx = getContext();
            requestId ??= ctx.contextId ?? null;
        }
        catch {
            /* no context active */
        }
        await this.db.execute({
            sql: `INSERT INTO audit_log
              (id, action, actor_id, actor_type, input, result_summary, status, error, duration_ms, request_id, started_at)
            VALUES (?, ?, ?, ?, ?, NULL, 'attempt', NULL, 0, ?, ?)`,
            args: [
                id,
                entry.action,
                entry.actorId ?? null,
                entry.actorType ?? 'system',
                JSON.stringify(entry.input),
                requestId ?? null,
                Math.floor(entry.startedAt.getTime() / 1000),
            ],
        });
        return id;
    }
    async recordSuccess(entry) {
        await this.db.execute({
            sql: `UPDATE audit_log
            SET status = 'success',
                result_summary = ?,
                duration_ms = ?
            WHERE id = ?`,
            args: [
                entry.resultSummary ? JSON.stringify(entry.resultSummary) : null,
                entry.durationMs,
                entry.auditId,
            ],
        });
    }
    async recordFailure(entry) {
        await this.db.execute({
            sql: `UPDATE audit_log
            SET status = 'failure',
                error = ?,
                duration_ms = ?
            WHERE id = ?`,
            args: [entry.error, entry.durationMs, entry.auditId],
        });
    }
    async query(filters) {
        const conditions = [];
        const args = [];
        if (filters.action) {
            conditions.push('action = ?');
            args.push(filters.action);
        }
        if (filters.actorId) {
            conditions.push('actor_id = ?');
            args.push(filters.actorId);
        }
        if (filters.actorType) {
            conditions.push('actor_type = ?');
            args.push(filters.actorType);
        }
        if (filters.status) {
            conditions.push('status = ?');
            args.push(filters.status);
        }
        if (filters.from) {
            conditions.push('started_at >= ?');
            args.push(Math.floor(filters.from.getTime() / 1000));
        }
        if (filters.to) {
            conditions.push('started_at <= ?');
            args.push(Math.floor(filters.to.getTime() / 1000));
        }
        const where = conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : '';
        const limit = filters.limit ?? 100;
        args.push(limit);
        const result = await this.db.execute({
            sql: `SELECT id, action, actor_id, actor_type, input, result_summary, status, error, duration_ms, request_id, started_at
            FROM audit_log
            ${where}
            ORDER BY started_at DESC
            LIMIT ?`,
            args,
        });
        return result.rows.map((row) => ({
            id: String(row['id']),
            action: String(row['action']),
            actorId: row['actor_id'] ?? null,
            actorType: String(row['actor_type'] ?? 'system'),
            input: String(row['input']),
            resultSummary: row['result_summary'] ?? null,
            status: row['status'] ?? 'attempt',
            error: row['error'] ?? null,
            durationMs: Number(row['duration_ms']),
            requestId: row['request_id'] ?? null,
            startedAt: new Date(Number(row['started_at']) * 1000),
        }));
    }
}
