// @illumiarq/http@1.0.0/dist/request/request.js
/**
 * Request abstraction over the Web Request object.
 */
export class HttpRequest {
    request;
    routeParams;
    constructor(request, routeParams = {}) {
        this.request = request;
        this.routeParams = routeParams;
    }
    raw() {
        return this.request;
    }
    params(name) {
        if (!name)
            return this.routeParams;
        return this.routeParams[name];
    }
    query(name) {
        const url = new URL(this.request.url);
        const values = url.searchParams.getAll(name ?? '');
        if (name) {
            if (values.length === 0)
                return undefined;
            if (values.length === 1)
                return values[0];
            return values;
        }
        const parsed = {};
        for (const key of url.searchParams.keys()) {
            const all = url.searchParams.getAll(key);
            parsed[key] = all.length <= 1 ? all[0] : all;
        }
        return parsed;
    }
    async body() {
        const contentType = this.request.headers.get('content-type') ?? '';
        if (contentType.includes('application/json')) {
            return (await this.request.json());
        }
        return (await this.request.text());
    }
}
