// @lumiarq/pack-django@1.0.1/dist/adapter.js
import { CURRENT_IR_VERSION, } from '@lumiarq/context';
import { loadDjangoTemplate, resolveDjangoTemplateName } from './templates/index.js';
import path from 'node:path';
import fs from 'node:fs';
const DJANGO_MIGRATION_INTENT_KINDS = new Set(['schema', 'migration', 'model']);
function toSnakeCase(value) {
    return value
        .replace(/([a-z0-9])([A-Z])/g, '$1_$2')
        .replace(/[^a-zA-Z0-9]+/g, '_')
        .replace(/^_+|_+$/g, '')
        .toLowerCase();
}
function normalizePath(value) {
    return value.replaceAll('\\', '/').replace(/\/+/g, '/');
}
function getMigrationPostStepPath(projectRoot) {
    return path.join(projectRoot, '.lumis', 'poststeps', 'django-migrations.md');
}
function isMigrationKind(kind) {
    return DJANGO_MIGRATION_INTENT_KINDS.has(kind);
}
function isDjangoMigrationFilePath(targetFile) {
    const normalized = normalizePath(targetFile);
    if (!normalized.includes('/migrations/') || !normalized.endsWith('.py')) {
        return false;
    }
    return !normalized.endsWith('/migrations/__init__.py');
}
function fileContains(projectRoot, fileName, pattern) {
    const filePath = path.join(projectRoot, fileName);
    if (!fs.existsSync(filePath)) {
        return false;
    }
    try {
        const content = fs.readFileSync(filePath, 'utf8');
        return pattern.test(content);
    }
    catch {
        return false;
    }
}
function hasDjangoExtensions(projectRoot) {
    const marker = /django-extensions/i;
    return (fileContains(projectRoot, 'requirements.txt', marker) ||
        fileContains(projectRoot, 'pyproject.toml', marker) ||
        fileContains(projectRoot, 'Pipfile', marker));
}
function listPotentialApps(projectRoot) {
    const ignored = new Set(['.git', '.lumis', '.venv', 'venv', '__pycache__', 'node_modules']);
    if (!fs.existsSync(projectRoot)) {
        return [];
    }
    return fs
        .readdirSync(projectRoot, { withFileTypes: true })
        .filter((entry) => entry.isDirectory())
        .filter((entry) => !ignored.has(entry.name))
        .map((entry) => path.join(projectRoot, entry.name))
        .filter((appDir) => {
        const hasAppsPy = fs.existsSync(path.join(appDir, 'apps.py'));
        const hasModelsPy = fs.existsSync(path.join(appDir, 'models.py'));
        return hasAppsPy || hasModelsPy;
    });
}
function listPythonFunctions(filePath) {
    if (!fs.existsSync(filePath)) {
        return [];
    }
    const content = fs.readFileSync(filePath, 'utf8');
    const functionRegex = /^def\s+([a-zA-Z_][\w]*)\s*\(/gm;
    const names = [];
    let match;
    do {
        match = functionRegex.exec(content);
        if (match?.[1]) {
            names.push(match[1]);
        }
    } while (match);
    return names;
}
function classifyDjangoArtifact(fileName, functionName) {
    const lower = functionName.toLowerCase();
    const isWriteVerb = /^(create|update|delete|save|upsert|perform)_/.test(lower);
    const isReadVerb = /^(get|list|fetch|select|find)_/.test(lower);
    if (fileName === 'services.py') {
        let confidence = 0.7;
        let reasoning = 'Defined in services.py';
        if (isWriteVerb) {
            confidence += 0.2;
            reasoning += ' and uses write-oriented naming';
        }
        if (isReadVerb) {
            confidence -= 0.2;
            reasoning += ', but uses read-oriented naming';
        }
        if (confidence < 0.8) {
            return {
                type: 'unknown',
                confidence,
                reasoning,
                alternatives: ['service', 'selector'],
            };
        }
        return { type: 'service', confidence, reasoning };
    }
    if (fileName === 'selectors.py') {
        let confidence = 0.7;
        let reasoning = 'Defined in selectors.py';
        if (isReadVerb) {
            confidence += 0.2;
            reasoning += ' and uses read-oriented naming';
        }
        if (isWriteVerb) {
            confidence -= 0.2;
            reasoning += ', but uses write-oriented naming';
        }
        if (confidence < 0.8) {
            return {
                type: 'unknown',
                confidence,
                reasoning,
                alternatives: ['service', 'selector'],
            };
        }
        return { type: 'selector', confidence, reasoning };
    }
    return {
        type: 'unknown',
        confidence: 0.4,
        reasoning: 'Unknown Django artifact source file',
        alternatives: ['service', 'selector'],
    };
}
function relativeFromRoot(projectRoot, filePath) {
    const rel = path.relative(projectRoot, filePath);
    return normalizePath(rel);
}
function buildMigrationPostStep(intent, ir) {
    const subject = intent.subjects[0];
    const modelName = subject?.name ?? 'example';
    const appName = subject?.attributes?.module?.toString() || 'app';
    const migrationName = `add_${toSnakeCase(modelName)}_changes`;
    const targetFile = getMigrationPostStepPath(ir.projectRoot);
    return {
        id: `poststep:migration:${toSnakeCase(modelName)}`,
        operation: 'append',
        artifactKind: 'migration-post-step',
        module: appName,
        name: migrationName,
        targetFile,
        templateName: resolveDjangoTemplateName('migration'),
        tokens: {
            AppName: appName,
            MigrationName: migrationName,
            ModelName: modelName,
        },
        dependsOn: [],
        requiresConfirmation: false,
        rollback: {
            operation: 'restore',
            targetFile,
        },
        rationale: `${intent.action} migration post-step in app "${appName}"`,
    };
}
const DJANGO_VERIFICATION_HOOKS = [
    {
        name: 'django.prevent-generated-migrations',
        when: 'before-write',
        run: async (ctx) => {
            if (!isDjangoMigrationFilePath(ctx.step.targetFile)) {
                return { passed: true, guarantee: 'django.migration-contract' };
            }
            return {
                passed: false,
                guarantee: 'django.migration-contract',
                message: 'Django migration generation is blocked by contract. Use the generated post-step guide and run manage.py makemigrations manually.',
            };
        },
    },
];
export const DjangoAdapter = {
    name: 'django',
    version: '1.0.0',
    language: 'python',
    detect: async (projectRoot) => {
        const settings = path.join(projectRoot, 'manage.py');
        const detected = fs.existsSync(settings);
        return {
            detected,
            confidence: detected ? 0.88 : 0,
            reason: detected ? 'manage.py found' : 'no manage.py',
        };
    },
    buildIR: async (projectRoot) => {
        const modules = listPotentialApps(projectRoot).map((appDir) => {
            const appName = path.basename(appDir);
            const artifacts = ['services.py', 'selectors.py']
                .flatMap((fileName) => {
                const filePath = path.join(appDir, fileName);
                const functions = listPythonFunctions(filePath);
                return functions.map((fnName) => {
                    const classification = classifyDjangoArtifact(fileName, fnName);
                    return {
                        id: `${appName}:${classification.type}:${fnName}`,
                        kind: classification.type,
                        name: fnName,
                        file: relativeFromRoot(projectRoot, filePath),
                        module: appName,
                        metadata: {
                            classification,
                            ambiguous: classification.confidence < 0.8,
                        },
                    };
                });
            });
            return {
                name: appName,
                directory: relativeFromRoot(projectRoot, appDir),
                artifacts,
                publicExports: [],
                metadata: {},
            };
        });
        return {
            irVersion: CURRENT_IR_VERSION,
            version: '1.0',
            projectRoot,
            framework: 'django',
            builtAt: new Date(),
            modules,
            dependencies: [],
            violations: [],
            metadata: {},
        };
    },
    updateIR: async (ir) => ir,
    placementRules: [],
    dependencyRules: [],
    namingRules: [],
    resolveIntent: async (intent, ir) => {
        if (intent.action !== 'create' || intent.subjects.length === 0) {
            return [];
        }
        const subject = intent.subjects[0];
        if (!subject) {
            return [];
        }
        const kind = subject.kind.trim().toLowerCase();
        if (isMigrationKind(kind)) {
            return [buildMigrationPostStep(intent, ir)];
        }
        const moduleName = (subject.name ?? 'module').toString().toLowerCase();
        const step = {
            id: 'step-1',
            operation: 'create',
            artifactKind: 'view',
            module: moduleName,
            name: `${moduleName}_view`,
            targetFile: `${ir.projectRoot}/${moduleName}/views.py`,
            templateName: resolveDjangoTemplateName('view'),
            tokens: { ModuleName: moduleName },
            dependsOn: [],
            requiresConfirmation: false,
            rollback: { operation: 'delete', targetFile: `${ir.projectRoot}/${moduleName}/views.py` },
            rationale: `${intent.action} view in app "${moduleName}"`,
        };
        return [step];
    },
    resolveTemplate: (templateName) => loadDjangoTemplate(templateName),
    verificationHooks: DJANGO_VERIFICATION_HOOKS,
    buildCommands: {},
    tinkerContext: async (projectRoot) => {
        const preferredShell = hasDjangoExtensions(projectRoot) ? 'shell_plus' : 'shell';
        const command = preferredShell === 'shell_plus' ? 'python manage.py shell_plus' : 'python manage.py shell';
        return {
            framework: 'django',
            preferredShell,
            command,
            fallbackShell: 'shell',
            migrationPolicy: 'manual-post-steps-only',
        };
    },
};
