// @lumiarq/lumis@1.2.0/dist/plugins/tinker-plugin.js
/**
 * Registry for loaded plugins.
 */
export class TinkerPluginRegistry {
    plugins = new Map();
    register(plugin) {
        if (this.plugins.has(plugin.name)) {
            throw new Error(`Plugin '${plugin.name}' already registered`);
        }
        this.plugins.set(plugin.name, plugin);
    }
    unregister(name) {
        return this.plugins.delete(name);
    }
    get(name) {
        return this.plugins.get(name);
    }
    getAll() {
        return Array.from(this.plugins.values());
    }
    getAllCommands() {
        const commands = [];
        for (const plugin of this.plugins.values()) {
            if (plugin.commands) {
                commands.push(...plugin.commands);
            }
        }
        return commands;
    }
    async callHook(hookName, args) {
        const results = [];
        for (const plugin of this.plugins.values()) {
            const hook = plugin[hookName];
            if (typeof hook === 'function') {
                const result = await hook.apply(plugin, args);
                results.push(result);
            }
        }
        return results;
    }
}
