// @lumiarq/verifier@1.0.1/dist/checks/placement.js
import path from 'node:path';
/**
 * Validate that a step writes inside the project root and has a usable target path.
 *
 * @param ctx - Verification context with current step and project metadata.
 * @returns Placement validation result.
 */
export async function checkPlacement(ctx) {
    const targetFile = ctx.step.targetFile.trim();
    if (!targetFile) {
        return {
            passed: false,
            guarantee: 'placement',
            message: 'Step targetFile is empty.',
            severity: 'critical',
        };
    }
    const projectRoot = path.resolve(ctx.projectRoot);
    const resolvedTargetFile = path.resolve(targetFile);
    const insideProject = resolvedTargetFile === projectRoot || resolvedTargetFile.startsWith(`${projectRoot}${path.sep}`);
    if (!insideProject) {
        return {
            passed: false,
            guarantee: 'placement',
            message: `Target file is outside project root: ${resolvedTargetFile}`,
            details: {
                projectRoot,
                resolvedTargetFile,
            },
            severity: 'critical',
        };
    }
    const targetDir = path.dirname(resolvedTargetFile);
    if (!targetDir || targetDir === '.') {
        return {
            passed: false,
            guarantee: 'placement',
            message: 'Unable to resolve a valid target directory for the step.',
            severity: 'critical',
        };
    }
    return {
        passed: true,
        guarantee: 'placement',
        message: 'Target file placement is valid.',
    };
}
