// @lumiarq/lumis@1.2.0/dist/lib/file-scanner.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { createHash } from 'node:crypto';
const DEFAULT_MAX_FILES = 300;
const DEFAULT_MAX_FILE_SIZE_BYTES = 200_000;
const ALLOWED_EXTENSIONS = new Set(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cjs', '.json', '.md']);
const SKIP_DIRS = new Set([
    '.git',
    '.next',
    'dist',
    'build',
    'coverage',
    'node_modules',
    '.turbo',
    '.cache',
]);
function toContentHash(content) {
    return createHash('sha256').update(content).digest('hex');
}
async function walkDirectory(root, current, files, maxFiles) {
    if (files.length >= maxFiles)
        return;
    const entries = await fs.readdir(current, { withFileTypes: true });
    for (const entry of entries) {
        if (files.length >= maxFiles)
            return;
        const absolutePath = path.join(current, entry.name);
        if (entry.isSymbolicLink())
            continue;
        if (entry.isDirectory()) {
            if (SKIP_DIRS.has(entry.name))
                continue;
            await walkDirectory(root, absolutePath, files, maxFiles);
            continue;
        }
        if (!entry.isFile())
            continue;
        const ext = path.extname(entry.name).toLowerCase();
        if (!ALLOWED_EXTENSIONS.has(ext))
            continue;
        const relative = path.relative(root, absolutePath);
        if (!relative || relative.startsWith('..'))
            continue;
        files.push(absolutePath);
    }
}
export async function scanProjectFiles(projectRoot, options = {}) {
    const maxFiles = options.maxFiles ?? DEFAULT_MAX_FILES;
    const maxFileSizeBytes = options.maxFileSizeBytes ?? DEFAULT_MAX_FILE_SIZE_BYTES;
    const files = [];
    await walkDirectory(projectRoot, projectRoot, files, maxFiles);
    const docs = [];
    for (const absolutePath of files) {
        const stat = await fs.stat(absolutePath);
        if (stat.size > maxFileSizeBytes)
            continue;
        const content = await fs.readFile(absolutePath, 'utf8');
        if (!content.trim())
            continue;
        const relative = path.relative(projectRoot, absolutePath);
        docs.push({
            id: relative,
            content,
            contentHash: toContentHash(content),
            metadata: { path: relative },
        });
    }
    return docs;
}
