// @lumiarq/lumis@1.2.0/dist/lib/tinker-replay.js
/**
 * Tinker Deterministic Replay and Audit Trail
 *
 * Records session interactions to JSONL for debugging, regression testing, and compliance.
 */
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
/**
 * Audit trail writer — appends entries to ~/.lumis/audit.jsonl
 */
export class AuditTrail {
    auditPath;
    constructor(projectRoot) {
        this.auditPath = path.join(projectRoot, '.lumis', 'audit.jsonl');
    }
    record(entry) {
        try {
            fs.mkdirSync(path.dirname(this.auditPath), { recursive: true });
            fs.appendFileSync(this.auditPath, `${JSON.stringify(entry)}\n`, 'utf8');
        }
        catch (err) {
            console.error(`[audit-trail] Failed to write: ${err instanceof Error ? err.message : String(err)}`);
        }
    }
    recordFileWrite(filePath, outcome, details) {
        this.record({
            timestamp: new Date().toISOString(),
            sessionId: process.env.TINKER_SESSION_ID || 'unknown',
            eventType: 'file-write',
            path: filePath,
            outcome,
            ...(details ? { details } : {}),
        });
    }
    recordFileDelete(filePath, outcome, details) {
        this.record({
            timestamp: new Date().toISOString(),
            sessionId: process.env.TINKER_SESSION_ID || 'unknown',
            eventType: 'file-delete',
            path: filePath,
            outcome,
            ...(details ? { details } : {}),
        });
    }
    recordToolCall(tool, outcome, details) {
        this.record({
            timestamp: new Date().toISOString(),
            sessionId: process.env.TINKER_SESSION_ID || 'unknown',
            eventType: 'tool-call',
            tool,
            outcome,
            ...(details ? { details } : {}),
        });
    }
    recordVerification(outcome, details) {
        this.record({
            timestamp: new Date().toISOString(),
            sessionId: process.env.TINKER_SESSION_ID || 'unknown',
            eventType: 'verification',
            outcome,
            ...(details ? { details } : {}),
        });
    }
}
/**
 * Session replay writer — appends turn entries to ~/.lumis/replays/<sessionId>.jsonl
 */
export class SessionReplayer {
    sessionId;
    replayPath;
    turnIndex = 0;
    constructor(projectRoot, sessionId) {
        this.sessionId = sessionId || generateSessionId();
        this.replayPath = path.join(projectRoot, '.lumis', 'replays', `${this.sessionId}.jsonl`);
    }
    recordUserMessage(content) {
        this.recordEntry('user', content);
    }
    recordAssistantMessage(content, toolCalls) {
        const entry = {
            timestamp: new Date().toISOString(),
            sessionId: this.sessionId,
            turnIndex: this.turnIndex++,
            role: 'assistant',
            content,
            ...(toolCalls ? { toolCalls } : {}),
        };
        try {
            fs.mkdirSync(path.dirname(this.replayPath), { recursive: true });
            fs.appendFileSync(this.replayPath, `${JSON.stringify(entry)}\n`, 'utf8');
        }
        catch (err) {
            console.error(`[session-replay] Failed to write: ${err instanceof Error ? err.message : String(err)}`);
        }
    }
    recordEntry(role, content) {
        const entry = {
            timestamp: new Date().toISOString(),
            sessionId: this.sessionId,
            turnIndex: this.turnIndex++,
            role,
            content,
        };
        try {
            fs.mkdirSync(path.dirname(this.replayPath), { recursive: true });
            fs.appendFileSync(this.replayPath, `${JSON.stringify(entry)}\n`, 'utf8');
        }
        catch (err) {
            console.error(`[session-replay] Failed to write: ${err instanceof Error ? err.message : String(err)}`);
        }
    }
    getReplayPath() {
        return this.replayPath;
    }
    getSessionId() {
        return this.sessionId;
    }
}
export function generateSessionId() {
    return `tinker-${Date.now()}-${crypto.randomBytes(4).toString('hex')}`;
}
