// @lumiarq/lumis@1.2.0/dist/lib/tinker-policy.js
/**
 * Tinker Enterprise Policy Controls
 *
 * Governs what operations are allowed in a session based on TINKER_POLICY env var.
 */
class ReadOnlyPolicy {
    canRead() {
        return true;
    }
    canWrite() {
        return false;
    }
    canExecute() {
        return false;
    }
    canToolCall() {
        return false;
    }
    describeLimitations() {
        return [
            'Read-only mode: file writes disabled',
            'Read-only mode: command execution disabled',
            'Read-only mode: tool calls disabled',
        ];
    }
}
class PlanOnlyPolicy {
    canRead() {
        return true;
    }
    canWrite() {
        return false;
    }
    canExecute() {
        return false;
    }
    canToolCall() {
        return true;
    }
    describeLimitations() {
        return [
            'Plan-only mode: file writes disabled',
            'Plan-only mode: command execution disabled',
            'Plan-only mode: tool calls allowed for planning only',
        ];
    }
}
class FullPolicy {
    canRead() {
        return true;
    }
    canWrite() {
        return true;
    }
    canExecute() {
        return true;
    }
    canToolCall() {
        return true;
    }
    describeLimitations() {
        return [];
    }
}
export function getPolicyGate(policy) {
    switch (policy) {
        case 'read-only':
            return new ReadOnlyPolicy();
        case 'plan-only':
            return new PlanOnlyPolicy();
        case 'full':
            return new FullPolicy();
    }
}
export function getTinkerPolicy() {
    const raw = process.env.TINKER_POLICY;
    if (raw === 'read-only' || raw === 'plan-only' || raw === 'full') {
        return raw;
    }
    return 'full'; // default to unrestricted
}
export function getPolicyForSession() {
    const policy = getTinkerPolicy();
    return getPolicyGate(policy);
}
/**
 * Check if an operation is allowed; throw if not.
 */
export function enforcePolicy(gate, operation) {
    const allowed = operation === 'read'
        ? gate.canRead()
        : operation === 'write'
            ? gate.canWrite()
            : operation === 'execute'
                ? gate.canExecute()
                : gate.canToolCall();
    if (!allowed) {
        const policy = getTinkerPolicy();
        throw new Error(`Operation '${operation}' is not allowed under TINKER_POLICY=${policy}. ${gate.describeLimitations().join(' ')}`);
    }
}
