// @lumiarq/lumis@1.2.0/dist/lib/tinker-memory.js
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import crypto from 'node:crypto';
const MAX_ENTRIES = 500;
export function getMemoryFilePath() {
    return path.join(os.homedir(), '.lumis', 'memory.json');
}
export function projectFingerprint(projectRoot) {
    return crypto.createHash('sha256').update(projectRoot).digest('hex').slice(0, 12);
}
export function loadMemoryStore() {
    const file = getMemoryFilePath();
    try {
        const raw = fs.readFileSync(file, 'utf8');
        const parsed = JSON.parse(raw);
        if (typeof parsed !== 'object' || parsed === null) {
            return { version: 1, entries: [] };
        }
        const obj = parsed;
        if (obj.version !== 1 || !Array.isArray(obj.entries)) {
            return { version: 1, entries: [] };
        }
        const entries = [];
        for (const item of obj.entries) {
            if (typeof item !== 'object' || item === null)
                continue;
            const e = item;
            if (typeof e.timestamp === 'string' &&
                typeof e.project === 'string' &&
                typeof e.user === 'string' &&
                typeof e.assistant === 'string' &&
                typeof e.summary === 'string') {
                entries.push({
                    timestamp: e.timestamp,
                    project: e.project,
                    user: e.user,
                    assistant: e.assistant,
                    summary: e.summary,
                });
            }
        }
        return { version: 1, entries };
    }
    catch {
        return { version: 1, entries: [] };
    }
}
export function saveMemoryStore(store) {
    const file = getMemoryFilePath();
    fs.mkdirSync(path.dirname(file), { recursive: true });
    fs.writeFileSync(file, JSON.stringify(store, null, 2), 'utf8');
}
export function appendMemory(store, projectRoot, user, assistant) {
    const trimmedUser = user.trim();
    const trimmedAssistant = assistant.trim();
    if (!trimmedUser || !trimmedAssistant)
        return store;
    const summary = summarizeTurn(trimmedUser, trimmedAssistant);
    const nextEntries = [...store.entries, {
            timestamp: new Date().toISOString(),
            project: projectFingerprint(projectRoot),
            user: trimmedUser,
            assistant: trimmedAssistant,
            summary,
        }];
    return {
        version: 1,
        entries: nextEntries.slice(-MAX_ENTRIES),
    };
}
export function clearMemoryStore() {
    return { version: 1, entries: [] };
}
export function retrieveMemoryContext(store, projectRoot, query, maxItems = 5) {
    const q = query.toLowerCase().trim();
    if (!q)
        return '';
    const projectId = projectFingerprint(projectRoot);
    const candidates = store.entries.filter((e) => e.project === projectId);
    if (candidates.length === 0)
        return '';
    const tokens = q.split(/\s+/).filter((t) => t.length >= 3);
    const scored = candidates
        .map((entry) => {
        const haystack = `${entry.summary}\n${entry.user}\n${entry.assistant}`.toLowerCase();
        const score = tokens.reduce((acc, token) => acc + (haystack.includes(token) ? 1 : 0), 0);
        return { entry, score };
    })
        .filter((x) => x.score > 0)
        .sort((a, b) => b.score - a.score)
        .slice(0, maxItems);
    if (scored.length === 0)
        return '';
    return [
        'Persistent memory context (previous related turns):',
        ...scored.map((s, idx) => `[m${idx + 1}] ${s.entry.summary}`),
    ].join('\n');
}
function summarizeTurn(user, assistant) {
    const u = user.replace(/\s+/g, ' ').trim();
    const a = assistant.replace(/\s+/g, ' ').trim();
    const uPart = u.length > 120 ? `${u.slice(0, 117)}...` : u;
    const aPart = a.length > 140 ? `${a.slice(0, 137)}...` : a;
    return `User: ${uPart} | Assistant: ${aPart}`;
}
