// @lumiarq/lumis@1.2.0/dist/lib/tinker-observer.js
/**
 * Tinker Observability
 *
 * Records agent turns, tool calls, and verification results to Trazze
 * and local structured logs for replay and debugging.
 */
/**
 * Default no-op observer.
 */
export const noopObserver = {
    recordTurn: async () => { },
    flush: async () => { },
};
/**
 * Composite observer that chains multiple observers.
 */
export class CompositeObserver {
    observers;
    constructor(observers) {
        this.observers = observers;
    }
    async recordTurn(event) {
        await Promise.all(this.observers.map((o) => o.recordTurn(event)));
    }
    async flush() {
        await Promise.all(this.observers.map((o) => o.flush()));
    }
}
/**
 * Trazze push observer — sends events to Trazze via HTTP/gRPC.
 * (Stub: full implementation depends on Trazze client availability.)
 */
export class TrazzeObserver {
    baseUrl;
    sessionId;
    constructor(baseUrl, sessionId) {
        this.baseUrl = baseUrl;
        this.sessionId = sessionId;
    }
    async recordTurn(event) {
        if (!process.env.TRAZZE_ENABLED)
            return;
        try {
            const response = await fetch(`${this.baseUrl}/events`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    event_type: 'tinker-turn',
                    payload: event,
                }),
            });
            if (!response.ok) {
                console.error(`[tinker-observer] Trazze push failed: ${response.status}`);
            }
        }
        catch (err) {
            console.error(`[tinker-observer] Trazze push error:`, err);
        }
    }
    async flush() {
        // Trazze flushes on its own schedule; this is a no-op.
    }
}
/**
 * Create an observer based on environment.
 */
export function createObserver(sessionId) {
    const observers = [];
    if (process.env.TRAZZE_ENABLED && process.env.TRAZZE_BASE_URL) {
        observers.push(new TrazzeObserver(process.env.TRAZZE_BASE_URL, sessionId));
    }
    if (observers.length === 0) {
        return noopObserver;
    }
    return new CompositeObserver(observers);
}
