// @illumiarq/http@1.0.0/dist/response/helpers.js
import { HTTP_STATUS_CODES } from './http-status-codes.js';
import { json } from './json.js';
function withStatus(status, options = {}) {
    return {
        ...options,
        status,
    };
}
export function ok(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.OK, options));
}
export function created(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.CREATED, options));
}
export function accepted(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.ACCEPTED, options));
}
export function noContent(options = {}) {
    return new Response(null, {
        status: HTTP_STATUS_CODES.NO_CONTENT,
        ...(options.headers ? { headers: options.headers } : {}),
    });
}
export function badRequest(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.BAD_REQUEST, options));
}
export function unauthorized(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.UNAUTHORIZED, options));
}
export function forbidden(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.FORBIDDEN, options));
}
export function notFound(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.NOT_FOUND, options));
}
export function unprocessableEntity(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.UNPROCESSABLE_ENTITY, options));
}
export function serverError(data, options = {}) {
    return json(data, withStatus(HTTP_STATUS_CODES.INTERNAL_SERVER_ERROR, options));
}
