// @lumiarq/lumis@1.2.0/dist/tools/index.js
import { readFileTool } from './read-file.js';
import { writeFileTool } from './write-file.js';
import { runCommandTool } from './run-command.js';
import { typecheckTool } from './typecheck.js';
import { runTestsTool } from './run-tests.js';
import { lintTool } from './lint.js';
import { gitDiffTool } from './git-diff.js';
export { rollbackLastWrite } from './write-file.js';
/**
 * Run all verification gates (typecheck + tests) and return a structured report.
 */
export async function runVerificationGates(projectRoot, gateLevel = 'standard', testPattern) {
    if (gateLevel === 'off') {
        return {
            passed: true,
            typecheckOutcome: 'skipped',
            testOutcome: 'skipped',
            errorCount: 0,
            warningCount: 0,
            summary: 'Verification gates disabled.',
            results: [],
        };
    }
    const results = [];
    const tcResult = await typecheckTool(projectRoot);
    results.push(tcResult);
    const testResult = await runTestsTool(projectRoot, testPattern);
    results.push(testResult);
    const allDiagnostics = results.flatMap((r) => r.diagnostics ?? []);
    const errorCount = allDiagnostics.filter((d) => d.severity === 'error').length;
    const warningCount = allDiagnostics.filter((d) => d.severity === 'warning').length;
    const typecheckOutcome = tcResult.outcome === 'success' ? 'pass' : 'fail';
    const testOutcome = testResult.outcome === 'success' ? 'pass' : 'fail';
    const failedGates = [];
    if (typecheckOutcome === 'fail')
        failedGates.push('typecheck');
    if (testOutcome === 'fail')
        failedGates.push('tests');
    const shouldFailOnWarnings = gateLevel === 'strict' && warningCount > 0;
    const passed = failedGates.length === 0 && !shouldFailOnWarnings;
    const summaryParts = [];
    summaryParts.push(`Typecheck: ${typecheckOutcome}`);
    summaryParts.push(`Tests: ${testOutcome}`);
    if (errorCount > 0)
        summaryParts.push(`${errorCount} error(s)`);
    if (warningCount > 0)
        summaryParts.push(`${warningCount} warning(s)`);
    if (!passed)
        summaryParts.push(`FAILED: ${failedGates.join(', ')}${shouldFailOnWarnings ? ', warnings (strict mode)' : ''}`);
    return {
        passed,
        typecheckOutcome,
        testOutcome,
        errorCount,
        warningCount,
        summary: summaryParts.join(' | '),
        results,
    };
}
/**
 * Execute a single ToolCall and return its ToolResult.
 */
export async function executeToolCall(call, projectRoot, writeLog) {
    switch (call.tool) {
        case 'read-file':
            return readFileTool(call.path, projectRoot);
        case 'write-file':
            return writeFileTool(call.path, call.content, projectRoot, writeLog);
        case 'run-command':
            return runCommandTool(call.cmd, call.cwd ?? projectRoot, call.timeoutMs);
        case 'typecheck':
            return typecheckTool(call.cwd ?? projectRoot);
        case 'run-tests':
            return runTestsTool(call.cwd ?? projectRoot, call.pattern);
        case 'lint':
            return lintTool(call.cwd ?? projectRoot, call.paths, call.fix);
        case 'git-diff':
            return gitDiffTool(projectRoot, call.base, call.path);
        case 'search-repo': {
            // Lightweight text grep — wraps ripgrep/grep; full RAG search is handled at chat layer
            const cmd = call.fileGlob
                ? ['grep', '-rn', '--include', call.fileGlob, '-m', String(call.maxResults ?? 20), call.query, '.']
                : ['grep', '-rn', '-m', String(call.maxResults ?? 20), call.query, '.'];
            return runCommandTool(cmd, projectRoot, 15_000);
        }
    }
}
