// @lumiarq/lumis@1.2.0/dist/tools/lint.js
import { runCommandTool } from './run-command.js';
/**
 * Parse eslint --format unix output into Diagnostic[].
 * Format: `path/to/file.ts:10:5: message [severity/rule]`
 */
function parseEslintOutput(output) {
    const diagnostics = [];
    const lineRegex = /^(.+):(\d+):(\d+):\s+(error|warning)\s+(.+?)(?:\s+(\S+\/\S+))?$/gm;
    let match;
    while ((match = lineRegex.exec(output)) !== null) {
        const [, file, line, column, severity, message, code] = match;
        if (!file || !line || !column || !severity || !message)
            continue;
        diagnostics.push({
            severity: severity,
            file: file.trim(),
            line: parseInt(line, 10),
            column: parseInt(column, 10),
            message: message.trim(),
            ...(code ? { code: code.trim() } : {}),
        });
    }
    return diagnostics;
}
export async function lintTool(projectRoot, paths = ['.'], fix = false) {
    const cmd = ['pnpm', 'lint', ...(fix ? ['--fix'] : []), ...paths];
    const result = await runCommandTool(cmd, projectRoot, 60_000);
    const diagnostics = parseEslintOutput(result.output);
    return {
        ...result,
        tool: 'lint',
        diagnostics,
    };
}
