// @lumiarq/lumis@1.2.0/dist/tools/read-file.js
import fs from 'node:fs/promises';
import path from 'node:path';
const MAX_BYTES = 250_000;
export async function readFileTool(filePath, projectRoot) {
    const start = Date.now();
    const absolute = path.isAbsolute(filePath) ? filePath : path.join(projectRoot, filePath);
    try {
        const stat = await fs.stat(absolute);
        if (!stat.isFile()) {
            return {
                tool: 'read-file',
                outcome: 'failure',
                output: '',
                error: `Not a file: ${filePath}`,
                durationMs: Date.now() - start,
            };
        }
        if (stat.size > MAX_BYTES) {
            return {
                tool: 'read-file',
                outcome: 'failure',
                output: '',
                error: `File too large (${stat.size} bytes, max ${MAX_BYTES})`,
                durationMs: Date.now() - start,
            };
        }
        const content = await fs.readFile(absolute, 'utf8');
        const lines = content.split('\n');
        const numbered = lines.map((line, i) => `${String(i + 1).padStart(4, ' ')}  ${line}`).join('\n');
        return {
            tool: 'read-file',
            outcome: 'success',
            output: numbered,
            durationMs: Date.now() - start,
        };
    }
    catch (error) {
        return {
            tool: 'read-file',
            outcome: 'failure',
            output: '',
            error: error instanceof Error ? error.message : String(error),
            durationMs: Date.now() - start,
        };
    }
}
