// @lumiarq/lumis@1.2.0/dist/tools/run-tests.js
import { runCommandTool } from './run-command.js';
/**
 * Parse vitest/jest output lines into Diagnostic[].
 * Captures FAIL file lines and lines like:
 *   ✗ test description  (AssertionError)
 *   × test description
 *   ● test description
 */
function parseTestOutput(output) {
    const diagnostics = [];
    // Lines starting with FAIL or failing test markers
    const failFileRegex = /^(FAIL|ERR!)\s+(.+)$/gm;
    let match;
    while ((match = failFileRegex.exec(output)) !== null) {
        const [, , file] = match;
        if (!file)
            continue;
        diagnostics.push({
            severity: 'error',
            file: file.trim(),
            message: `Test file failed: ${file.trim()}`,
        });
    }
    // Vitest × or jest ●
    const failTestRegex = /^\s+[×●✗]\s+(.+)$/gm;
    while ((match = failTestRegex.exec(output)) !== null) {
        const [, message] = match;
        if (!message)
            continue;
        diagnostics.push({ severity: 'error', message: message.trim() });
    }
    return diagnostics;
}
export async function runTestsTool(projectRoot, pattern) {
    const cmd = pattern
        ? ['pnpm', 'test', '--', '--reporter=verbose', pattern]
        : ['pnpm', 'test', '--', '--reporter=verbose'];
    const result = await runCommandTool(cmd, projectRoot, 120_000);
    const diagnostics = parseTestOutput(result.output);
    return {
        ...result,
        tool: 'run-tests',
        output: result.output.length > 12000
            ? result.output.slice(-12000) + '\n... (output truncated)'
            : result.output,
        diagnostics,
    };
}
