// @lumiarq/lumis@1.2.0/dist/tools/run-command.js
import { spawn } from 'node:child_process';
const DEFAULT_TIMEOUT_MS = 30_000;
const MAX_OUTPUT_CHARS = 40_000;
export async function runCommandTool(cmd, cwd, timeoutMs = DEFAULT_TIMEOUT_MS) {
    const start = Date.now();
    if (cmd.length === 0) {
        return { tool: 'run-command', outcome: 'failure', output: '', error: 'Empty command array', durationMs: 0 };
    }
    const executable = cmd[0];
    const args = cmd.slice(1);
    return new Promise((resolve) => {
        const chunks = [];
        let timedOut = false;
        const child = spawn(executable, args, {
            cwd,
            stdio: ['ignore', 'pipe', 'pipe'],
            shell: false,
        });
        const timer = setTimeout(() => {
            timedOut = true;
            child.kill('SIGTERM');
        }, timeoutMs);
        const onData = (data) => {
            const text = data.toString();
            chunks.push(text);
            // Truncate live to avoid holding huge buffers
            const total = chunks.reduce((acc, c) => acc + c.length, 0);
            if (total > MAX_OUTPUT_CHARS * 2) {
                chunks.splice(0, Math.floor(chunks.length / 2));
            }
        };
        child.stdout?.on('data', onData);
        child.stderr?.on('data', onData);
        child.on('close', (code) => {
            clearTimeout(timer);
            const rawOutput = chunks.join('');
            const output = rawOutput.length > MAX_OUTPUT_CHARS
                ? rawOutput.slice(-MAX_OUTPUT_CHARS) + `\n... (truncated, kept last ${MAX_OUTPUT_CHARS} chars)`
                : rawOutput;
            const durationMs = Date.now() - start;
            if (timedOut) {
                resolve({
                    tool: 'run-command',
                    outcome: 'failure',
                    output,
                    error: `Command timed out after ${timeoutMs}ms`,
                    durationMs,
                });
                return;
            }
            resolve({
                tool: 'run-command',
                outcome: code === 0 ? 'success' : 'failure',
                output,
                ...(code !== 0 ? { error: `Exit code ${code ?? 'null'}` } : {}),
                durationMs,
            });
        });
        child.on('error', (err) => {
            clearTimeout(timer);
            resolve({
                tool: 'run-command',
                outcome: 'failure',
                output: '',
                error: err.message,
                durationMs: Date.now() - start,
            });
        });
    });
}
