// @lumiarq/lumis@1.2.0/dist/tools/typecheck.js
import { runCommandTool } from './run-command.js';
/**
 * Parse tsc --noEmit output into structured Diagnostic[].
 * Handles the standard `file(line,col): error/warning TSxxxx: message` format.
 */
function parseTscOutput(output) {
    const diagnostics = [];
    // e.g.: src/foo.ts(10,5): error TS2322: Type 'string' is not assignable to type 'number'.
    const lineRegex = /^(.+?)\((\d+),(\d+)\):\s+(error|warning|info)\s+(TS\d+):\s+(.+)$/gm;
    let match;
    while ((match = lineRegex.exec(output)) !== null) {
        const [, file, line, column, severity, code, message] = match;
        if (!file || !line || !column || !severity || !message)
            continue;
        diagnostics.push({
            severity: severity,
            file: file.trim(),
            line: parseInt(line, 10),
            column: parseInt(column, 10),
            message: message.trim(),
            ...(code ? { code } : {}),
        });
    }
    return diagnostics;
}
export async function typecheckTool(projectRoot) {
    const result = await runCommandTool(['pnpm', 'typecheck'], projectRoot, 60_000);
    const diagnostics = parseTscOutput(result.output);
    const errorCount = diagnostics.filter((d) => d.severity === 'error').length;
    const warnCount = diagnostics.filter((d) => d.severity === 'warning').length;
    const summary = [
        result.outcome === 'success' ? 'Typecheck passed.' : `Typecheck failed.`,
        errorCount > 0 ? `${errorCount} error(s).` : '',
        warnCount > 0 ? `${warnCount} warning(s).` : '',
    ]
        .filter(Boolean)
        .join(' ');
    return {
        ...result,
        tool: 'typecheck',
        output: result.output.length > 8000
            ? result.output.slice(-8000) + '\n... (output truncated)'
            : result.output,
        diagnostics,
        ...(result.error ? { error: result.error + '\n' + summary } : {}),
    };
}
