// @lumiarq/lumis@1.2.0/dist/tools/write-file.js
import fs from 'node:fs/promises';
import path from 'node:path';
/**
 * Write `content` to `filePath` (relative to projectRoot or absolute).
 * Captures original content for rollback via the returned WriteRecord.
 */
export async function writeFileTool(filePath, content, projectRoot, writeLog) {
    const start = Date.now();
    const absolute = path.isAbsolute(filePath) ? filePath : path.join(projectRoot, filePath);
    try {
        let previousContent = null;
        try {
            previousContent = await fs.readFile(absolute, 'utf8');
        }
        catch {
            // File does not exist yet — previousContent stays null
        }
        await fs.mkdir(path.dirname(absolute), { recursive: true });
        await fs.writeFile(absolute, content, 'utf8');
        writeLog.push({ path: absolute, previousContent });
        return {
            tool: 'write-file',
            outcome: 'success',
            output: `Wrote ${content.length} chars to ${filePath}`,
            durationMs: Date.now() - start,
        };
    }
    catch (error) {
        return {
            tool: 'write-file',
            outcome: 'failure',
            output: '',
            error: error instanceof Error ? error.message : String(error),
            durationMs: Date.now() - start,
        };
    }
}
/**
 * Undo the last write captured by writeLog.
 */
export async function rollbackLastWrite(writeLog) {
    const record = writeLog.pop();
    if (!record)
        return null;
    if (record.previousContent === null) {
        await fs.unlink(record.path);
        return `Deleted new file: ${record.path}`;
    }
    await fs.writeFile(record.path, record.previousContent, 'utf8');
    return `Restored: ${record.path}`;
}
