// @illumiarq/lumis@1.2.8/dist/commands/app/command-runtime.js
import { spawnSync } from 'node:child_process';
import { existsSync, readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
export function readPackageJson(cwd) {
    const target = `${cwd}/package.json`;
    if (!existsSync(target)) {
        return {};
    }
    return JSON.parse(readFileSync(target, 'utf8'));
}
function bridgePath() {
    return fileURLToPath(new URL('../../bridges/project-bridge.ts', import.meta.url));
}
export function runBridge(args, cwd) {
    const result = spawnSync('pnpm', ['exec', 'tsx', bridgePath(), ...args], {
        cwd,
        encoding: 'utf8',
        env: { ...process.env, FORCE_COLOR: '1' },
    });
    return {
        status: result.status ?? 1,
        stdout: result.stdout ?? '',
        stderr: result.stderr ?? '',
    };
}
export function runPnpm(args, cwd, stdio = 'inherit') {
    const result = spawnSync('pnpm', args, {
        cwd,
        stdio,
        encoding: stdio === 'pipe' ? 'utf8' : undefined,
        env: { ...process.env, FORCE_COLOR: '1' },
    });
    if (stdio === 'pipe') {
        if (result.stdout)
            process.stdout.write(result.stdout);
        if (result.stderr)
            process.stderr.write(result.stderr);
    }
    return result.status ?? 1;
}
