// @lumiarq/pack-lumiarq@1.0.1/dist/ir/builder.js
import fs from 'node:fs';
import path from 'node:path';
import { LUMIARQ_FRAMEWORK } from '../frameworks/lumiarq.js';
import { buildComputedProperties } from './computed.js';
import { buildDependencyGraph } from './dependency-graph.js';
import { parseModule } from './module-parser.js';
import { getLumiArqNamingConventions } from './naming.js';
function getLayer(moduleRecord, layerName) {
    const layer = moduleRecord.layers[layerName];
    if (!layer) {
        throw new Error(`Missing layer "${layerName}" in module "${moduleRecord.name}"`);
    }
    return layer;
}
function applyLayerAliases(modules, framework) {
    if (!framework.layerAliases) {
        return modules;
    }
    return modules.map((moduleRecord) => {
        const aliasedLayers = { ...moduleRecord.layers };
        const aliases = framework.layerAliases ?? {};
        for (const [sourceLayer, aliasLayer] of Object.entries(aliases)) {
            if (!aliasLayer) {
                continue;
            }
            const sourceValue = aliasedLayers[sourceLayer];
            if (sourceValue && !aliasedLayers[aliasLayer]) {
                aliasedLayers[aliasLayer] = sourceValue;
            }
        }
        return {
            ...moduleRecord,
            layers: aliasedLayers,
        };
    });
}
function listModuleDirs(projectRoot) {
    const modulesRoot = path.join(projectRoot, 'src/modules');
    if (!fs.existsSync(modulesRoot)) {
        return [];
    }
    return fs
        .readdirSync(modulesRoot, { withFileTypes: true })
        .filter((entry) => entry.isDirectory())
        .map((entry) => path.join(modulesRoot, entry.name));
}
function extractRoutes(modules) {
    const routes = [];
    for (const moduleRecord of modules) {
        const httpLayer = getLayer(moduleRecord, 'http');
        for (const routeFile of httpLayer.routes) {
            if (routeFile.paths.length === 0) {
                routes.push({
                    module: moduleRecord.name,
                    path: '/',
                    method: 'GET',
                    handler: '',
                    loaders: [],
                    middleware: routeFile.middleware,
                    isApiRoute: routeFile.file.endsWith('api.route.ts'),
                    file: routeFile.file,
                });
                continue;
            }
            for (const routePath of routeFile.paths) {
                routes.push({
                    module: moduleRecord.name,
                    path: routePath,
                    method: 'GET',
                    handler: '',
                    loaders: [],
                    middleware: routeFile.middleware,
                    isApiRoute: routeFile.file.endsWith('api.route.ts'),
                    file: routeFile.file,
                });
            }
        }
    }
    return routes;
}
function extractSchemaRegistry(modules) {
    const registry = {};
    for (const moduleRecord of modules) {
        const databaseLayer = getLayer(moduleRecord, 'database');
        for (const schema of databaseLayer.schemas) {
            registry[`${schema.entity}Schema`] = {
                entity: schema.entity,
                module: moduleRecord.name,
                file: schema.file,
                fields: schema.fields,
                primaryKey: schema.fields.find((field) => field.primaryKey)?.name ?? 'id',
                foreignKeys: schema.foreignKeys,
                indexes: schema.indexes,
            };
        }
    }
    return registry;
}
function buildLayerBoundaries(modules, dependencyViolations) {
    const boundaries = {};
    for (const moduleRecord of modules) {
        boundaries[moduleRecord.name] = {
            contracts: { allowedImports: [], violations: [] },
            database: { allowedImports: ['contracts'], violations: [] },
            logic: { allowedImports: ['database', 'contracts'], violations: [] },
            events: { allowedImports: ['logic', 'contracts'], violations: [] },
            http: { allowedImports: ['logic', 'contracts'], violations: [] },
            ui: { allowedImports: ['http', 'contracts'], violations: [] },
        };
    }
    for (const violation of dependencyViolations) {
        const owningModule = modules.find((moduleRecord) => violation.from.includes(`/modules/${moduleRecord.name}/`));
        if (!owningModule) {
            continue;
        }
        boundaries[owningModule.name]?.logic?.violations.push({
            file: violation.from,
            importedFrom: violation.to,
            rule: violation.type,
        });
    }
    return boundaries;
}
export function buildIR(projectRoot, framework = LUMIARQ_FRAMEWORK) {
    const moduleDirs = listModuleDirs(projectRoot);
    const parsedModules = moduleDirs.map((moduleDir) => parseModule(projectRoot, moduleDir));
    const modules = applyLayerAliases(parsedModules, framework);
    const dependencies = buildDependencyGraph(modules);
    return {
        version: '1.0',
        projectRoot,
        framework: framework.name,
        frameworkDefinition: framework,
        builtAt: new Date(),
        modules,
        routes: extractRoutes(modules),
        layerBoundaries: buildLayerBoundaries(modules, dependencies.violations),
        naming: getLumiArqNamingConventions(),
        schemas: extractSchemaRegistry(modules),
        dependencies,
        computed: buildComputedProperties(modules, dependencies),
    };
}
