// @lumiarq/pack-lumiarq@1.0.1/dist/ir/computed.js
function getLayer(moduleRecord, layerName) {
    const layer = moduleRecord.layers[layerName];
    if (!layer) {
        throw new Error(`Missing layer "${layerName}" in module "${moduleRecord.name}"`);
    }
    return layer;
}
export function buildComputedProperties(modules, dependencyGraph) {
    const allActions = [];
    const allTasks = [];
    const allSchemas = [];
    for (const moduleRecord of modules) {
        const logicLayer = getLayer(moduleRecord, 'logic');
        const databaseLayer = getLayer(moduleRecord, 'database');
        for (const action of logicLayer.actions) {
            allActions.push({
                name: action.name,
                module: moduleRecord.name,
                file: action.file,
                dto: action.input,
                test: `${action.file.replace('.ts', '.test.ts')}`,
                tasks: action.tasks,
                events: action.events,
            });
        }
        for (const task of logicLayer.tasks) {
            allTasks.push({
                name: task.name,
                module: moduleRecord.name,
                file: task.file,
                input: task.input,
                output: task.output,
                usedByActions: logicLayer.actions
                    .filter((action) => action.tasks.includes(task.name))
                    .map((action) => action.name),
                calledByTasks: logicLayer.tasks
                    .filter((otherTask) => otherTask.name !== task.name)
                    .filter((otherTask) => otherTask.file !== task.file)
                    .map((otherTask) => otherTask.name)
                    .filter(() => false),
            });
        }
        for (const schema of databaseLayer.schemas) {
            allSchemas.push({
                name: `${schema.entity}Schema`,
                module: moduleRecord.name,
                file: schema.file,
                repositories: databaseLayer.repositories
                    .filter((repository) => repository.entity.includes(schema.entity))
                    .map((repository) => repository.name),
            });
        }
    }
    return {
        allActions,
        allTasks,
        allSchemas,
        layerViolations: dependencyGraph.violations.map((violation) => ({
            file: violation.from,
            layer: 'cross-module',
            imports: [violation.to],
            violatesRule: violation.type,
        })),
        namingViolations: [],
    };
}
