// @lumiarq/pack-lumiarq@1.0.1/dist/ir/dependency-graph.js
function detectTargetModule(source, moduleNames) {
    if (!source.startsWith('.')) {
        return null;
    }
    for (const moduleName of moduleNames) {
        const modulePattern = new RegExp(`(^|/)${moduleName}($|/)`);
        if (modulePattern.test(source)) {
            return moduleName;
        }
    }
    return null;
}
function isDirectInternalCrossModule(source) {
    return /\/(http|database|logic|events|contracts|ui|tests)\//.test(source);
}
function buildEdges(modules) {
    const moduleNames = new Set(modules.map((moduleRecord) => moduleRecord.name));
    const edges = [];
    for (const moduleRecord of modules) {
        for (const [fileName, importRecord] of Object.entries(moduleRecord.internalImports)) {
            for (const item of importRecord.imports) {
                if (item.type !== 'local') {
                    continue;
                }
                const targetModule = detectTargetModule(item.source, moduleNames);
                if (!targetModule || targetModule === moduleRecord.name) {
                    continue;
                }
                const via = isDirectInternalCrossModule(item.source) ? 'event-bus' : 'index.ts';
                edges.push({
                    from: moduleRecord.name,
                    to: targetModule,
                    via,
                    items: item.items,
                    source: fileName,
                });
            }
        }
    }
    return edges;
}
function detectCycles(modules, edges) {
    const adjacency = new Map();
    for (const moduleRecord of modules) {
        adjacency.set(moduleRecord.name, new Set());
    }
    for (const edge of edges) {
        adjacency.get(edge.from)?.add(edge.to);
    }
    const seenCycles = new Set();
    const cycles = [];
    function dfs(node, path, visiting) {
        if (visiting.has(node)) {
            const cycleStart = path.indexOf(node);
            if (cycleStart >= 0) {
                const cycleModules = path.slice(cycleStart);
                const key = [...cycleModules].sort().join('::');
                if (!seenCycles.has(key) && cycleModules.length > 1) {
                    seenCycles.add(key);
                    cycles.push({
                        modules: cycleModules,
                        via: 'import graph',
                    });
                }
            }
            return;
        }
        visiting.add(node);
        const nextNodes = adjacency.get(node);
        if (nextNodes) {
            for (const nextNode of nextNodes) {
                dfs(nextNode, [...path, node], visiting);
            }
        }
        visiting.delete(node);
    }
    for (const moduleRecord of modules) {
        dfs(moduleRecord.name, [], new Set());
    }
    return cycles;
}
export function buildDependencyGraph(modules) {
    const edges = buildEdges(modules);
    const graph = {
        modules: {},
        circularDependencies: [],
        violations: [],
    };
    for (const moduleRecord of modules) {
        graph.modules[moduleRecord.name] = {
            imports: [],
            exports: [
                ...moduleRecord.exports.actions,
                ...moduleRecord.exports.events,
                ...moduleRecord.exports.models,
                ...moduleRecord.exports.queries,
                ...moduleRecord.exports.types,
            ],
            eventBusListeners: [],
        };
    }
    for (const edge of edges) {
        graph.modules[edge.from]?.imports.push({
            module: edge.to,
            items: edge.items,
            via: edge.via,
        });
        if (edge.via === 'event-bus') {
            graph.violations.push({
                type: 'cross-module-direct-import',
                from: edge.source,
                to: edge.to,
                severity: 'error',
            });
        }
    }
    graph.circularDependencies = detectCycles(modules, edges);
    return graph;
}
