// @lumiarq/pack-lumiarq@1.0.1/dist/ir/module-parser.js
import fs from 'node:fs';
import path from 'node:path';
function readIfExists(filePath) {
    if (!fs.existsSync(filePath)) {
        return '';
    }
    return fs.readFileSync(filePath, 'utf8');
}
function listTsFiles(dirPath) {
    if (!fs.existsSync(dirPath)) {
        return [];
    }
    const entries = fs.readdirSync(dirPath, { withFileTypes: true });
    const files = [];
    for (const entry of entries) {
        const fullPath = path.join(dirPath, entry.name);
        if (entry.isDirectory()) {
            files.push(...listTsFiles(fullPath));
            continue;
        }
        if (entry.isFile() && fullPath.endsWith('.ts')) {
            files.push(fullPath);
        }
    }
    return files;
}
function fileToRelative(projectRoot, filePath) {
    return path.relative(projectRoot, filePath).replaceAll(path.sep, '/');
}
function uniqueMatches(content, regex) {
    const matches = new Set();
    for (const match of content.matchAll(regex)) {
        const value = match[1];
        if (value) {
            matches.add(value);
        }
    }
    return [...matches];
}
function extractClassName(content, fallbackFilePath) {
    const classMatch = content.match(/class\s+([A-Za-z0-9_]+)/);
    if (classMatch?.[1]) {
        return classMatch[1];
    }
    const base = path.basename(fallbackFilePath, '.ts');
    return base
        .split(/[.-]/)
        .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
        .join('');
}
function parseHttpLayer(projectRoot, moduleDir) {
    const handlersDir = path.join(moduleDir, 'http/handlers');
    const loadersDir = path.join(moduleDir, 'http/loaders');
    const routesDir = path.join(moduleDir, 'http/routes');
    const handlers = listTsFiles(handlersDir).map((filePath) => {
        const content = readIfExists(filePath);
        const action = content.match(/\b([A-Za-z0-9_]+Action)\b/)?.[1] ?? '';
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            action,
        };
    });
    const loaders = listTsFiles(loadersDir).map((filePath) => {
        const content = readIfExists(filePath);
        const queries = uniqueMatches(content, /\b([A-Za-z0-9_]+Query)\b/g);
        const schema = content.match(/\b([A-Za-z0-9_]+Schema)\b/)?.[1] ?? '';
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            queries,
            schema,
        };
    });
    const routes = listTsFiles(routesDir).map((filePath) => {
        const content = readIfExists(filePath);
        const paths = uniqueMatches(content, /path:\s*'([^']+)'/g);
        const middleware = uniqueMatches(content, /middleware:\s*'([^']+)'/g);
        return {
            file: fileToRelative(projectRoot, filePath),
            paths,
            middleware,
        };
    });
    return { handlers, loaders, routes };
}
function parseSchemaFields(content) {
    const fields = [];
    const fieldRegex = /\{\s*name:\s*'([^']+)'\s*,\s*columnName:\s*'([^']+)'\s*,\s*type:\s*'([^']+)'\s*,\s*nullable:\s*(true|false)\s*,\s*unique:\s*(true|false)\s*,\s*primaryKey:\s*(true|false)(?:\s*,\s*default:\s*([^,\n}]+))?\s*\}/g;
    for (const match of content.matchAll(fieldRegex)) {
        const typeValue = match[3];
        if (typeValue !== 'string' &&
            typeValue !== 'number' &&
            typeValue !== 'boolean' &&
            typeValue !== 'date' &&
            typeValue !== 'uuid' &&
            typeValue !== 'json') {
            continue;
        }
        const parsedField = {
            name: match[1] ?? '',
            columnName: match[2] ?? '',
            type: typeValue,
            nullable: match[4] === 'true',
            unique: match[5] === 'true',
            primaryKey: match[6] === 'true',
        };
        const maybeDefault = match[7]?.trim();
        if (maybeDefault) {
            parsedField.default = maybeDefault;
        }
        fields.push(parsedField);
    }
    return fields;
}
function parseDatabaseLayer(projectRoot, moduleDir) {
    const schemaFiles = listTsFiles(path.join(moduleDir, 'database/schemas'));
    const repositoryFiles = listTsFiles(path.join(moduleDir, 'database/repositories'));
    const factoryFiles = listTsFiles(path.join(moduleDir, 'database/factories'));
    const schemas = schemaFiles.map((filePath) => {
        const content = readIfExists(filePath);
        return {
            entity: content.match(/entity:\s*'([^']+)'/)?.[1] ?? extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            fields: parseSchemaFields(content),
            indexes: [],
            foreignKeys: [],
        };
    });
    const repositories = repositoryFiles.map((filePath) => {
        const content = readIfExists(filePath);
        const name = extractClassName(content, filePath);
        return {
            name,
            file: fileToRelative(projectRoot, filePath),
            entity: name.replace(/Repository$/, ''),
            methods: uniqueMatches(content, /(?:async\s+)?([A-Za-z0-9_]+)\s*\(/g)
                .filter((method) => method !== 'constructor')
                .map((method) => ({ name: method, returnType: 'unknown' })),
        };
    });
    const factories = factoryFiles.map((filePath) => {
        const content = readIfExists(filePath);
        const name = extractClassName(content, filePath);
        return {
            name,
            file: fileToRelative(projectRoot, filePath),
            entity: name.replace(/Factory$/, ''),
        };
    });
    return { schemas, repositories, factories };
}
function parseLogicLayer(projectRoot, moduleDir) {
    const actionFiles = listTsFiles(path.join(moduleDir, 'logic/actions'));
    const taskFiles = listTsFiles(path.join(moduleDir, 'logic/tasks'));
    const queryFiles = listTsFiles(path.join(moduleDir, 'logic/queries'));
    const policyFiles = listTsFiles(path.join(moduleDir, 'logic/policies'));
    const actions = actionFiles.map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            input: content.match(/execute\(\s*input:\s*([^\)]+)/)?.[1]?.trim() ?? 'unknown',
            tasks: uniqueMatches(content, /\b([A-Za-z0-9_]+Task)\b/g),
            events: uniqueMatches(content, /\b([A-Za-z0-9_]+Event)\b/g),
            queries: uniqueMatches(content, /\b([A-Za-z0-9_]+Query)\b/g),
        };
    });
    const tasks = taskFiles.map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            input: content.match(/run\(\s*([^\):]+)/)?.[1]?.trim() ?? 'unknown',
            output: 'unknown',
            callsRepository: /\b[A-Za-z0-9_]+Repository\b/.test(content),
            callsAction: /\b[A-Za-z0-9_]+Action\b/.test(content),
            crossModule: /\.\.\/.+\/[A-Z][A-Za-z0-9_]+\/logic\//.test(content),
        };
    });
    const queries = queryFiles.map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            input: content.match(/execute\(\s*input:\s*([^\)]+)/)?.[1]?.trim() ?? 'unknown',
            output: 'unknown',
            repository: content.match(/\b([A-Za-z0-9_]+Repository)\b/)?.[1] ?? '',
        };
    });
    const policies = policyFiles.map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            abilities: uniqueMatches(content, /\n\s*([A-Za-z0-9_]+)\(/g)
                .filter((name) => name !== 'constructor')
                .map((name) => ({ name })),
        };
    });
    return { actions, tasks, queries, policies };
}
function kebabToPascal(input) {
    return input
        .split('-')
        .filter(Boolean)
        .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
        .join('');
}
function parseEventsLayer(projectRoot, moduleDir) {
    const definitions = listTsFiles(path.join(moduleDir, 'events/definitions')).map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: content.match(/name:\s*'([^']+)'/)?.[1] ?? extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            schema: content.match(/schema:\s*'([^']+)'/)?.[1] ?? '',
            emittedBy: uniqueMatches(content, /\b([A-Za-z0-9_]+Action)\b/g),
        };
    });
    const listeners = listTsFiles(path.join(moduleDir, 'events/listeners')).map((filePath) => {
        const content = readIfExists(filePath);
        const base = path.basename(filePath).replace('.listener.ts', '');
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            listensTo: base.startsWith('on-') ? kebabToPascal(base.slice(3)) : kebabToPascal(base),
            tasks: uniqueMatches(content, /\b([A-Za-z0-9_]+Task)\b/g),
            isIdempotent: /idempotent/i.test(content),
        };
    });
    return { definitions, listeners };
}
function parseContractsLayer(projectRoot, moduleDir) {
    const models = listTsFiles(path.join(moduleDir, 'contracts/models')).map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: content.match(/(?:interface|type)\s+([A-Za-z0-9_]+)/)?.[1] ?? extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            fields: [...content.matchAll(/\n\s*([A-Za-z0-9_]+)\??:\s*([^\n;]+)/g)].map((match) => ({
                name: match[1] ?? '',
                type: (match[2] ?? '').trim(),
                nullable: (match[2] ?? '').includes('null'),
            })),
        };
    });
    const types = listTsFiles(path.join(moduleDir, 'contracts/types')).map((filePath) => {
        const content = readIfExists(filePath);
        const base = path.basename(filePath);
        let kind = 'dto';
        if (base.includes('.response.')) {
            kind = 'response';
        }
        else if (base.includes('.enum.')) {
            kind = 'enum';
        }
        else if (base.includes('.branded.')) {
            kind = 'branded';
        }
        return {
            name: content.match(/(?:type|enum|interface)\s+([A-Za-z0-9_]+)/)?.[1] ?? extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            kind,
        };
    });
    const validators = listTsFiles(path.join(moduleDir, 'contracts/validators')).map((filePath) => {
        const content = readIfExists(filePath);
        const name = content.match(/(?:const|class)\s+([A-Za-z0-9_]+)/)?.[1] ?? extractClassName(content, filePath);
        return {
            name,
            file: fileToRelative(projectRoot, filePath),
            schema: `${name}Schema`,
        };
    });
    return { models, types, validators };
}
function parseUiLayer(projectRoot, moduleDir) {
    const pages = listTsFiles(path.join(moduleDir, 'ui/web/pages')).map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            framework: 'react',
            loaders: uniqueMatches(content, /\b([A-Za-z0-9_]+Loader)\b/g),
        };
    });
    const components = listTsFiles(path.join(moduleDir, 'ui/web/components')).map((filePath) => {
        const content = readIfExists(filePath);
        return {
            name: extractClassName(content, filePath),
            file: fileToRelative(projectRoot, filePath),
            framework: 'react',
        };
    });
    if (pages.length === 0 && components.length === 0) {
        return undefined;
    }
    return { pages, components };
}
function parseTestsLayer(projectRoot, moduleDir) {
    const unit = listTsFiles(path.join(moduleDir, 'tests/unit')).map((filePath) => ({
        target: path.basename(filePath, '.ts'),
        file: fileToRelative(projectRoot, filePath),
    }));
    const integration = listTsFiles(path.join(moduleDir, 'tests/integration')).map((filePath) => ({
        target: path.basename(filePath, '.ts'),
        file: fileToRelative(projectRoot, filePath),
    }));
    const features = listTsFiles(path.join(moduleDir, 'tests/features')).map((filePath) => ({
        name: path.basename(filePath, '.ts'),
        file: fileToRelative(projectRoot, filePath),
    }));
    return { unit, integration, features };
}
function parseExports(indexFile) {
    const content = readIfExists(indexFile);
    const names = uniqueMatches(content, /export\s*\{\s*([A-Za-z0-9_]+)\s*\}/g);
    return {
        actions: names.filter((name) => name.endsWith('Action')),
        queries: names.filter((name) => name.endsWith('Query')),
        events: names.filter((name) => name.endsWith('Event')),
        types: names.filter((name) => name.endsWith('DTO') || name.endsWith('Type')),
        models: names.filter((name) => name.endsWith('Model')),
    };
}
export function parseInternalImports(projectRoot, moduleDir) {
    const files = listTsFiles(moduleDir);
    const map = {};
    for (const filePath of files) {
        const content = readIfExists(filePath);
        const imports = [];
        for (const match of content.matchAll(/import\s+([^;]+?)\s+from\s+['"]([^'"]+)['"]/g)) {
            const clause = (match[1] ?? '').trim();
            const source = (match[2] ?? '').trim();
            const items = uniqueMatches(clause, /([A-Za-z0-9_]+)/g);
            let type = 'external';
            if (source.startsWith('.')) {
                type = 'local';
            }
            else if (source.startsWith('@lumiarq/')) {
                type = 'framework';
            }
            imports.push({ type, source, items });
        }
        map[fileToRelative(projectRoot, filePath)] = { imports };
    }
    return map;
}
export function parseModule(projectRoot, moduleDir) {
    const name = path.basename(moduleDir);
    const ui = parseUiLayer(projectRoot, moduleDir);
    const layers = {
        http: parseHttpLayer(projectRoot, moduleDir),
        database: parseDatabaseLayer(projectRoot, moduleDir),
        logic: parseLogicLayer(projectRoot, moduleDir),
        events: parseEventsLayer(projectRoot, moduleDir),
        contracts: parseContractsLayer(projectRoot, moduleDir),
        tests: parseTestsLayer(projectRoot, moduleDir),
    };
    if (ui) {
        layers.ui = ui;
    }
    return {
        name,
        directory: fileToRelative(projectRoot, moduleDir),
        layers,
        artifacts: [],
        metadata: {
            adapter: 'lumiarq',
        },
        publicExports: [],
        exports: parseExports(path.join(moduleDir, 'index.ts')),
        internalImports: parseInternalImports(projectRoot, moduleDir),
        externalImports: {},
    };
}
