// @lumiarq/pack-lumiarq@1.0.1/dist/ir/naming.js
export function getLumiArqNamingConventions() {
    return {
        patterns: {
            action: {
                className: '{{Verb}}{{Entity}}Action',
                fileName: '{{verb}}-{{entity}}.action.ts',
                directory: 'logic/actions/',
            },
            task: {
                className: '{{Verb}}{{Entity}}Task',
                fileName: '{{verb}}-{{entity}}.task.ts',
                directory: 'logic/tasks/',
            },
            handler: {
                className: '{{Verb}}{{Entity}}Handler',
                fileName: '{{verb}}-{{entity}}.handler.ts',
                directory: 'http/handlers/',
            },
            loader: {
                className: 'Get{{Entity}}Loader',
                fileName: 'get-{{entity}}.loader.ts',
                directory: 'http/loaders/',
            },
            query: {
                className: 'Get{{Entity}}Query',
                fileName: 'get-{{entity}}.query.ts',
                directory: 'logic/queries/',
            },
            repository: {
                className: '{{Entity}}Repository',
                fileName: '{{entity}}.repository.ts',
                directory: 'database/repositories/',
            },
            schema: {
                className: '{{Entity}}Schema',
                fileName: '{{entity}}.schema.ts',
                directory: 'database/schemas/',
            },
            event: {
                className: '{{Entity}}{{PastTense}}Event',
                fileName: '{{entity}}-{{past-tense}}.event.ts',
                directory: 'events/definitions/',
            },
            validator: {
                className: '{{Verb}}{{Entity}}Validator',
                fileName: '{{verb}}-{{entity}}.validator.ts',
                directory: 'contracts/validators/',
            },
        },
        violations: [],
    };
}
