// @lumiarq/planner@1.0.1/dist/core-planner.js
import { topologicalSortCore } from './dag-core.js';
export class PlannerCycleError extends Error {
    constructor(message) {
        super(message);
        this.name = 'PlannerCycleError';
    }
}
function stablePlanId(intent, steps) {
    const payload = JSON.stringify({
        action: intent.action,
        raw: intent.raw,
        subjects: intent.subjects,
        constraints: intent.constraints,
        steps: steps.map((step) => ({
            id: step.id,
            operation: step.operation,
            artifactKind: step.artifactKind,
            module: step.module,
            name: step.name,
            targetFile: step.targetFile,
            templateName: step.templateName,
            dependsOn: step.dependsOn,
        })),
    });
    let hash = 0;
    for (let i = 0; i < payload.length; i += 1) {
        hash = (hash * 31 + payload.charCodeAt(i)) >>> 0;
    }
    return `plan_${hash.toString(16).padStart(8, '0')}`;
}
function assertNoCycle(steps) {
    const byId = new Map(steps.map((step) => [step.id, step]));
    const visiting = new Set();
    const visited = new Set();
    const visit = (id) => {
        if (visiting.has(id)) {
            throw new PlannerCycleError(`Cycle detected in execution plan at step "${id}".`);
        }
        if (visited.has(id)) {
            return;
        }
        visiting.add(id);
        const step = byId.get(id);
        if (step) {
            for (const dep of step.dependsOn) {
                if (byId.has(dep)) {
                    visit(dep);
                }
            }
        }
        visiting.delete(id);
        visited.add(id);
    };
    for (const step of steps) {
        visit(step.id);
    }
}
export async function createPlanWithAdapter(intent, ir, adapter) {
    const resolvedSteps = await adapter.resolveIntent(intent, ir);
    assertNoCycle(resolvedSteps);
    const steps = topologicalSortCore(resolvedSteps);
    return {
        id: stablePlanId(intent, steps),
        intent: intent.raw,
        steps,
        estimatedFiles: steps.filter((step) => step.operation === 'create').length,
    };
}
