// @lumiarq/planner@1.0.1/dist/plan-lock.js
// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------
export class PlanLockMismatchError extends Error {
    expected;
    actual;
    constructor(expected, actual) {
        super(`Plan lock mismatch — the plan content changed since it was locked.\n` +
            `  locked hash : ${expected}\n` +
            `  current hash: ${actual}\n` +
            `Re-generate the plan or update the lock to proceed.`);
        this.expected = expected;
        this.actual = actual;
        this.name = 'PlanLockMismatchError';
    }
}
// ---------------------------------------------------------------------------
// Hash
// ---------------------------------------------------------------------------
function stepFingerprint(step) {
    return {
        id: step.id,
        operation: step.operation,
        artifactKind: step.artifactKind,
        module: step.module,
        name: step.name,
        targetFile: step.targetFile,
        templateName: step.templateName,
        tokens: step.tokens,
        dependsOn: [...step.dependsOn].sort(),
    };
}
/**
 * Deterministic content hash over a plan's intent and steps.
 * The same plan content always produces the same hash regardless of
 * when `lockPlan` was called.
 */
export function planContentHash(plan) {
    const payload = JSON.stringify({
        intent: plan.intent,
        steps: plan.steps.map(stepFingerprint),
    });
    let hash = 0;
    for (let i = 0; i < payload.length; i += 1) {
        hash = (hash * 31 + payload.charCodeAt(i)) >>> 0;
    }
    return `lock_${hash.toString(16).padStart(8, '0')}`;
}
// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------
/**
 * Stamp a plan with a content-addressed lock.
 *
 * The returned plan is identical to the input except for the `planLock` field.
 * Calling `lockPlan` on the same plan content always produces the same hash,
 * making CI artefacts deterministic.
 *
 * @param plan - The plan to lock.
 * @returns New plan object with `planLock` populated.
 */
export function lockPlan(plan) {
    const lock = {
        hash: planContentHash(plan),
        lockedAt: new Date().toISOString(),
    };
    return { ...plan, planLock: lock };
}
/**
 * Verify that a locked plan's content has not drifted since it was locked.
 *
 * Call this before executing a plan when reproducibility is required (e.g. CI).
 * Throws `PlanLockMismatchError` if the hash does not match.
 *
 * No-op when the plan has no `planLock`.
 *
 * @param plan - The plan to validate.
 */
export function assertPlanUnchanged(plan) {
    if (!plan.planLock)
        return;
    const current = planContentHash(plan);
    if (current !== plan.planLock.hash) {
        throw new PlanLockMismatchError(plan.planLock.hash, current);
    }
}
