// @lumiarq/pack-laravel@1.0.1/dist/templates/index.js
import { createTemplateStubLoader, resolveDirectoryFromImportMeta } from '@lumiarq/context';
const LARAVEL_COMMAND_TEMPLATE_MAP = {
    migration: 'laravel-migration-post-step',
};
const loader = createTemplateStubLoader({
    namespace: 'Laravel',
    stubsDir: resolveDirectoryFromImportMeta(import.meta.url, 'stubs'),
    projectOverrideDirs: ['stubs/lumis/laravel', 'stubs/lumis'],
    templateToStubFile: {
        'laravel-migration-post-step': 'laravel-migration-post-step.stub',
    },
});
export function loadLaravelTemplate(templateName) {
    return loader.loadTemplate(templateName);
}
/**
 * Resolve the template name used for a normalized Laravel command kind.
 *
 * @param kind - Canonical command kind accepted by the Laravel adapter.
 * @returns Template catalog key used by `loadLaravelTemplate`.
 */
export function resolveLaravelTemplateName(kind) {
    return LARAVEL_COMMAND_TEMPLATE_MAP[kind];
}
export function listLaravelTemplateNames() {
    return loader.listTemplateNames();
}
