// @illumiarq/cache@1.1.0/dist/drivers/redis.js
async function createRedisClient(opts) {
    let IoRedis;
    try {
        IoRedis = (await import('ioredis'));
    }
    catch {
        throw new Error('[RedisCacheDriver] ioredis is not installed. Run: pnpm add ioredis');
    }
    // Handle both default export and named export shapes
    const Ctor = typeof IoRedis === 'function' ? IoRedis : IoRedis.default;
    return new Ctor({
        host: opts.host,
        port: opts.port,
        password: opts.password,
        db: opts.db ?? 0,
        keyPrefix: opts.keyPrefix ?? '',
    });
}
export class RedisCacheDriver {
    opts;
    client = null;
    constructor(opts) {
        this.opts = opts;
    }
    async getClient() {
        if (!this.client) {
            this.client = await createRedisClient(this.opts);
        }
        return this.client;
    }
    async get(key) {
        const redis = await this.getClient();
        const raw = await redis.get(key);
        if (raw === null)
            return null;
        try {
            return JSON.parse(raw);
        }
        catch {
            return raw;
        }
    }
    async set(key, value, ttl) {
        const redis = await this.getClient();
        const serialized = JSON.stringify(value);
        if (ttl && ttl > 0) {
            await redis.set(key, serialized, 'EX', ttl);
        }
        else {
            await redis.set(key, serialized);
        }
    }
    async forget(key) {
        const redis = await this.getClient();
        await redis.del(key);
    }
    async remember(key, ttl, fn) {
        const existing = await this.get(key);
        if (existing !== null)
            return existing;
        const value = await fn();
        await this.set(key, value, ttl);
        return value;
    }
    /** Cleanly close the Redis connection. */
    async disconnect() {
        if (this.client) {
            await this.client.quit();
            this.client = null;
        }
    }
}
