// @illumiarq/http@1.1.0/dist/request/form-request.js
/**
 * Formats a ZodError into a flat `{ field: string[] }` errors map.
 * Nested fields are dot-joined (e.g. `"address.city": ["Required"]`).
 */
function formatZodError(err) {
    const errors = {};
    for (const issue of err.issues) {
        const path = issue.path.join('.') || '_root';
        (errors[path] ??= []).push(issue.message);
    }
    return errors;
}
/** Produces a 422 Response with `{ message, errors }` JSON body. */
function validationResponse(errors) {
    return new Response(JSON.stringify({ message: 'Validation failed.', errors }), {
        status: 422,
        headers: { 'Content-Type': 'application/json' },
    });
}
/**
 * Parse and validate the request body as JSON.
 *
 * @throws {Response} 422 Unprocessable Entity if validation fails.
 * @throws {Response} 400 Bad Request if body is not valid JSON.
 */
export async function parseBody(req, schema) {
    let raw;
    try {
        raw = await req.json();
    }
    catch {
        throw new Response(JSON.stringify({ message: 'Request body must be valid JSON.' }), {
            status: 400,
            headers: { 'Content-Type': 'application/json' },
        });
    }
    const result = schema.safeParse(raw);
    if (!result.success) {
        throw validationResponse(formatZodError(result.error));
    }
    return result.data;
}
/**
 * Parse and validate URL query parameters.
 * All query values arrive as strings — use Zod's `.coerce` for numbers/booleans.
 *
 * @throws {Response} 422 Unprocessable Entity if validation fails.
 *
 * @example
 * const ListSchema = z.object({
 *   page: z.coerce.number().int().positive().default(1),
 *   limit: z.coerce.number().int().max(100).default(20),
 * });
 * const query = await parseQuery(req, ListSchema);
 */
export async function parseQuery(req, schema) {
    const url = new URL(req.url);
    const raw = {};
    for (const [key, value] of url.searchParams.entries()) {
        const existing = raw[key];
        if (existing === undefined) {
            raw[key] = value;
        }
        else if (Array.isArray(existing)) {
            existing.push(value);
        }
        else {
            raw[key] = [existing, value];
        }
    }
    const result = schema.safeParse(raw);
    if (!result.success) {
        throw validationResponse(formatZodError(result.error));
    }
    return result.data;
}
