// @illumiarq/http@1.1.0/dist/request/helpers.js
import { HttpRequest } from './request.js';
export function query(req, name) {
    return new HttpRequest(req).query(name);
}
export function params(req, routeParams = {}, name) {
    return new HttpRequest(req, routeParams).params(name);
}
export function body(req) {
    return new HttpRequest(req).body();
}
function appendField(target, key, value) {
    const existing = target[key];
    if (existing === undefined) {
        target[key] = value;
        return;
    }
    if (Array.isArray(existing)) {
        existing.push(value);
        return;
    }
    target[key] = [existing, value];
}
function parseMultipartTextFields(body, contentType) {
    const parsed = {};
    const boundaryMatch = contentType.match(/boundary=(?:"([^"]+)"|([^;]+))/i);
    const boundary = boundaryMatch?.[1] ?? boundaryMatch?.[2];
    if (!boundary)
        return parsed;
    for (const rawPart of body.split(`--${boundary}`)) {
        const part = rawPart.trim();
        if (!part || part === '--')
            continue;
        const normalized = rawPart.replace(/^\r\n/, '').replace(/\r\n$/, '');
        const separatorIndex = normalized.indexOf('\r\n\r\n');
        if (separatorIndex === -1)
            continue;
        const headerBlock = normalized.slice(0, separatorIndex);
        if (/filename=/i.test(headerBlock))
            continue;
        const fieldName = headerBlock.match(/name="([^"]+)"/i)?.[1];
        if (!fieldName)
            continue;
        const value = normalized.slice(separatorIndex + 4).replace(/\r\n$/, '');
        appendField(parsed, fieldName, value);
    }
    return parsed;
}
async function readInputBody(req) {
    const contentType = req.headers.get('content-type') ?? '';
    if (contentType.includes('application/json')) {
        try {
            const parsed = await req.clone().json();
            if (parsed !== null && typeof parsed === 'object' && !Array.isArray(parsed)) {
                return parsed;
            }
        }
        catch {
            return {};
        }
        return {};
    }
    if (contentType.includes('application/x-www-form-urlencoded')) {
        try {
            const text = await req.clone().text();
            const params = new URLSearchParams(text);
            const parsed = {};
            for (const key of new Set(params.keys())) {
                const values = params.getAll(key);
                parsed[key] = values.length === 1 ? values[0] : values;
            }
            return parsed;
        }
        catch {
            return {};
        }
    }
    if (contentType.includes('multipart/form-data')) {
        try {
            const text = await req.clone().text();
            return parseMultipartTextFields(text, contentType);
        }
        catch {
            return {};
        }
    }
    return {};
}
/**
 * Reads merged input from query string and request body.
 * When a key is provided, returns that value (query takes precedence, then body, then fallback).
 * When no key is provided, returns the full merged input object.
 */
export async function input(req, key, fallback) {
    const url = new URL(req.url);
    const queryData = {};
    for (const k of url.searchParams.keys()) {
        const all = url.searchParams.getAll(k);
        queryData[k] = all.length === 1 ? all[0] : all;
    }
    const bodyData = await readInputBody(req);
    const merged = { ...bodyData, ...queryData };
    if (key === undefined)
        return merged;
    return Object.prototype.hasOwnProperty.call(merged, key) ? merged[key] : fallback;
}
/** Returns the value of a named request header, or `fallback` if absent. */
export function header(req, name, fallback) {
    return req.headers.get(name) ?? fallback;
}
/**
 * Extracts the Bearer token from the Authorization header.
 * Returns `null` if the header is absent or uses a different scheme.
 */
export function bearerToken(req) {
    const auth = req.headers.get('authorization');
    if (!auth?.startsWith('Bearer '))
        return null;
    return auth.slice(7);
}
/** Returns `true` if the client signals it accepts a JSON response (via Accept header). */
export function wantsJson(req) {
    const accept = req.headers.get('accept') ?? '';
    return accept.includes('application/json');
}
/** Returns `true` if the request body is JSON (via Content-Type header). */
export function expectsJson(req) {
    const contentType = req.headers.get('content-type') ?? '';
    return contentType.includes('application/json');
}
/** Returns the full URL string of the request. */
export function fullUrl(req) {
    return req.url;
}
/**
 * Returns the full URL with the given query parameters merged in.
 * Existing params with the same key are overwritten.
 */
export function fullUrlWithQuery(req, patch) {
    const url = new URL(req.url);
    for (const [k, v] of Object.entries(patch)) {
        url.searchParams.set(k, v);
    }
    return url.toString();
}
