// @lumiarq/pack-django@1.0.1/dist/templates/index.js
import { createTemplateStubLoader, resolveDirectoryFromImportMeta } from '@lumiarq/context';
const DJANGO_COMMAND_TEMPLATE_MAP = {
    view: 'django-view',
    migration: 'django-migration-post-step',
};
const loader = createTemplateStubLoader({
    namespace: 'Django',
    stubsDir: resolveDirectoryFromImportMeta(import.meta.url, 'stubs'),
    projectOverrideDirs: ['stubs/lumis/django', 'stubs/lumis'],
    templateToStubFile: {
        'django-view': 'django-view.stub',
        'django-migration-post-step': 'django-migration-post-step.stub',
    },
});
export function loadDjangoTemplate(templateName) {
    return loader.loadTemplate(templateName);
}
/**
 * Resolve the template name used for a normalized Django command kind.
 *
 * @param kind - Canonical command kind resolved during Django planning.
 * @returns Template catalog key used by `loadDjangoTemplate`.
 */
export function resolveDjangoTemplateName(kind) {
    return DJANGO_COMMAND_TEMPLATE_MAP[kind];
}
export function listDjangoTemplateNames() {
    return loader.listTemplateNames();
}
