// @lumiarq/rag@0.1.0/dist/grounding/index.js
/**
 * Creates a `RAGGenerator` backed by a plain async generation function.
 *
 * The generator injects retrieved chunk text as numbered sources in the system
 * prompt, calls the provided `generate` function, and extracts `CitationSpan[]`
 * by matching source numbers in the response.
 *
 * Usage:
 * ```ts
 * const generator = createGroundingGenerator(
 *   async (sys, user) => myAIGateway.complete({ task: 'answer-ground', systemPrompt: sys, prompt: user }).then(r => r.text),
 * )
 * ```
 */
export function createGroundingGenerator(generate, opts = {}, streamGenerate) {
    return {
        async generate(query, chunks) {
            const start = Date.now();
            const { systemPrompt, contextChunks } = buildGroundingPrompt(query, chunks, opts);
            const answerText = await generate(systemPrompt, query);
            const citations = extractCitations(answerText, contextChunks);
            const usage = {
                embedTokens: 0, // filled in by pipeline
                generationTokens: Math.ceil(answerText.length / 4),
                chunksRetrieved: chunks.length,
                chunksAfterRerank: contextChunks.length,
                latencyMs: Date.now() - start,
            };
            return { text: answerText, citations, chunks: contextChunks, usage };
        },
        async *generateStream(query, chunks) {
            if (!streamGenerate) {
                // Fallback: complete and yield as single chunk
                const result = await this.generate(query, chunks);
                yield result.text;
                return;
            }
            const { systemPrompt, contextChunks: _contextChunks } = buildGroundingPrompt(query, chunks, opts);
            yield* streamGenerate(systemPrompt, query);
        },
    };
}
// ── Helpers ───────────────────────────────────────────────────────────────────
function buildGroundingPrompt(_query, chunks, opts) {
    const maxCtx = opts.maxContextChunks ?? chunks.length;
    const contextChunks = chunks.slice(0, maxCtx);
    const prefix = opts.systemPromptPrefix ??
        'You are a helpful assistant. Answer the user question using ONLY the provided sources. ' +
            'For every claim in your answer, include a citation in the format [N] where N is the source number.';
    const sources = contextChunks
        .map((c, i) => `[${i + 1}] ${c.chunk.text.trim()}`)
        .join('\n\n');
    const systemPrompt = `${prefix}\n\nSources:\n${sources}`;
    return { systemPrompt, contextChunks };
}
/**
 * Extract CitationSpan[] from an answer that contains [N] markers.
 * Matches patterns like [1], [2], [1,2], [1, 2], [1, 2, 3].
 * Each number in a multi-citation bracket is expanded into its own CitationSpan.
 */
function extractCitations(text, chunks) {
    const citations = [];
    // Match brackets containing one or more comma-separated numbers (with optional spaces)
    const citationRegex = /\[(\d+(?:\s*,\s*\d+)*)\]/g;
    let match;
    while ((match = citationRegex.exec(text)) !== null) {
        const raw = match[1];
        const numbers = raw.split(',').map((s) => parseInt(s.trim(), 10));
        for (const num of numbers) {
            const n = num - 1; // 0-based index
            const chunk = chunks[n];
            if (!chunk)
                continue;
            citations.push({
                chunkId: chunk.chunk.id,
                documentId: chunk.chunk.documentId,
                text: chunk.chunk.text.slice(0, 200), // first 200 chars of source
                score: chunk.score,
                startChar: match.index,
                endChar: match.index + match[0].length,
            });
        }
    }
    // Deduplicate by chunkId (keep first occurrence)
    const seen = new Set();
    return citations.filter((c) => {
        if (seen.has(c.chunkId))
            return false;
        seen.add(c.chunkId);
        return true;
    });
}
