// @illumiarq/runtime@1.1.2/dist/mail/resend-mailer.js
async function createResendClient(apiKey) {
    let ResendModule;
    try {
        ResendModule = (await import('resend'));
    }
    catch {
        throw new Error('[ResendMailer] resend is not installed. Run: pnpm add resend');
    }
    return new ResendModule.Resend(apiKey);
}
export class ResendMailer {
    opts;
    client = null;
    constructor(opts) {
        this.opts = opts;
    }
    async getClient() {
        if (!this.client) {
            this.client = await createResendClient(this.opts.apiKey);
        }
        return this.client;
    }
    async send(message) {
        const resend = await this.getClient();
        const result = await resend.emails.send({
            from: message.from ?? this.opts.from,
            to: message.to,
            cc: message.cc,
            bcc: message.bcc,
            reply_to: message.replyTo,
            subject: message.subject,
            // Template rendering would produce html here in a full implementation.
            text: `Template: ${message.template}\nPayload: ${JSON.stringify(message.payload)}`,
        });
        if (result.error) {
            throw new Error(`[ResendMailer] send failed: ${result.error.message}`);
        }
    }
    async queue(message) {
        if (!this.opts.queue) {
            return this.send(message);
        }
        await this.opts.queue.dispatch({ name: 'mail.send', data: message }, { queue: 'mail' });
    }
}
