// @illumiarq/runtime@1.1.2/dist/mail/smtp-mailer.js
async function createTransporter(config) {
    let nodemailer;
    try {
        nodemailer = await import('nodemailer');
    }
    catch {
        throw new Error('[SMTPMailer] nodemailer is not installed. Run: pnpm add nodemailer && pnpm add -D @types/nodemailer');
    }
    return nodemailer.createTransport({
        host: config.host,
        port: config.port,
        secure: config.secure,
        auth: { user: config.user, pass: config.pass },
    });
}
export class SMTPMailer {
    opts;
    transporter = null;
    constructor(opts) {
        this.opts = opts;
    }
    async getTransporter() {
        if (!this.transporter) {
            this.transporter = await createTransporter(this.opts.config);
        }
        return this.transporter;
    }
    async send(message) {
        const transport = await this.getTransporter();
        await transport.sendMail({
            from: message.from ?? this.opts.from,
            to: Array.isArray(message.to) ? message.to.join(', ') : message.to,
            cc: message.cc,
            bcc: message.bcc,
            replyTo: message.replyTo,
            subject: message.subject,
            // In a full implementation, template would be rendered to HTML here.
            // For now, pass the template name and payload as plain text for debugging.
            text: `Template: ${message.template}\nPayload: ${JSON.stringify(message.payload)}`,
        });
    }
    async queue(message) {
        if (!this.opts.queue) {
            // No queue configured — fall back to sending immediately
            return this.send(message);
        }
        await this.opts.queue.dispatch({ name: 'mail.send', data: message }, { queue: 'mail' });
    }
}
