// @lumiarq/ai@1.2.0/dist/config/errors.js
/**
 * Structured error types for @lumiarq/ai.
 * All errors extend Error with a `code` discriminant for programmatic handling.
 */
export class AITimeoutError extends Error {
    providerId;
    timeoutMs;
    code = 'AI_TIMEOUT';
    constructor(providerId, timeoutMs) {
        super(`AI request to provider "${providerId}" timed out after ${timeoutMs}ms`);
        this.providerId = providerId;
        this.timeoutMs = timeoutMs;
        this.name = 'AITimeoutError';
    }
}
export class AIProviderError extends Error {
    providerId;
    status;
    code = 'AI_PROVIDER_ERROR';
    constructor(providerId, message, status) {
        super(message);
        this.providerId = providerId;
        this.status = status;
        this.name = 'AIProviderError';
    }
}
export class AIAllProvidersFailedError extends Error {
    attempted;
    code = 'AI_ALL_PROVIDERS_FAILED';
    constructor(attempted) {
        super(`All AI providers failed. Attempted: ${attempted.join(', ')}`);
        this.attempted = attempted;
        this.name = 'AIAllProvidersFailedError';
    }
}
export class AINoEmbeddingProviderError extends Error {
    code = 'AI_NO_EMBEDDING_PROVIDER';
    constructor() {
        super('No AI provider with embedding support is registered in this gateway.');
        this.name = 'AINoEmbeddingProviderError';
    }
}
export class AINoProvidersError extends Error {
    code = 'AI_NO_PROVIDERS';
    constructor() {
        super('No AI providers registered in gateway.');
        this.name = 'AINoProvidersError';
    }
}
export class AINoModelForTaskError extends Error {
    code = 'AI_NO_MODEL_FOR_TASK';
    constructor(task, availableProviders) {
        super(`No model available for task "${task}" among providers: [${availableProviders.join(', ')}]. ` +
            'Add the required provider or configure taskModels for this task.');
        this.name = 'AINoModelForTaskError';
    }
}
export class AIMiddlewareError extends Error {
    middlewareIndex;
    phase;
    code = 'AI_MIDDLEWARE_ERROR';
    constructor(middlewareIndex, phase, cause) {
        const msg = cause instanceof Error ? cause.message : String(cause);
        super(`Middleware[${middlewareIndex}] ${phase} hook threw: ${msg}`);
        this.middlewareIndex = middlewareIndex;
        this.phase = phase;
        this.name = 'AIMiddlewareError';
        if (cause instanceof Error)
            this.cause = cause;
    }
}
