// @lumiarq/ai@1.2.0/dist/config/models.js
/**
 * Centralized model registry.
 *
 * All built-in default models are defined here.  Users can extend or override
 * any entry by passing `modelPolicy.taskModels` to `createAIGateway()` or
 * `createGatewayFromEnv()`, or by supplying a custom `ModelRegistry` to
 * `createGatewayFromEnv({ models: myRegistry })`.
 *
 * Provider-prefixed keys (`openai:gpt-4.1`) are used so the gateway's
 * `parseModelRef()` can route them to the correct provider automatically.
 */
// ── Per-provider default models ────────────────────────────────────────────────
/**
 * Full model catalog per provider, organized by task.
 * Values are bare model names (no provider prefix).
 */
export const PROVIDER_DEFAULT_MODELS = {
    openai: {
        'intent-parse': 'gpt-4.1-mini',
        'plan-propose': 'gpt-4.1',
        'token-suggest': 'gpt-4.1-mini',
        'template-variant': 'gpt-4.1',
        embed: 'text-embedding-3-small',
        rerank: 'gpt-4.1-mini',
        'query-rewrite': 'gpt-4.1-mini',
        'answer-ground': 'gpt-4.1',
    },
    anthropic: {
        'intent-parse': 'claude-haiku-4-5',
        'plan-propose': 'claude-sonnet-4-5',
        'token-suggest': 'claude-haiku-4-5',
        'template-variant': 'claude-sonnet-4-5',
        embed: 'claude-haiku-4-5',
        rerank: 'claude-haiku-4-5',
        'query-rewrite': 'claude-haiku-4-5',
        'answer-ground': 'claude-sonnet-4-5',
    },
    gemini: {
        'intent-parse': 'gemini-2.0-flash',
        'plan-propose': 'gemini-2.5-pro',
        'token-suggest': 'gemini-2.0-flash',
        'template-variant': 'gemini-2.5-pro',
        embed: 'text-embedding-004',
        rerank: 'gemini-2.0-flash',
        'query-rewrite': 'gemini-2.0-flash',
        'answer-ground': 'gemini-2.5-pro',
    },
    /** Groq — free-tier, ultra-fast inference via OpenAI-compatible API. */
    groq: {
        'intent-parse': 'llama-3.3-70b-versatile',
        'plan-propose': 'llama-3.3-70b-versatile',
        'token-suggest': 'llama-3.1-8b-instant',
        'template-variant': 'llama-3.3-70b-versatile',
        rerank: 'llama-3.3-70b-versatile',
        'query-rewrite': 'llama-3.1-8b-instant',
        'answer-ground': 'llama-3.3-70b-versatile',
    },
    /** OpenRouter — unified gateway to 200+ models. */
    openrouter: {
        'intent-parse': 'anthropic/claude-haiku-4-5',
        'plan-propose': 'anthropic/claude-sonnet-4-5',
        'token-suggest': 'meta-llama/llama-3.1-8b-instruct:free',
        'template-variant': 'openai/gpt-4.1',
        rerank: 'meta-llama/llama-3.1-8b-instruct:free',
        'query-rewrite': 'meta-llama/llama-3.1-8b-instruct:free',
        'answer-ground': 'anthropic/claude-sonnet-4-5',
    },
    /** DeepSeek — OpenAI-compatible, cost-effective reasoning models. */
    deepseek: {
        'intent-parse': 'deepseek-chat',
        'plan-propose': 'deepseek-reasoner',
        'token-suggest': 'deepseek-chat',
        'template-variant': 'deepseek-chat',
        rerank: 'deepseek-chat',
        'query-rewrite': 'deepseek-chat',
        'answer-ground': 'deepseek-chat',
    },
    /** Mistral — European sovereign AI, OpenAI-compatible. */
    mistral: {
        'intent-parse': 'mistral-small-latest',
        'plan-propose': 'mistral-large-latest',
        'token-suggest': 'mistral-small-latest',
        'template-variant': 'mistral-large-latest',
        embed: 'mistral-embed',
        rerank: 'mistral-small-latest',
        'query-rewrite': 'mistral-small-latest',
        'answer-ground': 'mistral-large-latest',
    },
    /** Cohere — strong at RAG tasks (rerank, embed). */
    cohere: {
        'intent-parse': 'command-r-plus',
        'plan-propose': 'command-r-plus',
        'token-suggest': 'command-r',
        'template-variant': 'command-r-plus',
        embed: 'embed-english-v3.0',
        rerank: 'rerank-english-v3.0',
        'query-rewrite': 'command-r',
        'answer-ground': 'command-r-plus',
    },
    /** Ollama — local LLM runtime (user must have Ollama running). */
    ollama: {
        'intent-parse': 'llama3.2',
        'plan-propose': 'llama3.2',
        'token-suggest': 'llama3.2',
        'template-variant': 'llama3.2',
        embed: 'nomic-embed-text',
        rerank: 'llama3.2',
        'query-rewrite': 'llama3.2',
        'answer-ground': 'llama3.2',
    },
    /** Local deterministic fallback — zero-latency, zero-cost rule-based responses. */
    local: {
        'intent-parse': 'deterministic',
        'plan-propose': 'deterministic',
        'token-suggest': 'deterministic',
        'template-variant': 'deterministic',
        rerank: 'deterministic',
        'query-rewrite': 'deterministic',
        'answer-ground': 'deterministic',
    },
};
/**
 * Well-known model names grouped by provider for documentation and auto-complete.
 * These are the most capable and cost-effective choices per provider.
 * Not exhaustive — providers frequently release new models.
 */
export const KNOWN_MODELS = {
    openai: {
        /** Latest frontier model — best for complex reasoning and planning. */
        GPT_4_1: 'gpt-4.1',
        /** Fast + cheap — ideal for classification, parsing, and suggestions. */
        GPT_4_1_MINI: 'gpt-4.1-mini',
        /** Most cost-effective embedding model. */
        TEXT_EMBEDDING_3_SMALL: 'text-embedding-3-small',
        /** Highest-quality embedding model. */
        TEXT_EMBEDDING_3_LARGE: 'text-embedding-3-large',
    },
    anthropic: {
        CLAUDE_OPUS_4_5: 'claude-opus-4-5',
        CLAUDE_SONNET_4_5: 'claude-sonnet-4-5',
        CLAUDE_HAIKU_4_5: 'claude-haiku-4-5',
    },
    gemini: {
        GEMINI_2_5_PRO: 'gemini-2.5-pro',
        GEMINI_2_0_FLASH: 'gemini-2.0-flash',
        TEXT_EMBEDDING_004: 'text-embedding-004',
    },
    groq: {
        LLAMA_3_3_70B: 'llama-3.3-70b-versatile',
        LLAMA_3_1_8B_INSTANT: 'llama-3.1-8b-instant',
        MIXTRAL_8X7B: 'mixtral-8x7b-32768',
    },
    deepseek: {
        DEEPSEEK_CHAT: 'deepseek-chat',
        DEEPSEEK_REASONER: 'deepseek-reasoner',
    },
    mistral: {
        MISTRAL_LARGE: 'mistral-large-latest',
        MISTRAL_SMALL: 'mistral-small-latest',
        MISTRAL_EMBED: 'mistral-embed',
    },
    openrouter: {
        /** Free Llama 3.1 8B via OpenRouter — good for low-cost tasks. */
        LLAMA_3_1_8B_FREE: 'meta-llama/llama-3.1-8b-instruct:free',
        /** Claude Haiku via OpenRouter. */
        CLAUDE_HAIKU: 'anthropic/claude-haiku-4-5',
    },
};
/**
 * Merge a user-supplied registry on top of the built-in defaults.
 * User values win at the task level.
 */
export function mergeModelRegistry(user) {
    if (!user)
        return { ...PROVIDER_DEFAULT_MODELS };
    const merged = { ...PROVIDER_DEFAULT_MODELS };
    for (const [providerId, tasks] of Object.entries(user)) {
        if (tasks) {
            merged[providerId] = { ...merged[providerId], ...tasks };
        }
    }
    return merged;
}
