// @lumiarq/ai@1.2.0/dist/providers/anthropic.js
import { AI_PROVIDER_ENDPOINT_PATH, AI_PROVIDER_URL_PATH } from '../config/paths.js';
function mapStopReason(raw) {
    const map = {
        end_turn: 'stop',
        max_tokens: 'length',
        stop_sequence: 'stop',
        tool_use: 'tool-calls',
    };
    return map[raw] ?? 'unknown';
}
function buildAnthropicBody(request, model) {
    if (request.messages && request.messages.length > 0) {
        const systemParts = request.messages.filter((m) => m.role === 'system').map((m) => m.content);
        const msgs = request.messages
            .filter((m) => m.role !== 'system')
            .map((m) => ({ role: m.role, content: m.content }));
        return {
            model,
            max_tokens: 4096,
            messages: msgs,
            ...(systemParts.length > 0 ? { system: systemParts.join('\n\n') } : {}),
            ...(request.temperature !== undefined ? { temperature: request.temperature } : {}),
        };
    }
    return {
        model,
        max_tokens: 4096,
        messages: [{ role: 'user', content: request.prompt }],
        ...(request.systemPrompt && request.systemPrompt.trim().length > 0 ? { system: request.systemPrompt } : {}),
        ...(request.temperature !== undefined ? { temperature: request.temperature } : {}),
    };
}
const TASK_DEFAULT_MODELS = {
    'intent-parse': 'claude-haiku-4-5',
    'plan-propose': 'claude-sonnet-4-5',
    'token-suggest': 'claude-haiku-4-5',
    'template-variant': 'claude-sonnet-4-5',
    'embed': 'claude-haiku-4-5',
    'rerank': 'claude-haiku-4-5',
    'query-rewrite': 'claude-haiku-4-5',
    'answer-ground': 'claude-sonnet-4-5',
};
function resolveModel(task, model, options) {
    const stripped = model?.startsWith('anthropic:') ? model.slice('anthropic:'.length).trim() : model?.trim();
    if (stripped && stripped.length > 0 && stripped !== model) {
        return stripped;
    }
    if (stripped && stripped.length > 0 && !model?.includes(':')) {
        return stripped;
    }
    const taskSpecific = options.taskModels?.[task];
    if (taskSpecific && taskSpecific.trim().length > 0) {
        return taskSpecific.trim();
    }
    if (options.defaultModel && options.defaultModel.trim().length > 0) {
        return options.defaultModel.trim();
    }
    return TASK_DEFAULT_MODELS[task];
}
export function createAnthropicProvider(options) {
    const baseUrl = options.baseUrl ?? AI_PROVIDER_URL_PATH.anthropic;
    return {
        id: 'anthropic',
        complete: async (request) => {
            const model = resolveModel(request.task, request.model, options);
            const body = buildAnthropicBody(request, model);
            const response = await fetch(`${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.anthropic.messages}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'x-api-key': options.apiKey,
                    'anthropic-version': '2023-06-01',
                },
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok) {
                const text = await response.text().catch(() => '');
                throw new Error(`Anthropic request failed (${response.status}): ${text}`);
            }
            const json = (await response.json());
            const content = json.content.find((block) => block.type === 'text')?.text ?? '';
            const result = {
                provider: 'anthropic',
                model: json.model,
                text: content,
                ...(json.stop_reason !== undefined ? { finishReason: mapStopReason(json.stop_reason) } : {}),
            };
            if (json.usage) {
                const usage = {};
                if (json.usage.input_tokens !== undefined)
                    usage.inputTokens = json.usage.input_tokens;
                if (json.usage.output_tokens !== undefined)
                    usage.outputTokens = json.usage.output_tokens;
                result.usage = usage;
            }
            return result;
        },
        async *completeStream(request) {
            const model = resolveModel(request.task, request.model, options);
            const body = { ...buildAnthropicBody(request, model), stream: true };
            const response = await fetch(`${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.anthropic.messages}`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'x-api-key': options.apiKey,
                    'anthropic-version': '2023-06-01',
                },
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok || !response.body) {
                const text = await response.text().catch(() => '');
                throw new Error(`Anthropic streaming request failed (${response.status}): ${text}`);
            }
            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            let buffer = '';
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed.startsWith('data: '))
                        continue;
                    try {
                        const parsed = JSON.parse(trimmed.slice(6));
                        if (parsed.type === 'content_block_delta' && parsed.delta?.type === 'text_delta') {
                            const chunk = parsed.delta.text;
                            if (typeof chunk === 'string')
                                yield chunk;
                        }
                    }
                    catch {
                        // skip malformed SSE chunks
                    }
                }
            }
        },
    };
}
