// @lumiarq/ai@1.2.0/dist/providers/gemini.js
import { AI_PROVIDER_ENDPOINT_PATH, AI_PROVIDER_URL_PATH } from '../config/paths.js';
function mapGeminiFinishReason(raw) {
    const map = {
        STOP: 'stop',
        MAX_TOKENS: 'length',
        SAFETY: 'content-filter',
        OTHER: 'unknown',
    };
    return map[raw] ?? 'unknown';
}
function buildGeminiBody(request) {
    if (request.messages && request.messages.length > 0) {
        const systemMessages = request.messages.filter((m) => m.role === 'system');
        const contents = request.messages
            .filter((m) => m.role !== 'system')
            .map((m) => ({
            role: m.role === 'assistant' ? 'model' : 'user',
            parts: [{ text: m.content }],
        }));
        return {
            contents,
            ...(systemMessages.length > 0
                ? { systemInstruction: { parts: systemMessages.map((m) => ({ text: m.content })) } }
                : {}),
            ...(request.temperature !== undefined
                ? { generationConfig: { temperature: request.temperature, maxOutputTokens: 8192 } }
                : {}),
        };
    }
    return {
        contents: [{ role: 'user', parts: [{ text: request.prompt }] }],
        ...(request.systemPrompt && request.systemPrompt.trim().length > 0
            ? { systemInstruction: { parts: [{ text: request.systemPrompt }] } }
            : {}),
        ...(request.temperature !== undefined
            ? { generationConfig: { temperature: request.temperature, maxOutputTokens: 8192 } }
            : {}),
    };
}
const TASK_DEFAULT_MODELS = {
    'intent-parse': 'gemini-2.0-flash',
    'plan-propose': 'gemini-2.5-pro',
    'token-suggest': 'gemini-2.0-flash',
    'template-variant': 'gemini-2.5-pro',
    'embed': 'text-embedding-004',
    'rerank': 'gemini-2.0-flash',
    'query-rewrite': 'gemini-2.0-flash',
    'answer-ground': 'gemini-2.5-pro',
};
function resolveModel(task, model, options) {
    const stripped = model?.startsWith('gemini:') ? model.slice('gemini:'.length).trim() : model?.trim();
    if (stripped && stripped.length > 0 && stripped !== model) {
        return stripped;
    }
    if (stripped && stripped.length > 0 && !model?.includes(':')) {
        return stripped;
    }
    const taskSpecific = options.taskModels?.[task];
    if (taskSpecific && taskSpecific.trim().length > 0) {
        return taskSpecific.trim();
    }
    if (options.defaultModel && options.defaultModel.trim().length > 0) {
        return options.defaultModel.trim();
    }
    return TASK_DEFAULT_MODELS[task];
}
/**
 * Gemini provider — Google's advanced language models.
 *
 * https://ai.google.dev/gemini
 * Env: GEMINI_API_KEY or GOOGLE_AI_API_KEY
 *
 * @param options Configuration options for the Gemini provider.
 * @returns An AIProvider instance configured for Gemini.
 *
 * Supports both chat completions and embeddings via the same API, with provider-prefixed model names
 * (e.g. `gemini:text-embedding-004`) or auto-resolved models based on task type.
 */
export function createGeminiProvider(options) {
    const baseUrl = options.baseUrl ?? AI_PROVIDER_URL_PATH.gemini;
    return {
        id: 'gemini',
        complete: async (request) => {
            const model = resolveModel(request.task, request.model, options);
            const url = `${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.gemini.modelBase}/${model}${AI_PROVIDER_ENDPOINT_PATH.gemini.generateContent}?key=${options.apiKey}`;
            const body = buildGeminiBody(request);
            const response = await fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok) {
                const text = await response.text().catch(() => '');
                throw new Error(`Gemini request failed (${response.status}): ${text}`);
            }
            const json = (await response.json());
            const content = json.candidates[0]?.content.parts.map((p) => p.text).join('') ?? '';
            const resolvedModel = json.modelVersion ?? model;
            const rawFinish = json.candidates[0]?.finishReason ?? 'unknown';
            const result = {
                provider: 'gemini',
                model: resolvedModel,
                text: content,
                finishReason: mapGeminiFinishReason(rawFinish),
            };
            if (json.usageMetadata) {
                const usage = {};
                if (json.usageMetadata.promptTokenCount !== undefined)
                    usage.inputTokens = json.usageMetadata.promptTokenCount;
                if (json.usageMetadata.candidatesTokenCount !== undefined)
                    usage.outputTokens = json.usageMetadata.candidatesTokenCount;
                result.usage = usage;
            }
            return result;
        },
        async *completeStream(request) {
            const model = resolveModel(request.task, request.model, options);
            const url = `${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.gemini.modelBase}/${model}${AI_PROVIDER_ENDPOINT_PATH.gemini.streamGenerateContent}?key=${options.apiKey}&alt=sse`;
            const body = buildGeminiBody(request);
            const response = await fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok || !response.body) {
                const text = await response.text().catch(() => '');
                throw new Error(`Gemini streaming request failed (${response.status}): ${text}`);
            }
            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            let buffer = '';
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed.startsWith('data: '))
                        continue;
                    try {
                        const parsed = JSON.parse(trimmed.slice(6));
                        const chunk = parsed.candidates?.[0]?.content.parts.map((p) => p.text).join('') ?? '';
                        if (chunk)
                            yield chunk;
                    }
                    catch {
                        // skip malformed SSE chunks
                    }
                }
            }
        },
        async embed(request) {
            const model = resolveModel('embed', request.model, options);
            const inputs = Array.isArray(request.input) ? request.input : [request.input];
            const url = `${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.gemini.modelBase}/${model}${AI_PROVIDER_ENDPOINT_PATH.gemini.batchEmbedContents}?key=${options.apiKey}`;
            const body = {
                requests: inputs.map((text) => ({
                    model: `models/${model}`,
                    content: { parts: [{ text }] },
                })),
            };
            const response = await fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok) {
                const text = await response.text().catch(() => '');
                throw new Error(`Gemini embed request failed (${response.status}): ${text}`);
            }
            const json = (await response.json());
            return {
                embeddings: json.embeddings.map((e) => e.values),
                provider: 'gemini',
                model,
            };
        },
    };
}
