// @lumiarq/ai@1.2.0/dist/providers/index.js
/**
 * Provider factory.
 *
 * Creates the appropriate AIProvider implementation from available credentials.
 * Falls back to the deterministic provider when no credentials are found.
 * Zero runtime dependencies — each adapter uses the global `fetch` built into Node 18+.
 */
import { createDeterministicProvider } from '../deterministic.js';
import { createAIGateway } from '../gateway.js';
import { PROVIDER_DEFAULT_MODELS, mergeModelRegistry } from '../config/models.js';
import { createOpenAIProvider } from './openai.js';
import { createAnthropicProvider } from './anthropic.js';
import { createGeminiProvider } from './gemini.js';
import { createOllamaProvider } from './ollama.js';
import { createOpenRouterProvider } from './openrouter.js';
import { createDeepSeekProvider } from './deepseek.js';
import { createMistralProvider } from './mistral.js';
import { AI_PROVIDER_URL_PATH } from '../config/paths.js';
/**
 * Create the available provider instances from environment API keys.
 * Each provider is only registered when its key is present and non-empty.
 * A deterministic fallback provider is always included.
 */
export function createProvidersFromEnv(options = {}) {
    const env = options.env ?? process.env;
    const providers = [];
    const openaiKey = env['OPENAI_API_KEY'];
    if (openaiKey && openaiKey.trim().length > 0) {
        providers.push(createOpenAIProvider({ apiKey: openaiKey.trim() }));
    }
    const anthropicKey = env['ANTHROPIC_API_KEY'];
    if (anthropicKey && anthropicKey.trim().length > 0) {
        providers.push(createAnthropicProvider({ apiKey: anthropicKey.trim() }));
    }
    const geminiKey = env['GEMINI_API_KEY'] ?? env['GOOGLE_AI_API_KEY'];
    if (geminiKey && geminiKey.trim().length > 0) {
        providers.push(createGeminiProvider({ apiKey: geminiKey.trim() }));
    }
    // Groq — OpenAI-compatible, free tier, fast inference.
    const groqKey = env['GROQ_API_KEY'];
    if (groqKey && groqKey.trim().length > 0) {
        const groqTaskModels = PROVIDER_DEFAULT_MODELS['groq'];
        providers.push(createOpenAIProvider({
            id: 'groq',
            apiKey: groqKey.trim(),
            baseUrl: AI_PROVIDER_URL_PATH.groq,
            defaultModel: groqTaskModels?.['token-suggest'] ?? 'llama-3.3-70b-versatile',
            ...(groqTaskModels !== undefined && { taskModels: groqTaskModels }),
        }));
    }
    // OpenRouter — gateway to 200+ models, including free-tier community models.
    const openrouterKey = env['OPENROUTER_API_KEY'];
    if (openrouterKey && openrouterKey.trim().length > 0) {
        const openrouterAppUrl = env['OPENROUTER_APP_URL'];
        const openrouterAppTitle = env['OPENROUTER_APP_TITLE'];
        providers.push(createOpenRouterProvider({
            apiKey: openrouterKey.trim(),
            ...(openrouterAppUrl && openrouterAppUrl.trim().length > 0 ? { appUrl: openrouterAppUrl.trim() } : {}),
            ...(openrouterAppTitle && openrouterAppTitle.trim().length > 0 ? { appTitle: openrouterAppTitle.trim() } : {}),
        }));
    }
    // DeepSeek — cost-effective reasoning models.
    const deepseekKey = env['DEEPSEEK_API_KEY'];
    if (deepseekKey && deepseekKey.trim().length > 0) {
        providers.push(createDeepSeekProvider({ apiKey: deepseekKey.trim() }));
    }
    // Mistral AI.
    const mistralKey = env['MISTRAL_API_KEY'];
    if (mistralKey && mistralKey.trim().length > 0) {
        providers.push(createMistralProvider({ apiKey: mistralKey.trim() }));
    }
    // Ollama — local LLM runtime. No API key required; detected by presence of OLLAMA_BASE_URL
    // or (on non-production) always added with the default localhost URL.
    const ollamaUrl = env['OLLAMA_BASE_URL'];
    if (ollamaUrl && ollamaUrl.trim().length > 0) {
        providers.push(createOllamaProvider({ baseUrl: ollamaUrl.trim() }));
    }
    // Always include the local deterministic fallback.
    providers.push(createDeterministicProvider());
    return providers;
}
/**
 * Infer the best default provider from the set of available providers.
 * Preference: groq (free) → openrouter → anthropic → openai → deepseek → mistral → gemini → ollama → local
 * Free / open-weight tier first for cost-conscious users.
 */
export function inferDefaultProvider(providers) {
    const ids = new Set(providers.map((p) => p.id));
    const preference = ['groq', 'openrouter', 'anthropic', 'openai', 'deepseek', 'mistral', 'gemini', 'ollama', 'local'];
    for (const preferred of preference) {
        if (ids.has(preferred))
            return preferred;
    }
    return 'local';
}
/**
 * The default task-to-model routing table for `mode=auto`.
 * Sourced from the centralized model registry in `models.ts`.
 * Developers can override any entry via `modelPolicy.taskModels` or `CreateGatewayFromEnvOptions.models`.
 */
export const AUTO_POLICY_TASK_MODELS = PROVIDER_DEFAULT_MODELS;
/**
 * Build task model entries for the given provider prefixed with `provider:model`
 * so the gateway's `parseModelRef` routes them correctly.
 */
function buildPrefixedTaskModels(providerId, registry) {
    const providerTable = registry[providerId];
    if (!providerTable)
        return {};
    const out = {};
    for (const [task, model] of Object.entries(providerTable)) {
        if (model)
            out[task] = `${providerId}:${model}`;
    }
    return out;
}
/**
 * Create an AIGateway fully wired from environment credentials.
 * In `mode=auto`, task-specific models are selected per provider automatically.
 * Callers may pass `modelPolicy` to override any auto-selected entry, or
 * `models` to supply provider-level model overrides merged on top of the defaults.
 */
export function createGatewayFromEnv(options = {}) {
    const envProviders = options.env ? createProvidersFromEnv({ env: options.env }) : createProvidersFromEnv();
    const allProviders = options.extraProviders ? [...envProviders, ...options.extraProviders] : envProviders;
    const defaultProviderId = options.defaultProviderId ?? inferDefaultProvider(allProviders);
    // Merge any caller-supplied per-provider model overrides.
    const registry = mergeModelRegistry(options.models);
    if (options.debug) {
        const ids = allProviders.map((p) => p.id);
        console.log(`[ai] providers registered: [${ids.join(', ')}]`);
        console.log(`[ai] default provider: ${defaultProviderId}`);
    }
    const callerPolicy = options.modelPolicy ?? {};
    const isAuto = callerPolicy.mode === 'auto' || callerPolicy.mode === undefined;
    const autoTaskModels = isAuto ? buildPrefixedTaskModels(defaultProviderId, registry) : {};
    const mergedTaskModels = {
        ...autoTaskModels,
        ...(callerPolicy.taskModels ?? {}),
    };
    const modelPolicy = {
        mode: callerPolicy.mode ?? 'auto',
        ...(callerPolicy.defaultModel ? { defaultModel: callerPolicy.defaultModel } : {}),
        taskModels: mergedTaskModels,
    };
    return createAIGateway({
        providers: allProviders,
        defaultProviderId,
        modelPolicy,
        ...(options.retryPolicy !== undefined ? { retryPolicy: options.retryPolicy } : {}),
    });
}
