// @lumiarq/ai@1.2.0/dist/providers/openai.js
import { AI_PROVIDER_ENDPOINT_PATH, AI_PROVIDER_URL_PATH } from '../config/paths.js';
function mapFinishReason(raw) {
    const map = {
        stop: 'stop',
        length: 'length',
        content_filter: 'content-filter',
        tool_calls: 'tool-calls',
    };
    return map[raw] ?? 'unknown';
}
function buildMessages(request) {
    if (request.messages && request.messages.length > 0) {
        return request.messages.map((m) => ({ role: m.role, content: m.content }));
    }
    const msgs = [];
    if (request.systemPrompt && request.systemPrompt.trim().length > 0) {
        msgs.push({ role: 'system', content: request.systemPrompt });
    }
    msgs.push({ role: 'user', content: request.prompt });
    return msgs;
}
const TASK_DEFAULT_MODELS = {
    'intent-parse': 'gpt-4.1-mini',
    'plan-propose': 'gpt-4.1',
    'token-suggest': 'gpt-4.1-mini',
    'template-variant': 'gpt-4.1',
    'embed': 'text-embedding-3-small',
    'rerank': 'gpt-4.1-mini',
    'query-rewrite': 'gpt-4.1-mini',
    'answer-ground': 'gpt-4.1',
};
function resolveModel(task, model, options) {
    if (model && model.trim().length > 0 && !model.startsWith('openai:')) {
        return model.trim();
    }
    const stripped = model?.startsWith('openai:') ? model.slice('openai:'.length).trim() : undefined;
    if (stripped && stripped.length > 0) {
        return stripped;
    }
    const taskSpecific = options.taskModels?.[task];
    if (taskSpecific && taskSpecific.trim().length > 0) {
        return taskSpecific.trim();
    }
    if (options.defaultModel && options.defaultModel.trim().length > 0) {
        return options.defaultModel.trim();
    }
    return TASK_DEFAULT_MODELS[task];
}
export function createOpenAIProvider(options) {
    const baseUrl = options.baseUrl ?? AI_PROVIDER_URL_PATH.openai;
    const extra = options.extraHeaders ?? {};
    const authHeaders = () => ({
        'Content-Type': 'application/json',
        Authorization: `Bearer ${options.apiKey}`,
        ...extra,
    });
    return {
        id: options.id ?? 'openai',
        complete: async (request) => {
            const model = resolveModel(request.task, request.model, options);
            const messages = buildMessages(request);
            const body = {
                model,
                messages,
                ...(request.temperature !== undefined ? { temperature: request.temperature } : {}),
                max_tokens: request.maxTokens ?? 8192,
            };
            const response = await fetch(`${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.openai.chatCompletions}`, {
                method: 'POST',
                headers: authHeaders(),
                body: JSON.stringify(body),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok) {
                const text = await response.text().catch(() => '');
                throw new Error(`OpenAI request failed (${response.status}): ${text}`);
            }
            const json = (await response.json());
            const content = json.choices[0]?.message.content ?? '';
            const rawFinish = json.choices[0]?.finish_reason ?? 'unknown';
            const result = {
                provider: options.id ?? 'openai',
                model: json.model,
                text: content,
                finishReason: mapFinishReason(rawFinish),
            };
            if (json.usage) {
                const usage = {};
                if (json.usage.prompt_tokens !== undefined)
                    usage.inputTokens = json.usage.prompt_tokens;
                if (json.usage.completion_tokens !== undefined)
                    usage.outputTokens = json.usage.completion_tokens;
                result.usage = usage;
            }
            return result;
        },
        async *completeStream(request) {
            const model = resolveModel(request.task, request.model, options);
            const messages = buildMessages(request);
            const response = await fetch(`${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.openai.chatCompletions}`, {
                method: 'POST',
                headers: authHeaders(),
                body: JSON.stringify({
                    model,
                    messages,
                    ...(request.temperature !== undefined ? { temperature: request.temperature } : {}),
                    max_tokens: request.maxTokens ?? 8192,
                    stream: true,
                }),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok || !response.body) {
                const text = await response.text().catch(() => '');
                throw new Error(`OpenAI streaming request failed (${response.status}): ${text}`);
            }
            const reader = response.body.getReader();
            const decoder = new TextDecoder();
            let buffer = '';
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() ?? '';
                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed.startsWith('data: '))
                        continue;
                    const data = trimmed.slice(6);
                    if (data === '[DONE]')
                        return;
                    try {
                        const parsed = JSON.parse(data);
                        const chunk = parsed.choices?.[0]?.delta?.content;
                        if (typeof chunk === 'string')
                            yield chunk;
                    }
                    catch {
                        // skip malformed SSE chunks
                    }
                }
            }
        },
        async embed(request) {
            const model = resolveModel('embed', request.model, options);
            const input = Array.isArray(request.input) ? request.input : [request.input];
            const response = await fetch(`${baseUrl}${AI_PROVIDER_ENDPOINT_PATH.openai.embeddings}`, {
                method: 'POST',
                headers: authHeaders(),
                body: JSON.stringify({ model, input }),
                ...(request.signal !== undefined ? { signal: request.signal } : {}),
            });
            if (!response.ok) {
                const text = await response.text().catch(() => '');
                throw new Error(`OpenAI embed request failed (${response.status}): ${text}`);
            }
            const json = (await response.json());
            const result = {
                embeddings: json.data.map((d) => d.embedding),
                provider: options.id ?? 'openai',
                model: json.model,
            };
            if (json.usage?.prompt_tokens !== undefined) {
                result.usage = { inputTokens: json.usage.prompt_tokens };
            }
            return result;
        },
    };
}
