// @lumiarq/ai@1.2.0/dist/providers/openrouter.js
import { createOpenAIProvider } from './openai.js';
import { PROVIDER_DEFAULT_MODELS } from '../config/models.js';
import { AI_PROVIDER_URL_PATH } from '../config/paths.js';
/**
 * Create an OpenRouter provider instance.
 *
 * @param options Configuration options for the OpenRouter provider.
 * @returns An AIProvider instance configured for OpenRouter.
 */
export function createOpenRouterProvider(options) {
    const extraHeaders = {};
    if (options.appUrl)
        extraHeaders['HTTP-Referer'] = options.appUrl;
    if (options.appTitle)
        extraHeaders['X-Title'] = options.appTitle;
    return createOpenAIProvider({
        id: 'openrouter',
        apiKey: options.apiKey,
        baseUrl: AI_PROVIDER_URL_PATH.openrouter,
        defaultModel: options.defaultModel ?? (PROVIDER_DEFAULT_MODELS['openrouter']?.['token-suggest'] ?? 'meta-llama/llama-3.1-8b-instruct:free'),
        taskModels: { ...PROVIDER_DEFAULT_MODELS['openrouter'], ...options.taskModels },
        extraHeaders,
    });
}
