// @illumiarq/runtime@1.1.2/dist/scheduler/cron-scheduler.js
async function loadNodeCron() {
    try {
        return (await import('node-cron'));
    }
    catch {
        throw new Error('[CronScheduler] node-cron is not installed. Run: pnpm add node-cron && pnpm add -D @types/node-cron');
    }
}
function getActionName(action) {
    if (typeof action === 'function')
        return action.name || 'AnonymousAction';
    if (action && typeof action === 'object' && 'name' in action) {
        return String(action.name) || 'AnonymousAction';
    }
    return 'AnonymousAction';
}
export class CronScheduler {
    _jobs = [];
    started = false;
    call(action, cron, options = {}) {
        if (!cron || typeof cron !== 'string') {
            throw new Error(`[CronScheduler] Invalid cron expression: ${String(cron)}`);
        }
        const name = getActionName(action);
        this._jobs.push({
            action,
            name,
            cron,
            options,
            task: null,
            lastRunAt: null,
            lastStatus: 'pending',
        });
    }
    /**
     * Start all registered cron jobs.
     * Call this once from bootstrap/worker.ts after all jobs are registered.
     */
    async start() {
        if (this.started)
            return;
        const cron = await loadNodeCron();
        for (const entry of this._jobs) {
            const task = cron.schedule(entry.cron, async () => {
                entry.lastRunAt = new Date();
                try {
                    // Execute the action — supports both plain functions and defineAction() results
                    const action = entry.action;
                    if (typeof action === 'function') {
                        await action();
                    }
                    else if (action &&
                        typeof action === 'object' &&
                        'execute' in action &&
                        typeof action.execute === 'function') {
                        await action.execute();
                    }
                    entry.lastStatus = 'success';
                }
                catch (err) {
                    entry.lastStatus = 'failed';
                    console.error(`[CronScheduler] Job "${entry.name}" failed:`, err);
                }
            }, {
                scheduled: true,
                ...(entry.options.timezone ? { timezone: entry.options.timezone } : {}),
            });
            entry.task = task;
        }
        this.started = true;
    }
    /** Stop all running cron tasks. */
    stop() {
        for (const entry of this._jobs) {
            entry.task?.stop();
        }
        this.started = false;
    }
    jobs() {
        return this._jobs.map((entry) => ({
            action: entry.name,
            cron: entry.cron,
            options: entry.options,
            nextRunAt: computeNextRun(entry.cron),
            lastRunAt: entry.lastRunAt,
            lastStatus: entry.lastStatus,
        }));
    }
    due(now = new Date()) {
        // Return jobs that are past their computed next run time
        return this.jobs().filter((job) => job.nextRunAt <= now);
    }
    /** Run a specific job immediately by name (for lumis schedule:run). */
    async runByName(name) {
        const entry = this._jobs.find((j) => j.name === name);
        if (!entry)
            throw new Error(`[CronScheduler] No job found with name "${name}"`);
        const action = entry.action;
        if (typeof action === 'function') {
            await action();
        }
        else if (action &&
            typeof action === 'object' &&
            'execute' in action &&
            typeof action.execute === 'function') {
            await action.execute();
        }
        entry.lastRunAt = new Date();
        entry.lastStatus = 'success';
    }
}
/** Naive next-run computation — for display purposes only. */
function computeNextRun(_cron) {
    const next = new Date();
    next.setSeconds(0, 0);
    next.setMinutes(next.getMinutes() + 1);
    return next;
}
